
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genomio/reference/genomio/genome_metadata/">
      
      
        <link rel="prev" href="../genbank/">
      
      
        <link rel="next" href="../genome_stats/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.8">
    
    
      
        <title>genome_metadata - Ensembl-GenomIO</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.f2e4d321.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ensembl.io.genomio.genome_metadata" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Ensembl-GenomIO" class="md-header__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../img/ebang.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-GenomIO
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              genome_metadata
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M272 384c9.6-31.9 29.5-59.1 49.2-86.2 5.2-7.1 10.4-14.2 15.4-21.4 19.8-28.5 31.4-63 31.4-100.3C368 78.8 289.2 0 192 0S16 78.8 16 176c0 37.3 11.6 71.9 31.4 100.3 5 7.2 10.2 14.3 15.4 21.4 19.8 27.1 39.7 54.4 49.2 86.2h160zm-80 128c44.2 0 80-35.8 80-80v-16H112v16c0 44.2 35.8 80 80 80zm-80-336c0 8.8-7.2 16-16 16s-16-7.2-16-16c0-61.9 50.1-112 112-112 8.8 0 16 7.2 16 16s-7.2 16-16 16c-44.2 0-80 35.8-80 80z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M297.2 248.9c14.4-20.6 22.8-45.7 22.8-72.9 0-70.7-57.3-128-128-128S64 105.3 64 176c0 27.2 8.4 52.3 22.8 72.9 3.7 5.3 8.1 11.3 12.8 17.7 12.9 17.7 28.3 38.9 39.8 59.8 10.4 19 15.7 38.8 18.3 57.5H109c-2.2-12-5.9-23.7-11.8-34.5-9.9-18-22.2-34.9-34.5-51.8-5.2-7.1-10.4-14.2-15.4-21.4C27.6 247.9 16 213.3 16 176 16 78.8 94.8 0 192 0s176 78.8 176 176c0 37.3-11.6 71.9-31.4 100.3-5 7.2-10.2 14.3-15.4 21.4-12.3 16.8-24.6 33.7-34.5 51.8-5.9 10.8-9.6 22.5-11.8 34.5h-48.5c2.6-18.7 7.9-38.6 18.3-57.5 11.5-20.9 26.9-42.1 39.8-59.8 4.7-6.4 9-12.4 12.7-17.7zM192 128c-26.5 0-48 21.5-48 48 0 8.8-7.2 16-16 16s-16-7.2-16-16c0-44.2 35.8-80 80-80 8.8 0 16 7.2 16 16s-7.2 16-16 16zm0 384c-44.2 0-80-35.8-80-80v-16h160v16c0 44.2-35.8 80-80 80z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl-GenomIO
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Ensembl-GenomIO" class="md-nav__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../img/ebang.png" alt="logo">

    </a>
    Ensembl-GenomIO
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl-GenomIO
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ensembl
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    brc4
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            brc4
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../runnable/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    runnable
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_1" id="__nav_4_1_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            runnable
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/compare_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_fasta.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/compare_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_report.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/core_server/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    core_server.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/download_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download_genbank.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/dump_stable_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_stable_ids.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/extract_from_gb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_gb.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/fill_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fill_metadata.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/gff3_specifier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3_specifier.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/load_sequence_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load_sequence_data.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/read_json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    read_json.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/say_accession/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    say_accession.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/seqregion_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seqregion_parser.py
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    io
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            io
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_1" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genomio
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_1" id="__nav_4_1_2_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            genomio
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    database
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../events/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fasta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;genome_metadata
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" genome_metadata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.DEFAULT_API_URL" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;DEFAULT_API_URL
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.PROVIDER_DATA" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;PROVIDER_DATA
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.MetadataError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MetadataError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.MissingDataError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MissingDataError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.MissingNodeError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MissingNodeError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.add_assembly_version" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_assembly_version
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.add_genebuild_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_genebuild_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.add_provider" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_provider
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.add_species_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_species_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.amend_genomic_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;amend_genomic_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.check_assembly_version" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_assembly_version
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.filter_genome_meta" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;filter_genome_meta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_additions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_additions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_gbff_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_gbff_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_genome_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_genome_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_report_regions_names" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_report_regions_names
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_taxonomy_from_accession" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_taxonomy_from_accession
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.prepare_genome_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;prepare_genome_metadata
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genome_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_1_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../schemas/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    schemas
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_1_10" id="__nav_4_1_2_1_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_2_1_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2_1_10">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seq_region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Nextflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;genome_metadata
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" genome_metadata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.DEFAULT_API_URL" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;DEFAULT_API_URL
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.PROVIDER_DATA" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;PROVIDER_DATA
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.MetadataError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MetadataError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.MissingDataError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MissingDataError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.MissingNodeError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MissingNodeError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.add_assembly_version" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_assembly_version
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.add_genebuild_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_genebuild_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.add_provider" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_provider
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.add_species_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_species_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.amend_genomic_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;amend_genomic_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.check_assembly_version" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_assembly_version
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.filter_genome_meta" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;filter_genome_meta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_additions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_additions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_gbff_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_gbff_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_genome_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_genome_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_report_regions_names" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_report_regions_names
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.get_taxonomy_from_accession" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_taxonomy_from_accession
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.genome_metadata.prepare_genome_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;prepare_genome_metadata
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>genome_metadata</h1>

<div class="doc doc-object doc-module">



<h2 id="ensembl.io.genomio.genome_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>          <code>ensembl.io.genomio.genome_metadata</code>


<a href="#ensembl.io.genomio.genome_metadata" class="headerlink" title="Permanent link">&para;</a></h2>

  <div class="doc doc-contents first">
  
      <p>Genome metadata handling module.</p>

  

  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.genome_metadata.DEFAULT_API_URL" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>          <code class="highlight language-python"><span class="n">DEFAULT_API_URL</span> <span class="o">=</span> <span class="s1">&#39;https://www.ebi.ac.uk/ena/browser/api/xml/&#39;</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.genome_metadata.DEFAULT_API_URL" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.genome_metadata.PROVIDER_DATA" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>          <code class="highlight language-python"><span class="n">PROVIDER_DATA</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;GenBank&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;assembly&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;provider_name&#39;</span><span class="p">:</span> <span class="s1">&#39;GenBank&#39;</span><span class="p">,</span> <span class="s1">&#39;provider_url&#39;</span><span class="p">:</span> <span class="s1">&#39;https://www.ncbi.nlm.nih.gov/assembly&#39;</span><span class="p">},</span> <span class="s1">&#39;annotation&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;provider_name&#39;</span><span class="p">:</span> <span class="s1">&#39;GenBank&#39;</span><span class="p">,</span> <span class="s1">&#39;provider_url&#39;</span><span class="p">:</span> <span class="s1">&#39;https://www.ncbi.nlm.nih.gov/assembly&#39;</span><span class="p">}},</span> <span class="s1">&#39;RefSeq&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;assembly&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;provider_name&#39;</span><span class="p">:</span> <span class="s1">&#39;RefSeq&#39;</span><span class="p">,</span> <span class="s1">&#39;provider_url&#39;</span><span class="p">:</span> <span class="s1">&#39;https://www.ncbi.nlm.nih.gov/refseq&#39;</span><span class="p">},</span> <span class="s1">&#39;annotation&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;provider_name&#39;</span><span class="p">:</span> <span class="s1">&#39;RefSeq&#39;</span><span class="p">,</span> <span class="s1">&#39;provider_url&#39;</span><span class="p">:</span> <span class="s1">&#39;https://www.ncbi.nlm.nih.gov/refseq&#39;</span><span class="p">}}}</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.genome_metadata.PROVIDER_DATA" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  </div>

</div>


<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.genome_metadata.MetadataError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>MetadataError</code>


<a href="#ensembl.io.genomio.genome_metadata.MetadataError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>When a metadata value is not expected.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/prepare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">74</span>
<span class="normal">75</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">MetadataError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;When a metadata value is not expected.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.genome_metadata.MissingDataError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>MissingDataError</code>


<a href="#ensembl.io.genomio.genome_metadata.MissingDataError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>Used if some data is missing from the report file.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/extend.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">MissingDataError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Used if some data is missing from the report file.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">report_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">msg</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">report_msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Can&#39;t get data for </span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2"> in report </span><span class="si">{</span><span class="n">report_path</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="n">msg</span><span class="p">:</span>
            <span class="n">report_msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">report_msg</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">msg</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">msg</span> <span class="o">=</span> <span class="n">report_msg</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">











  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.genome_metadata.MissingNodeError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>MissingNodeError</code>


<a href="#ensembl.io.genomio.genome_metadata.MissingNodeError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>When a taxon XML node cannot be found.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/prepare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">70</span>
<span class="normal">71</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">MissingNodeError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;When a taxon XML node cannot be found.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>



<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.add_assembly_version" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">add_assembly_version</span><span class="p">(</span><span class="n">genome_data</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.add_assembly_version" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Adds version number to the genome's assembly if one is not present already.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>genome_data</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Genome information of assembly, accession and annotation.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_assembly_version</span><span class="p">(</span><span class="n">genome_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds version number to the genome&#39;s assembly if one is not present already.</span>

<span class="sd">    Args:</span>
<span class="sd">        genome_data: Genome information of assembly, accession and annotation.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">assembly</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="s2">&quot;version&quot;</span> <span class="ow">in</span> <span class="n">assembly</span><span class="p">:</span>
        <span class="n">accession</span> <span class="o">=</span> <span class="n">assembly</span><span class="p">[</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">accession</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)</span> <span class="ow">and</span> <span class="n">values</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">assembly</span><span class="p">[</span><span class="s2">&quot;version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">values</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.add_genebuild_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">add_genebuild_metadata</span><span class="p">(</span><span class="n">genome_data</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.add_genebuild_metadata" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Adds missing genebuild metadata.</p>
<p>The default convention is to use the current date as <code>version</code> and <code>start_date</code>.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>genome_data</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Genome information of assembly, accession and annotation.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_genebuild_metadata</span><span class="p">(</span><span class="n">genome_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds missing genebuild metadata.</span>

<span class="sd">    The default convention is to use the current date as ``version`` and ``start_date``.</span>

<span class="sd">    Args:</span>
<span class="sd">        genome_data: Genome information of assembly, accession and annotation.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genebuild</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;genebuild&quot;</span><span class="p">]</span>
    <span class="n">current_date</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">date</span><span class="o">.</span><span class="n">today</span><span class="p">()</span><span class="o">.</span><span class="n">isoformat</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="s2">&quot;version&quot;</span> <span class="ow">in</span> <span class="n">genebuild</span><span class="p">:</span>
        <span class="n">genebuild</span><span class="p">[</span><span class="s2">&quot;version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">current_date</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="s2">&quot;start_date&quot;</span> <span class="ow">in</span> <span class="n">genebuild</span><span class="p">:</span>
        <span class="n">genebuild</span><span class="p">[</span><span class="s2">&quot;start_date&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">current_date</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.add_provider" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">add_provider</span><span class="p">(</span><span class="n">genome_data</span><span class="p">,</span> <span class="n">gff3_file</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.add_provider" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Adds provider metadata for assembly and gene models in <code>genome_data</code>.</p>
<p>Assembly provider metadata will only be added if it is missing, i.e. neither <code>provider_name</code> or
<code>provider_url</code> are present. The gene model metadata will only be added if <code>gff3_file</code> is provided.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>genome_data</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Genome information of assembly, accession and annotation.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>gff3_file</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="os.PathLike">PathLike</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to GFF3 file to use as annotation source for this genome.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_provider</span><span class="p">(</span><span class="n">genome_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">gff3_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds provider metadata for assembly and gene models in `genome_data`.</span>

<span class="sd">    Assembly provider metadata will only be added if it is missing, i.e. neither ``provider_name`` or</span>
<span class="sd">    ``provider_url`` are present. The gene model metadata will only be added if `gff3_file` is provided.</span>

<span class="sd">    Args:</span>
<span class="sd">        genome_data: Genome information of assembly, accession and annotation.</span>
<span class="sd">        gff3_file: Path to GFF3 file to use as annotation source for this genome.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get accession provider</span>
    <span class="n">accession</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">accession</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GCF&quot;</span><span class="p">):</span>
        <span class="n">provider</span> <span class="o">=</span> <span class="n">PROVIDER_DATA</span><span class="p">[</span><span class="s2">&quot;RefSeq&quot;</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">accession</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GCA&quot;</span><span class="p">):</span>
        <span class="n">provider</span> <span class="o">=</span> <span class="n">PROVIDER_DATA</span><span class="p">[</span><span class="s2">&quot;GenBank&quot;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">MetadataError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Accession doesn&#39;t look like an INSDC or RefSeq accession: </span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Add assembly provider (if missing)</span>
    <span class="n">assembly</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="s2">&quot;provider_name&quot;</span> <span class="ow">in</span> <span class="n">assembly</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="ow">not</span> <span class="s2">&quot;provider_url&quot;</span> <span class="ow">in</span> <span class="n">assembly</span><span class="p">):</span>
        <span class="n">assembly</span><span class="p">[</span><span class="s2">&quot;provider_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">provider</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;provider_name&quot;</span><span class="p">]</span>
        <span class="n">assembly</span><span class="p">[</span><span class="s2">&quot;provider_url&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">provider</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;provider_url&quot;</span><span class="p">]</span>

    <span class="c1"># Add annotation provider if there are gene models</span>
    <span class="k">if</span> <span class="n">gff3_file</span><span class="p">:</span>
        <span class="n">annotation</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="s2">&quot;annotation&quot;</span> <span class="ow">in</span> <span class="n">genome_data</span><span class="p">:</span>
            <span class="n">annotation</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;annotation&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;provider_name&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotation</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;provider_url&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotation</span><span class="p">):</span>
            <span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;provider_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">provider</span><span class="p">[</span><span class="s2">&quot;annotation&quot;</span><span class="p">][</span><span class="s2">&quot;provider_name&quot;</span><span class="p">]</span>
            <span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;provider_url&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">provider</span><span class="p">[</span><span class="s2">&quot;annotation&quot;</span><span class="p">][</span><span class="s2">&quot;provider_url&quot;</span><span class="p">]</span>
        <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;annotation&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotation</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.add_species_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">add_species_metadata</span><span class="p">(</span><span class="n">genome_data</span><span class="p">,</span> <span class="n">base_api_url</span><span class="o">=</span><span class="n">DEFAULT_API_URL</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.add_species_metadata" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Adds missing species metadata based on the genome's accession.</p>
<p>The <code>taxonomy_id</code>, <code>strain</code> and <code>scientific_name</code> will be fetched from the taxonomy information
linked to the given accession.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>genome_data</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Genome information of assembly, accession and annotation.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>base_api_url</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Base API URL to fetch the taxonomy data from.</p>
            </div>
          </td>
          <td>
                <code><a class="autorefs autorefs-internal" title="ensembl.io.genomio.genome_metadata.prepare.DEFAULT_API_URL" href="#ensembl.io.genomio.genome_metadata.DEFAULT_API_URL">DEFAULT_API_URL</a></code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_species_metadata</span><span class="p">(</span><span class="n">genome_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">base_api_url</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">DEFAULT_API_URL</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds missing species metadata based on the genome&#39;s accession.</span>

<span class="sd">    The ``taxonomy_id``, ``strain`` and ``scientific_name`` will be fetched from the taxonomy information</span>
<span class="sd">    linked to the given accession.</span>

<span class="sd">    Args:</span>
<span class="sd">        genome_data: Genome information of assembly, accession and annotation.</span>
<span class="sd">        base_api_url: Base API URL to fetch the taxonomy data from.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">species</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;species&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="s2">&quot;taxonomy_id&quot;</span> <span class="ow">in</span> <span class="n">species</span><span class="p">:</span>
        <span class="n">accession</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span>
        <span class="n">taxonomy</span> <span class="o">=</span> <span class="n">get_taxonomy_from_accession</span><span class="p">(</span><span class="n">accession</span><span class="p">,</span> <span class="n">base_api_url</span><span class="p">)</span>
        <span class="n">species</span><span class="p">[</span><span class="s2">&quot;taxonomy_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="s2">&quot;strain&quot;</span> <span class="ow">in</span> <span class="n">species</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;strain&quot;</span> <span class="ow">in</span> <span class="n">taxonomy</span><span class="p">):</span>
            <span class="n">species</span><span class="p">[</span><span class="s2">&quot;strain&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="p">[</span><span class="s2">&quot;strain&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="s2">&quot;scientific_name&quot;</span> <span class="ow">in</span> <span class="n">species</span><span class="p">:</span>
            <span class="n">species</span><span class="p">[</span><span class="s2">&quot;scientific_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="p">[</span><span class="s2">&quot;scientific_name&quot;</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.amend_genomic_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">amend_genomic_metadata</span><span class="p">(</span><span class="n">genome_infile</span><span class="p">,</span> <span class="n">genome_outfile</span><span class="p">,</span> <span class="n">report_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">genbank_infile</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.amend_genomic_metadata" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>genome_infile</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Genome data following the src/python/ensembl/io/genomio/data/schemas/genome.json.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>genome_outfile</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Amended genome data file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>report_file</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="os.PathLike">PathLike</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>INSDC/RefSeq sequences report file.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>genbank_infile</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="os.PathLike">PathLike</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>INSDC/RefSeq GBFF file.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/extend.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">amend_genomic_metadata</span><span class="p">(</span>
    <span class="n">genome_infile</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">genome_outfile</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">report_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">genbank_infile</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Args:</span>
<span class="sd">        genome_infile: Genome data following the src/python/ensembl/io/genomio/data/schemas/genome.json.</span>
<span class="sd">        genome_outfile: Amended genome data file.</span>
<span class="sd">        report_file: INSDC/RefSeq sequences report file.</span>
<span class="sd">        genbank_infile: INSDC/RefSeq GBFF file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome_metadata</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">genome_infile</span><span class="p">)</span>

    <span class="c1"># Get additional sequences in the assembly but not in the data</span>
    <span class="k">if</span> <span class="n">report_file</span><span class="p">:</span>
        <span class="n">gbff_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">genbank_infile</span><span class="p">)</span> <span class="k">if</span> <span class="n">genbank_infile</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="n">additions</span> <span class="o">=</span> <span class="n">get_additions</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">report_file</span><span class="p">),</span> <span class="n">gbff_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">additions</span><span class="p">:</span>
            <span class="n">genome_metadata</span><span class="p">[</span><span class="s2">&quot;added_seq&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;region_name&quot;</span><span class="p">:</span> <span class="n">additions</span><span class="p">}</span>

    <span class="c1"># Print out the file</span>
    <span class="n">genome_outfile</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">genome_outfile</span><span class="p">)</span>
    <span class="n">print_json</span><span class="p">(</span><span class="n">genome_outfile</span><span class="p">,</span> <span class="n">genome_metadata</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.check_assembly_version" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">check_assembly_version</span><span class="p">(</span><span class="n">gmeta_out</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.check_assembly_version" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Update the assembly version of the genome metadata provided to use an integer.
Get the version from the assembly accession as alternative.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gmeta</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Nested metadata key values from the core metadata table.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_assembly_version</span><span class="p">(</span><span class="n">gmeta_out</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Update the assembly version of the genome metadata provided to use an integer.</span>
<span class="sd">    Get the version from the assembly accession as alternative.</span>

<span class="sd">    Args:</span>
<span class="sd">        gmeta (Dict[str, Any]): Nested metadata key values from the core metadata table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">assembly</span> <span class="o">=</span> <span class="n">gmeta_out</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">]</span>
    <span class="n">version</span> <span class="o">=</span> <span class="n">assembly</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;version&quot;</span><span class="p">)</span>

    <span class="c1"># Check the version is an integer</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">assembly</span><span class="p">[</span><span class="s2">&quot;version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">version</span><span class="p">)</span>
    <span class="k">except</span> <span class="p">(</span><span class="ne">ValueError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">)</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
        <span class="c1"># Get the version from the assembly accession</span>
        <span class="n">accession</span> <span class="o">=</span> <span class="n">assembly</span><span class="p">[</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="n">accession</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">isdigit</span><span class="p">():</span>
            <span class="n">version</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">assembly</span><span class="p">[</span><span class="s2">&quot;version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">version</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s1">&#39;Asm version [v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s1">] obtained from: assembly accession (</span><span class="si">{</span><span class="n">assembly</span><span class="p">[</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">).&#39;</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Assembly version is not an integer in </span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">exc</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Located version [v</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">version</span><span class="p">)</span><span class="si">}</span><span class="s2">] info from meta data.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.filter_genome_meta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">filter_genome_meta</span><span class="p">(</span><span class="n">gmeta</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.filter_genome_meta" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns a filtered metadata dict with only predefined keys.
Also converts expected numbers to integers (to follow the genome json schema).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gmeta</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Nested metadata key values from the core metadata table.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">filter_genome_meta</span><span class="p">(</span><span class="n">gmeta</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a filtered metadata dict with only predefined keys.</span>
<span class="sd">    Also converts expected numbers to integers (to follow the genome json schema).</span>

<span class="sd">    Args:</span>
<span class="sd">        gmeta (Dict[str, Any]): Nested metadata key values from the core metadata table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">meta_list</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;taxonomy_id&quot;</span><span class="p">,</span>
            <span class="s2">&quot;production_name&quot;</span><span class="p">,</span>
            <span class="s2">&quot;scientific_name&quot;</span><span class="p">,</span>
            <span class="s2">&quot;strain&quot;</span><span class="p">,</span>
            <span class="s2">&quot;display_name&quot;</span><span class="p">,</span>
            <span class="s2">&quot;division&quot;</span><span class="p">,</span>
            <span class="s2">&quot;alias&quot;</span><span class="p">,</span>
            <span class="s2">&quot;annotation_source&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;assembly&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;accession&quot;</span><span class="p">,</span> <span class="s2">&quot;date&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;version&quot;</span><span class="p">,</span> <span class="s2">&quot;provider_name&quot;</span><span class="p">,</span> <span class="s2">&quot;provider_url&quot;</span><span class="p">},</span>
        <span class="s2">&quot;genebuild&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;version&quot;</span><span class="p">,</span> <span class="s2">&quot;method&quot;</span><span class="p">,</span> <span class="s2">&quot;start_date&quot;</span><span class="p">,</span> <span class="s2">&quot;method_display&quot;</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">},</span>
        <span class="s2">&quot;annotation&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;provider_name&quot;</span><span class="p">,</span> <span class="s2">&quot;provider_url&quot;</span><span class="p">},</span>
        <span class="s2">&quot;BRC4&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;organism_abbrev&quot;</span><span class="p">,</span> <span class="s2">&quot;component&quot;</span><span class="p">},</span>
        <span class="s2">&quot;added_seq&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;region_name&quot;</span><span class="p">},</span>
    <span class="p">}</span>
    <span class="n">is_integer</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;taxonomy_id&quot;</span><span class="p">},</span> <span class="s2">&quot;assembly&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;version&quot;</span><span class="p">}}</span>

    <span class="n">gmeta_out</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">key1</span><span class="p">,</span> <span class="n">subkeys</span> <span class="ow">in</span> <span class="n">meta_list</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">key1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gmeta</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">subkeys</span><span class="p">:</span>
            <span class="n">gmeta_out</span><span class="p">[</span><span class="n">key1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">key2</span> <span class="ow">in</span> <span class="n">subkeys</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">key2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gmeta</span><span class="p">[</span><span class="n">key1</span><span class="p">]:</span>
                    <span class="k">continue</span>
                <span class="n">value</span> <span class="o">=</span> <span class="n">gmeta</span><span class="p">[</span><span class="n">key1</span><span class="p">][</span><span class="n">key2</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">key2</span> <span class="ow">in</span> <span class="n">is_integer</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">key1</span><span class="p">,</span> <span class="p">{}):</span>
                        <span class="n">value</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
                <span class="n">gmeta_out</span><span class="p">[</span><span class="n">key1</span><span class="p">][</span><span class="n">key2</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">value</span> <span class="o">=</span> <span class="n">gmeta</span><span class="p">[</span><span class="n">key1</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">value</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">is_integer</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">key1</span><span class="p">):</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
            <span class="n">gmeta_out</span><span class="p">[</span><span class="n">key1</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

    <span class="n">check_assembly_version</span><span class="p">(</span><span class="n">gmeta_out</span><span class="p">)</span>
    <span class="n">check_genebuild_version</span><span class="p">(</span><span class="n">gmeta_out</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gmeta_out</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.get_additions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_additions</span><span class="p">(</span><span class="n">report_path</span><span class="p">,</span> <span class="n">gbff_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.get_additions" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns all <code>seq_regions</code> that are mentioned in the report but that are not in the data.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>report_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to the report file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>gbff_path</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="pathlib.Path">Path</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to the GBFF file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/extend.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_additions</span><span class="p">(</span><span class="n">report_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">gbff_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns all `seq_regions` that are mentioned in the report but that are not in the data.</span>

<span class="sd">    Args:</span>
<span class="sd">        report_path: Path to the report file.</span>
<span class="sd">        gbff_path: Path to the GBFF file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gbff_regions</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">get_gbff_regions</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">))</span>
    <span class="n">report_regions</span> <span class="o">=</span> <span class="n">get_report_regions_names</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span>

    <span class="n">additions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">rep_seq</span> <span class="ow">in</span> <span class="n">report_regions</span><span class="p">:</span>
        <span class="p">(</span><span class="n">rs_seq</span><span class="p">,</span> <span class="n">gb_seq</span><span class="p">)</span> <span class="o">=</span> <span class="n">rep_seq</span>
        <span class="k">if</span> <span class="n">rs_seq</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gbff_regions</span> <span class="ow">and</span> <span class="n">gb_seq</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gbff_regions</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">rs_seq</span><span class="p">:</span>
                <span class="n">additions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rs_seq</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">additions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gb_seq</span><span class="p">)</span>
    <span class="n">additions</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">additions</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">additions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.get_gbff_regions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_gbff_regions</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.get_gbff_regions" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns the <code>seq_region</code> data from the GBFF file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gbff_path</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="pathlib.Path">Path</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Gbff file path to use.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/extend.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_gbff_regions</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the `seq_region` data from the GBFF file.</span>

<span class="sd">    Args:</span>
<span class="sd">        gbff_path: Gbff file path to use.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">gbff_path</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>

    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gbff_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gbff_file</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
            <span class="n">record_id</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">_VERSION_END</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="n">seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record_id</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.get_genome_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_genome_metadata</span><span class="p">(</span><span class="n">session</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.get_genome_metadata" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a select list of metadata from the core database.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>session</code></td>
          <td>
                <code><span title="sqlalchemy.orm.Session">Session</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Session for the current core.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A nested dict.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_genome_metadata</span><span class="p">(</span><span class="n">session</span><span class="p">:</span> <span class="n">Session</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieve a select list of metadata from the core database.</span>

<span class="sd">    Args:</span>
<span class="sd">        session: Session for the current core.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A nested dict.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gmeta</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="n">gmeta_st</span> <span class="o">=</span> <span class="n">select</span><span class="p">(</span><span class="n">Meta</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">session</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">gmeta_st</span><span class="p">)</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
        <span class="n">dat</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">meta_key</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">meta_key</span>
        <span class="n">meta_value</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">meta_value</span>

        <span class="k">if</span> <span class="s2">&quot;.&quot;</span> <span class="ow">in</span> <span class="n">meta_key</span><span class="p">:</span>
            <span class="p">(</span><span class="n">high_key</span><span class="p">,</span> <span class="n">low_key</span><span class="p">)</span> <span class="o">=</span> <span class="n">meta_key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">high_key</span> <span class="ow">in</span> <span class="n">gmeta</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">low_key</span> <span class="ow">in</span> <span class="n">gmeta</span><span class="p">[</span><span class="n">high_key</span><span class="p">]:</span>
                    <span class="n">gmeta</span><span class="p">[</span><span class="n">high_key</span><span class="p">][</span><span class="n">low_key</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">meta_value</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">gmeta</span><span class="p">[</span><span class="n">high_key</span><span class="p">][</span><span class="n">low_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">meta_value</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gmeta</span><span class="p">[</span><span class="n">high_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">gmeta</span><span class="p">[</span><span class="n">high_key</span><span class="p">][</span><span class="n">low_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">meta_value</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">meta_key</span> <span class="ow">in</span> <span class="n">gmeta</span><span class="p">:</span>
                <span class="n">gmeta</span><span class="p">[</span><span class="n">meta_key</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">meta_value</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gmeta</span><span class="p">[</span><span class="n">meta_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">meta_value</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">gmeta</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.get_report_regions_names" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_report_regions_names</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.get_report_regions_names" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns a list of <code>seq_region</code> names from the report file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>report_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to the seq_regions report from INSDC/RefSeq.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/extend.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_report_regions_names</span><span class="p">(</span><span class="n">report_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a list of `seq_region` names from the report file.</span>

<span class="sd">    Args:</span>
<span class="sd">        report_path: Path to the seq_regions report from INSDC/RefSeq.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get the report in a CSV format, easier to manipulate</span>
    <span class="n">report_csv</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">_report_to_csv</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span>

    <span class="c1"># Feed the csv string to the CSV reader</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">report_csv</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_NONE</span><span class="p">)</span>

    <span class="c1"># Create the seq_regions</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">reader</span><span class="p">:</span>
        <span class="n">refseq_name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;RefSeq-Accn&quot;</span><span class="p">]</span>
        <span class="n">genbank_name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;GenBank-Accn&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">refseq_name</span> <span class="o">==</span> <span class="s2">&quot;na&quot;</span><span class="p">:</span>
            <span class="n">refseq_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">if</span> <span class="n">genbank_name</span> <span class="o">==</span> <span class="s2">&quot;na&quot;</span><span class="p">:</span>
            <span class="n">genbank_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">refseq_name</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">_VERSION_END</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">refseq_name</span><span class="p">)</span>
        <span class="n">genbank_name</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">_VERSION_END</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">genbank_name</span><span class="p">)</span>
        <span class="n">seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">genbank_name</span><span class="p">,</span> <span class="n">refseq_name</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.get_taxonomy_from_accession" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_taxonomy_from_accession</span><span class="p">(</span><span class="n">accession</span><span class="p">,</span> <span class="n">base_api_url</span><span class="o">=</span><span class="n">DEFAULT_API_URL</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.get_taxonomy_from_accession" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns the taxonomy metadata associated to the given accession.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>accession</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>INSDC accession ID.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>base_api_url</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Base API URL to fetch the taxonomy data from.</p>
            </div>
          </td>
          <td>
                <code><a class="autorefs autorefs-internal" title="ensembl.io.genomio.genome_metadata.prepare.DEFAULT_API_URL" href="#ensembl.io.genomio.genome_metadata.DEFAULT_API_URL">DEFAULT_API_URL</a></code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Dictionary with key-value pairs for <code>taxon_id</code> and <code>scientific_name</code>. <code>strain</code> will be added</p>
            </div>
          </td>
        </tr>
        <tr>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>only if present in the fetched taxonomy data.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>MissinDataException</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If <code>TAXON_ID</code> or <code>SCIENTIFIC_NAME</code> are missing in the taxonomy data fetched.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_taxonomy_from_accession</span><span class="p">(</span><span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">base_api_url</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">DEFAULT_API_URL</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the taxonomy metadata associated to the given accession.</span>

<span class="sd">    Args:</span>
<span class="sd">        accession: INSDC accession ID.</span>
<span class="sd">        base_api_url: Base API URL to fetch the taxonomy data from.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dictionary with key-value pairs for ``taxon_id`` and ``scientific_name``. ``strain`` will be added</span>
<span class="sd">        only if present in the fetched taxonomy data.</span>

<span class="sd">    Raises:</span>
<span class="sd">        MissinDataException: If ``TAXON_ID`` or ``SCIENTIFIC_NAME`` are missing in the taxonomy data fetched.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Use the GenBank accession without version</span>
    <span class="n">gb_accession</span> <span class="o">=</span> <span class="n">accession</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;GCF&quot;</span><span class="p">,</span> <span class="s2">&quot;GCA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_api_url</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gb_accession</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">60</span><span class="p">)</span>
    <span class="n">entry</span> <span class="o">=</span> <span class="n">ElementTree</span><span class="o">.</span><span class="n">fromstring</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>

    <span class="n">taxon_node</span> <span class="o">=</span> <span class="n">entry</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;.//TAXON&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">taxon_node</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">MissingNodeError</span><span class="p">(</span><span class="s2">&quot;Can&#39;t find the TAXON node&quot;</span><span class="p">)</span>

    <span class="c1"># Fetch taxon ID, scientific_name and strain</span>
    <span class="n">taxon_id</span> <span class="o">=</span> <span class="n">_get_node_text</span><span class="p">(</span><span class="n">taxon_node</span><span class="p">,</span> <span class="s2">&quot;TAXON_ID&quot;</span><span class="p">)</span>
    <span class="n">scientific_name</span> <span class="o">=</span> <span class="n">_get_node_text</span><span class="p">(</span><span class="n">taxon_node</span><span class="p">,</span> <span class="s2">&quot;SCIENTIFIC_NAME&quot;</span><span class="p">)</span>
    <span class="n">strain</span> <span class="o">=</span> <span class="n">_get_node_text</span><span class="p">(</span><span class="n">taxon_node</span><span class="p">,</span> <span class="s2">&quot;STRAIN&quot;</span><span class="p">,</span> <span class="n">optional</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">taxon_id</span> <span class="ow">and</span> <span class="n">scientific_name</span><span class="p">:</span>
        <span class="n">taxonomy</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;taxon_id&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">),</span>
            <span class="s2">&quot;scientific_name&quot;</span><span class="p">:</span> <span class="n">scientific_name</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="k">if</span> <span class="n">strain</span><span class="p">:</span>
        <span class="n">taxonomy</span><span class="p">[</span><span class="s2">&quot;strain&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">strain</span>
    <span class="k">return</span> <span class="n">taxonomy</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.genome_metadata.prepare_genome_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">prepare_genome_metadata</span><span class="p">(</span><span class="n">input_file</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="n">gff3_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">base_api_url</span><span class="o">=</span><span class="n">DEFAULT_API_URL</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.genome_metadata.prepare_genome_metadata" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Updates the genome metadata JSON file with additional information.</p>
<p>In particular, more information is added about the provider, the assembly and its gene build version,
and the taxonomy.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>input_file</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to JSON file with genome metadata.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>output_file</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Output directory where to generate the final <code>genome.json</code> file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>gff3_file</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="os.PathLike">PathLike</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to GFF3 file to use as annotation source for this genome.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>base_api_url</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Base API URL to fetch the taxonomy data from.</p>
            </div>
          </td>
          <td>
                <code><a class="autorefs autorefs-internal" title="ensembl.io.genomio.genome_metadata.prepare.DEFAULT_API_URL" href="#ensembl.io.genomio.genome_metadata.DEFAULT_API_URL">DEFAULT_API_URL</a></code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/genome_metadata/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">prepare_genome_metadata</span><span class="p">(</span>
    <span class="n">input_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">output_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">gff3_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">base_api_url</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">DEFAULT_API_URL</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Updates the genome metadata JSON file with additional information.</span>

<span class="sd">    In particular, more information is added about the provider, the assembly and its gene build version,</span>
<span class="sd">    and the taxonomy.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_file: Path to JSON file with genome metadata.</span>
<span class="sd">        output_file: Output directory where to generate the final `genome.json` file.</span>
<span class="sd">        gff3_file: Path to GFF3 file to use as annotation source for this genome.</span>
<span class="sd">        base_api_url: Base API URL to fetch the taxonomy data from.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome_data</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
    <span class="c1"># Amend any missing metadata</span>
    <span class="n">add_provider</span><span class="p">(</span><span class="n">genome_data</span><span class="p">,</span> <span class="n">gff3_file</span><span class="p">)</span>
    <span class="n">add_assembly_version</span><span class="p">(</span><span class="n">genome_data</span><span class="p">)</span>
    <span class="n">add_genebuild_metadata</span><span class="p">(</span><span class="n">genome_data</span><span class="p">)</span>
    <span class="n">add_species_metadata</span><span class="p">(</span><span class="n">genome_data</span><span class="p">,</span> <span class="n">base_api_url</span><span class="p">)</span>
    <span class="c1"># Dump updated genome metadata</span>
    <span class="n">print_json</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">genome_data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-2023] EMBL-European Bioinformatics Institute
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.8fd75fb4.min.js"></script>
      
    
  </body>
</html>