Search.setIndex({"docnames": ["ensembl", "ensembl.brc4", "ensembl.brc4.runnable", "ensembl.io", "ensembl.io.genomio", "ensembl.io.genomio.assembly", "ensembl.io.genomio.database", "ensembl.io.genomio.events", "ensembl.io.genomio.fasta", "ensembl.io.genomio.genbank", "ensembl.io.genomio.genome_metadata", "ensembl.io.genomio.genome_stats", "ensembl.io.genomio.gff3", "ensembl.io.genomio.manifest", "ensembl.io.genomio.schemas", "ensembl.io.genomio.schemas.json", "ensembl.io.genomio.seq_region", "ensembl.io.genomio.utils", "index", "install", "license", "modules", "usage"], "filenames": ["ensembl.rst", "ensembl.brc4.rst", "ensembl.brc4.runnable.rst", "ensembl.io.rst", "ensembl.io.genomio.rst", "ensembl.io.genomio.assembly.rst", "ensembl.io.genomio.database.rst", "ensembl.io.genomio.events.rst", "ensembl.io.genomio.fasta.rst", "ensembl.io.genomio.genbank.rst", "ensembl.io.genomio.genome_metadata.rst", "ensembl.io.genomio.genome_stats.rst", "ensembl.io.genomio.gff3.rst", "ensembl.io.genomio.manifest.rst", "ensembl.io.genomio.schemas.rst", "ensembl.io.genomio.schemas.json.rst", "ensembl.io.genomio.seq_region.rst", "ensembl.io.genomio.utils.rst", "index.rst", "install.rst", "license.rst", "modules.rst", "usage.rst"], "titles": ["ensembl namespace", "ensembl.brc4 package", "ensembl.brc4.runnable package", "ensembl.io namespace", "ensembl.io.genomio package", "ensembl.io.genomio.assembly package", "ensembl.io.genomio.database package", "ensembl.io.genomio.events package", "ensembl.io.genomio.fasta package", "ensembl.io.genomio.genbank package", "ensembl.io.genomio.genome_metadata package", "ensembl.io.genomio.genome_stats package", "ensembl.io.genomio.gff3 package", "ensembl.io.genomio.manifest package", "ensembl.io.genomio.schemas package", "ensembl.io.genomio.schemas.json package", "ensembl.io.genomio.seq_region package", "ensembl.io.genomio.utils package", "Ensembl-genomio", "API Setup and installation", "License", "ensembl", "Usage"], "terms": {"brc4": [0, 6, 7, 16, 18, 21], "packag": [0, 3, 21], "runnabl": [0, 1], "submodul": [0, 1, 3, 4, 14], "compare_fasta": [0, 1], "modul": [0, 1, 3, 4, 14, 18, 22], "compare_report": [0, 1], "core_serv": [0, 1], "download_genbank": [0, 1, 4, 9], "dump_stable_id": [0, 1], "extract_from_gb": [0, 1], "fill_metadata": [0, 1], "gff3_specifi": [0, 1], "integr": [0, 1, 13], "json_schema_factori": [0, 1, 14, 15], "load_sequence_data": [0, 1], "manifest": [0, 1, 3, 4, 6, 15], "manifest_stat": [0, 1, 4, 13], "prepare_genom": [0, 1], "read_json": [0, 1], "say_access": [0, 1], "seqregion_pars": [0, 1], "seqgroup": [1, 2], "add_id": [1, 2, 4, 13], "add_report_to_map": [1, 2], "assembly_level": [1, 2], "build_seq_dict": [1, 2], "compare_seq": [1, 2], "find_common_group": [1, 2], "get_fasta": [1, 2], "get_json": [1, 2, 4, 17], "get_map": [1, 2], "organellar_assembli": [1, 2], "param_default": [1, 2], "print_map": [1, 2], "run": [1, 2, 4, 13, 18, 22], "coreserv": [1, 2, 6, 7], "host": [1, 2, 6], "port": [1, 2], "user": [1, 2, 18], "password": [1, 2], "connect": [1, 2, 7], "get_all_cor": [1, 2], "get_cor": [1, 2], "get_cursor": [1, 2], "get_db_metadata": [1, 2], "get_table_data": [1, 2], "set_databas": [1, 2], "downloaderror": [1, 2, 4, 9], "dumpstableid": [1, 2, 4, 7], "server": [1, 2, 4, 6, 7], "extend_ev": [1, 2, 4, 7], "get_histori": [1, 2, 4, 7], "get_mapping_sess": [1, 2, 4, 7], "get_pair": [1, 2, 4, 7], "make_ev": [1, 2, 4, 7], "print_ev": [1, 2, 4, 7], "stableidev": [1, 2, 4, 7], "from_list": [1, 2, 4, 7], "to_list": [1, 2, 4, 7], "releas": [1, 2, 4, 7], "date": [1, 2, 4, 7, 10, 20], "name": [1, 2, 4, 5, 6, 7, 10, 12, 13, 15, 16, 20], "pair": [1, 2, 4, 7, 10], "add_from": [1, 2, 4, 7], "add_pair": [1, 2, 4, 7], "add_to": [1, 2, 4, 7], "brc_format_1": [1, 2, 4, 7], "brc_format_2": [1, 2, 4, 7], "clean_pair": [1, 2, 4, 7], "clean_set": [1, 2, 4, 7], "get_full_releas": [1, 2, 4, 7], "get_nam": [1, 2, 4, 7], "set_dat": [1, 2, 4, 7], "set_releas": [1, 2, 4, 7], "unsupportedev": [1, 2, 4, 7], "formattedfilesgener": [1, 2, 4, 9], "allowed_feat_typ": [1, 2, 4, 9], "locat": [1, 2, 4, 9, 13, 16], "parse_genbank": [1, 2, 4, 9], "set_prefix": [1, 2, 4, 9], "set_production_nam": [1, 2, 4, 9], "db_string": [1, 2], "flattern": [1, 2], "run_sql_req": [1, 2], "allowed_attribut": [1, 2], "allowed_biotyp": [1, 2], "get_gff3": [1, 2, 4, 13], "ignored_biotyp": [1, 2], "parse_gff3": [1, 2], "check_id": [1, 2, 4, 13], "check_length": [1, 2, 4, 13], "check_md5sum": [1, 2], "check_seq_region_length": [1, 2, 4, 13], "get_agp_seq_region": [1, 2, 4, 13], "get_fasta_length": [1, 2, 4, 13], "get_functional_annot": [1, 2, 4, 13], "dataflow_json": [1, 2], "add_asm_map": [1, 2], "add_contig_ena_attrib": [1, 2], "add_karyotype_band": [1, 2], "add_karyotype_data": [1, 2], "add_karyotype_rank_based_on_assembly_metadata": [1, 2], "add_karyotype_rank_for_chromosom": [1, 2], "add_karyotype_rank_from_bands_info": [1, 2], "add_sr_attrib": [1, 2], "add_sr_ebi_brc4_nam": [1, 2], "add_sr_synonym": [1, 2], "agp_prun": [1, 2], "asm_nam": [1, 2], "chunk_contig": [1, 2], "coord_sys_ord": [1, 2], "copy_sr_name_to_syn": [1, 2], "filter_already_loaded_regions_from_agp": [1, 2], "flattern_seq_region_item": [1, 2], "from_param": [1, 2], "get_external_db_map": [1, 2, 4, 16], "get_toplevel_from_c": [1, 2], "id_from_map_or_di": [1, 2], "initial_sequence_load": [1, 2], "insert_to_db": [1, 2], "is_gz": [1, 2], "load_agp": [1, 2], "load_cs_data": [1, 2], "load_map_from_core_db": [1, 2], "load_map_from_sql_stdout": [1, 2], "load_seq_data": [1, 2], "load_seq_region": [1, 2], "load_seq_region_synonyms_trios_from_core_db": [1, 2], "name_and_id_from_seq_region_item": [1, 2], "nullify_ctg_cs_vers": [1, 2], "order_agp_level": [1, 2], "param_bool": [1, 2], "pjc": [1, 2], "prune_agp": [1, 2], "quote_or_nul": [1, 2], "remove_iupac": [1, 2], "remove_components_from_toplevel": [1, 2], "set_toplevel": [1, 2], "sr_name_unvers": [1, 2], "unversion_scaffold": [1, 2], "update_db_single_group": [1, 2], "used_cs_rank": [1, 2], "copy_fil": [1, 2], "create_manifest": [1, 2, 4, 13], "get_genome_data": [1, 2], "get_md5sum": [1, 2], "biotypecollect": [1, 2], "add_featur": [1, 2, 4, 12], "get_count": [1, 2], "sorted_stat": [1, 2], "featurecount": [1, 2], "exampl": [1, 2, 13, 15, 20], "unique_count": [1, 2, 4, 13], "check_ncbi_stat": [1, 2, 4, 13], "compare_ncbi_count": [1, 2, 4, 13], "get_gff3_stat": [1, 2, 4, 13], "get_manifest": [1, 2, 4, 13], "get_seq_region_stat": [1, 2, 4, 13], "check_db_name_format": [1, 2], "get_assembly_vers": [1, 2], "get_genom": [1, 2], "get_scientific_nam": [1, 2], "make_db_nam": [1, 2], "make_production_nam": [1, 2], "update_access": [1, 2], "update_assembly_nam": [1, 2], "seqregionpars": [1, 2], "get_report_region": [1, 2, 4, 16], "make_seq_region": [1, 2, 4, 16], "molecule_loc": [1, 2, 16], "report_to_csv": [1, 2, 4, 16], "synonym_map": [1, 2, 16], "class": [2, 7, 9, 11, 12, 13, 16, 20], "sequenc": [2, 8, 9, 10, 13, 16], "identifi": [2, 12, 20], "none": [2, 5, 6, 7, 8, 9, 10, 12, 13, 15, 16, 17], "sourc": [2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 15, 16, 17, 20, 22], "base": [2, 5, 7, 8, 9, 10, 11, 12, 13, 16, 20], "object": [2, 7, 9, 11, 12, 13, 16, 17, 20], "read_fileno": 2, "write_fileno": 2, "debug": 2, "baserunn": 2, "seq_map": 2, "dict": [2, 5, 6, 7, 9, 10, 11, 12, 13, 16], "report_seq": 2, "list": [2, 5, 6, 7, 8, 10, 12, 13, 15, 16, 20], "ani": [2, 7, 10, 11, 12, 13, 16, 17, 20], "core_data": 2, "tupl": [2, 7, 10, 16], "str": [2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 15, 16], "seq": [2, 8, 13, 16], "build": [2, 7, 10, 13], "take": [2, 8, 16], "duplic": 2, "account": 2, "seq1": 2, "seq2": 2, "seqs1": 2, "seqs2": 2, "fasta_path": [2, 13], "map_dna": 2, "json_path": [2, 13], "map_path": 2, "data": [2, 5, 7, 9, 10, 12, 13, 16, 17, 18], "return": [2, 5, 6, 7, 9, 10, 11, 12, 13, 16], "default": [2, 10, 13, 17], "paramet": [2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 15, 16, 17], "thi": [2, 7, 10, 12, 13, 16, 17, 20, 22], "map_fil": [2, 7, 16], "report_fil": [2, 10, 16], "access": [2, 5, 9, 10, 13, 16], "basic": [2, 18], "interfac": [2, 6, 11, 20], "mysql": [2, 19], "core": [2, 6, 7, 9, 10, 11, 16, 18], "databas": [2, 3, 4, 7, 9, 10, 11, 16, 18], "allow": [2, 5, 13, 18], "get": [2, 5, 7, 10, 12, 13, 16], "provid": [2, 7, 10, 11, 12, 13, 15, 16, 20], "them": [2, 7, 12, 13, 16], "type": [2, 6, 7, 11, 12, 13, 15, 16, 20], "option": [2, 8, 12], "To": [2, 16, 20], "specif": [2, 13, 16, 18, 20], "1": [2, 13, 19, 20], "creat": [2, 7, 9, 12, 13, 17, 18], "2": [2, 7, 20], "set": [2, 5, 7, 8, 9, 13, 16, 18], "dbname": 2, "3": [2, 5, 6, 19], "retriev": [2, 5, 7, 10, 13, 16], "cursor": 2, "queri": 2, "all": [2, 5, 6, 7, 9, 10, 12, 13, 16, 20], "look": 2, "like": 2, "prefix": [2, 9, 12], "version": [2, 5, 7, 10, 12, 20], "dbname_r": 2, "filter": [2, 10], "request": [2, 12, 16, 19], "param": [2, 12], "_": 2, "i": [2, 5, 7, 9, 10, 11, 12, 13, 16, 17, 20], "ad": [2, 7, 10, 13, 16], "automat": [2, 7, 12], "regular": 2, "express": [2, 20], "A": [2, 7, 10, 12, 13, 16, 17, 18, 20], "metadata": [2, 6, 9, 10, 12, 13, 15, 16], "from": [2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 15, 16, 18, 19, 20], "kei": [2, 5, 6, 10, 16, 17], "meta_kei": 2, "valu": [2, 5, 6, 7, 10, 11, 13, 16], "meta_valu": 2, "tabl": [2, 10, 11, 16], "field": [2, 16, 20], "constraint": 2, "row": 2, "contain": [2, 9, 12, 13, 16, 17, 20], "each": [2, 7, 11, 13, 15, 16, 20], "db_name": 2, "except": [2, 5, 7, 8, 9, 10, 12, 13, 16, 20], "msg": [2, 9, 10], "In": [2, 9, 10, 20], "case": [2, 9, 12, 13], "download": [2, 3, 4], "fail": [2, 5, 9, 13], "static": [2, 7, 12, 13], "dl_dir": [2, 5], "given": [2, 5, 7, 8, 9, 10, 11, 12, 13, 16], "genbank": [2, 3, 4, 8, 16], "correspond": [2, 7, 9, 13, 16], "file": [2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 15, 16, 17, 18, 20], "format": [2, 3, 4, 5, 9, 12, 13, 14, 16, 18, 20], "an": [2, 5, 7, 9, 10, 12, 16, 17, 18, 19, 20], "processor": [2, 7], "event": [2, 3, 4, 13, 20], "id": [2, 7, 8, 9, 10, 12, 13, 16], "can": [2, 6, 7, 9, 13, 16, 17], "print": [2, 7, 12, 16, 17, 20], "those": [2, 7, 12, 20], "out": [2, 7, 12, 18, 20], "aggreg": [2, 7], "whole": [2, 7, 20], "group": [2, 7], "extend": [2, 3, 4, 7], "which": [2, 5, 7, 12, 16, 20], "have": [2, 5, 7, 12, 13, 15, 20], "been": [2, 7, 12, 20], "remov": [2, 7, 12, 16], "map": [2, 6, 7, 16], "session": [2, 7, 10, 11, 16], "session_id": [2, 7], "int": [2, 5, 7, 11, 12, 13, 16], "old_id": [2, 7], "new_id": [2, 7], "output_fil": [2, 7, 9, 10], "path": [2, 5, 7, 9, 10, 12, 13, 15, 16, 17, 22], "brc": [2, 7, 13], "genom": [2, 4, 5, 7, 9, 10, 11, 12, 13, 16, 18], "where": [2, 6, 7, 10, 12, 13, 15, 16, 17, 20], "datetim": [2, 7], "repres": [2, 7, 10, 11, 16, 20], "stabl": [2, 7, 12], "one": [2, 6, 7, 10, 12, 15, 16, 20], "gene": [2, 7, 9, 10, 12, 13], "anoth": [2, 7], "variou": [2, 7, 11, 13, 18], "new": [2, 7, 12], "delet": [2, 7, 12], "merg": [2, 4, 7, 12, 16], "sever": [2, 7], "split": [2, 7, 12], "mix": [2, 7], "previou": [2, 7], "updat": [2, 7, 10], "befor": [2, 7], "2019": [2, 7], "09": [2, 7], "dub": [2, 7], "pre": [2, 7], "from_id": [2, 4, 7], "store": [2, 5, 7, 9, 13, 16, 17], "from_set": [2, 7], "keep": [2, 5, 7, 13], "record": [2, 4, 7, 9, 12, 13, 16], "dictionari": [2, 6, 10, 13, 16], "rais": [2, 7, 9, 10, 12, 13, 16], "valueerror": [2, 7], "when": [2, 5, 7, 9, 10, 12, 13, 16], "empti": [2, 7, 13, 16], "either": [2, 20], "ar": [2, 5, 6, 7, 8, 10, 12, 13, 16, 20], "us": [2, 5, 7, 9, 10, 12, 13, 16, 17, 18, 20], "to_id": [2, 4, 7], "line": [2, 7, 12, 16], "per": [2, 6, 7, 15], "initi": [2, 7], "follow": [2, 7, 9, 10, 12, 16, 18, 20], "tsv": [2, 7], "old": [2, 7], "comma": [2, 7], "separ": [2, 7, 12, 20], "combin": [2, 7, 16, 20], "this_list": [2, 7], "element": [2, 6, 7, 12, 16], "item": [2, 7, 16], "so": [2, 7, 12, 13, 16], "The": [2, 5, 7, 10, 12, 13, 16, 17, 20], "clean": [2, 7, 8, 9, 12], "expand": [2, 7, 10], "beforehand": [2, 7], "output_dir": [2, 15], "parser": [2, 9, 13], "load": [2, 3, 4, 9, 12, 13, 16, 17, 22], "output": [2, 4, 8, 9, 10, 12, 13, 16], "our": [2, 9, 11], "schema": [2, 3, 4, 9, 10, 16], "input": [2, 4, 7, 8, 9], "transcript": [2, 9, 12, 13], "trna": [2, 9, 12], "rrna": [2, 9, 12], "cd": [2, 9, 12, 13], "apicoplast": [2, 9], "apicoplast_chromosom": [2, 9], "chloroplast": [2, 9], "chloroplast_chromosom": [2, 9], "chromoplast": [2, 9], "chromoplast_chromosom": [2, 9], "cyanel": [2, 9], "cyanelle_chromosom": [2, 9], "leucoplast": [2, 9], "leucoplast_chromosom": [2, 9], "mitochondrion": [2, 9], "mitochondrial_chromosom": [2, 9], "gb_file": [2, 9], "defin": [2, 5, 9, 12, 16, 20], "add": [2, 7, 9, 10, 12, 13, 16, 20], "featur": [2, 9, 11, 12, 13], "prod_nam": [2, 9], "production_nam": [2, 6, 9], "ignore_lst": 2, "pfx": 2, "sql": 2, "log_pfx": 2, "from_fil": 2, "fals": [2, 5, 6, 7, 8, 12, 13, 16], "parent": [2, 12], "derives_from": 2, "biotyp": [2, 11, 12, 13], "mrna": [2, 12], "exon": [2, 12], "three_prime_utr": 2, "five_prime_utr": 2, "pseudogenic_transcript": 2, "pseudogenic_exon": 2, "pseudogen": [2, 12], "gff3_path": [2, 13], "biological_region": 2, "supercontig": 2, "chromosom": [2, 10], "scaffold": [2, 13], "contig": [2, 13], "region": [2, 8, 13, 16], "gff3_handl": 2, "tmph": 2, "check": [2, 5, 7, 12, 13, 15, 18], "annot": [2, 10, 11, 12, 13, 20], "list1": [2, 13], "list2": [2, 13], "compar": [2, 3, 4, 13], "function": [2, 7, 12, 13], "json": [2, 4, 6, 9, 10, 11, 12, 13, 14, 16, 17, 18], "length": [2, 13], "gff": [2, 8, 12, 13, 19], "string": [2, 5, 13, 16], "error": [2, 8, 9, 12, 13, 16], "do": [2, 12, 13, 20, 22], "match": [2, 12, 13, 16], "allowed_len_diff": [2, 13], "special_diff": [2, 13], "differ": [2, 7, 12, 13, 14, 16, 20], "between": [2, 13], "There": [2, 13], "few": [2, 13], "special": [2, 12, 13, 20], "here": [2, 13], "we": [2, 5, 12, 13, 20], "certain": [2, 13], "asymmetri": [2, 13], "chang": [2, 12, 13, 16, 20], "argument": [2, 13], "fasta": [2, 3, 4, 9, 13, 18], "want": [2, 5, 12, 13], "accept": [2, 13, 20], "how": [2, 13, 18, 20], "much": [2, 13], "expect": [2, 5, 6, 7, 9, 10, 12, 13, 17], "report": [2, 5, 10, 13, 16], "common": [2, 13, 20], "bp": [2, 13], "md5sum": 2, "verifi": [2, 13], "md5": [2, 13], "hash": [2, 13], "gener": [2, 3, 4, 6, 10, 11, 12, 15, 17, 20], "doe": [2, 5, 17, 20], "seqr": [2, 13], "feat": [2, 12, 13], "circular": [2, 13], "seq_region": [2, 3, 4, 9, 10, 13, 15], "It": [2, 12, 13], "synonym": [2, 13, 16], "consist": [2, 13, 20], "agp_dict": [2, 13], "agp": [2, 13], "describ": [2, 13, 20], "assembli": [2, 3, 4, 10, 11, 13, 16], "larger": [2, 13], "smaller": [2, 13], "eg": [2, 13], "inform": [2, 6, 10, 12, 13, 16, 18, 20], "about": [2, 10, 11, 13], "onli": [2, 5, 10, 12, 13, 16, 20], "older": [2, 13], "current": [2, 5, 7, 9, 10, 12, 13, 16], "process": [2, 3, 4, 13, 16], "ignore_final_stop": [2, 13], "correct": [2, 13], "stop": [2, 13], "codon": [2, 13, 16], "fasta_dna": [2, 13], "fasta_pep": [2, 13], "non": [2, 13, 20], "uniqu": [2, 12, 13], "found": [2, 10, 12, 13, 16], "gene_id": [2, 12, 13], "translat": [2, 13, 16, 20], "descript": [2, 12, 13, 20], "functional_annot": [2, 12, 13], "gff3": [2, 3, 4, 9, 10, 13, 18], "abov": [2, 20], "metadata_typ": [2, 15], "metadata_json": 2, "atrribut": 2, "ehiv": 2, "sequnc": 2, "analysi": [2, 19], "perl": [2, 19], "script": [2, 6, 16], "coord_system": [2, 13, 16], "attribut": [2, 20], "command": 2, "through": [2, 20], "code": [2, 11, 16, 20], "replac": [2, 20], "proper": [2, 12], "python": [2, 22], "api": [2, 10, 18, 22], "some": [2, 8, 9, 10, 13], "point": [2, 6, 16], "cs_pair": 2, "meta": 2, "nullifi": 2, "asm_map": 2, "well": [2, 6, 12], "don": [2, 20], "t": [2, 7, 9, 20], "toplevel": 2, "doesn": 2, "singl": 2, "c": 2, "cs_name": 2, "ena": [2, 10], "attrib": [2, 16], "nno": 2, "sequence_level": 2, "just": [2, 12], "see": [2, 20, 22], "datacheck": 2, "lib": 2, "bio": 2, "seqregionnamesinsdc": 2, "pm": 2, "seq_region_fil": 2, "seq_region_map": 2, "attrib_type_map": 2, "work_dir": 2, "unvers": 2, "bool": [2, 5, 6, 7, 8, 12, 13, 16], "karyotype_bands_properti": 2, "karyotype_band": 2, "karyotyp": [2, 16], "metafil": 2, "seq_region_schema": [2, 16], "compat": 2, "seq_region_nam": 2, "seq_region_id": 2, "unversioned_nam": 2, "trio": 2, "band": [2, 16], "info": 2, "If": [2, 5, 6, 7, 9, 10, 12, 16, 17, 20], "true": [2, 6, 12, 16], "would": 2, "possibl": [2, 20], "too": 2, "close": 2, "db": [2, 6, 11, 16], "present": [2, 10, 16], "karyotype_rank": 2, "genome_data": [2, 10], "arrai": 2, "chromosome_display_ord": 2, "coord_system_tag": 2, "cs_tag_for_ord": 2, "underfin": 2, "forc": 2, "force_update_coord_system_tag": 2, "chromosome_coord_system_nam": 2, "coordin": [2, 11, 12, 13], "system": [2, 11, 13, 20], "Not": [2, 20], "alter": 2, "tag": 2, "regions_with_karyotype_band": 2, "other_regions_with_rank": 2, "seq_region_attrib": 2, "": [2, 10, 12, 20, 22], "explicit": 2, "taken": 2, "sr_attrib_typ": 2, "json_properti": 2, "attrib_typ": [2, 11], "its": [2, 10, 13, 16, 20], "treat": 2, "delimet": 2, "e": [2, 10, 15, 22], "added_sequ": 2, "assembly_provid": 2, "No": 2, "simpl": [2, 7, 12], "flattabl": 2, "ebi": [2, 10], "_seq_region_nam": 2, "origin": [2, 20], "seq_region_synonym": 2, "deriv": [2, 12, 20], "external_db_map": [2, 16], "unversionable_sources_set": 2, "frozenset": 2, "insdc": [2, 5, 9, 10, 12, 16], "refseq": [2, 5, 10, 16], "alreadi": [2, 10, 12], "exinst": 2, "ones": [2, 16], "to_fil": 2, "compon": [2, 6], "thei": [2, 12], "seen": 2, "cs_rank": 2, "chunk_siz": 2, "0": [2, 5, 12, 13, 20], "chunks_cs_nam": 2, "ensembl_intern": 2, "chunk": 2, "dna": [2, 8, 9, 13], "50k": 2, "cs_order_str": 2, "x_db": 2, "cood_systen": 2, "extern": 2, "src_file": [2, 17], "dst_file": 2, "loaded_region": 2, "new_region": 2, "prop_nam": 2, "path_attrib_id_map": 2, "sep": 2, "properti": 2, "attrib_id": 2, "work": [2, 9, 20], "implemet": 2, "tree": 2, "travers": 2, "util": [2, 3, 4], "method": [2, 12, 16], "not_throw": 2, "external_db": [2, 16], "coord_system_nam": 2, "coord": 2, "map_dict": [2, 7], "name_for_pan": 2, "prepar": [2, 3, 4, 5, 13, 20], "list_of_tupl": 2, "table_nam": 2, "col_nam": 2, "ignor": 2, "insert": [2, 7], "definit": [2, 20], "below": [2, 16, 20], "put": [2, 12], "filenam": 2, "asm_v": 2, "pl": 2, "rank": 2, "seq_level": 2, "seqregion": [2, 16], "col": 2, "in_fil": 2, "skip_head": 2, "skip": 2, "header": 2, "accordingli": 2, "additional_load": 2, "seq_region_item": 2, "try_unvers": 2, "throw_miss": 2, "struct": 2, "throw": 2, "abl": 2, "cs_order": 2, "everi": 2, "than": 2, "part": [2, 16, 20], "join": 2, "try": [2, 12], "directori": [2, 5, 10, 13], "last": 2, "agps_pruned_dir": 2, "prune": 2, "val": 2, "quot": 2, "null": 2, "strings_onli": 2, "wrap": 2, "instanc": 2, "valid": [2, 4, 12, 14], "symbol": 2, "sed": 2, "result": [2, 20], "entri": [2, 6, 16], "done": 2, "ignored_c": 2, "6": [2, 19], "tbl": 2, "fld": 2, "suffix": 2, "d": 2, "suffic": 2, "log": [2, 13], "need": [2, 12], "syn": 2, "lower": 2, "level": [2, 16], "rest": 2, "dict_of_col_to_valu": 2, "dict_of_cols_to_valu": 2, "condit": [2, 20], "claus": 2, "noagp_default": 2, "origin_path": 2, "dir_path": 2, "speci": [2, 6, 10, 16], "genome_nam": 2, "genome_dir": 2, "file_path": [2, 5, 17], "feature_id": [2, 13], "feature_typ": 2, "dataset": [2, 11, 13], "tool": [2, 13], "ncbi": [2, 9, 11, 13], "stat": [2, 11, 13], "what": [2, 12, 13], "manifest_path": [2, 13], "seq_region_path": [2, 13], "textio": [2, 17], "taxon_id": [2, 10, 16], "read": [2, 13], "flow": 2, "pipelin": [2, 22], "dataflow": 2, "associ": [2, 10, 13, 20], "bring": 2, "genome_schema": [2, 10], "main": [2, 4, 6, 7, 13, 16], "report_path": [2, 10, 16], "use_refseq": 2, "etc": [2, 11, 12], "nuclear_chromosom": 2, "kinetoplast": 2, "linkag": 2, "linkage_group": 2, "plasmid": 2, "csv": [2, 16], "head": [2, 16], "assign": [2, 6], "molecul": 2, "accn": 2, "insdc_submitted_nam": 2, "genomio": [3, 19, 22], "factori": [3, 4, 14], "dump": [3, 4, 18], "extract_data": [3, 4], "genome_metadata": [3, 4], "genome_stat": [3, 4], "extract_annot": [3, 4], "check_integr": [3, 4], "compute_stat": [3, 4], "archive_util": [3, 4], "json_util": [3, 4], "handl": [4, 7, 10, 11, 13, 14, 15, 16, 18], "librari": 4, "filedownloaderror": [4, 5], "unsupportedformaterror": [4, 5], "download_fil": [4, 5], "get_checksum": [4, 5], "get_files_select": [4, 5], "get_root_nam": [4, 5], "md5_file": [4, 5], "retrieve_assembly_data": [4, 5], "format_db_data": [4, 6], "get_metadata_valu": [4, 6], "get_pairs_from_to": [4, 7], "has_new_id": [4, 7], "has_old_id": [4, 7], "is_empti": [4, 7], "idsmapp": [4, 7], "load_list": [4, 7], "eventcollect": [4, 7], "add_delet": [4, 7], "load_ev": [4, 7], "load_events_from_gene_diff": [4, 7], "remap_to_id": [4, 7], "write_events_to_db": [4, 7], "write_events_to_fil": [4, 7], "idev": [4, 7], "is_chang": [4, 7], "release_d": [4, 7], "mapsess": [4, 7], "add_ev": [4, 7], "gffparsererror": [4, 8, 12], "get_peptides_to_exclud": [4, 8], "prep_fasta_data": [4, 8], "extract_gb": [4, 9], "gbparseerror": [4, 9], "genomefil": [4, 9], "unsupporteddata": [4, 9], "check_assembly_vers": [4, 10], "filter_genome_meta": [4, 10], "get_genome_metadata": [4, 10], "missingdataerror": [4, 10], "amend_genomic_metadata": [4, 10], "get_addit": [4, 10], "get_gbff_region": [4, 10], "get_report_regions_nam": [4, 10], "metadataerror": [4, 10], "missingnodeerror": [4, 10], "add_assembly_vers": [4, 10], "add_genebuild_metadata": [4, 10], "add_provid": [4, 10], "add_species_metadata": [4, 10], "get_taxonomy_from_access": [4, 10], "prepare_genome_metadata": [4, 10], "compare_annot": [4, 11], "compare_assembli": [4, 11], "compare_stat": [4, 11], "statsgener": [4, 11], "get_annotation_stat": [4, 11], "get_assembly_stat": [4, 11], "get_attrib_count": [4, 11], "get_biotyp": [4, 11], "get_feature_stat": [4, 11], "get_stat": [4, 11], "annotationerror": [4, 12], "duplicateiderror": [4, 12], "functionalannot": [4, 12], "add_par": [4, 12], "get_featur": [4, 12], "get_par": [4, 12], "product_is_inform": [4, 12], "to_json": [4, 12], "missingparenterror": [4, 12], "gffgenemerg": [4, 12], "gffsimplifi": [4, 12], "allow_pseudogene_with_cd": [4, 12], "build_transcript": [4, 12], "cds_gene": [4, 12], "clean_extra_exon": [4, 12], "current_stable_id_numb": [4, 12], "exclude_seq_region": [4, 12, 16], "format_gene_seg": [4, 12], "format_mobile_el": [4, 12], "gene_cds_skip_oth": [4, 12], "gene_to_cd": [4, 12], "gene_to_exon": [4, 12], "generate_stable_id": [4, 12], "min_id_length": [4, 12], "move_cds_to_mrna": [4, 12], "normalize_cds_id": [4, 12], "normalize_gen": [4, 12], "normalize_gene_id": [4, 12], "normalize_pseudogene_cd": [4, 12], "normalize_transcript_id": [4, 12], "remove_cds_from_pseudogen": [4, 12], "remove_prefix": [4, 12], "simpler_gff3": [4, 12], "skip_unrecogn": [4, 12], "stable_id_prefix": [4, 12], "transcript_for_gen": [4, 12], "transcript_gen": [4, 12], "valid_id": [4, 12], "validate_gene_id": [4, 12], "to_gff": [4, 12], "integritytool": [4, 13], "add_error": [4, 13], "set_brc_mod": [4, 13], "set_ignore_final_stop": [4, 13], "invalidintegrityerror": [4, 13], "get_circular": [4, 13], "get_length": [4, 13], "has_length": [4, 13], "prepare_integrity_data": [4, 13], "biotypecount": [4, 13], "statserror": [4, 13], "biotypes_stat": [4, 13], "coord_systems_stat": [4, 13], "count_biotyp": [4, 13], "increment_biotyp": [4, 13], "seq_region_special_stat": [4, 13], "manifestmak": [4, 13], "alias_nam": [4, 13], "get_files_checksum": [4, 13], "same_nam": [4, 13], "mapformaterror": [4, 16], "add_attrib": [4, 16], "get_attrib": [4, 16], "get_coord_system": [4, 16], "get_karyotyp": [4, 16], "get_seq_region": [4, 16], "get_synonym": [4, 16], "unknownmetadata": [4, 16], "add_insdc_seq_region_nam": [4, 16], "add_mitochondrial_codon_t": [4, 16], "add_translation_t": [4, 16], "get_codon_t": [4, 16], "get_gbff_seq_region": [4, 16], "get_genbank_id": [4, 16], "get_organel": [4, 16], "merge_seq_region": [4, 16], "prepare_seq_region_metadata": [4, 16], "extract_fil": [4, 17], "open_gz_fil": [4, 17], "print_json": [4, 17], "problem": 5, "max_redo": 5, "avail": [5, 20], "ftp": 5, "dir": 5, "checksum_path": 5, "checksum": [5, 13], "sum": [5, 13], "among": 5, "subset": 5, "control": [5, 16, 20], "rel": 5, "_file_end": 5, "root": 5, "miss": [5, 10, 16], "ha": [5, 7, 13, 20], "wrong": [5, 12], "download_dir": 5, "pathlik": [5, 7, 8, 9, 10, 12, 13, 15, 16, 17], "max_incr": 5, "todo": 5, "you": [5, 20], "max": 5, "number": [5, 10, 11, 12, 13, 16], "time": 5, "retri": 5, "insid": [6, 15], "includ": [6, 10, 11, 15, 16, 19, 20], "itself": [6, 15], "import": [6, 20], "call": 6, "same": [6, 12, 13, 16, 20], "outcom": 6, "brc_mode": [6, 13, 16], "organism_abbrev": [6, 12], "divis": 6, "otherwis": [6, 12, 20], "first": 6, "search": [6, 18], "extract": [7, 8, 9, 11, 12, 13, 15, 16, 17], "represent": [7, 9, 12, 13], "test": 7, "support": [7, 9, 12, 16, 20], "entrypoint": [7, 13], "mapper": 7, "cleanli": 7, "list_fil": 7, "cf": 7, "tab": 7, "collect": 7, "loader": [7, 17, 18], "writer": 7, "release_nam": 7, "input_fil": [7, 10], "column": [7, 16], "event_nam": 7, "gene_diff": 7, "remap": 7, "write": [7, 12, 20], "exist": [7, 12, 17], "mapping_sess": 7, "peptid": [8, 9, 13], "up": [8, 9, 12, 16], "exclud": [8, 16, 20], "while": [8, 20], "pars": [8, 9, 12, 16], "genbank_path": 8, "seqr_to_exclud": 8, "fasta_infil": 8, "genbank_infil": [8, 10], "fasta_outfil": 8, "peptide_mod": 8, "fasta_fil": 8, "protein": 8, "gbff": [8, 10, 16], "instead": 8, "fetch": [9, 10, 16], "manipul": 9, "entrez": 9, "servic": [9, 20], "model": [9, 10, 12, 13], "gffparseerror": 9, "structur": [9, 12, 16], "gb": 9, "rtype": 9, "json_output": 9, "out_dir": 9, "posixpath": 9, "home": 9, "runner": 9, "doc": 9, "gmeta_out": 10, "integ": 10, "altern": 10, "gmeta": 10, "nest": 10, "predefin": 10, "also": [10, 12, 18, 20], "convert": 10, "select": 10, "more": [10, 20], "g": [10, 15], "mt": 10, "genome_infil": 10, "genome_outfil": 10, "amend": 10, "gbff_path": [10, 16], "mention": 10, "detail": [10, 22], "taxonomi": [10, 19], "taxon": [10, 16], "xml": 10, "node": 10, "cannot": [10, 12, 20], "genebuild": 10, "convent": 10, "start_dat": 10, "gff3_file": 10, "neither": 10, "provider_nam": 10, "provider_url": 10, "base_api_url": 10, "http": [10, 19, 20], "www": [10, 20], "ac": 10, "uk": 10, "browser": 10, "taxonomy_id": 10, "strain": 10, "scientific_nam": 10, "link": [10, 20], "url": 10, "missindataexcept": 10, "addit": [10, 20], "particular": [10, 20], "final": [10, 16], "statist": [11, 13], "section": [11, 18, 20, 22], "comparison": 11, "v": 11, "count": [11, 13], "diff": 11, "compil": [11, 20], "anyth": 12, "happen": 12, "seqfeatur": 12, "feat_typ": 12, "parent_id": 12, "relatioship": 12, "parent_typ": 12, "child_id": 12, "child": 12, "relationship": 12, "product": [12, 19, 20], "feat_id": 12, "word": 12, "consid": 12, "uninform": 12, "hypothet": 12, "out_path": 12, "without": [12, 20], "standard": 12, "gffparsercommon": 12, "prior": 12, "further": [12, 18], "in_gff_path": 12, "out_gff_path": 12, "multipl": 12, "genome_path": 12, "make_missing_stable_id": 12, "ggf3": 12, "fix": 12, "lone": 12, "extra": 12, "usual": [12, 17], "start": [12, 16], "detect": 12, "equival": 12, "segment": 12, "c_gene_seg": 12, "v_gene_seg": 12, "mobile_genetic_el": 12, "transform": [12, 20], "transposable_el": 12, "children": 12, "intermediari": 12, "direct": [12, 20], "stable_id": 12, "made": [12, 20], "auto": 12, "increment": 12, "prepend": 12, "tmp_": 12, "8": [12, 19], "move": 12, "cdss": 12, "might": 12, "recogn": [12, 16], "cds_id": 12, "unnecessari": 12, "around": 12, "fail_typ": 12, "normal": [12, 20], "append": [12, 13], "ensur": [12, 13], "make": [12, 20], "assum": [12, 20], "sub": 12, "unalt": 12, "rewrit": 12, "simpler": 12, "ncrna": 12, "other": [12, 20], "iter": 12, "seqrecord": [12, 16], "content": [13, 20], "sync": 13, "manifest_fil": 13, "arg": 13, "show": 13, "whether": [13, 20], "mode": 13, "calcul": 13, "is_circular": 13, "flag": 13, "later": 13, "comput": [13, 20], "counter": 13, "total": 13, "far": 13, "could": 13, "manifest_dir": [13, 15], "datasets_bin": 13, "order": [13, 16, 22], "printabl": 13, "For": [13, 20, 22], "minimum": 13, "maximum": 13, "mean": [13, 20], "feature_biotyp": 13, "respect": [13, 19], "stats_path": 13, "txt": [13, 22], "folder": [13, 15, 16], "codon_t": 13, "f": 13, "9f": 13, "depend": 13, "end": [13, 16, 20], "gene_model": 13, "pep": 13, "seq_attrib": 13, "verif": [14, 15], "json_schema_valid": [14, 15], "json_fil": 15, "json_schema": 15, "against": [15, 20], "check_json_schema": 15, "attrib_dict": 16, "modifi": [16, 20], "coordsystem": 16, "stain": 16, "top": 16, "export": [16, 19], "construct": 16, "synonym_sourc": 16, "brc4_seq_region_nam": 16, "ebi_seq_region_nam": 16, "former": 16, "whilst": 16, "latter": 16, "prefer": [16, 20], "mitochondri": 16, "genet": 16, "noth": [16, 20], "location_codon": 16, "known": 16, "to_exclud": 16, "exclus": [16, 20], "comment": [16, 20], "organel": 16, "keyerror": 16, "vocabulari": 16, "is_refseq": 16, "recognis": 16, "role": 16, "left": 16, "right": [16, 20], "two": 16, "preced": 16, "over": 16, "sort": [16, 17], "genome_fil": 16, "dst_dir": [16, 17], "gbff_file": 16, "both": 16, "deal": 17, "archiv": [17, 20], "copi": [17, 20], "unpack": 17, "yield": 17, "open": 17, "even": [17, 20], "compress": 17, "gzip": 17, "text": [17, 20], "src_path": 17, "kwarg": 17, "dst_path": 17, "dumper": [17, 18], "pretti": 17, "indent": 17, "4": 17, "repositori": [18, 22], "dedic": 18, "turn": 18, "usag": 18, "requir": [18, 20, 22], "flat": 18, "instal": [18, 22], "project": 18, "setup": 18, "licens": 18, "page": [18, 20], "ensembl": [19, 22], "checkout": 19, "src": 19, "perl5lib": 19, "pythonpath": 19, "hive": 19, "dev": 19, "hive_mast": 19, "branch": 19, "orm": 19, "bioperl": 19, "9": [19, 20], "bcbio": 19, "biopython": 19, "jsonschema": 19, "connector": 19, "redmin": 19, "directli": 19, "github": 19, "git": 19, "clone": 19, "com": 19, "b": 19, "apach": 20, "januari": 20, "2004": 20, "org": 20, "shall": 20, "term": 20, "reproduct": 20, "distribut": 20, "document": 20, "licensor": 20, "copyright": 20, "owner": 20, "entiti": 20, "author": 20, "grant": 20, "legal": 20, "union": 20, "act": 20, "under": 20, "purpos": 20, "power": 20, "indirect": 20, "caus": 20, "manag": 20, "contract": 20, "ii": 20, "ownership": 20, "fifti": 20, "percent": 20, "50": 20, "outstand": 20, "share": 20, "iii": 20, "benefici": 20, "your": 20, "individu": 20, "exercis": 20, "permiss": 20, "form": 20, "modif": 20, "limit": 20, "softwar": 20, "configur": 20, "mechan": 20, "convers": 20, "media": 20, "authorship": 20, "indic": 20, "notic": 20, "attach": 20, "appendix": 20, "editori": 20, "revis": 20, "elabor": 20, "remain": 20, "mere": 20, "bind": 20, "thereof": 20, "contribut": 20, "intention": 20, "submit": 20, "inclus": 20, "behalf": 20, "electron": 20, "verbal": 20, "written": 20, "commun": 20, "sent": 20, "mail": 20, "issu": 20, "track": 20, "discuss": 20, "improv": 20, "conspicu": 20, "mark": 20, "design": 20, "contributor": 20, "whom": 20, "receiv": 20, "subsequ": 20, "incorpor": 20, "within": 20, "subject": 20, "herebi": 20, "perpetu": 20, "worldwid": 20, "charg": 20, "royalti": 20, "free": 20, "irrevoc": 20, "reproduc": 20, "publicli": 20, "displai": 20, "perform": 20, "sublicens": 20, "patent": 20, "state": 20, "offer": 20, "sell": 20, "transfer": 20, "appli": 20, "claim": 20, "necessarili": 20, "infring": 20, "alon": 20, "wa": 20, "institut": 20, "litig": 20, "cross": 20, "counterclaim": 20, "lawsuit": 20, "alleg": 20, "constitut": 20, "contributori": 20, "termin": 20, "redistribut": 20, "mai": 20, "medium": 20, "meet": 20, "must": 20, "give": 20, "recipi": 20, "carri": 20, "promin": 20, "retain": 20, "trademark": 20, "pertain": 20, "readabl": 20, "least": 20, "place": 20, "along": 20, "wherev": 20, "third": 20, "parti": 20, "appear": 20, "own": 20, "alongsid": 20, "addendum": 20, "constru": 20, "statement": 20, "compli": 20, "submiss": 20, "unless": 20, "explicitli": 20, "notwithstand": 20, "herein": 20, "supersed": 20, "agreement": 20, "execut": 20, "regard": 20, "trade": 20, "reason": 20, "customari": 20, "disclaim": 20, "warranti": 20, "applic": 20, "law": 20, "agre": 20, "AS": 20, "basi": 20, "OR": 20, "OF": 20, "kind": 20, "impli": 20, "titl": 20, "merchant": 20, "fit": 20, "FOR": 20, "sole": 20, "respons": 20, "determin": 20, "appropri": 20, "risk": 20, "liabil": 20, "theori": 20, "tort": 20, "neglig": 20, "deliber": 20, "grossli": 20, "liabl": 20, "damag": 20, "incident": 20, "consequenti": 20, "charact": 20, "aris": 20, "inabl": 20, "loss": 20, "goodwil": 20, "stoppag": 20, "failur": 20, "malfunct": 20, "commerci": 20, "advis": 20, "choos": 20, "fee": 20, "indemn": 20, "oblig": 20, "howev": 20, "indemnifi": 20, "defend": 20, "hold": 20, "harmless": 20, "incur": 20, "assert": 20, "AND": 20, "boilerpl": 20, "enclos": 20, "bracket": 20, "should": 20, "syntax": 20, "recommend": 20, "easier": 20, "identif": 20, "yyyi": 20, "complianc": 20, "obtain": 20, "languag": 20, "govern": 20, "namespac": 21, "subpackag": 21, "full": 22, "python3": 22, "7": 22, "m": 22, "venv": 22, "virtual_env": 22, "bin": 22, "activ": 22, "pip": 22, "r": 22, "forget": 22}, "objects": {"": [[0, 0, 0, "-", "ensembl"]], "ensembl": [[1, 0, 0, "-", "brc4"], [3, 0, 0, "-", "io"]], "ensembl.brc4": [[2, 0, 0, "-", "runnable"]], "ensembl.brc4.runnable": [[2, 0, 0, "-", "compare_fasta"], [2, 0, 0, "-", "compare_report"], [2, 0, 0, "-", "core_server"], [2, 0, 0, "-", "download_genbank"], [2, 0, 0, "-", "dump_stable_ids"], [2, 0, 0, "-", "extract_from_gb"], [2, 0, 0, "-", "fill_metadata"], [2, 0, 0, "-", "gff3_specifier"], [2, 0, 0, "-", "integrity"], [2, 0, 0, "-", "json_schema_factory"], [2, 0, 0, "-", "load_sequence_data"], [2, 0, 0, "-", "manifest"], [2, 0, 0, "-", "manifest_stats"], [2, 0, 0, "-", "prepare_genome"], [2, 0, 0, "-", "read_json"], [2, 0, 0, "-", "say_accession"], [2, 0, 0, "-", "seqregion_parser"]], "ensembl.brc4.runnable.compare_fasta": [[2, 1, 1, "", "SeqGroup"], [2, 1, 1, "", "compare_fasta"]], "ensembl.brc4.runnable.compare_fasta.SeqGroup": [[2, 2, 1, "", "add_id"]], "ensembl.brc4.runnable.compare_fasta.compare_fasta": [[2, 2, 1, "", "add_report_to_map"], [2, 2, 1, "", "assembly_level"], [2, 2, 1, "", "build_seq_dict"], [2, 2, 1, "", "compare_seqs"], [2, 2, 1, "", "find_common_groups"], [2, 2, 1, "", "get_fasta"], [2, 2, 1, "", "get_json"], [2, 2, 1, "", "get_map"], [2, 2, 1, "", "organellar_assembly"], [2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "print_map"], [2, 2, 1, "", "run"]], "ensembl.brc4.runnable.compare_report": [[2, 1, 1, "", "compare_report"]], "ensembl.brc4.runnable.compare_report.compare_report": [[2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "run"]], "ensembl.brc4.runnable.core_server": [[2, 1, 1, "", "CoreServer"]], "ensembl.brc4.runnable.core_server.CoreServer": [[2, 2, 1, "", "connect"], [2, 2, 1, "", "get_all_cores"], [2, 2, 1, "", "get_cores"], [2, 2, 1, "", "get_cursor"], [2, 2, 1, "", "get_db_metadata"], [2, 2, 1, "", "get_table_data"], [2, 3, 1, "", "host"], [2, 3, 1, "", "password"], [2, 3, 1, "", "port"], [2, 2, 1, "", "set_database"], [2, 3, 1, "", "user"]], "ensembl.brc4.runnable.download_genbank": [[2, 4, 1, "", "DownloadError"], [2, 1, 1, "", "download_genbank"]], "ensembl.brc4.runnable.download_genbank.download_genbank": [[2, 2, 1, "", "download_genbank"], [2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "run"]], "ensembl.brc4.runnable.dump_stable_ids": [[2, 1, 1, "", "DumpStableIDs"], [2, 1, 1, "", "StableIdEvent"], [2, 4, 1, "", "UnsupportedEvent"]], "ensembl.brc4.runnable.dump_stable_ids.DumpStableIDs": [[2, 2, 1, "", "extend_event"], [2, 2, 1, "", "get_history"], [2, 2, 1, "", "get_mapping_sessions"], [2, 2, 1, "", "get_pairs"], [2, 2, 1, "", "make_events"], [2, 2, 1, "", "print_events"], [2, 3, 1, "", "server"]], "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent": [[2, 2, 1, "", "add_from"], [2, 2, 1, "", "add_pair"], [2, 2, 1, "", "add_pairs"], [2, 2, 1, "", "add_to"], [2, 2, 1, "", "brc_format_1"], [2, 2, 1, "", "brc_format_2"], [2, 2, 1, "", "clean_pairs"], [2, 2, 1, "", "clean_set"], [2, 3, 1, "", "date"], [2, 3, 1, "", "from_list"], [2, 2, 1, "", "get_full_release"], [2, 2, 1, "", "get_name"], [2, 3, 1, "", "name"], [2, 3, 1, "", "pairs"], [2, 3, 1, "", "release"], [2, 2, 1, "", "set_date"], [2, 2, 1, "", "set_release"], [2, 3, 1, "", "to_list"]], "ensembl.brc4.runnable.extract_from_gb": [[2, 1, 1, "", "FormattedFilesGenerator"], [2, 1, 1, "", "extract_from_gb"]], "ensembl.brc4.runnable.extract_from_gb.FormattedFilesGenerator": [[2, 3, 1, "", "allowed_feat_types"], [2, 3, 1, "", "locations"], [2, 2, 1, "", "parse_genbank"], [2, 2, 1, "", "set_prefix"], [2, 2, 1, "", "set_production_name"]], "ensembl.brc4.runnable.extract_from_gb.extract_from_gb": [[2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "run"]], "ensembl.brc4.runnable.fill_metadata": [[2, 1, 1, "", "fill_metadata"]], "ensembl.brc4.runnable.fill_metadata.fill_metadata": [[2, 2, 1, "", "db_string"], [2, 2, 1, "", "flattern"], [2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "run"], [2, 2, 1, "", "run_sql_req"]], "ensembl.brc4.runnable.gff3_specifier": [[2, 1, 1, "", "gff3_specifier"]], "ensembl.brc4.runnable.gff3_specifier.gff3_specifier": [[2, 3, 1, "", "allowed_attributes"], [2, 3, 1, "", "allowed_biotypes"], [2, 2, 1, "", "get_gff3"], [2, 3, 1, "", "ignored_biotypes"], [2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "parse_gff3"], [2, 2, 1, "", "run"]], "ensembl.brc4.runnable.integrity": [[2, 1, 1, "", "integrity"]], "ensembl.brc4.runnable.integrity.integrity": [[2, 2, 1, "", "check_ids"], [2, 2, 1, "", "check_lengths"], [2, 2, 1, "", "check_md5sum"], [2, 2, 1, "", "check_seq_region_lengths"], [2, 2, 1, "", "get_agp_seq_regions"], [2, 2, 1, "", "get_fasta_lengths"], [2, 2, 1, "", "get_functional_annotation"], [2, 2, 1, "", "get_gff3"], [2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "parse_gff3"], [2, 2, 1, "", "run"]], "ensembl.brc4.runnable.json_schema_factory": [[2, 1, 1, "", "json_schema_factory"]], "ensembl.brc4.runnable.json_schema_factory.json_schema_factory": [[2, 2, 1, "", "dataflow_json"], [2, 2, 1, "", "run"]], "ensembl.brc4.runnable.load_sequence_data": [[2, 1, 1, "", "load_sequence_data"]], "ensembl.brc4.runnable.load_sequence_data.load_sequence_data": [[2, 2, 1, "", "add_asm_mappings"], [2, 2, 1, "", "add_contig_ena_attrib"], [2, 2, 1, "", "add_karyotype_bands"], [2, 2, 1, "", "add_karyotype_data"], [2, 2, 1, "", "add_karyotype_rank_based_on_assembly_metadata"], [2, 2, 1, "", "add_karyotype_rank_for_chromosomes"], [2, 2, 1, "", "add_karyotype_rank_from_bands_info"], [2, 2, 1, "", "add_sr_attribs"], [2, 2, 1, "", "add_sr_ebi_brc4_names"], [2, 2, 1, "", "add_sr_synonyms"], [2, 2, 1, "", "agp_prune"], [2, 2, 1, "", "asm_name"], [2, 2, 1, "", "chunk_contigs"], [2, 2, 1, "", "coord_sys_order"], [2, 2, 1, "", "copy_sr_name_to_syn"], [2, 2, 1, "", "db_string"], [2, 2, 1, "", "filter_already_loaded_regions_from_agp"], [2, 2, 1, "", "flattern_seq_region_item"], [2, 2, 1, "", "from_param"], [2, 2, 1, "", "get_external_db_mapping"], [2, 2, 1, "", "get_toplevel_from_cs"], [2, 2, 1, "", "id_from_map_or_die"], [2, 2, 1, "", "initial_sequence_loading"], [2, 2, 1, "", "insert_to_db"], [2, 2, 1, "", "is_gz"], [2, 2, 1, "", "load_agp"], [2, 2, 1, "", "load_cs_data"], [2, 2, 1, "", "load_map_from_core_db"], [2, 2, 1, "", "load_map_from_sql_stdout"], [2, 2, 1, "", "load_seq_data"], [2, 2, 1, "", "load_seq_region"], [2, 2, 1, "", "load_seq_region_synonyms_trios_from_core_db"], [2, 2, 1, "", "name_and_id_from_seq_region_item"], [2, 2, 1, "", "nullify_ctg_cs_version"], [2, 2, 1, "", "order_agp_levels"], [2, 2, 1, "", "param_bool"], [2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "pjc"], [2, 2, 1, "", "prune_agps"], [2, 2, 1, "", "quote_or_null"], [2, 2, 1, "", "remove_IUPAC"], [2, 2, 1, "", "remove_components_from_toplevel"], [2, 2, 1, "", "run"], [2, 2, 1, "", "run_sql_req"], [2, 2, 1, "", "set_toplevel"], [2, 2, 1, "", "sr_name_unversion"], [2, 2, 1, "", "unversion_scaffolds"], [2, 2, 1, "", "update_db_single_group"], [2, 2, 1, "", "used_cs_ranks"]], "ensembl.brc4.runnable.manifest": [[2, 1, 1, "", "manifest"]], "ensembl.brc4.runnable.manifest.manifest": [[2, 2, 1, "", "copy_file"], [2, 2, 1, "", "copy_files"], [2, 2, 1, "", "create_manifest"], [2, 2, 1, "", "get_genome_data"], [2, 2, 1, "", "get_md5sum"], [2, 2, 1, "", "run"]], "ensembl.brc4.runnable.manifest_stats": [[2, 1, 1, "", "BiotypeCollection"], [2, 1, 1, "", "FeatureCounter"], [2, 1, 1, "", "manifest_stats"]], "ensembl.brc4.runnable.manifest_stats.BiotypeCollection": [[2, 2, 1, "", "add_feature"], [2, 2, 1, "", "get_count"], [2, 2, 1, "", "sorted_stats"]], "ensembl.brc4.runnable.manifest_stats.FeatureCounter": [[2, 2, 1, "", "add_feature"], [2, 2, 1, "", "example"], [2, 2, 1, "", "unique_count"]], "ensembl.brc4.runnable.manifest_stats.manifest_stats": [[2, 2, 1, "", "check_ncbi_stats"], [2, 2, 1, "", "compare_ncbi_counts"], [2, 2, 1, "", "get_gff3_stats"], [2, 2, 1, "", "get_manifest"], [2, 2, 1, "", "get_seq_region_stats"], [2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "parse_gff3"], [2, 2, 1, "", "run"]], "ensembl.brc4.runnable.prepare_genome": [[2, 1, 1, "", "prepare_genome"]], "ensembl.brc4.runnable.prepare_genome.prepare_genome": [[2, 2, 1, "", "check_db_name_format"], [2, 2, 1, "", "get_assembly_version"], [2, 2, 1, "", "get_genome"], [2, 2, 1, "", "get_manifest"], [2, 2, 1, "", "get_scientific_name"], [2, 2, 1, "", "make_db_name"], [2, 2, 1, "", "make_production_name"], [2, 2, 1, "", "param_defaults"], [2, 2, 1, "", "run"], [2, 2, 1, "", "update_accession"], [2, 2, 1, "", "update_assembly_name"]], "ensembl.brc4.runnable.read_json": [[2, 1, 1, "", "read_json"]], "ensembl.brc4.runnable.read_json.read_json": [[2, 2, 1, "", "run"]], "ensembl.brc4.runnable.say_accession": [[2, 1, 1, "", "say_accession"]], "ensembl.brc4.runnable.say_accession.say_accession": [[2, 2, 1, "", "run"]], "ensembl.brc4.runnable.seqregion_parser": [[2, 1, 1, "", "SeqregionParser"]], "ensembl.brc4.runnable.seqregion_parser.SeqregionParser": [[2, 2, 1, "", "get_report_regions"], [2, 2, 1, "", "make_seq_region"], [2, 3, 1, "", "molecule_location"], [2, 2, 1, "", "report_to_csv"], [2, 3, 1, "", "synonym_map"]], "ensembl.io": [[4, 0, 0, "-", "genomio"]], "ensembl.io.genomio": [[5, 0, 0, "-", "assembly"], [6, 0, 0, "-", "database"], [7, 0, 0, "-", "events"], [8, 0, 0, "-", "fasta"], [9, 0, 0, "-", "genbank"], [10, 0, 0, "-", "genome_metadata"], [11, 0, 0, "-", "genome_stats"], [12, 0, 0, "-", "gff3"], [13, 0, 0, "-", "manifest"], [14, 0, 0, "-", "schemas"], [16, 0, 0, "-", "seq_region"], [17, 0, 0, "-", "utils"]], "ensembl.io.genomio.assembly": [[5, 0, 0, "-", "download"]], "ensembl.io.genomio.assembly.download": [[5, 4, 1, "", "FileDownloadError"], [5, 4, 1, "", "UnsupportedFormatError"], [5, 5, 1, "", "download_files"], [5, 5, 1, "", "get_checksums"], [5, 5, 1, "", "get_files_selection"], [5, 5, 1, "", "get_root_name"], [5, 5, 1, "", "md5_files"], [5, 5, 1, "", "retrieve_assembly_data"]], "ensembl.io.genomio.database": [[6, 0, 0, "-", "factory"]], "ensembl.io.genomio.database.factory": [[6, 5, 1, "", "format_db_data"], [6, 5, 1, "", "get_metadata_value"], [6, 5, 1, "", "main"]], "ensembl.io.genomio.events": [[7, 0, 0, "-", "dump"], [7, 0, 0, "-", "format"], [7, 0, 0, "-", "load"]], "ensembl.io.genomio.events.dump": [[7, 1, 1, "", "DumpStableIDs"], [7, 1, 1, "", "Pair"], [7, 1, 1, "", "StableIdEvent"], [7, 4, 1, "", "UnsupportedEvent"], [7, 5, 1, "", "main"]], "ensembl.io.genomio.events.dump.DumpStableIDs": [[7, 2, 1, "", "extend_event"], [7, 2, 1, "", "get_history"], [7, 2, 1, "", "get_mapping_sessions"], [7, 2, 1, "", "get_pairs"], [7, 2, 1, "", "get_pairs_from_to"], [7, 2, 1, "", "make_events"], [7, 2, 1, "", "print_events"], [7, 3, 1, "", "server"]], "ensembl.io.genomio.events.dump.Pair": [[7, 2, 1, "", "has_new_id"], [7, 2, 1, "", "has_old_id"], [7, 2, 1, "", "is_empty"]], "ensembl.io.genomio.events.dump.StableIdEvent": [[7, 2, 1, "", "add_from"], [7, 2, 1, "", "add_pair"], [7, 2, 1, "", "add_pairs"], [7, 2, 1, "", "add_to"], [7, 2, 1, "", "brc_format_1"], [7, 2, 1, "", "brc_format_2"], [7, 2, 1, "", "clean_pairs"], [7, 2, 1, "", "clean_set"], [7, 3, 1, "", "date"], [7, 3, 1, "", "from_list"], [7, 2, 1, "", "get_full_release"], [7, 2, 1, "", "get_name"], [7, 3, 1, "", "name"], [7, 3, 1, "", "pairs"], [7, 3, 1, "", "release"], [7, 2, 1, "", "set_date"], [7, 2, 1, "", "set_release"], [7, 3, 1, "", "to_list"]], "ensembl.io.genomio.events.format": [[7, 1, 1, "", "IdsMapper"], [7, 5, 1, "", "load_list"]], "ensembl.io.genomio.events.load": [[7, 1, 1, "", "EventCollection"], [7, 1, 1, "", "IdEvent"], [7, 1, 1, "", "MapSession"], [7, 5, 1, "", "main"]], "ensembl.io.genomio.events.load.EventCollection": [[7, 2, 1, "", "add_deletes"], [7, 2, 1, "", "load_events"], [7, 2, 1, "", "load_events_from_gene_diff"], [7, 2, 1, "", "remap_to_ids"], [7, 2, 1, "", "write_events_to_db"], [7, 2, 1, "", "write_events_to_file"]], "ensembl.io.genomio.events.load.IdEvent": [[7, 3, 1, "", "event"], [7, 3, 1, "", "from_id"], [7, 2, 1, "", "is_change"], [7, 3, 1, "", "release"], [7, 3, 1, "", "release_date"], [7, 3, 1, "", "to_id"]], "ensembl.io.genomio.events.load.MapSession": [[7, 2, 1, "", "add_event"]], "ensembl.io.genomio.fasta": [[8, 0, 0, "-", "process"]], "ensembl.io.genomio.fasta.process": [[8, 4, 1, "", "GFFParserError"], [8, 5, 1, "", "get_peptides_to_exclude"], [8, 5, 1, "", "prep_fasta_data"]], "ensembl.io.genomio.genbank": [[9, 0, 0, "-", "download"], [9, 0, 0, "-", "extract_data"]], "ensembl.io.genomio.genbank.download": [[9, 4, 1, "", "DownloadError"], [9, 5, 1, "", "download_genbank"]], "ensembl.io.genomio.genbank.extract_data": [[9, 1, 1, "", "FormattedFilesGenerator"], [9, 4, 1, "", "GBParseError"], [9, 1, 1, "", "GenomeFiles"], [9, 4, 1, "", "UnsupportedData"]], "ensembl.io.genomio.genbank.extract_data.FormattedFilesGenerator": [[9, 3, 1, "", "allowed_feat_types"], [9, 2, 1, "", "extract_gb"], [9, 3, 1, "", "locations"], [9, 2, 1, "", "parse_genbank"], [9, 2, 1, "", "set_prefix"], [9, 2, 1, "", "set_production_name"]], "ensembl.io.genomio.genome_metadata": [[10, 0, 0, "-", "dump"], [10, 0, 0, "-", "extend"], [10, 0, 0, "-", "prepare"]], "ensembl.io.genomio.genome_metadata.dump": [[10, 5, 1, "", "check_assembly_version"], [10, 5, 1, "", "filter_genome_meta"], [10, 5, 1, "", "get_genome_metadata"]], "ensembl.io.genomio.genome_metadata.extend": [[10, 4, 1, "", "MissingDataError"], [10, 5, 1, "", "amend_genomic_metadata"], [10, 5, 1, "", "get_additions"], [10, 5, 1, "", "get_gbff_regions"], [10, 5, 1, "", "get_report_regions_names"]], "ensembl.io.genomio.genome_metadata.prepare": [[10, 4, 1, "", "MetadataError"], [10, 4, 1, "", "MissingNodeError"], [10, 5, 1, "", "add_assembly_version"], [10, 5, 1, "", "add_genebuild_metadata"], [10, 5, 1, "", "add_provider"], [10, 5, 1, "", "add_species_metadata"], [10, 5, 1, "", "get_taxonomy_from_accession"], [10, 5, 1, "", "prepare_genome_metadata"]], "ensembl.io.genomio.genome_stats": [[11, 0, 0, "-", "compare"], [11, 0, 0, "-", "dump"]], "ensembl.io.genomio.genome_stats.compare": [[11, 5, 1, "", "compare_annotation"], [11, 5, 1, "", "compare_assembly"], [11, 5, 1, "", "compare_stats"]], "ensembl.io.genomio.genome_stats.dump": [[11, 1, 1, "", "StatsGenerator"]], "ensembl.io.genomio.genome_stats.dump.StatsGenerator": [[11, 2, 1, "", "get_annotation_stats"], [11, 2, 1, "", "get_assembly_stats"], [11, 2, 1, "", "get_attrib_counts"], [11, 2, 1, "", "get_biotypes"], [11, 2, 1, "", "get_feature_stats"], [11, 2, 1, "", "get_stats"]], "ensembl.io.genomio.gff3": [[12, 0, 0, "-", "extract_annotation"], [12, 0, 0, "-", "process"]], "ensembl.io.genomio.gff3.extract_annotation": [[12, 4, 1, "", "AnnotationError"], [12, 4, 1, "", "DuplicateIdError"], [12, 1, 1, "", "FunctionalAnnotations"], [12, 4, 1, "", "MissingParentError"]], "ensembl.io.genomio.gff3.extract_annotation.FunctionalAnnotations": [[12, 2, 1, "", "add_feature"], [12, 2, 1, "", "add_parent"], [12, 2, 1, "", "get_features"], [12, 2, 1, "", "get_parent"], [12, 2, 1, "", "product_is_informative"], [12, 2, 1, "", "to_json"]], "ensembl.io.genomio.gff3.process": [[12, 1, 1, "", "GFFGeneMerger"], [12, 4, 1, "", "GFFParserError"], [12, 1, 1, "", "GFFSimplifier"], [12, 1, 1, "", "Records"]], "ensembl.io.genomio.gff3.process.GFFGeneMerger": [[12, 2, 1, "", "merge"]], "ensembl.io.genomio.gff3.process.GFFSimplifier": [[12, 3, 1, "", "allow_pseudogene_with_CDS"], [12, 2, 1, "", "build_transcript"], [12, 2, 1, "", "cds_gene"], [12, 2, 1, "", "clean_extra_exons"], [12, 3, 1, "", "current_stable_id_number"], [12, 3, 1, "", "exclude_seq_regions"], [12, 2, 1, "", "format_gene_segments"], [12, 2, 1, "", "format_mobile_element"], [12, 3, 1, "", "gene_cds_skip_others"], [12, 2, 1, "", "gene_to_cds"], [12, 2, 1, "", "gene_to_exon"], [12, 2, 1, "", "generate_stable_id"], [12, 3, 1, "", "min_id_length"], [12, 2, 1, "", "move_cds_to_mrna"], [12, 2, 1, "", "normalize_cds_id"], [12, 2, 1, "", "normalize_gene"], [12, 2, 1, "", "normalize_gene_id"], [12, 2, 1, "", "normalize_pseudogene_cds"], [12, 2, 1, "", "normalize_transcript_id"], [12, 2, 1, "", "remove_cds_from_pseudogene"], [12, 2, 1, "", "remove_prefixes"], [12, 2, 1, "", "simpler_gff3"], [12, 3, 1, "", "skip_unrecognized"], [12, 3, 1, "", "stable_id_prefix"], [12, 2, 1, "", "transcript_for_gene"], [12, 2, 1, "", "transcript_gene"], [12, 2, 1, "", "valid_id"], [12, 3, 1, "", "validate_gene_id"]], "ensembl.io.genomio.gff3.process.Records": [[12, 2, 1, "", "to_gff"]], "ensembl.io.genomio.manifest": [[13, 0, 0, "-", "check_integrity"], [13, 0, 0, "-", "compute_stats"], [13, 0, 0, "-", "generate"]], "ensembl.io.genomio.manifest.check_integrity": [[13, 1, 1, "", "IntegrityTool"], [13, 4, 1, "", "InvalidIntegrityError"], [13, 1, 1, "", "Manifest"], [13, 5, 1, "", "main"]], "ensembl.io.genomio.manifest.check_integrity.IntegrityTool": [[13, 2, 1, "", "add_errors"], [13, 2, 1, "", "check_ids"], [13, 2, 1, "", "check_integrity"], [13, 2, 1, "", "check_lengths"], [13, 2, 1, "", "check_seq_region_lengths"], [13, 2, 1, "", "set_brc_mode"], [13, 2, 1, "", "set_ignore_final_stops"]], "ensembl.io.genomio.manifest.check_integrity.Manifest": [[13, 2, 1, "", "get_agp_seq_regions"], [13, 2, 1, "", "get_circular"], [13, 2, 1, "", "get_fasta_lengths"], [13, 2, 1, "", "get_functional_annotation"], [13, 2, 1, "", "get_gff3"], [13, 2, 1, "", "get_lengths"], [13, 2, 1, "", "get_manifest"], [13, 2, 1, "", "has_lengths"], [13, 2, 1, "", "prepare_integrity_data"]], "ensembl.io.genomio.manifest.compute_stats": [[13, 1, 1, "", "BiotypeCounter"], [13, 4, 1, "", "StatsError"], [13, 5, 1, "", "main"], [13, 1, 1, "", "manifest_stats"]], "ensembl.io.genomio.manifest.compute_stats.BiotypeCounter": [[13, 2, 1, "", "add_id"], [13, 2, 1, "", "unique_count"]], "ensembl.io.genomio.manifest.compute_stats.manifest_stats": [[13, 2, 1, "", "biotypes_stats"], [13, 2, 1, "", "check_ncbi_stats"], [13, 2, 1, "", "compare_ncbi_counts"], [13, 2, 1, "", "coord_systems_stats"], [13, 2, 1, "", "count_biotypes"], [13, 2, 1, "", "get_gff3_stats"], [13, 2, 1, "", "get_manifest"], [13, 2, 1, "", "get_seq_region_stats"], [13, 2, 1, "", "increment_biotype"], [13, 2, 1, "", "run"], [13, 2, 1, "", "seq_region_special_stats"]], "ensembl.io.genomio.manifest.generate": [[13, 1, 1, "", "ManifestMaker"], [13, 5, 1, "", "main"]], "ensembl.io.genomio.manifest.generate.ManifestMaker": [[13, 3, 1, "", "alias_names"], [13, 2, 1, "", "create_manifest"], [13, 2, 1, "", "get_files_checksums"], [13, 3, 1, "", "names"], [13, 3, 1, "", "same_names"]], "ensembl.io.genomio.schemas": [[15, 0, 0, "-", "json"]], "ensembl.io.genomio.schemas.json": [[15, 0, 0, "-", "factory"], [15, 0, 0, "-", "validate"]], "ensembl.io.genomio.schemas.json.factory": [[15, 5, 1, "", "json_schema_factory"]], "ensembl.io.genomio.schemas.json.validate": [[15, 5, 1, "", "json_schema_validator"]], "ensembl.io.genomio.seq_region": [[16, 0, 0, "-", "dump"], [16, 0, 0, "-", "prepare"]], "ensembl.io.genomio.seq_region.dump": [[16, 4, 1, "", "MapFormatError"], [16, 5, 1, "", "add_attribs"], [16, 5, 1, "", "get_attribs"], [16, 5, 1, "", "get_coord_systems"], [16, 5, 1, "", "get_external_db_map"], [16, 5, 1, "", "get_karyotype"], [16, 5, 1, "", "get_seq_regions"], [16, 5, 1, "", "get_synonyms"], [16, 5, 1, "", "main"]], "ensembl.io.genomio.seq_region.prepare": [[16, 4, 1, "", "UnknownMetadata"], [16, 5, 1, "", "add_insdc_seq_region_name"], [16, 5, 1, "", "add_mitochondrial_codon_table"], [16, 5, 1, "", "add_translation_table"], [16, 5, 1, "", "exclude_seq_regions"], [16, 5, 1, "", "get_codon_table"], [16, 5, 1, "", "get_gbff_seq_regions"], [16, 5, 1, "", "get_genbank_id"], [16, 5, 1, "", "get_organelle"], [16, 5, 1, "", "get_report_regions"], [16, 5, 1, "", "make_seq_region"], [16, 5, 1, "", "merge_seq_regions"], [16, 5, 1, "", "prepare_seq_region_metadata"], [16, 5, 1, "", "report_to_csv"]], "ensembl.io.genomio.utils": [[17, 0, 0, "-", "archive_utils"], [17, 0, 0, "-", "json_utils"]], "ensembl.io.genomio.utils.archive_utils": [[17, 5, 1, "", "extract_file"], [17, 5, 1, "", "open_gz_file"]], "ensembl.io.genomio.utils.json_utils": [[17, 5, 1, "", "get_json"], [17, 5, 1, "", "print_json"]]}, "objtypes": {"0": "py:module", "1": "py:class", "2": "py:method", "3": "py:attribute", "4": "py:exception", "5": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "class", "Python class"], "2": ["py", "method", "Python method"], "3": ["py", "attribute", "Python attribute"], "4": ["py", "exception", "Python exception"], "5": ["py", "function", "Python function"]}, "titleterms": {"ensembl": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 21], "namespac": [0, 3], "subpackag": [0, 1, 3, 4, 14], "brc4": [1, 2], "packag": [1, 2, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17], "runnabl": 2, "submodul": [2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 15, 16, 17], "compare_fasta": 2, "modul": [2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 15, 16, 17, 19], "compare_report": 2, "core_serv": 2, "download_genbank": 2, "dump_stable_id": 2, "extract_from_gb": 2, "fill_metadata": 2, "gff3_specifi": 2, "integr": 2, "json_schema_factori": 2, "load_sequence_data": 2, "manifest": [2, 13], "manifest_stat": 2, "prepare_genom": 2, "read_json": 2, "say_access": 2, "seqregion_pars": 2, "io": [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17], "genomio": [4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18], "assembli": 5, "download": [5, 9], "databas": 6, "factori": [6, 15], "event": 7, "dump": [7, 10, 11, 16], "format": 7, "load": 7, "fasta": 8, "process": [8, 12], "genbank": 9, "extract_data": 9, "genome_metadata": 10, "extend": 10, "prepar": [10, 16], "genome_stat": 11, "compar": 11, "gff3": 12, "extract_annot": 12, "check_integr": 13, "compute_stat": 13, "gener": 13, "schema": [14, 15], "json": 15, "valid": 15, "seq_region": 16, "util": 17, "archive_util": 17, "json_util": 17, "ehiv": 18, "pipelin": 18, "content": 18, "index": 18, "indic": 18, "tabl": 18, "api": 19, "setup": [19, 22], "instal": 19, "requir": 19, "softwar": 19, "python": 19, "licens": 20, "usag": 22, "environ": 22}, "envversion": {"sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 58}, "alltitles": {"ensembl namespace": [[0, "module-ensembl"]], "Subpackages": [[0, "subpackages"], [1, "subpackages"], [3, "subpackages"], [4, "subpackages"], [14, "subpackages"]], "ensembl.brc4 package": [[1, "module-ensembl.brc4"]], "ensembl.brc4.runnable package": [[2, "module-ensembl.brc4.runnable"]], "Submodules": [[2, "submodules"], [5, "submodules"], [6, "submodules"], [7, "submodules"], [8, "submodules"], [9, "submodules"], [10, "submodules"], [11, "submodules"], [12, "submodules"], [13, "submodules"], [15, "submodules"], [16, "submodules"], [17, "submodules"]], "ensembl.brc4.runnable.compare_fasta module": [[2, "module-ensembl.brc4.runnable.compare_fasta"]], "ensembl.brc4.runnable.compare_report module": [[2, "module-ensembl.brc4.runnable.compare_report"]], "ensembl.brc4.runnable.core_server module": [[2, "module-ensembl.brc4.runnable.core_server"]], "ensembl.brc4.runnable.download_genbank module": [[2, "module-ensembl.brc4.runnable.download_genbank"]], "ensembl.brc4.runnable.dump_stable_ids module": [[2, "module-ensembl.brc4.runnable.dump_stable_ids"]], "ensembl.brc4.runnable.extract_from_gb module": [[2, "module-ensembl.brc4.runnable.extract_from_gb"]], "ensembl.brc4.runnable.fill_metadata module": [[2, "module-ensembl.brc4.runnable.fill_metadata"]], "ensembl.brc4.runnable.gff3_specifier module": [[2, "module-ensembl.brc4.runnable.gff3_specifier"]], "ensembl.brc4.runnable.integrity module": [[2, "module-ensembl.brc4.runnable.integrity"]], "ensembl.brc4.runnable.json_schema_factory module": [[2, "module-ensembl.brc4.runnable.json_schema_factory"]], "ensembl.brc4.runnable.load_sequence_data module": [[2, "module-ensembl.brc4.runnable.load_sequence_data"]], "ensembl.brc4.runnable.manifest module": [[2, "module-ensembl.brc4.runnable.manifest"]], "ensembl.brc4.runnable.manifest_stats module": [[2, "module-ensembl.brc4.runnable.manifest_stats"]], "ensembl.brc4.runnable.prepare_genome module": [[2, "module-ensembl.brc4.runnable.prepare_genome"]], "ensembl.brc4.runnable.read_json module": [[2, "module-ensembl.brc4.runnable.read_json"]], "ensembl.brc4.runnable.say_accession module": [[2, "module-ensembl.brc4.runnable.say_accession"]], "ensembl.brc4.runnable.seqregion_parser module": [[2, "module-ensembl.brc4.runnable.seqregion_parser"]], "ensembl.io namespace": [[3, "module-ensembl.io"]], "ensembl.io.genomio package": [[4, "module-ensembl.io.genomio"]], "ensembl.io.genomio.assembly package": [[5, "module-ensembl.io.genomio.assembly"]], "ensembl.io.genomio.assembly.download module": [[5, "module-ensembl.io.genomio.assembly.download"]], "ensembl.io.genomio.database package": [[6, "module-ensembl.io.genomio.database"]], "ensembl.io.genomio.database.factory module": [[6, "module-ensembl.io.genomio.database.factory"]], "ensembl.io.genomio.events package": [[7, "module-ensembl.io.genomio.events"]], "ensembl.io.genomio.events.dump module": [[7, "module-ensembl.io.genomio.events.dump"]], "ensembl.io.genomio.events.format module": [[7, "module-ensembl.io.genomio.events.format"]], "ensembl.io.genomio.events.load module": [[7, "module-ensembl.io.genomio.events.load"]], "ensembl.io.genomio.fasta package": [[8, "module-ensembl.io.genomio.fasta"]], "ensembl.io.genomio.fasta.process module": [[8, "module-ensembl.io.genomio.fasta.process"]], "ensembl.io.genomio.genbank package": [[9, "module-ensembl.io.genomio.genbank"]], "ensembl.io.genomio.genbank.download module": [[9, "module-ensembl.io.genomio.genbank.download"]], "ensembl.io.genomio.genbank.extract_data module": [[9, "module-ensembl.io.genomio.genbank.extract_data"]], "ensembl.io.genomio.genome_metadata package": [[10, "module-ensembl.io.genomio.genome_metadata"]], "ensembl.io.genomio.genome_metadata.dump module": [[10, "module-ensembl.io.genomio.genome_metadata.dump"]], "ensembl.io.genomio.genome_metadata.extend module": [[10, "module-ensembl.io.genomio.genome_metadata.extend"]], "ensembl.io.genomio.genome_metadata.prepare module": [[10, "module-ensembl.io.genomio.genome_metadata.prepare"]], "ensembl.io.genomio.genome_stats package": [[11, "module-ensembl.io.genomio.genome_stats"]], "ensembl.io.genomio.genome_stats.compare module": [[11, "module-ensembl.io.genomio.genome_stats.compare"]], "ensembl.io.genomio.genome_stats.dump module": [[11, "module-ensembl.io.genomio.genome_stats.dump"]], "ensembl.io.genomio.gff3 package": [[12, "module-ensembl.io.genomio.gff3"]], "ensembl.io.genomio.gff3.extract_annotation module": [[12, "module-ensembl.io.genomio.gff3.extract_annotation"]], "ensembl.io.genomio.gff3.process module": [[12, "module-ensembl.io.genomio.gff3.process"]], "ensembl.io.genomio.manifest package": [[13, "module-ensembl.io.genomio.manifest"]], "ensembl.io.genomio.manifest.check_integrity module": [[13, "module-ensembl.io.genomio.manifest.check_integrity"]], "ensembl.io.genomio.manifest.compute_stats module": [[13, "module-ensembl.io.genomio.manifest.compute_stats"]], "ensembl.io.genomio.manifest.generate module": [[13, "module-ensembl.io.genomio.manifest.generate"]], "ensembl.io.genomio.schemas package": [[14, "module-ensembl.io.genomio.schemas"]], "ensembl.io.genomio.schemas.json package": [[15, "module-ensembl.io.genomio.schemas.json"]], "ensembl.io.genomio.schemas.json.factory module": [[15, "module-ensembl.io.genomio.schemas.json.factory"]], "ensembl.io.genomio.schemas.json.validate module": [[15, "module-ensembl.io.genomio.schemas.json.validate"]], "ensembl.io.genomio.seq_region package": [[16, "module-ensembl.io.genomio.seq_region"]], "ensembl.io.genomio.seq_region.dump module": [[16, "module-ensembl.io.genomio.seq_region.dump"]], "ensembl.io.genomio.seq_region.prepare module": [[16, "module-ensembl.io.genomio.seq_region.prepare"]], "ensembl.io.genomio.utils package": [[17, "module-ensembl.io.genomio.utils"]], "ensembl.io.genomio.utils.archive_utils module": [[17, "module-ensembl.io.genomio.utils.archive_utils"]], "ensembl.io.genomio.utils.json_utils module": [[17, "module-ensembl.io.genomio.utils.json_utils"]], "Ensembl-genomio": [[18, "ensembl-genomio"]], "Ehive pipelines": [[18, "ehive-pipelines"]], "Contents": [[18, "contents"]], "Index": [[18, null]], "Indices and tables": [[18, "indices-and-tables"]], "API Setup and installation": [[19, "api-setup-and-installation"]], "Requirements": [[19, "requirements"]], "Software": [[19, "software"]], "Python Modules": [[19, "python-modules"]], "Installation": [[19, "installation"]], "License": [[20, "license"]], "ensembl": [[21, "ensembl"]], "Usage": [[22, "usage"]], "Environment setup": [[22, "environment-setup"]]}, "indexentries": {"ensembl": [[0, "module-ensembl"]], "module": [[0, "module-ensembl"], [1, "module-ensembl.brc4"], [2, "module-ensembl.brc4.runnable"], [2, "module-ensembl.brc4.runnable.compare_fasta"], [2, "module-ensembl.brc4.runnable.compare_report"], [2, "module-ensembl.brc4.runnable.core_server"], [2, "module-ensembl.brc4.runnable.download_genbank"], [2, "module-ensembl.brc4.runnable.dump_stable_ids"], [2, "module-ensembl.brc4.runnable.extract_from_gb"], [2, "module-ensembl.brc4.runnable.fill_metadata"], [2, "module-ensembl.brc4.runnable.gff3_specifier"], [2, "module-ensembl.brc4.runnable.integrity"], [2, "module-ensembl.brc4.runnable.json_schema_factory"], [2, "module-ensembl.brc4.runnable.load_sequence_data"], [2, "module-ensembl.brc4.runnable.manifest"], [2, "module-ensembl.brc4.runnable.manifest_stats"], [2, "module-ensembl.brc4.runnable.prepare_genome"], [2, "module-ensembl.brc4.runnable.read_json"], [2, "module-ensembl.brc4.runnable.say_accession"], [2, "module-ensembl.brc4.runnable.seqregion_parser"], [3, "module-ensembl.io"], [4, "module-ensembl.io.genomio"], [5, "module-ensembl.io.genomio.assembly"], [5, "module-ensembl.io.genomio.assembly.download"], [6, "module-ensembl.io.genomio.database"], [6, "module-ensembl.io.genomio.database.factory"], [7, "module-ensembl.io.genomio.events"], [7, "module-ensembl.io.genomio.events.dump"], [7, "module-ensembl.io.genomio.events.format"], [7, "module-ensembl.io.genomio.events.load"], [8, "module-ensembl.io.genomio.fasta"], [8, "module-ensembl.io.genomio.fasta.process"], [9, "module-ensembl.io.genomio.genbank"], [9, "module-ensembl.io.genomio.genbank.download"], [9, "module-ensembl.io.genomio.genbank.extract_data"], [10, "module-ensembl.io.genomio.genome_metadata"], [10, "module-ensembl.io.genomio.genome_metadata.dump"], [10, "module-ensembl.io.genomio.genome_metadata.extend"], [10, "module-ensembl.io.genomio.genome_metadata.prepare"], [11, "module-ensembl.io.genomio.genome_stats"], [11, "module-ensembl.io.genomio.genome_stats.compare"], [11, "module-ensembl.io.genomio.genome_stats.dump"], [12, "module-ensembl.io.genomio.gff3"], [12, "module-ensembl.io.genomio.gff3.extract_annotation"], [12, "module-ensembl.io.genomio.gff3.process"], [13, "module-ensembl.io.genomio.manifest"], [13, "module-ensembl.io.genomio.manifest.check_integrity"], [13, "module-ensembl.io.genomio.manifest.compute_stats"], [13, "module-ensembl.io.genomio.manifest.generate"], [14, "module-ensembl.io.genomio.schemas"], [15, "module-ensembl.io.genomio.schemas.json"], [15, "module-ensembl.io.genomio.schemas.json.factory"], [15, "module-ensembl.io.genomio.schemas.json.validate"], [16, "module-ensembl.io.genomio.seq_region"], [16, "module-ensembl.io.genomio.seq_region.dump"], [16, "module-ensembl.io.genomio.seq_region.prepare"], [17, "module-ensembl.io.genomio.utils"], [17, "module-ensembl.io.genomio.utils.archive_utils"], [17, "module-ensembl.io.genomio.utils.json_utils"]], "ensembl.brc4": [[1, "module-ensembl.brc4"]], "biotypecollection (class in ensembl.brc4.runnable.manifest_stats)": [[2, "ensembl.brc4.runnable.manifest_stats.BiotypeCollection"]], "coreserver (class in ensembl.brc4.runnable.core_server)": [[2, "ensembl.brc4.runnable.core_server.CoreServer"]], "downloaderror": [[2, "ensembl.brc4.runnable.download_genbank.DownloadError"], [9, "ensembl.io.genomio.genbank.download.DownloadError"]], "dumpstableids (class in ensembl.brc4.runnable.dump_stable_ids)": [[2, "ensembl.brc4.runnable.dump_stable_ids.DumpStableIDs"]], "featurecounter (class in ensembl.brc4.runnable.manifest_stats)": [[2, "ensembl.brc4.runnable.manifest_stats.FeatureCounter"]], "formattedfilesgenerator (class in ensembl.brc4.runnable.extract_from_gb)": [[2, "ensembl.brc4.runnable.extract_from_gb.FormattedFilesGenerator"]], "seqgroup (class in ensembl.brc4.runnable.compare_fasta)": [[2, "ensembl.brc4.runnable.compare_fasta.SeqGroup"]], "seqregionparser (class in ensembl.brc4.runnable.seqregion_parser)": [[2, "ensembl.brc4.runnable.seqregion_parser.SeqregionParser"]], "stableidevent (class in ensembl.brc4.runnable.dump_stable_ids)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent"]], "unsupportedevent": [[2, "ensembl.brc4.runnable.dump_stable_ids.UnsupportedEvent"], [7, "ensembl.io.genomio.events.dump.UnsupportedEvent"]], "add_asm_mappings() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_asm_mappings"]], "add_contig_ena_attrib() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_contig_ena_attrib"]], "add_feature() (ensembl.brc4.runnable.manifest_stats.biotypecollection method)": [[2, "ensembl.brc4.runnable.manifest_stats.BiotypeCollection.add_feature"]], "add_feature() (ensembl.brc4.runnable.manifest_stats.featurecounter method)": [[2, "ensembl.brc4.runnable.manifest_stats.FeatureCounter.add_feature"]], "add_from() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.add_from"]], "add_id() (ensembl.brc4.runnable.compare_fasta.seqgroup method)": [[2, "ensembl.brc4.runnable.compare_fasta.SeqGroup.add_id"]], "add_karyotype_bands() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_karyotype_bands"]], "add_karyotype_data() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_karyotype_data"]], "add_karyotype_rank_based_on_assembly_metadata() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_karyotype_rank_based_on_assembly_metadata"]], "add_karyotype_rank_for_chromosomes() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_karyotype_rank_for_chromosomes"]], "add_karyotype_rank_from_bands_info() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_karyotype_rank_from_bands_info"]], "add_pair() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.add_pair"]], "add_pairs() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.add_pairs"]], "add_report_to_map() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.add_report_to_map"]], "add_sr_attribs() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_sr_attribs"]], "add_sr_ebi_brc4_names() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_sr_ebi_brc4_names"]], "add_sr_synonyms() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.add_sr_synonyms"]], "add_to() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.add_to"]], "agp_prune() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.agp_prune"]], "allowed_attributes (ensembl.brc4.runnable.gff3_specifier.gff3_specifier attribute)": [[2, "ensembl.brc4.runnable.gff3_specifier.gff3_specifier.allowed_attributes"]], "allowed_biotypes (ensembl.brc4.runnable.gff3_specifier.gff3_specifier attribute)": [[2, "ensembl.brc4.runnable.gff3_specifier.gff3_specifier.allowed_biotypes"]], "allowed_feat_types (ensembl.brc4.runnable.extract_from_gb.formattedfilesgenerator attribute)": [[2, "ensembl.brc4.runnable.extract_from_gb.FormattedFilesGenerator.allowed_feat_types"]], "asm_name() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.asm_name"]], "assembly_level() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.assembly_level"]], "brc_format_1() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.brc_format_1"]], "brc_format_2() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.brc_format_2"]], "build_seq_dict() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.build_seq_dict"]], "check_db_name_format() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.check_db_name_format"]], "check_ids() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.check_ids"]], "check_lengths() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.check_lengths"]], "check_md5sum() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.check_md5sum"]], "check_ncbi_stats() (ensembl.brc4.runnable.manifest_stats.manifest_stats method)": [[2, "ensembl.brc4.runnable.manifest_stats.manifest_stats.check_ncbi_stats"]], "check_seq_region_lengths() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.check_seq_region_lengths"]], "chunk_contigs() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.chunk_contigs"]], "clean_pairs() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.clean_pairs"]], "clean_set() (ensembl.brc4.runnable.dump_stable_ids.stableidevent static method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.clean_set"]], "compare_fasta (class in ensembl.brc4.runnable.compare_fasta)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta"]], "compare_ncbi_counts() (ensembl.brc4.runnable.manifest_stats.manifest_stats method)": [[2, "ensembl.brc4.runnable.manifest_stats.manifest_stats.compare_ncbi_counts"]], "compare_report (class in ensembl.brc4.runnable.compare_report)": [[2, "ensembl.brc4.runnable.compare_report.compare_report"]], "compare_seqs() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.compare_seqs"]], "connect() (ensembl.brc4.runnable.core_server.coreserver method)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.connect"]], "coord_sys_order() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.coord_sys_order"]], "copy_file() (ensembl.brc4.runnable.manifest.manifest method)": [[2, "ensembl.brc4.runnable.manifest.manifest.copy_file"]], "copy_files() (ensembl.brc4.runnable.manifest.manifest method)": [[2, "ensembl.brc4.runnable.manifest.manifest.copy_files"]], "copy_sr_name_to_syn() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.copy_sr_name_to_syn"]], "create_manifest() (ensembl.brc4.runnable.manifest.manifest method)": [[2, "ensembl.brc4.runnable.manifest.manifest.create_manifest"]], "dataflow_json() (ensembl.brc4.runnable.json_schema_factory.json_schema_factory method)": [[2, "ensembl.brc4.runnable.json_schema_factory.json_schema_factory.dataflow_json"]], "date (ensembl.brc4.runnable.dump_stable_ids.stableidevent attribute)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.date"]], "db_string() (ensembl.brc4.runnable.fill_metadata.fill_metadata method)": [[2, "ensembl.brc4.runnable.fill_metadata.fill_metadata.db_string"]], "db_string() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.db_string"]], "download_genbank (class in ensembl.brc4.runnable.download_genbank)": [[2, "ensembl.brc4.runnable.download_genbank.download_genbank"]], "download_genbank() (ensembl.brc4.runnable.download_genbank.download_genbank static method)": [[2, "ensembl.brc4.runnable.download_genbank.download_genbank.download_genbank"]], "ensembl.brc4.runnable": [[2, "module-ensembl.brc4.runnable"]], "ensembl.brc4.runnable.compare_fasta": [[2, "module-ensembl.brc4.runnable.compare_fasta"]], "ensembl.brc4.runnable.compare_report": [[2, "module-ensembl.brc4.runnable.compare_report"]], "ensembl.brc4.runnable.core_server": [[2, "module-ensembl.brc4.runnable.core_server"]], "ensembl.brc4.runnable.download_genbank": [[2, "module-ensembl.brc4.runnable.download_genbank"]], "ensembl.brc4.runnable.dump_stable_ids": [[2, "module-ensembl.brc4.runnable.dump_stable_ids"]], "ensembl.brc4.runnable.extract_from_gb": [[2, "module-ensembl.brc4.runnable.extract_from_gb"]], "ensembl.brc4.runnable.fill_metadata": [[2, "module-ensembl.brc4.runnable.fill_metadata"]], "ensembl.brc4.runnable.gff3_specifier": [[2, "module-ensembl.brc4.runnable.gff3_specifier"]], "ensembl.brc4.runnable.integrity": [[2, "module-ensembl.brc4.runnable.integrity"]], "ensembl.brc4.runnable.json_schema_factory": [[2, "module-ensembl.brc4.runnable.json_schema_factory"]], "ensembl.brc4.runnable.load_sequence_data": [[2, "module-ensembl.brc4.runnable.load_sequence_data"]], "ensembl.brc4.runnable.manifest": [[2, "module-ensembl.brc4.runnable.manifest"]], "ensembl.brc4.runnable.manifest_stats": [[2, "module-ensembl.brc4.runnable.manifest_stats"]], "ensembl.brc4.runnable.prepare_genome": [[2, "module-ensembl.brc4.runnable.prepare_genome"]], "ensembl.brc4.runnable.read_json": [[2, "module-ensembl.brc4.runnable.read_json"]], "ensembl.brc4.runnable.say_accession": [[2, "module-ensembl.brc4.runnable.say_accession"]], "ensembl.brc4.runnable.seqregion_parser": [[2, "module-ensembl.brc4.runnable.seqregion_parser"]], "example() (ensembl.brc4.runnable.manifest_stats.featurecounter method)": [[2, "ensembl.brc4.runnable.manifest_stats.FeatureCounter.example"]], "extend_event() (ensembl.brc4.runnable.dump_stable_ids.dumpstableids method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.DumpStableIDs.extend_event"]], "extract_from_gb (class in ensembl.brc4.runnable.extract_from_gb)": [[2, "ensembl.brc4.runnable.extract_from_gb.extract_from_gb"]], "fill_metadata (class in ensembl.brc4.runnable.fill_metadata)": [[2, "ensembl.brc4.runnable.fill_metadata.fill_metadata"]], "filter_already_loaded_regions_from_agp() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.filter_already_loaded_regions_from_agp"]], "find_common_groups() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.find_common_groups"]], "flattern() (ensembl.brc4.runnable.fill_metadata.fill_metadata method)": [[2, "ensembl.brc4.runnable.fill_metadata.fill_metadata.flattern"]], "flattern_seq_region_item() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.flattern_seq_region_item"]], "from_list (ensembl.brc4.runnable.dump_stable_ids.stableidevent attribute)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.from_list"]], "from_param() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.from_param"]], "get_agp_seq_regions() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.get_agp_seq_regions"]], "get_all_cores() (ensembl.brc4.runnable.core_server.coreserver method)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.get_all_cores"]], "get_assembly_version() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.get_assembly_version"]], "get_cores() (ensembl.brc4.runnable.core_server.coreserver method)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.get_cores"]], "get_count() (ensembl.brc4.runnable.manifest_stats.biotypecollection method)": [[2, "ensembl.brc4.runnable.manifest_stats.BiotypeCollection.get_count"]], "get_cursor() (ensembl.brc4.runnable.core_server.coreserver method)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.get_cursor"]], "get_db_metadata() (ensembl.brc4.runnable.core_server.coreserver method)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.get_db_metadata"]], "get_external_db_mapping() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.get_external_db_mapping"]], "get_fasta() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.get_fasta"]], "get_fasta_lengths() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.get_fasta_lengths"]], "get_full_release() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.get_full_release"]], "get_functional_annotation() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.get_functional_annotation"]], "get_genome() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.get_genome"]], "get_genome_data() (ensembl.brc4.runnable.manifest.manifest method)": [[2, "ensembl.brc4.runnable.manifest.manifest.get_genome_data"]], "get_gff3() (ensembl.brc4.runnable.gff3_specifier.gff3_specifier method)": [[2, "ensembl.brc4.runnable.gff3_specifier.gff3_specifier.get_gff3"]], "get_gff3() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.get_gff3"]], "get_gff3_stats() (ensembl.brc4.runnable.manifest_stats.manifest_stats method)": [[2, "ensembl.brc4.runnable.manifest_stats.manifest_stats.get_gff3_stats"]], "get_history() (ensembl.brc4.runnable.dump_stable_ids.dumpstableids method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.DumpStableIDs.get_history"]], "get_json() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.get_json"]], "get_manifest() (ensembl.brc4.runnable.manifest_stats.manifest_stats method)": [[2, "ensembl.brc4.runnable.manifest_stats.manifest_stats.get_manifest"]], "get_manifest() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.get_manifest"]], "get_map() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.get_map"]], "get_mapping_sessions() (ensembl.brc4.runnable.dump_stable_ids.dumpstableids method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.DumpStableIDs.get_mapping_sessions"]], "get_md5sum() (ensembl.brc4.runnable.manifest.manifest method)": [[2, "ensembl.brc4.runnable.manifest.manifest.get_md5sum"]], "get_name() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.get_name"]], "get_pairs() (ensembl.brc4.runnable.dump_stable_ids.dumpstableids method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.DumpStableIDs.get_pairs"]], "get_report_regions() (ensembl.brc4.runnable.seqregion_parser.seqregionparser method)": [[2, "ensembl.brc4.runnable.seqregion_parser.SeqregionParser.get_report_regions"]], "get_scientific_name() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.get_scientific_name"]], "get_seq_region_stats() (ensembl.brc4.runnable.manifest_stats.manifest_stats method)": [[2, "ensembl.brc4.runnable.manifest_stats.manifest_stats.get_seq_region_stats"]], "get_table_data() (ensembl.brc4.runnable.core_server.coreserver method)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.get_table_data"]], "get_toplevel_from_cs() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.get_toplevel_from_cs"]], "gff3_specifier (class in ensembl.brc4.runnable.gff3_specifier)": [[2, "ensembl.brc4.runnable.gff3_specifier.gff3_specifier"]], "host (ensembl.brc4.runnable.core_server.coreserver attribute)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.host"]], "id_from_map_or_die() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.id_from_map_or_die"]], "ignored_biotypes (ensembl.brc4.runnable.gff3_specifier.gff3_specifier attribute)": [[2, "ensembl.brc4.runnable.gff3_specifier.gff3_specifier.ignored_biotypes"]], "initial_sequence_loading() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.initial_sequence_loading"]], "insert_to_db() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.insert_to_db"]], "integrity (class in ensembl.brc4.runnable.integrity)": [[2, "ensembl.brc4.runnable.integrity.integrity"]], "is_gz() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.is_gz"]], "json_schema_factory (class in ensembl.brc4.runnable.json_schema_factory)": [[2, "ensembl.brc4.runnable.json_schema_factory.json_schema_factory"]], "load_agp() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.load_agp"]], "load_cs_data() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.load_cs_data"]], "load_map_from_core_db() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.load_map_from_core_db"]], "load_map_from_sql_stdout() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.load_map_from_sql_stdout"]], "load_seq_data() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.load_seq_data"]], "load_seq_region() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.load_seq_region"]], "load_seq_region_synonyms_trios_from_core_db() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.load_seq_region_synonyms_trios_from_core_db"]], "load_sequence_data (class in ensembl.brc4.runnable.load_sequence_data)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data"]], "locations (ensembl.brc4.runnable.extract_from_gb.formattedfilesgenerator attribute)": [[2, "ensembl.brc4.runnable.extract_from_gb.FormattedFilesGenerator.locations"]], "make_db_name() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.make_db_name"]], "make_events() (ensembl.brc4.runnable.dump_stable_ids.dumpstableids method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.DumpStableIDs.make_events"]], "make_production_name() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.make_production_name"]], "make_seq_region() (ensembl.brc4.runnable.seqregion_parser.seqregionparser method)": [[2, "ensembl.brc4.runnable.seqregion_parser.SeqregionParser.make_seq_region"]], "manifest (class in ensembl.brc4.runnable.manifest)": [[2, "ensembl.brc4.runnable.manifest.manifest"]], "manifest_stats (class in ensembl.brc4.runnable.manifest_stats)": [[2, "ensembl.brc4.runnable.manifest_stats.manifest_stats"]], "molecule_location (ensembl.brc4.runnable.seqregion_parser.seqregionparser attribute)": [[2, "ensembl.brc4.runnable.seqregion_parser.SeqregionParser.molecule_location"]], "name (ensembl.brc4.runnable.dump_stable_ids.stableidevent attribute)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.name"]], "name_and_id_from_seq_region_item() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.name_and_id_from_seq_region_item"]], "nullify_ctg_cs_version() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.nullify_ctg_cs_version"]], "order_agp_levels() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.order_agp_levels"]], "organellar_assembly() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.organellar_assembly"]], "pairs (ensembl.brc4.runnable.dump_stable_ids.stableidevent attribute)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.pairs"]], "param_bool() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.param_bool"]], "param_defaults() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.param_defaults"]], "param_defaults() (ensembl.brc4.runnable.compare_report.compare_report method)": [[2, "ensembl.brc4.runnable.compare_report.compare_report.param_defaults"]], "param_defaults() (ensembl.brc4.runnable.download_genbank.download_genbank method)": [[2, "ensembl.brc4.runnable.download_genbank.download_genbank.param_defaults"]], "param_defaults() (ensembl.brc4.runnable.extract_from_gb.extract_from_gb method)": [[2, "ensembl.brc4.runnable.extract_from_gb.extract_from_gb.param_defaults"]], "param_defaults() (ensembl.brc4.runnable.fill_metadata.fill_metadata method)": [[2, "ensembl.brc4.runnable.fill_metadata.fill_metadata.param_defaults"]], "param_defaults() (ensembl.brc4.runnable.gff3_specifier.gff3_specifier method)": [[2, "ensembl.brc4.runnable.gff3_specifier.gff3_specifier.param_defaults"]], "param_defaults() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.param_defaults"]], "param_defaults() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.param_defaults"]], "param_defaults() (ensembl.brc4.runnable.manifest_stats.manifest_stats method)": [[2, "ensembl.brc4.runnable.manifest_stats.manifest_stats.param_defaults"]], "param_defaults() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.param_defaults"]], "parse_genbank() (ensembl.brc4.runnable.extract_from_gb.formattedfilesgenerator method)": [[2, "ensembl.brc4.runnable.extract_from_gb.FormattedFilesGenerator.parse_genbank"]], "parse_gff3() (ensembl.brc4.runnable.gff3_specifier.gff3_specifier method)": [[2, "ensembl.brc4.runnable.gff3_specifier.gff3_specifier.parse_gff3"]], "parse_gff3() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.parse_gff3"]], "parse_gff3() (ensembl.brc4.runnable.manifest_stats.manifest_stats method)": [[2, "ensembl.brc4.runnable.manifest_stats.manifest_stats.parse_gff3"]], "password (ensembl.brc4.runnable.core_server.coreserver attribute)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.password"]], "pjc() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.pjc"]], "port (ensembl.brc4.runnable.core_server.coreserver attribute)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.port"]], "prepare_genome (class in ensembl.brc4.runnable.prepare_genome)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome"]], "print_events() (ensembl.brc4.runnable.dump_stable_ids.dumpstableids method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.DumpStableIDs.print_events"]], "print_map() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.print_map"]], "prune_agps() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.prune_agps"]], "quote_or_null() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.quote_or_null"]], "read_json (class in ensembl.brc4.runnable.read_json)": [[2, "ensembl.brc4.runnable.read_json.read_json"]], "release (ensembl.brc4.runnable.dump_stable_ids.stableidevent attribute)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.release"]], "remove_iupac() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.remove_IUPAC"]], "remove_components_from_toplevel() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.remove_components_from_toplevel"]], "report_to_csv() (ensembl.brc4.runnable.seqregion_parser.seqregionparser method)": [[2, "ensembl.brc4.runnable.seqregion_parser.SeqregionParser.report_to_csv"]], "run() (ensembl.brc4.runnable.compare_fasta.compare_fasta method)": [[2, "ensembl.brc4.runnable.compare_fasta.compare_fasta.run"]], "run() (ensembl.brc4.runnable.compare_report.compare_report method)": [[2, "ensembl.brc4.runnable.compare_report.compare_report.run"]], "run() (ensembl.brc4.runnable.download_genbank.download_genbank method)": [[2, "ensembl.brc4.runnable.download_genbank.download_genbank.run"]], "run() (ensembl.brc4.runnable.extract_from_gb.extract_from_gb method)": [[2, "ensembl.brc4.runnable.extract_from_gb.extract_from_gb.run"]], "run() (ensembl.brc4.runnable.fill_metadata.fill_metadata method)": [[2, "ensembl.brc4.runnable.fill_metadata.fill_metadata.run"]], "run() (ensembl.brc4.runnable.gff3_specifier.gff3_specifier method)": [[2, "ensembl.brc4.runnable.gff3_specifier.gff3_specifier.run"]], "run() (ensembl.brc4.runnable.integrity.integrity method)": [[2, "ensembl.brc4.runnable.integrity.integrity.run"]], "run() (ensembl.brc4.runnable.json_schema_factory.json_schema_factory method)": [[2, "ensembl.brc4.runnable.json_schema_factory.json_schema_factory.run"]], "run() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.run"]], "run() (ensembl.brc4.runnable.manifest.manifest method)": [[2, "ensembl.brc4.runnable.manifest.manifest.run"]], "run() (ensembl.brc4.runnable.manifest_stats.manifest_stats method)": [[2, "ensembl.brc4.runnable.manifest_stats.manifest_stats.run"]], "run() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.run"]], "run() (ensembl.brc4.runnable.read_json.read_json method)": [[2, "ensembl.brc4.runnable.read_json.read_json.run"]], "run() (ensembl.brc4.runnable.say_accession.say_accession method)": [[2, "ensembl.brc4.runnable.say_accession.say_accession.run"]], "run_sql_req() (ensembl.brc4.runnable.fill_metadata.fill_metadata method)": [[2, "ensembl.brc4.runnable.fill_metadata.fill_metadata.run_sql_req"]], "run_sql_req() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.run_sql_req"]], "say_accession (class in ensembl.brc4.runnable.say_accession)": [[2, "ensembl.brc4.runnable.say_accession.say_accession"]], "server (ensembl.brc4.runnable.dump_stable_ids.dumpstableids attribute)": [[2, "ensembl.brc4.runnable.dump_stable_ids.DumpStableIDs.server"]], "set_database() (ensembl.brc4.runnable.core_server.coreserver method)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.set_database"]], "set_date() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.set_date"]], "set_prefix() (ensembl.brc4.runnable.extract_from_gb.formattedfilesgenerator method)": [[2, "ensembl.brc4.runnable.extract_from_gb.FormattedFilesGenerator.set_prefix"]], "set_production_name() (ensembl.brc4.runnable.extract_from_gb.formattedfilesgenerator method)": [[2, "ensembl.brc4.runnable.extract_from_gb.FormattedFilesGenerator.set_production_name"]], "set_release() (ensembl.brc4.runnable.dump_stable_ids.stableidevent method)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.set_release"]], "set_toplevel() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.set_toplevel"]], "sorted_stats() (ensembl.brc4.runnable.manifest_stats.biotypecollection method)": [[2, "ensembl.brc4.runnable.manifest_stats.BiotypeCollection.sorted_stats"]], "sr_name_unversion() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.sr_name_unversion"]], "synonym_map (ensembl.brc4.runnable.seqregion_parser.seqregionparser attribute)": [[2, "ensembl.brc4.runnable.seqregion_parser.SeqregionParser.synonym_map"]], "to_list (ensembl.brc4.runnable.dump_stable_ids.stableidevent attribute)": [[2, "ensembl.brc4.runnable.dump_stable_ids.StableIdEvent.to_list"]], "unique_count() (ensembl.brc4.runnable.manifest_stats.featurecounter method)": [[2, "ensembl.brc4.runnable.manifest_stats.FeatureCounter.unique_count"]], "unversion_scaffolds() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.unversion_scaffolds"]], "update_accession() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.update_accession"]], "update_assembly_name() (ensembl.brc4.runnable.prepare_genome.prepare_genome method)": [[2, "ensembl.brc4.runnable.prepare_genome.prepare_genome.update_assembly_name"]], "update_db_single_group() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.update_db_single_group"]], "used_cs_ranks() (ensembl.brc4.runnable.load_sequence_data.load_sequence_data method)": [[2, "ensembl.brc4.runnable.load_sequence_data.load_sequence_data.used_cs_ranks"]], "user (ensembl.brc4.runnable.core_server.coreserver attribute)": [[2, "ensembl.brc4.runnable.core_server.CoreServer.user"]], "ensembl.io": [[3, "module-ensembl.io"]], "ensembl.io.genomio": [[4, "module-ensembl.io.genomio"]], "filedownloaderror": [[5, "ensembl.io.genomio.assembly.download.FileDownloadError"]], "unsupportedformaterror": [[5, "ensembl.io.genomio.assembly.download.UnsupportedFormatError"]], "download_files() (in module ensembl.io.genomio.assembly.download)": [[5, "ensembl.io.genomio.assembly.download.download_files"]], "ensembl.io.genomio.assembly": [[5, "module-ensembl.io.genomio.assembly"]], "ensembl.io.genomio.assembly.download": [[5, "module-ensembl.io.genomio.assembly.download"]], "get_checksums() (in module ensembl.io.genomio.assembly.download)": [[5, "ensembl.io.genomio.assembly.download.get_checksums"]], "get_files_selection() (in module ensembl.io.genomio.assembly.download)": [[5, "ensembl.io.genomio.assembly.download.get_files_selection"]], "get_root_name() (in module ensembl.io.genomio.assembly.download)": [[5, "ensembl.io.genomio.assembly.download.get_root_name"]], "md5_files() (in module ensembl.io.genomio.assembly.download)": [[5, "ensembl.io.genomio.assembly.download.md5_files"]], "retrieve_assembly_data() (in module ensembl.io.genomio.assembly.download)": [[5, "ensembl.io.genomio.assembly.download.retrieve_assembly_data"]], "ensembl.io.genomio.database": [[6, "module-ensembl.io.genomio.database"]], "ensembl.io.genomio.database.factory": [[6, "module-ensembl.io.genomio.database.factory"]], "format_db_data() (in module ensembl.io.genomio.database.factory)": [[6, "ensembl.io.genomio.database.factory.format_db_data"]], "get_metadata_value() (in module ensembl.io.genomio.database.factory)": [[6, "ensembl.io.genomio.database.factory.get_metadata_value"]], "main() (in module ensembl.io.genomio.database.factory)": [[6, "ensembl.io.genomio.database.factory.main"]], "dumpstableids (class in ensembl.io.genomio.events.dump)": [[7, "ensembl.io.genomio.events.dump.DumpStableIDs"]], "eventcollection (class in ensembl.io.genomio.events.load)": [[7, "ensembl.io.genomio.events.load.EventCollection"]], "idevent (class in ensembl.io.genomio.events.load)": [[7, "ensembl.io.genomio.events.load.IdEvent"]], "idsmapper (class in ensembl.io.genomio.events.format)": [[7, "ensembl.io.genomio.events.format.IdsMapper"]], "mapsession (class in ensembl.io.genomio.events.load)": [[7, "ensembl.io.genomio.events.load.MapSession"]], "pair (class in ensembl.io.genomio.events.dump)": [[7, "ensembl.io.genomio.events.dump.Pair"]], "stableidevent (class in ensembl.io.genomio.events.dump)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent"]], "add_deletes() (ensembl.io.genomio.events.load.eventcollection method)": [[7, "ensembl.io.genomio.events.load.EventCollection.add_deletes"]], "add_event() (ensembl.io.genomio.events.load.mapsession method)": [[7, "ensembl.io.genomio.events.load.MapSession.add_event"]], "add_from() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.add_from"]], "add_pair() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.add_pair"]], "add_pairs() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.add_pairs"]], "add_to() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.add_to"]], "brc_format_1() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.brc_format_1"]], "brc_format_2() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.brc_format_2"]], "clean_pairs() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.clean_pairs"]], "clean_set() (ensembl.io.genomio.events.dump.stableidevent static method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.clean_set"]], "date (ensembl.io.genomio.events.dump.stableidevent attribute)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.date"]], "ensembl.io.genomio.events": [[7, "module-ensembl.io.genomio.events"]], "ensembl.io.genomio.events.dump": [[7, "module-ensembl.io.genomio.events.dump"]], "ensembl.io.genomio.events.format": [[7, "module-ensembl.io.genomio.events.format"]], "ensembl.io.genomio.events.load": [[7, "module-ensembl.io.genomio.events.load"]], "event (ensembl.io.genomio.events.load.idevent attribute)": [[7, "ensembl.io.genomio.events.load.IdEvent.event"]], "extend_event() (ensembl.io.genomio.events.dump.dumpstableids method)": [[7, "ensembl.io.genomio.events.dump.DumpStableIDs.extend_event"]], "from_id (ensembl.io.genomio.events.load.idevent attribute)": [[7, "ensembl.io.genomio.events.load.IdEvent.from_id"]], "from_list (ensembl.io.genomio.events.dump.stableidevent attribute)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.from_list"]], "get_full_release() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.get_full_release"]], "get_history() (ensembl.io.genomio.events.dump.dumpstableids method)": [[7, "ensembl.io.genomio.events.dump.DumpStableIDs.get_history"]], "get_mapping_sessions() (ensembl.io.genomio.events.dump.dumpstableids method)": [[7, "ensembl.io.genomio.events.dump.DumpStableIDs.get_mapping_sessions"]], "get_name() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.get_name"]], "get_pairs() (ensembl.io.genomio.events.dump.dumpstableids method)": [[7, "ensembl.io.genomio.events.dump.DumpStableIDs.get_pairs"]], "get_pairs_from_to() (ensembl.io.genomio.events.dump.dumpstableids static method)": [[7, "ensembl.io.genomio.events.dump.DumpStableIDs.get_pairs_from_to"]], "has_new_id() (ensembl.io.genomio.events.dump.pair method)": [[7, "ensembl.io.genomio.events.dump.Pair.has_new_id"]], "has_old_id() (ensembl.io.genomio.events.dump.pair method)": [[7, "ensembl.io.genomio.events.dump.Pair.has_old_id"]], "is_change() (ensembl.io.genomio.events.load.idevent method)": [[7, "ensembl.io.genomio.events.load.IdEvent.is_change"]], "is_empty() (ensembl.io.genomio.events.dump.pair method)": [[7, "ensembl.io.genomio.events.dump.Pair.is_empty"]], "load_events() (ensembl.io.genomio.events.load.eventcollection method)": [[7, "ensembl.io.genomio.events.load.EventCollection.load_events"]], "load_events_from_gene_diff() (ensembl.io.genomio.events.load.eventcollection method)": [[7, "ensembl.io.genomio.events.load.EventCollection.load_events_from_gene_diff"]], "load_list() (in module ensembl.io.genomio.events.format)": [[7, "ensembl.io.genomio.events.format.load_list"]], "main() (in module ensembl.io.genomio.events.dump)": [[7, "ensembl.io.genomio.events.dump.main"]], "main() (in module ensembl.io.genomio.events.load)": [[7, "ensembl.io.genomio.events.load.main"]], "make_events() (ensembl.io.genomio.events.dump.dumpstableids method)": [[7, "ensembl.io.genomio.events.dump.DumpStableIDs.make_events"]], "name (ensembl.io.genomio.events.dump.stableidevent attribute)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.name"]], "pairs (ensembl.io.genomio.events.dump.stableidevent attribute)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.pairs"]], "print_events() (ensembl.io.genomio.events.dump.dumpstableids method)": [[7, "ensembl.io.genomio.events.dump.DumpStableIDs.print_events"]], "release (ensembl.io.genomio.events.dump.stableidevent attribute)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.release"]], "release (ensembl.io.genomio.events.load.idevent attribute)": [[7, "ensembl.io.genomio.events.load.IdEvent.release"]], "release_date (ensembl.io.genomio.events.load.idevent attribute)": [[7, "ensembl.io.genomio.events.load.IdEvent.release_date"]], "remap_to_ids() (ensembl.io.genomio.events.load.eventcollection method)": [[7, "ensembl.io.genomio.events.load.EventCollection.remap_to_ids"]], "server (ensembl.io.genomio.events.dump.dumpstableids attribute)": [[7, "ensembl.io.genomio.events.dump.DumpStableIDs.server"]], "set_date() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.set_date"]], "set_release() (ensembl.io.genomio.events.dump.stableidevent method)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.set_release"]], "to_id (ensembl.io.genomio.events.load.idevent attribute)": [[7, "ensembl.io.genomio.events.load.IdEvent.to_id"]], "to_list (ensembl.io.genomio.events.dump.stableidevent attribute)": [[7, "ensembl.io.genomio.events.dump.StableIdEvent.to_list"]], "write_events_to_db() (ensembl.io.genomio.events.load.eventcollection method)": [[7, "ensembl.io.genomio.events.load.EventCollection.write_events_to_db"]], "write_events_to_file() (ensembl.io.genomio.events.load.eventcollection method)": [[7, "ensembl.io.genomio.events.load.EventCollection.write_events_to_file"]], "gffparsererror": [[8, "ensembl.io.genomio.fasta.process.GFFParserError"], [12, "ensembl.io.genomio.gff3.process.GFFParserError"]], "ensembl.io.genomio.fasta": [[8, "module-ensembl.io.genomio.fasta"]], "ensembl.io.genomio.fasta.process": [[8, "module-ensembl.io.genomio.fasta.process"]], "get_peptides_to_exclude() (in module ensembl.io.genomio.fasta.process)": [[8, "ensembl.io.genomio.fasta.process.get_peptides_to_exclude"]], "prep_fasta_data() (in module ensembl.io.genomio.fasta.process)": [[8, "ensembl.io.genomio.fasta.process.prep_fasta_data"]], "formattedfilesgenerator (class in ensembl.io.genomio.genbank.extract_data)": [[9, "ensembl.io.genomio.genbank.extract_data.FormattedFilesGenerator"]], "gbparseerror": [[9, "ensembl.io.genomio.genbank.extract_data.GBParseError"]], "genomefiles (class in ensembl.io.genomio.genbank.extract_data)": [[9, "ensembl.io.genomio.genbank.extract_data.GenomeFiles"]], "unsupporteddata": [[9, "ensembl.io.genomio.genbank.extract_data.UnsupportedData"]], "allowed_feat_types (ensembl.io.genomio.genbank.extract_data.formattedfilesgenerator attribute)": [[9, "ensembl.io.genomio.genbank.extract_data.FormattedFilesGenerator.allowed_feat_types"]], "download_genbank() (in module ensembl.io.genomio.genbank.download)": [[9, "ensembl.io.genomio.genbank.download.download_genbank"]], "ensembl.io.genomio.genbank": [[9, "module-ensembl.io.genomio.genbank"]], "ensembl.io.genomio.genbank.download": [[9, "module-ensembl.io.genomio.genbank.download"]], "ensembl.io.genomio.genbank.extract_data": [[9, "module-ensembl.io.genomio.genbank.extract_data"]], "extract_gb() (ensembl.io.genomio.genbank.extract_data.formattedfilesgenerator method)": [[9, "ensembl.io.genomio.genbank.extract_data.FormattedFilesGenerator.extract_gb"]], "locations (ensembl.io.genomio.genbank.extract_data.formattedfilesgenerator attribute)": [[9, "ensembl.io.genomio.genbank.extract_data.FormattedFilesGenerator.locations"]], "parse_genbank() (ensembl.io.genomio.genbank.extract_data.formattedfilesgenerator method)": [[9, "ensembl.io.genomio.genbank.extract_data.FormattedFilesGenerator.parse_genbank"]], "set_prefix() (ensembl.io.genomio.genbank.extract_data.formattedfilesgenerator method)": [[9, "ensembl.io.genomio.genbank.extract_data.FormattedFilesGenerator.set_prefix"]], "set_production_name() (ensembl.io.genomio.genbank.extract_data.formattedfilesgenerator method)": [[9, "ensembl.io.genomio.genbank.extract_data.FormattedFilesGenerator.set_production_name"]], "metadataerror": [[10, "ensembl.io.genomio.genome_metadata.prepare.MetadataError"]], "missingdataerror": [[10, "ensembl.io.genomio.genome_metadata.extend.MissingDataError"]], "missingnodeerror": [[10, "ensembl.io.genomio.genome_metadata.prepare.MissingNodeError"]], "add_assembly_version() (in module ensembl.io.genomio.genome_metadata.prepare)": [[10, "ensembl.io.genomio.genome_metadata.prepare.add_assembly_version"]], "add_genebuild_metadata() (in module ensembl.io.genomio.genome_metadata.prepare)": [[10, "ensembl.io.genomio.genome_metadata.prepare.add_genebuild_metadata"]], "add_provider() (in module ensembl.io.genomio.genome_metadata.prepare)": [[10, "ensembl.io.genomio.genome_metadata.prepare.add_provider"]], "add_species_metadata() (in module ensembl.io.genomio.genome_metadata.prepare)": [[10, "ensembl.io.genomio.genome_metadata.prepare.add_species_metadata"]], "amend_genomic_metadata() (in module ensembl.io.genomio.genome_metadata.extend)": [[10, "ensembl.io.genomio.genome_metadata.extend.amend_genomic_metadata"]], "check_assembly_version() (in module ensembl.io.genomio.genome_metadata.dump)": [[10, "ensembl.io.genomio.genome_metadata.dump.check_assembly_version"]], "ensembl.io.genomio.genome_metadata": [[10, "module-ensembl.io.genomio.genome_metadata"]], "ensembl.io.genomio.genome_metadata.dump": [[10, "module-ensembl.io.genomio.genome_metadata.dump"]], "ensembl.io.genomio.genome_metadata.extend": [[10, "module-ensembl.io.genomio.genome_metadata.extend"]], "ensembl.io.genomio.genome_metadata.prepare": [[10, "module-ensembl.io.genomio.genome_metadata.prepare"]], "filter_genome_meta() (in module ensembl.io.genomio.genome_metadata.dump)": [[10, "ensembl.io.genomio.genome_metadata.dump.filter_genome_meta"]], "get_additions() (in module ensembl.io.genomio.genome_metadata.extend)": [[10, "ensembl.io.genomio.genome_metadata.extend.get_additions"]], "get_gbff_regions() (in module ensembl.io.genomio.genome_metadata.extend)": [[10, "ensembl.io.genomio.genome_metadata.extend.get_gbff_regions"]], "get_genome_metadata() (in module ensembl.io.genomio.genome_metadata.dump)": [[10, "ensembl.io.genomio.genome_metadata.dump.get_genome_metadata"]], "get_report_regions_names() (in module ensembl.io.genomio.genome_metadata.extend)": [[10, "ensembl.io.genomio.genome_metadata.extend.get_report_regions_names"]], "get_taxonomy_from_accession() (in module ensembl.io.genomio.genome_metadata.prepare)": [[10, "ensembl.io.genomio.genome_metadata.prepare.get_taxonomy_from_accession"]], "prepare_genome_metadata() (in module ensembl.io.genomio.genome_metadata.prepare)": [[10, "ensembl.io.genomio.genome_metadata.prepare.prepare_genome_metadata"]], "statsgenerator (class in ensembl.io.genomio.genome_stats.dump)": [[11, "ensembl.io.genomio.genome_stats.dump.StatsGenerator"]], "compare_annotation() (in module ensembl.io.genomio.genome_stats.compare)": [[11, "ensembl.io.genomio.genome_stats.compare.compare_annotation"]], "compare_assembly() (in module ensembl.io.genomio.genome_stats.compare)": [[11, "ensembl.io.genomio.genome_stats.compare.compare_assembly"]], "compare_stats() (in module ensembl.io.genomio.genome_stats.compare)": [[11, "ensembl.io.genomio.genome_stats.compare.compare_stats"]], "ensembl.io.genomio.genome_stats": [[11, "module-ensembl.io.genomio.genome_stats"]], "ensembl.io.genomio.genome_stats.compare": [[11, "module-ensembl.io.genomio.genome_stats.compare"]], "ensembl.io.genomio.genome_stats.dump": [[11, "module-ensembl.io.genomio.genome_stats.dump"]], "get_annotation_stats() (ensembl.io.genomio.genome_stats.dump.statsgenerator method)": [[11, "ensembl.io.genomio.genome_stats.dump.StatsGenerator.get_annotation_stats"]], "get_assembly_stats() (ensembl.io.genomio.genome_stats.dump.statsgenerator method)": [[11, "ensembl.io.genomio.genome_stats.dump.StatsGenerator.get_assembly_stats"]], "get_attrib_counts() (ensembl.io.genomio.genome_stats.dump.statsgenerator method)": [[11, "ensembl.io.genomio.genome_stats.dump.StatsGenerator.get_attrib_counts"]], "get_biotypes() (ensembl.io.genomio.genome_stats.dump.statsgenerator method)": [[11, "ensembl.io.genomio.genome_stats.dump.StatsGenerator.get_biotypes"]], "get_feature_stats() (ensembl.io.genomio.genome_stats.dump.statsgenerator method)": [[11, "ensembl.io.genomio.genome_stats.dump.StatsGenerator.get_feature_stats"]], "get_stats() (ensembl.io.genomio.genome_stats.dump.statsgenerator method)": [[11, "ensembl.io.genomio.genome_stats.dump.StatsGenerator.get_stats"]], "annotationerror": [[12, "ensembl.io.genomio.gff3.extract_annotation.AnnotationError"]], "duplicateiderror": [[12, "ensembl.io.genomio.gff3.extract_annotation.DuplicateIdError"]], "functionalannotations (class in ensembl.io.genomio.gff3.extract_annotation)": [[12, "ensembl.io.genomio.gff3.extract_annotation.FunctionalAnnotations"]], "gffgenemerger (class in ensembl.io.genomio.gff3.process)": [[12, "ensembl.io.genomio.gff3.process.GFFGeneMerger"]], "gffsimplifier (class in ensembl.io.genomio.gff3.process)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier"]], "missingparenterror": [[12, "ensembl.io.genomio.gff3.extract_annotation.MissingParentError"]], "records (class in ensembl.io.genomio.gff3.process)": [[12, "ensembl.io.genomio.gff3.process.Records"]], "add_feature() (ensembl.io.genomio.gff3.extract_annotation.functionalannotations method)": [[12, "ensembl.io.genomio.gff3.extract_annotation.FunctionalAnnotations.add_feature"]], "add_parent() (ensembl.io.genomio.gff3.extract_annotation.functionalannotations method)": [[12, "ensembl.io.genomio.gff3.extract_annotation.FunctionalAnnotations.add_parent"]], "allow_pseudogene_with_cds (ensembl.io.genomio.gff3.process.gffsimplifier attribute)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.allow_pseudogene_with_CDS"]], "build_transcript() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.build_transcript"]], "cds_gene() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.cds_gene"]], "clean_extra_exons() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.clean_extra_exons"]], "current_stable_id_number (ensembl.io.genomio.gff3.process.gffsimplifier attribute)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.current_stable_id_number"]], "ensembl.io.genomio.gff3": [[12, "module-ensembl.io.genomio.gff3"]], "ensembl.io.genomio.gff3.extract_annotation": [[12, "module-ensembl.io.genomio.gff3.extract_annotation"]], "ensembl.io.genomio.gff3.process": [[12, "module-ensembl.io.genomio.gff3.process"]], "exclude_seq_regions (ensembl.io.genomio.gff3.process.gffsimplifier attribute)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.exclude_seq_regions"]], "format_gene_segments() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.format_gene_segments"]], "format_mobile_element() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.format_mobile_element"]], "gene_cds_skip_others (ensembl.io.genomio.gff3.process.gffsimplifier attribute)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.gene_cds_skip_others"]], "gene_to_cds() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.gene_to_cds"]], "gene_to_exon() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.gene_to_exon"]], "generate_stable_id() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.generate_stable_id"]], "get_features() (ensembl.io.genomio.gff3.extract_annotation.functionalannotations method)": [[12, "ensembl.io.genomio.gff3.extract_annotation.FunctionalAnnotations.get_features"]], "get_parent() (ensembl.io.genomio.gff3.extract_annotation.functionalannotations method)": [[12, "ensembl.io.genomio.gff3.extract_annotation.FunctionalAnnotations.get_parent"]], "merge() (ensembl.io.genomio.gff3.process.gffgenemerger method)": [[12, "ensembl.io.genomio.gff3.process.GFFGeneMerger.merge"]], "min_id_length (ensembl.io.genomio.gff3.process.gffsimplifier attribute)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.min_id_length"]], "move_cds_to_mrna() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.move_cds_to_mrna"]], "normalize_cds_id() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_cds_id"]], "normalize_gene() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_gene"]], "normalize_gene_id() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_gene_id"]], "normalize_pseudogene_cds() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_pseudogene_cds"]], "normalize_transcript_id() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_transcript_id"]], "product_is_informative() (ensembl.io.genomio.gff3.extract_annotation.functionalannotations static method)": [[12, "ensembl.io.genomio.gff3.extract_annotation.FunctionalAnnotations.product_is_informative"]], "remove_cds_from_pseudogene() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.remove_cds_from_pseudogene"]], "remove_prefixes() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.remove_prefixes"]], "simpler_gff3() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.simpler_gff3"]], "skip_unrecognized (ensembl.io.genomio.gff3.process.gffsimplifier attribute)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.skip_unrecognized"]], "stable_id_prefix (ensembl.io.genomio.gff3.process.gffsimplifier attribute)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.stable_id_prefix"]], "to_gff() (ensembl.io.genomio.gff3.process.records method)": [[12, "ensembl.io.genomio.gff3.process.Records.to_gff"]], "to_json() (ensembl.io.genomio.gff3.extract_annotation.functionalannotations method)": [[12, "ensembl.io.genomio.gff3.extract_annotation.FunctionalAnnotations.to_json"]], "transcript_for_gene() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.transcript_for_gene"]], "transcript_gene() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.transcript_gene"]], "valid_id() (ensembl.io.genomio.gff3.process.gffsimplifier method)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.valid_id"]], "validate_gene_id (ensembl.io.genomio.gff3.process.gffsimplifier attribute)": [[12, "ensembl.io.genomio.gff3.process.GFFSimplifier.validate_gene_id"]], "biotypecounter (class in ensembl.io.genomio.manifest.compute_stats)": [[13, "ensembl.io.genomio.manifest.compute_stats.BiotypeCounter"]], "integritytool (class in ensembl.io.genomio.manifest.check_integrity)": [[13, "ensembl.io.genomio.manifest.check_integrity.IntegrityTool"]], "invalidintegrityerror": [[13, "ensembl.io.genomio.manifest.check_integrity.InvalidIntegrityError"]], "manifest (class in ensembl.io.genomio.manifest.check_integrity)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest"]], "manifestmaker (class in ensembl.io.genomio.manifest.generate)": [[13, "ensembl.io.genomio.manifest.generate.ManifestMaker"]], "statserror": [[13, "ensembl.io.genomio.manifest.compute_stats.StatsError"]], "add_errors() (ensembl.io.genomio.manifest.check_integrity.integritytool method)": [[13, "ensembl.io.genomio.manifest.check_integrity.IntegrityTool.add_errors"]], "add_id() (ensembl.io.genomio.manifest.compute_stats.biotypecounter method)": [[13, "ensembl.io.genomio.manifest.compute_stats.BiotypeCounter.add_id"]], "alias_names (ensembl.io.genomio.manifest.generate.manifestmaker attribute)": [[13, "ensembl.io.genomio.manifest.generate.ManifestMaker.alias_names"]], "biotypes_stats() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.biotypes_stats"]], "check_ids() (ensembl.io.genomio.manifest.check_integrity.integritytool method)": [[13, "ensembl.io.genomio.manifest.check_integrity.IntegrityTool.check_ids"]], "check_integrity() (ensembl.io.genomio.manifest.check_integrity.integritytool method)": [[13, "ensembl.io.genomio.manifest.check_integrity.IntegrityTool.check_integrity"]], "check_lengths() (ensembl.io.genomio.manifest.check_integrity.integritytool method)": [[13, "ensembl.io.genomio.manifest.check_integrity.IntegrityTool.check_lengths"]], "check_ncbi_stats() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.check_ncbi_stats"]], "check_seq_region_lengths() (ensembl.io.genomio.manifest.check_integrity.integritytool method)": [[13, "ensembl.io.genomio.manifest.check_integrity.IntegrityTool.check_seq_region_lengths"]], "compare_ncbi_counts() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.compare_ncbi_counts"]], "coord_systems_stats() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.coord_systems_stats"]], "count_biotypes() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.count_biotypes"]], "create_manifest() (ensembl.io.genomio.manifest.generate.manifestmaker method)": [[13, "ensembl.io.genomio.manifest.generate.ManifestMaker.create_manifest"]], "ensembl.io.genomio.manifest": [[13, "module-ensembl.io.genomio.manifest"]], "ensembl.io.genomio.manifest.check_integrity": [[13, "module-ensembl.io.genomio.manifest.check_integrity"]], "ensembl.io.genomio.manifest.compute_stats": [[13, "module-ensembl.io.genomio.manifest.compute_stats"]], "ensembl.io.genomio.manifest.generate": [[13, "module-ensembl.io.genomio.manifest.generate"]], "get_agp_seq_regions() (ensembl.io.genomio.manifest.check_integrity.manifest method)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest.get_agp_seq_regions"]], "get_circular() (ensembl.io.genomio.manifest.check_integrity.manifest method)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest.get_circular"]], "get_fasta_lengths() (ensembl.io.genomio.manifest.check_integrity.manifest method)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest.get_fasta_lengths"]], "get_files_checksums() (ensembl.io.genomio.manifest.generate.manifestmaker method)": [[13, "ensembl.io.genomio.manifest.generate.ManifestMaker.get_files_checksums"]], "get_functional_annotation() (ensembl.io.genomio.manifest.check_integrity.manifest method)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest.get_functional_annotation"]], "get_gff3() (ensembl.io.genomio.manifest.check_integrity.manifest method)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest.get_gff3"]], "get_gff3_stats() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.get_gff3_stats"]], "get_lengths() (ensembl.io.genomio.manifest.check_integrity.manifest method)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest.get_lengths"]], "get_manifest() (ensembl.io.genomio.manifest.check_integrity.manifest method)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest.get_manifest"]], "get_manifest() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.get_manifest"]], "get_seq_region_stats() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.get_seq_region_stats"]], "has_lengths() (ensembl.io.genomio.manifest.check_integrity.manifest method)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest.has_lengths"]], "increment_biotype() (ensembl.io.genomio.manifest.compute_stats.manifest_stats static method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.increment_biotype"]], "main() (in module ensembl.io.genomio.manifest.check_integrity)": [[13, "ensembl.io.genomio.manifest.check_integrity.main"]], "main() (in module ensembl.io.genomio.manifest.compute_stats)": [[13, "ensembl.io.genomio.manifest.compute_stats.main"]], "main() (in module ensembl.io.genomio.manifest.generate)": [[13, "ensembl.io.genomio.manifest.generate.main"]], "manifest_stats (class in ensembl.io.genomio.manifest.compute_stats)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats"]], "names (ensembl.io.genomio.manifest.generate.manifestmaker attribute)": [[13, "ensembl.io.genomio.manifest.generate.ManifestMaker.names"]], "prepare_integrity_data() (ensembl.io.genomio.manifest.check_integrity.manifest method)": [[13, "ensembl.io.genomio.manifest.check_integrity.Manifest.prepare_integrity_data"]], "run() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.run"]], "same_names (ensembl.io.genomio.manifest.generate.manifestmaker attribute)": [[13, "ensembl.io.genomio.manifest.generate.ManifestMaker.same_names"]], "seq_region_special_stats() (ensembl.io.genomio.manifest.compute_stats.manifest_stats method)": [[13, "ensembl.io.genomio.manifest.compute_stats.manifest_stats.seq_region_special_stats"]], "set_brc_mode() (ensembl.io.genomio.manifest.check_integrity.integritytool method)": [[13, "ensembl.io.genomio.manifest.check_integrity.IntegrityTool.set_brc_mode"]], "set_ignore_final_stops() (ensembl.io.genomio.manifest.check_integrity.integritytool method)": [[13, "ensembl.io.genomio.manifest.check_integrity.IntegrityTool.set_ignore_final_stops"]], "unique_count() (ensembl.io.genomio.manifest.compute_stats.biotypecounter method)": [[13, "ensembl.io.genomio.manifest.compute_stats.BiotypeCounter.unique_count"]], "ensembl.io.genomio.schemas": [[14, "module-ensembl.io.genomio.schemas"]], "ensembl.io.genomio.schemas.json": [[15, "module-ensembl.io.genomio.schemas.json"]], "ensembl.io.genomio.schemas.json.factory": [[15, "module-ensembl.io.genomio.schemas.json.factory"]], "ensembl.io.genomio.schemas.json.validate": [[15, "module-ensembl.io.genomio.schemas.json.validate"]], "json_schema_factory() (in module ensembl.io.genomio.schemas.json.factory)": [[15, "ensembl.io.genomio.schemas.json.factory.json_schema_factory"]], "json_schema_validator() (in module ensembl.io.genomio.schemas.json.validate)": [[15, "ensembl.io.genomio.schemas.json.validate.json_schema_validator"]], "mapformaterror": [[16, "ensembl.io.genomio.seq_region.dump.MapFormatError"]], "unknownmetadata": [[16, "ensembl.io.genomio.seq_region.prepare.UnknownMetadata"]], "add_attribs() (in module ensembl.io.genomio.seq_region.dump)": [[16, "ensembl.io.genomio.seq_region.dump.add_attribs"]], "add_insdc_seq_region_name() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.add_insdc_seq_region_name"]], "add_mitochondrial_codon_table() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.add_mitochondrial_codon_table"]], "add_translation_table() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.add_translation_table"]], "ensembl.io.genomio.seq_region": [[16, "module-ensembl.io.genomio.seq_region"]], "ensembl.io.genomio.seq_region.dump": [[16, "module-ensembl.io.genomio.seq_region.dump"]], "ensembl.io.genomio.seq_region.prepare": [[16, "module-ensembl.io.genomio.seq_region.prepare"]], "exclude_seq_regions() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.exclude_seq_regions"]], "get_attribs() (in module ensembl.io.genomio.seq_region.dump)": [[16, "ensembl.io.genomio.seq_region.dump.get_attribs"]], "get_codon_table() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.get_codon_table"]], "get_coord_systems() (in module ensembl.io.genomio.seq_region.dump)": [[16, "ensembl.io.genomio.seq_region.dump.get_coord_systems"]], "get_external_db_map() (in module ensembl.io.genomio.seq_region.dump)": [[16, "ensembl.io.genomio.seq_region.dump.get_external_db_map"]], "get_gbff_seq_regions() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.get_gbff_seq_regions"]], "get_genbank_id() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.get_genbank_id"]], "get_karyotype() (in module ensembl.io.genomio.seq_region.dump)": [[16, "ensembl.io.genomio.seq_region.dump.get_karyotype"]], "get_organelle() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.get_organelle"]], "get_report_regions() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.get_report_regions"]], "get_seq_regions() (in module ensembl.io.genomio.seq_region.dump)": [[16, "ensembl.io.genomio.seq_region.dump.get_seq_regions"]], "get_synonyms() (in module ensembl.io.genomio.seq_region.dump)": [[16, "ensembl.io.genomio.seq_region.dump.get_synonyms"]], "main() (in module ensembl.io.genomio.seq_region.dump)": [[16, "ensembl.io.genomio.seq_region.dump.main"]], "make_seq_region() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.make_seq_region"]], "merge_seq_regions() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.merge_seq_regions"]], "prepare_seq_region_metadata() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.prepare_seq_region_metadata"]], "report_to_csv() (in module ensembl.io.genomio.seq_region.prepare)": [[16, "ensembl.io.genomio.seq_region.prepare.report_to_csv"]], "ensembl.io.genomio.utils": [[17, "module-ensembl.io.genomio.utils"]], "ensembl.io.genomio.utils.archive_utils": [[17, "module-ensembl.io.genomio.utils.archive_utils"]], "ensembl.io.genomio.utils.json_utils": [[17, "module-ensembl.io.genomio.utils.json_utils"]], "extract_file() (in module ensembl.io.genomio.utils.archive_utils)": [[17, "ensembl.io.genomio.utils.archive_utils.extract_file"]], "get_json() (in module ensembl.io.genomio.utils.json_utils)": [[17, "ensembl.io.genomio.utils.json_utils.get_json"]], "open_gz_file() (in module ensembl.io.genomio.utils.archive_utils)": [[17, "ensembl.io.genomio.utils.archive_utils.open_gz_file"]], "print_json() (in module ensembl.io.genomio.utils.json_utils)": [[17, "ensembl.io.genomio.utils.json_utils.print_json"]]}})