
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genomio/reference/genomio/gff3/">
      
      
        <link rel="prev" href="../genome_stats/">
      
      
        <link rel="next" href="../manifest/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.8">
    
    
      
        <title>gff3 - Ensembl-GenomIO</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.f2e4d321.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ensembl.io.genomio.gff3" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Ensembl-GenomIO" class="md-header__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../img/ebang.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-GenomIO
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              gff3
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M272 384c9.6-31.9 29.5-59.1 49.2-86.2 5.2-7.1 10.4-14.2 15.4-21.4 19.8-28.5 31.4-63 31.4-100.3C368 78.8 289.2 0 192 0S16 78.8 16 176c0 37.3 11.6 71.9 31.4 100.3 5 7.2 10.2 14.3 15.4 21.4 19.8 27.1 39.7 54.4 49.2 86.2h160zm-80 128c44.2 0 80-35.8 80-80v-16H112v16c0 44.2 35.8 80 80 80zm-80-336c0 8.8-7.2 16-16 16s-16-7.2-16-16c0-61.9 50.1-112 112-112 8.8 0 16 7.2 16 16s-7.2 16-16 16c-44.2 0-80 35.8-80 80z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M297.2 248.9c14.4-20.6 22.8-45.7 22.8-72.9 0-70.7-57.3-128-128-128S64 105.3 64 176c0 27.2 8.4 52.3 22.8 72.9 3.7 5.3 8.1 11.3 12.8 17.7 12.9 17.7 28.3 38.9 39.8 59.8 10.4 19 15.7 38.8 18.3 57.5H109c-2.2-12-5.9-23.7-11.8-34.5-9.9-18-22.2-34.9-34.5-51.8-5.2-7.1-10.4-14.2-15.4-21.4C27.6 247.9 16 213.3 16 176 16 78.8 94.8 0 192 0s176 78.8 176 176c0 37.3-11.6 71.9-31.4 100.3-5 7.2-10.2 14.3-15.4 21.4-12.3 16.8-24.6 33.7-34.5 51.8-5.9 10.8-9.6 22.5-11.8 34.5h-48.5c2.6-18.7 7.9-38.6 18.3-57.5 11.5-20.9 26.9-42.1 39.8-59.8 4.7-6.4 9-12.4 12.7-17.7zM192 128c-26.5 0-48 21.5-48 48 0 8.8-7.2 16-16 16s-16-7.2-16-16c0-44.2 35.8-80 80-80 8.8 0 16 7.2 16 16s-7.2 16-16 16zm0 384c-44.2 0-80-35.8-80-80v-16h160v16c0 44.2-35.8 80-80 80z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl-GenomIO
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Ensembl-GenomIO" class="md-nav__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../img/ebang.png" alt="logo">

    </a>
    Ensembl-GenomIO
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl-GenomIO
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ensembl
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    brc4
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            brc4
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../runnable/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    runnable
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_1" id="__nav_4_1_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            runnable
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/compare_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_fasta.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/compare_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_report.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/core_server/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    core_server.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/download_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download_genbank.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/dump_stable_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_stable_ids.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/extract_from_gb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_gb.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/fill_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fill_metadata.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/gff3_specifier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3_specifier.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/load_sequence_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load_sequence_data.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/read_json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    read_json.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/say_accession/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    say_accession.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/seqregion_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seqregion_parser.py
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    io
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            io
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_1" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genomio
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_1" id="__nav_4_1_2_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            genomio
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    database
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../events/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fasta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genome_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    gff3
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    gff3
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;gff3
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" gff3">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.Annotation" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;Annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.AnnotationError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;AnnotationError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;ArgumentParser
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" ArgumentParser">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;__init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument_dst_path" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_dst_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument_src_path" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_src_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument_url" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_url
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_log_arguments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_log_arguments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_server_arguments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_server_arguments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.DuplicateIdError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;DuplicateIdError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;FunctionalAnnotations
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" FunctionalAnnotations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.add_feature" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_feature
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.add_parent_link" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_parent_link
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.get_features" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_features
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.get_parent" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_parent
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.product_is_informative" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;product_is_informative
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.to_json" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;to_json
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.transfer_descriptions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;transfer_descriptions
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFGeneMerger" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;GFFGeneMerger
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" GFFGeneMerger">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFGeneMerger.merge" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;merge
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFParserError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;GFFParserError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;GFFSimplifier
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" GFFSimplifier">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.build_transcript" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;build_transcript
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.cds_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;cds_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.clean_extra_exons" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;clean_extra_exons
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.clean_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;clean_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.format_gene_segments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;format_gene_segments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.format_mobile_element" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;format_mobile_element
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.gene_to_cds" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;gene_to_cds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.gene_to_exon" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;gene_to_exon
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.generate_stable_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;generate_stable_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.move_cds_to_mrna" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;move_cds_to_mrna
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_cds_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_cds_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_gene_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_gene_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_pseudogene_cds" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_pseudogene_cds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_transcript_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_transcript_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.remove_cds_from_pseudogene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;remove_cds_from_pseudogene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.remove_prefixes" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;remove_prefixes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.simpler_gff3" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;simpler_gff3
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.simpler_gff3_feature" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;simpler_gff3_feature
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.store_gene_annotations" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;store_gene_annotations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.transcript_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;transcript_for_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.transcript_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;transcript_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.valid_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;valid_id
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.MissingParentError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MissingParentError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.Records" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;Records
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" Records">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.Records.to_gff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;to_gff
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.init_logging_with_args" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;init_logging_with_args
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.main" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_1_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../schemas/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    schemas
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_1_10" id="__nav_4_1_2_1_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_2_1_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2_1_10">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seq_region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Nextflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;gff3
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" gff3">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.Annotation" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;Annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.AnnotationError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;AnnotationError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;ArgumentParser
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" ArgumentParser">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;__init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument_dst_path" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_dst_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument_src_path" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_src_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument_url" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_url
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_log_arguments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_log_arguments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.add_server_arguments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_server_arguments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.ArgumentParser.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.DuplicateIdError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;DuplicateIdError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;FunctionalAnnotations
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" FunctionalAnnotations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.add_feature" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_feature
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.add_parent_link" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_parent_link
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.get_features" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_features
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.get_parent" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_parent
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.product_is_informative" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;product_is_informative
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.to_json" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;to_json
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.transfer_descriptions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;transfer_descriptions
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFGeneMerger" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;GFFGeneMerger
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" GFFGeneMerger">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFGeneMerger.merge" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;merge
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFParserError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;GFFParserError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;GFFSimplifier
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" GFFSimplifier">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.build_transcript" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;build_transcript
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.cds_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;cds_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.clean_extra_exons" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;clean_extra_exons
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.clean_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;clean_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.format_gene_segments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;format_gene_segments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.format_mobile_element" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;format_mobile_element
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.gene_to_cds" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;gene_to_cds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.gene_to_exon" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;gene_to_exon
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.generate_stable_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;generate_stable_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.move_cds_to_mrna" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;move_cds_to_mrna
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_cds_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_cds_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_gene_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_gene_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_pseudogene_cds" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_pseudogene_cds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_transcript_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;normalize_transcript_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.remove_cds_from_pseudogene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;remove_cds_from_pseudogene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.remove_prefixes" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;remove_prefixes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.simpler_gff3" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;simpler_gff3
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.simpler_gff3_feature" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;simpler_gff3_feature
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.store_gene_annotations" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;store_gene_annotations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.transcript_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;transcript_for_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.transcript_gene" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;transcript_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.GFFSimplifier.valid_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;valid_id
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.MissingParentError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MissingParentError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.Records" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;Records
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" Records">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.Records.to_gff" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;to_gff
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.init_logging_with_args" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;init_logging_with_args
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.gff3.main" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>gff3</h1>

<div class="doc doc-object doc-module">



<h2 id="ensembl.io.genomio.gff3" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>          <code>ensembl.io.genomio.gff3</code>


<a href="#ensembl.io.genomio.gff3" class="headerlink" title="Permanent link">&para;</a></h2>

  <div class="doc doc-contents first">
  
      <p>GFF3 files processing module.</p>

  

  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.gff3.Annotation" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>          <code class="highlight language-python"><span class="n">Annotation</span> <span class="o">=</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.Annotation" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  </div>

</div>


<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.AnnotationError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>AnnotationError</code>


<a href="#ensembl.io.genomio.gff3.AnnotationError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>If anything wrong happens when recording annotations.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">51</span>
<span class="normal">52</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">AnnotationError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;If anything wrong happens when recording annotations.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.ArgumentParser" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>ArgumentParser</code>


<a href="#ensembl.io.genomio.gff3.ArgumentParser" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><span title="argparse.ArgumentParser">ArgumentParser</span></code></p>

  
      <p>Extends :class:<code>argparse.ArgumentParser</code> with additional methods and functionality.</p>
<p>The default behaviour of the help text will be to display the default values on every non-required
argument, i.e. optional arguments with "required=False".</p>

            <details class="quote">
              <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">ArgumentParser</span><span class="p">(</span><span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extends :class:`argparse.ArgumentParser` with additional methods and functionality.</span>

<span class="sd">    The default behaviour of the help text will be to display the default values on every non-required</span>
<span class="sd">    argument, i.e. optional arguments with &quot;required=False&quot;.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extends the base class to include the information about default argument values by default.&quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">formatter_class</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentDefaultsHelpFormatter</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="nf">_validate_src_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">src_path</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the path if exists and it is readable, raises an error through the parser otherwise.</span>

<span class="sd">        Args:</span>
<span class="sd">            src_path: File or directory path to check.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">src_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">src_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">src_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">src_path</span><span class="si">}</span><span class="s2">&#39; not found&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">src_path</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">R_OK</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">src_path</span><span class="si">}</span><span class="s2">&#39; not readable&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">src_path</span>

    <span class="k">def</span> <span class="nf">_validate_dst_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dst_path</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">PathLike</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the path if it is writable, raises an error through the parser otherwise.</span>

<span class="sd">        Args:</span>
<span class="sd">            dst_path: File or directory path to check.</span>
<span class="sd">            exists_ok: Do not raise an error if the destination path already exists.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">dst_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">dst_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">dst_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">dst_path</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">W_OK</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">exists_ok</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">dst_path</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">dst_path</span><span class="si">}</span><span class="s2">&#39; already exists&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">dst_path</span><span class="si">}</span><span class="s2">&#39; is not writable&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">parent_path</span> <span class="ow">in</span> <span class="n">dst_path</span><span class="o">.</span><span class="n">parents</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">parent_path</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">parent_path</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">W_OK</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">dst_path</span><span class="si">}</span><span class="s2">&#39; is not writable&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">dst_path</span>

    <span class="k">def</span> <span class="nf">add_argument</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extends the parent function by excluding the default value in the help text when not provided.</span>

<span class="sd">        Only applied to required arguments without a default value, i.e. positional arguments or optional</span>
<span class="sd">        arguments with &quot;required=True&quot;.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;required&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
            <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;default&quot;</span><span class="p">,</span> <span class="n">argparse</span><span class="o">.</span><span class="n">SUPPRESS</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_argument_src_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds :class:`pathlib.Path` argument, checking if it exists and it is readable at parsing time.</span>

<span class="sd">        If &quot;metavar&quot; is not defined it is added with &quot;PATH&quot; as value to improve help text readability.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;PATH&quot;</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_src_path</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_argument_dst_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds :class:`pathlib.Path` argument, checking if it is writable at parsing time.</span>

<span class="sd">        If &quot;metavar&quot; is not defined it is added with &quot;PATH&quot; as value to improve help text readability.</span>

<span class="sd">        Args:</span>
<span class="sd">            exists_ok: Do not raise an error if the destination path already exists.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;PATH&quot;</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dst_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_argument_url</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds :class:`sqlalchemy.engine.URL` argument.</span>

<span class="sd">        If &quot;metavar&quot; is not defined it is added with &quot;URI&quot; as value to improve help text readability.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;URI&quot;</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">make_url</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_server_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">include_database</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds the usual set of arguments needed to connect to a server.</span>

<span class="sd">        This means: &quot;--host&quot;, &quot;--port&quot;, &quot;--user&quot; and &quot;--password&quot; (optional).</span>

<span class="sd">        Args:</span>
<span class="sd">            prefix: Prefix to add the each argument, e.g. if prefix is &quot;src_&quot;, the arguments will be</span>
<span class="sd">                &quot;--src_host&quot;, etc.</span>
<span class="sd">            include_database: Include &quot;--database&quot; argument.</span>
<span class="sd">            help: Description message to include for this set of arguments.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">group</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">server connection arguments&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="n">help</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">host&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;HOST&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;host name&quot;</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">port&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PORT&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;port number&quot;</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">user&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;USER&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;user name&quot;</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">password&quot;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PWD&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;host password&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">include_database</span><span class="p">:</span>
            <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">database&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;NAME&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;database name&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_log_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">add_log_file</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds the usual set of arguments required to set and initialise a logging system.</span>

<span class="sd">        The current set includes a mutually exclusive group for the default logging level: &quot;--verbose&quot;,</span>
<span class="sd">        &quot;--debug&quot; or &quot;--log LEVEL&quot;.</span>

<span class="sd">        Args:</span>
<span class="sd">            add_log_file: Add arguments to allow storing messages into a file, i.e. &quot;--log_file&quot; and</span>
<span class="sd">                &quot;--log_file_level&quot;.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Define the list of log levels available</span>
        <span class="n">log_levels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span> <span class="s2">&quot;INFO&quot;</span><span class="p">,</span> <span class="s2">&quot;WARNING&quot;</span><span class="p">,</span> <span class="s2">&quot;ERROR&quot;</span><span class="p">,</span> <span class="s2">&quot;CRITICAL&quot;</span><span class="p">]</span>
        <span class="c1"># NOTE: from 3.11 this list can be changed to: logging.getLevelNamesMapping().keys()</span>
        <span class="c1"># Create logging arguments group</span>
        <span class="n">group</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;logging arguments&quot;</span><span class="p">)</span>
        <span class="c1"># Add 3 mutually exclusive options to set the logging level</span>
        <span class="n">subgroup</span> <span class="o">=</span> <span class="n">group</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">()</span>
        <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;-v&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--verbose&quot;</span><span class="p">,</span>
            <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
            <span class="n">const</span><span class="o">=</span><span class="s2">&quot;INFO&quot;</span><span class="p">,</span>
            <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;verbose mode, i.e. &#39;INFO&#39; log level&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--debug&quot;</span><span class="p">,</span>
            <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
            <span class="n">const</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
            <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;debugging mode, i.e. &#39;DEBUG&#39; log level&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--log&quot;</span><span class="p">,</span>
            <span class="n">choices</span><span class="o">=</span><span class="n">log_levels</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">,</span>
            <span class="n">default</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
            <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;LEVEL&quot;</span><span class="p">,</span>
            <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;level of the events to track: </span><span class="si">%(choices)s</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">subgroup</span><span class="o">.</span><span class="n">set_defaults</span><span class="p">(</span><span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">add_log_file</span><span class="p">:</span>
            <span class="c1"># Add log file-related arguments</span>
            <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
                <span class="s2">&quot;--log_file&quot;</span><span class="p">,</span>
                <span class="nb">type</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dst_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">exists_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PATH&quot;</span><span class="p">,</span>
                <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">help</span><span class="o">=</span><span class="s2">&quot;log file path&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
                <span class="s2">&quot;--log_file_level&quot;</span><span class="p">,</span>
                <span class="n">choices</span><span class="o">=</span><span class="n">log_levels</span><span class="p">,</span>
                <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">,</span>
                <span class="n">default</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
                <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;LEVEL&quot;</span><span class="p">,</span>
                <span class="n">help</span><span class="o">=</span><span class="s2">&quot;level of the events to track in the log file: </span><span class="si">%(choices)s</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">parse_args</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">argparse</span><span class="o">.</span><span class="n">Namespace</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extends the parent function by adding a new URL argument for every server group added.</span>

<span class="sd">        The type of this new argument will be :class:`sqlalchemy.engine.URL`. It also logs all the parsed</span>
<span class="sd">        arguments for debugging purposes when logging arguments have been added.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">args</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="c1"># Build and add an sqlalchemy.engine.URL object for every server group added</span>
        <span class="k">for</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span><span class="p">:</span>
            <span class="c1"># Raise an error rather than overwriting when the URL argument is already present</span>
            <span class="k">if</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&quot;</span> <span class="ow">in</span> <span class="n">args</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;argument &#39;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&#39; is already present&quot;</span><span class="p">)</span>
            <span class="n">server_url</span> <span class="o">=</span> <span class="n">URL</span><span class="o">.</span><span class="n">create</span><span class="p">(</span>
                <span class="s2">&quot;mysql&quot;</span><span class="p">,</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">user&quot;</span><span class="p">),</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">password&quot;</span><span class="p">),</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">host&quot;</span><span class="p">),</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">port&quot;</span><span class="p">),</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">database&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&quot;</span><span class="p">,</span> <span class="n">server_url</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">args</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.ArgumentParser.__init__" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.ArgumentParser.__init__" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Extends the base class to include the information about default argument values by default.</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extends the base class to include the information about default argument values by default.&quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">formatter_class</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentDefaultsHelpFormatter</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span> <span class="o">=</span> <span class="p">[]</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.ArgumentParser.add_argument" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Extends the parent function by excluding the default value in the help text when not provided.</p>
<p>Only applied to required arguments without a default value, i.e. positional arguments or optional
arguments with "required=True".</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_argument</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extends the parent function by excluding the default value in the help text when not provided.</span>

<span class="sd">    Only applied to required arguments without a default value, i.e. positional arguments or optional</span>
<span class="sd">    arguments with &quot;required=True&quot;.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;required&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;default&quot;</span><span class="p">,</span> <span class="n">argparse</span><span class="o">.</span><span class="n">SUPPRESS</span><span class="p">)</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.ArgumentParser.add_argument_dst_path" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_argument_dst_path</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">exists_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument_dst_path" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds :class:<code>pathlib.Path</code> argument, checking if it is writable at parsing time.</p>
<p>If "metavar" is not defined it is added with "PATH" as value to improve help text readability.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>exists_ok</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Do not raise an error if the destination path already exists.</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_argument_dst_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds :class:`pathlib.Path` argument, checking if it is writable at parsing time.</span>

<span class="sd">    If &quot;metavar&quot; is not defined it is added with &quot;PATH&quot; as value to improve help text readability.</span>

<span class="sd">    Args:</span>
<span class="sd">        exists_ok: Do not raise an error if the destination path already exists.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;PATH&quot;</span><span class="p">)</span>
    <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dst_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.ArgumentParser.add_argument_src_path" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_argument_src_path</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument_src_path" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds :class:<code>pathlib.Path</code> argument, checking if it exists and it is readable at parsing time.</p>
<p>If "metavar" is not defined it is added with "PATH" as value to improve help text readability.</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_argument_src_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds :class:`pathlib.Path` argument, checking if it exists and it is readable at parsing time.</span>

<span class="sd">    If &quot;metavar&quot; is not defined it is added with &quot;PATH&quot; as value to improve help text readability.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;PATH&quot;</span><span class="p">)</span>
    <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_src_path</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.ArgumentParser.add_argument_url" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_argument_url</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.ArgumentParser.add_argument_url" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds :class:<code>sqlalchemy.engine.URL</code> argument.</p>
<p>If "metavar" is not defined it is added with "URI" as value to improve help text readability.</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_argument_url</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds :class:`sqlalchemy.engine.URL` argument.</span>

<span class="sd">    If &quot;metavar&quot; is not defined it is added with &quot;URI&quot; as value to improve help text readability.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;URI&quot;</span><span class="p">)</span>
    <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">make_url</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.ArgumentParser.add_log_arguments" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_log_arguments</span><span class="p">(</span><span class="n">add_log_file</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.ArgumentParser.add_log_arguments" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds the usual set of arguments required to set and initialise a logging system.</p>
<p>The current set includes a mutually exclusive group for the default logging level: "--verbose",
"--debug" or "--log LEVEL".</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>add_log_file</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Add arguments to allow storing messages into a file, i.e. "--log_file" and
"--log_file_level".</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_log_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">add_log_file</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the usual set of arguments required to set and initialise a logging system.</span>

<span class="sd">    The current set includes a mutually exclusive group for the default logging level: &quot;--verbose&quot;,</span>
<span class="sd">    &quot;--debug&quot; or &quot;--log LEVEL&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        add_log_file: Add arguments to allow storing messages into a file, i.e. &quot;--log_file&quot; and</span>
<span class="sd">            &quot;--log_file_level&quot;.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define the list of log levels available</span>
    <span class="n">log_levels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span> <span class="s2">&quot;INFO&quot;</span><span class="p">,</span> <span class="s2">&quot;WARNING&quot;</span><span class="p">,</span> <span class="s2">&quot;ERROR&quot;</span><span class="p">,</span> <span class="s2">&quot;CRITICAL&quot;</span><span class="p">]</span>
    <span class="c1"># NOTE: from 3.11 this list can be changed to: logging.getLevelNamesMapping().keys()</span>
    <span class="c1"># Create logging arguments group</span>
    <span class="n">group</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;logging arguments&quot;</span><span class="p">)</span>
    <span class="c1"># Add 3 mutually exclusive options to set the logging level</span>
    <span class="n">subgroup</span> <span class="o">=</span> <span class="n">group</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">()</span>
    <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-v&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--verbose&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
        <span class="n">const</span><span class="o">=</span><span class="s2">&quot;INFO&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;verbose mode, i.e. &#39;INFO&#39; log level&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--debug&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
        <span class="n">const</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;debugging mode, i.e. &#39;DEBUG&#39; log level&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--log&quot;</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="n">log_levels</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;LEVEL&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;level of the events to track: </span><span class="si">%(choices)s</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">subgroup</span><span class="o">.</span><span class="n">set_defaults</span><span class="p">(</span><span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">add_log_file</span><span class="p">:</span>
        <span class="c1"># Add log file-related arguments</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--log_file&quot;</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dst_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">exists_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
            <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PATH&quot;</span><span class="p">,</span>
            <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;log file path&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--log_file_level&quot;</span><span class="p">,</span>
            <span class="n">choices</span><span class="o">=</span><span class="n">log_levels</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">,</span>
            <span class="n">default</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
            <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;LEVEL&quot;</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;level of the events to track in the log file: </span><span class="si">%(choices)s</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.ArgumentParser.add_server_arguments" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_server_arguments</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">include_database</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.ArgumentParser.add_server_arguments" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds the usual set of arguments needed to connect to a server.</p>
<p>This means: "--host", "--port", "--user" and "--password" (optional).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>prefix</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Prefix to add the each argument, e.g. if prefix is "src_", the arguments will be
"--src_host", etc.</p>
            </div>
          </td>
          <td>
                <code>&#39;&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>include_database</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Include "--database" argument.</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>help</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Description message to include for this set of arguments.</p>
            </div>
          </td>
          <td>
                <code>&#39;&#39;</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_server_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">include_database</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the usual set of arguments needed to connect to a server.</span>

<span class="sd">    This means: &quot;--host&quot;, &quot;--port&quot;, &quot;--user&quot; and &quot;--password&quot; (optional).</span>

<span class="sd">    Args:</span>
<span class="sd">        prefix: Prefix to add the each argument, e.g. if prefix is &quot;src_&quot;, the arguments will be</span>
<span class="sd">            &quot;--src_host&quot;, etc.</span>
<span class="sd">        include_database: Include &quot;--database&quot; argument.</span>
<span class="sd">        help: Description message to include for this set of arguments.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">group</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">server connection arguments&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="n">help</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">host&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;HOST&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;host name&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">port&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PORT&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;port number&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">user&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;USER&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;user name&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">password&quot;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PWD&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;host password&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">include_database</span><span class="p">:</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">database&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;NAME&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;database name&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.ArgumentParser.parse_args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">parse_args</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.ArgumentParser.parse_args" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Extends the parent function by adding a new URL argument for every server group added.</p>
<p>The type of this new argument will be :class:<code>sqlalchemy.engine.URL</code>. It also logs all the parsed
arguments for debugging purposes when logging arguments have been added.</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">parse_args</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">argparse</span><span class="o">.</span><span class="n">Namespace</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extends the parent function by adding a new URL argument for every server group added.</span>

<span class="sd">    The type of this new argument will be :class:`sqlalchemy.engine.URL`. It also logs all the parsed</span>
<span class="sd">    arguments for debugging purposes when logging arguments have been added.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="c1"># Build and add an sqlalchemy.engine.URL object for every server group added</span>
    <span class="k">for</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span><span class="p">:</span>
        <span class="c1"># Raise an error rather than overwriting when the URL argument is already present</span>
        <span class="k">if</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&quot;</span> <span class="ow">in</span> <span class="n">args</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;argument &#39;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&#39; is already present&quot;</span><span class="p">)</span>
        <span class="n">server_url</span> <span class="o">=</span> <span class="n">URL</span><span class="o">.</span><span class="n">create</span><span class="p">(</span>
            <span class="s2">&quot;mysql&quot;</span><span class="p">,</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">user&quot;</span><span class="p">),</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">password&quot;</span><span class="p">),</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">host&quot;</span><span class="p">),</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">port&quot;</span><span class="p">),</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">database&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&quot;</span><span class="p">,</span> <span class="n">server_url</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">args</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.DuplicateIdError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>DuplicateIdError</code>


<a href="#ensembl.io.genomio.gff3.DuplicateIdError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>Trying to add a feature with an ID already in use.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">43</span>
<span class="normal">44</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">DuplicateIdError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Trying to add a feature with an ID already in use.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.FunctionalAnnotations" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>FunctionalAnnotations</code>


<a href="#ensembl.io.genomio.gff3.FunctionalAnnotations" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">

  
      <p>List of annotations extracted from a GFF3 file.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">FunctionalAnnotations</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List of annotations extracted from a GFF3 file.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Annotation</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Annotated features</span>
        <span class="c1"># Under each feature, each dict&#39;s key is a feature ID</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Annotation</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;transcript&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;translation&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="p">}</span>
        <span class="c1"># Keep parent info: key is the feature ID, value is the parent ID</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parents</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;transcript&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="p">}</span>

    <span class="k">def</span> <span class="nf">get_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Annotation</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get all feature annotations for the requested type.&quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">feat_type</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No such feature type </span><span class="si">{</span><span class="n">feat_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>

    <span class="k">def</span> <span class="nf">add_parent_link</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parent_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">parent_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">child_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Record a parent-child IDs relationship for a given parent biotype.&quot;&quot;&quot;</span>
        <span class="n">features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">parent_type</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">parent_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">MissingParentError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Parent </span><span class="si">{</span><span class="n">parent_type</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">parent_id</span><span class="si">}</span><span class="s2"> not found for </span><span class="si">{</span><span class="n">child_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parents</span><span class="p">[</span><span class="n">parent_type</span><span class="p">][</span><span class="n">child_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent_id</span>

    <span class="k">def</span> <span class="nf">get_parent</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parent_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">child_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the parent ID of a given child for a given parent biotype.&quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">parents</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parents</span><span class="p">[</span><span class="n">parent_type</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported parent type </span><span class="si">{</span><span class="n">parent_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>

        <span class="n">parent_id</span> <span class="o">=</span> <span class="n">parents</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">child_id</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">parent_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">MissingParentError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t find </span><span class="si">{</span><span class="n">parent_type</span><span class="si">}</span><span class="s2"> parent for </span><span class="si">{</span><span class="n">child_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">parent_id</span>

    <span class="k">def</span> <span class="nf">add_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">feat_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">parent_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add annotation for a feature of a given type. If a parent_id is provided, record the relatioship.</span>

<span class="sd">        Args:</span>
<span class="sd">            feature: The feature to create an annotation.</span>
<span class="sd">            feat_type: Type of the feature to annotate.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">feat_type</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">AnnotationError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Feature </span><span class="si">{</span><span class="n">feat_type</span><span class="si">}</span><span class="s2"> ID </span><span class="si">{</span><span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> already added&quot;</span><span class="p">)</span>

        <span class="n">feature_object</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_feature</span><span class="p">(</span><span class="n">feature</span><span class="p">,</span> <span class="n">feat_type</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">feat_type</span><span class="p">][</span><span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">feature_object</span>

        <span class="k">if</span> <span class="n">parent_id</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat_type</span> <span class="ow">in</span> <span class="n">_PARENTS</span><span class="p">:</span>
                <span class="n">parent_type</span> <span class="o">=</span> <span class="n">_PARENTS</span><span class="p">[</span><span class="n">feat_type</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">add_parent_link</span><span class="p">(</span><span class="n">parent_type</span><span class="p">,</span> <span class="n">parent_id</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">AnnotationError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No parent possible for </span><span class="si">{</span><span class="n">feat_type</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_generic_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">feat_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create a feature object following the specifications.</span>

<span class="sd">        Args:</span>
<span class="sd">            feature: The SeqFeature to add to the list.</span>
<span class="sd">            feat_type: Feature type of the feature to store (e.g. gene, transcript, translation).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">feature_object</span><span class="p">:</span> <span class="n">Annotation</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;object_type&quot;</span><span class="p">:</span> <span class="n">feat_type</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="p">}</span>

        <span class="c1"># Description?</span>
        <span class="k">if</span> <span class="s2">&quot;product&quot;</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
            <span class="n">description</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;product&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">product_is_informative</span><span class="p">(</span><span class="n">description</span><span class="p">):</span>
                <span class="n">feature_object</span><span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">description</span>

        <span class="k">if</span> <span class="s2">&quot;Name&quot;</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span> <span class="ow">and</span> <span class="s2">&quot;description&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">feature_object</span><span class="p">:</span>
            <span class="n">feature_object</span><span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># Don&#39;t keep useless description</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;description&quot;</span> <span class="ow">in</span> <span class="n">feature_object</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">product_is_informative</span><span class="p">(</span>
            <span class="n">feature_object</span><span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">],</span> <span class="n">feature</span><span class="o">.</span><span class="n">id</span>
        <span class="p">):</span>
            <span class="k">del</span> <span class="n">feature_object</span><span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">]</span>

        <span class="c1"># Synonyms?</span>
        <span class="k">if</span> <span class="s2">&quot;Name&quot;</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
            <span class="n">feat_name</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">feat_name</span> <span class="o">!=</span> <span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="p">:</span>
                <span class="n">feature_object</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;synonym&quot;</span><span class="p">:</span> <span class="n">feat_name</span><span class="p">,</span> <span class="s2">&quot;default&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>

        <span class="c1"># is_pseudogene?</span>
        <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;pseudogen&quot;</span><span class="p">):</span>
            <span class="n">feature_object</span><span class="p">[</span><span class="s2">&quot;is_pseudogene&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">return</span> <span class="n">feature_object</span>

    <span class="k">def</span> <span class="nf">transfer_descriptions</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Transfers the feature descriptions in 2 steps:</span>
<span class="sd">        - from translations to transcripts (if the transcript description is empty)</span>
<span class="sd">        - from transcripts to genes (same case)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_transfer_description_up</span><span class="p">(</span><span class="s2">&quot;translation&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_transfer_description_up</span><span class="p">(</span><span class="s2">&quot;transcript&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_transfer_description_up</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">child_feature</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Transfer descriptions from all feature of a given type, up to their parent.</span>

<span class="sd">        Args:</span>
<span class="sd">            child_feature: Either &quot;translation&quot; (transfer to transcript) or &quot;transcript&quot; (to gene).</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">children_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">child_feature</span><span class="p">)</span>
        <span class="n">parent_type</span> <span class="o">=</span> <span class="n">_PARENTS</span><span class="p">[</span><span class="n">child_feature</span><span class="p">]</span>
        <span class="n">parent_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">parent_type</span><span class="p">)</span>

        <span class="c1"># Transfer description from children to their parent</span>
        <span class="k">for</span> <span class="n">child_id</span><span class="p">,</span> <span class="n">child</span> <span class="ow">in</span> <span class="n">children_features</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">child_description</span> <span class="o">=</span> <span class="n">child</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;description&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">child_description</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># Check parent</span>
                <span class="n">parent_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_parent</span><span class="p">(</span><span class="n">parent_type</span><span class="p">,</span> <span class="n">child_id</span><span class="p">)</span>
                <span class="n">parent</span> <span class="o">=</span> <span class="n">parent_features</span><span class="p">[</span><span class="n">parent_id</span><span class="p">]</span>
                <span class="n">parent_description</span> <span class="o">=</span> <span class="n">parent</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;description&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">parent_description</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">parent</span><span class="p">[</span><span class="s2">&quot;description&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">child_description</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">product_is_informative</span><span class="p">(</span><span class="n">product</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">feat_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns True if the product name contains informative words, False otherwise.</span>

<span class="sd">        It is considered uninformative when the description contains words such as &quot;hypothetical&quot; or</span>
<span class="sd">        or &quot;putative&quot;. If a feature ID is provided, consider it uninformative as well (we do not want</span>
<span class="sd">        descriptions to be just the ID).</span>

<span class="sd">        Args:</span>
<span class="sd">            product: A product name.</span>
<span class="sd">            feat_id: Feature ID (optional).</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">non_informative_words</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;hypothetical&quot;</span><span class="p">,</span>
            <span class="s2">&quot;putative&quot;</span><span class="p">,</span>
            <span class="s2">&quot;uncharacterized&quot;</span><span class="p">,</span>
            <span class="s2">&quot;unspecified&quot;</span><span class="p">,</span>
            <span class="s2">&quot;unknown&quot;</span><span class="p">,</span>
            <span class="sa">r</span><span class="s2">&quot;(of )?unknown function&quot;</span><span class="p">,</span>
            <span class="s2">&quot;conserved&quot;</span><span class="p">,</span>
            <span class="s2">&quot;predicted&quot;</span><span class="p">,</span>
            <span class="s2">&quot;fragment&quot;</span><span class="p">,</span>
            <span class="s2">&quot;product&quot;</span><span class="p">,</span>
            <span class="s2">&quot;function&quot;</span><span class="p">,</span>
            <span class="s2">&quot;protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
            <span class="s2">&quot;RNA&quot;</span><span class="p">,</span>
            <span class="sa">r</span><span class="s2">&quot;variant( \d+)?&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">non_informative_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">non_informative_words</span><span class="p">),</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">)</span>

        <span class="c1"># Remove the feature ID if it&#39;s in the description</span>
        <span class="k">if</span> <span class="n">feat_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">feat_id_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">feat_id</span><span class="p">,</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">)</span>
            <span class="n">product</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">feat_id_re</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">product</span><span class="p">)</span>

        <span class="c1"># Remove punctuations</span>
        <span class="n">punct_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[,;: _()-]+&quot;</span><span class="p">)</span>
        <span class="n">product</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">punct_re</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">product</span><span class="p">)</span>

        <span class="c1"># Then remove non informative words</span>
        <span class="n">product</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">non_informative_re</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">product</span><span class="p">)</span>

        <span class="c1"># Anything (informative) left?</span>
        <span class="n">empty_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^[ ]*$&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="ow">not</span> <span class="nb">bool</span><span class="p">(</span><span class="n">empty_re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">product</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">_to_list</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">all_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">feat_dict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">all_list</span> <span class="o">+=</span> <span class="n">feat_dict</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">all_list</span>

    <span class="k">def</span> <span class="nf">to_json</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Print out the current annotation list in a json file.</span>

<span class="sd">        Args:</span>
<span class="sd">            out_path: JSON file path where to write the data.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">transfer_descriptions</span><span class="p">()</span>
        <span class="n">feats_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_to_list</span><span class="p">()</span>
        <span class="n">print_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">out_path</span><span class="p">),</span> <span class="n">feats_list</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.FunctionalAnnotations.add_feature" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_feature</span><span class="p">(</span><span class="n">feature</span><span class="p">,</span> <span class="n">feat_type</span><span class="p">,</span> <span class="n">parent_id</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.add_feature" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Add annotation for a feature of a given type. If a parent_id is provided, record the relatioship.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>feature</code></td>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The feature to create an annotation.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>feat_type</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Type of the feature to annotate.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">feat_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">parent_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add annotation for a feature of a given type. If a parent_id is provided, record the relatioship.</span>

<span class="sd">    Args:</span>
<span class="sd">        feature: The feature to create an annotation.</span>
<span class="sd">        feat_type: Type of the feature to annotate.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">feat_type</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">AnnotationError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Feature </span><span class="si">{</span><span class="n">feat_type</span><span class="si">}</span><span class="s2"> ID </span><span class="si">{</span><span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> already added&quot;</span><span class="p">)</span>

    <span class="n">feature_object</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_feature</span><span class="p">(</span><span class="n">feature</span><span class="p">,</span> <span class="n">feat_type</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">feat_type</span><span class="p">][</span><span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">feature_object</span>

    <span class="k">if</span> <span class="n">parent_id</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">feat_type</span> <span class="ow">in</span> <span class="n">_PARENTS</span><span class="p">:</span>
            <span class="n">parent_type</span> <span class="o">=</span> <span class="n">_PARENTS</span><span class="p">[</span><span class="n">feat_type</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_parent_link</span><span class="p">(</span><span class="n">parent_type</span><span class="p">,</span> <span class="n">parent_id</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">AnnotationError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No parent possible for </span><span class="si">{</span><span class="n">feat_type</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">feature</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.FunctionalAnnotations.add_parent_link" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_parent_link</span><span class="p">(</span><span class="n">parent_type</span><span class="p">,</span> <span class="n">parent_id</span><span class="p">,</span> <span class="n">child_id</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.add_parent_link" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Record a parent-child IDs relationship for a given parent biotype.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_parent_link</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parent_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">parent_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">child_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Record a parent-child IDs relationship for a given parent biotype.&quot;&quot;&quot;</span>
    <span class="n">features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">parent_type</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">parent_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">MissingParentError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Parent </span><span class="si">{</span><span class="n">parent_type</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">parent_id</span><span class="si">}</span><span class="s2"> not found for </span><span class="si">{</span><span class="n">child_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">parents</span><span class="p">[</span><span class="n">parent_type</span><span class="p">][</span><span class="n">child_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.FunctionalAnnotations.get_features" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_features</span><span class="p">(</span><span class="n">feat_type</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.get_features" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Get all feature annotations for the requested type.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Annotation</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get all feature annotations for the requested type.&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">feat_type</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No such feature type </span><span class="si">{</span><span class="n">feat_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.FunctionalAnnotations.get_parent" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_parent</span><span class="p">(</span><span class="n">parent_type</span><span class="p">,</span> <span class="n">child_id</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.get_parent" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns the parent ID of a given child for a given parent biotype.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span>
<span class="normal">99</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_parent</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parent_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">child_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the parent ID of a given child for a given parent biotype.&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">parents</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parents</span><span class="p">[</span><span class="n">parent_type</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported parent type </span><span class="si">{</span><span class="n">parent_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>

    <span class="n">parent_id</span> <span class="o">=</span> <span class="n">parents</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">child_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">parent_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">MissingParentError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t find </span><span class="si">{</span><span class="n">parent_type</span><span class="si">}</span><span class="s2"> parent for </span><span class="si">{</span><span class="n">child_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">parent_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.FunctionalAnnotations.product_is_informative" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">product_is_informative</span><span class="p">(</span><span class="n">product</span><span class="p">,</span> <span class="n">feat_id</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.product_is_informative" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns True if the product name contains informative words, False otherwise.</p>
<p>It is considered uninformative when the description contains words such as "hypothetical" or
or "putative". If a feature ID is provided, consider it uninformative as well (we do not want
descriptions to be just the ID).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>product</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A product name.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>feat_id</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Feature ID (optional).</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">product_is_informative</span><span class="p">(</span><span class="n">product</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">feat_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns True if the product name contains informative words, False otherwise.</span>

<span class="sd">    It is considered uninformative when the description contains words such as &quot;hypothetical&quot; or</span>
<span class="sd">    or &quot;putative&quot;. If a feature ID is provided, consider it uninformative as well (we do not want</span>
<span class="sd">    descriptions to be just the ID).</span>

<span class="sd">    Args:</span>
<span class="sd">        product: A product name.</span>
<span class="sd">        feat_id: Feature ID (optional).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">non_informative_words</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;hypothetical&quot;</span><span class="p">,</span>
        <span class="s2">&quot;putative&quot;</span><span class="p">,</span>
        <span class="s2">&quot;uncharacterized&quot;</span><span class="p">,</span>
        <span class="s2">&quot;unspecified&quot;</span><span class="p">,</span>
        <span class="s2">&quot;unknown&quot;</span><span class="p">,</span>
        <span class="sa">r</span><span class="s2">&quot;(of )?unknown function&quot;</span><span class="p">,</span>
        <span class="s2">&quot;conserved&quot;</span><span class="p">,</span>
        <span class="s2">&quot;predicted&quot;</span><span class="p">,</span>
        <span class="s2">&quot;fragment&quot;</span><span class="p">,</span>
        <span class="s2">&quot;product&quot;</span><span class="p">,</span>
        <span class="s2">&quot;function&quot;</span><span class="p">,</span>
        <span class="s2">&quot;protein&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RNA&quot;</span><span class="p">,</span>
        <span class="sa">r</span><span class="s2">&quot;variant( \d+)?&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">non_informative_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">non_informative_words</span><span class="p">),</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">)</span>

    <span class="c1"># Remove the feature ID if it&#39;s in the description</span>
    <span class="k">if</span> <span class="n">feat_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">feat_id_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">feat_id</span><span class="p">,</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">)</span>
        <span class="n">product</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">feat_id_re</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">product</span><span class="p">)</span>

    <span class="c1"># Remove punctuations</span>
    <span class="n">punct_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[,;: _()-]+&quot;</span><span class="p">)</span>
    <span class="n">product</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">punct_re</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">product</span><span class="p">)</span>

    <span class="c1"># Then remove non informative words</span>
    <span class="n">product</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">non_informative_re</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">product</span><span class="p">)</span>

    <span class="c1"># Anything (informative) left?</span>
    <span class="n">empty_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^[ ]*$&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="ow">not</span> <span class="nb">bool</span><span class="p">(</span><span class="n">empty_re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">product</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.FunctionalAnnotations.to_json" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">to_json</span><span class="p">(</span><span class="n">out_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.to_json" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Print out the current annotation list in a json file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>out_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>JSON file path where to write the data.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">to_json</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Print out the current annotation list in a json file.</span>

<span class="sd">    Args:</span>
<span class="sd">        out_path: JSON file path where to write the data.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">transfer_descriptions</span><span class="p">()</span>
    <span class="n">feats_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_to_list</span><span class="p">()</span>
    <span class="n">print_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">out_path</span><span class="p">),</span> <span class="n">feats_list</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.FunctionalAnnotations.transfer_descriptions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">transfer_descriptions</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.gff3.FunctionalAnnotations.transfer_descriptions" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Transfers the feature descriptions in 2 steps:
- from translations to transcripts (if the transcript description is empty)
- from transcripts to genes (same case)</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">transfer_descriptions</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Transfers the feature descriptions in 2 steps:</span>
<span class="sd">    - from translations to transcripts (if the transcript description is empty)</span>
<span class="sd">    - from transcripts to genes (same case)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_transfer_description_up</span><span class="p">(</span><span class="s2">&quot;translation&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_transfer_description_up</span><span class="p">(</span><span class="s2">&quot;transcript&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.GFFGeneMerger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>GFFGeneMerger</code>


<a href="#ensembl.io.genomio.gff3.GFFGeneMerger" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">

  
      <p>Specialized class to merge split genes in a GFF3 file, prior to further parsing.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/gene_merger.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GFFGeneMerger</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Specialized class to merge split genes in a GFF3 file, prior to further parsing.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">biotypes_json</span> <span class="o">=</span> <span class="n">files</span><span class="p">(</span><span class="n">ensembl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">genomio</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">gff3</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;biotypes.json&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">biotypes_json</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merge genes in a gff that are split in multiple lines.</span>

<span class="sd">        Args:</span>
<span class="sd">            in_gff_path: Input GFF3 that may have split merge.</span>
<span class="sd">            out_gff_path: Output GFF3 with those genes merged.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List of all merged genes, each represented as a string of the GFF3 lines of all their parts.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">to_merge</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">merged</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">in_gff_fh</span><span class="p">:</span>
                <span class="c1"># Skip comments</span>
                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;##FASTA&quot;</span><span class="p">):</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;This GFF3 file contains FASTA sequences&quot;</span><span class="p">)</span>
                        <span class="k">break</span>
                    <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Parse one line</span>
                    <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span>
                    <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">attr_fields</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
                    <span class="n">attrs</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">attr_fields</span><span class="p">:</span>
                        <span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">)</span>
                        <span class="n">attrs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

                    <span class="c1"># Check this is a gene to merge; cache it then</span>
                    <span class="k">if</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span> <span class="ow">and</span> <span class="p">(</span>
                        <span class="s2">&quot;part&quot;</span> <span class="ow">in</span> <span class="n">attrs</span> <span class="ow">or</span> <span class="s2">&quot;is_ordered&quot;</span> <span class="ow">in</span> <span class="n">attrs</span>
                    <span class="p">):</span>
                        <span class="n">to_merge</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span>

                    <span class="c1"># If not, merge previous gene if needed, and print the line</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">to_merge</span><span class="p">:</span>
                            <span class="n">merged_str</span> <span class="o">=</span> <span class="p">[]</span>
                            <span class="k">for</span> <span class="n">line_to_merge</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
                                <span class="n">merged_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_to_merge</span><span class="p">))</span>
                            <span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

                            <span class="n">new_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge_genes</span><span class="p">(</span><span class="n">to_merge</span><span class="p">)</span>
                            <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>
                            <span class="n">to_merge</span> <span class="o">=</span> <span class="p">[]</span>
                        <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Print last merged gene if there is one</span>
            <span class="k">if</span> <span class="n">to_merge</span><span class="p">:</span>
                <span class="n">merged_str</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">line_to_merge</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
                    <span class="n">merged_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_to_merge</span><span class="p">))</span>
                <span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="n">new_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge_genes</span><span class="p">(</span><span class="n">to_merge</span><span class="p">)</span>
                <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged lines: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">merged</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">merged</span>

    <span class="k">def</span> <span class="nf">_merge_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">to_merge</span><span class="p">:</span> <span class="n">List</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a single gene gff3 line merged from separate parts.</span>

<span class="sd">        Args:</span>
<span class="sd">            to_merge: List of gff3 lines with gene parts.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">min_start</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="n">max_end</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>

            <span class="k">if</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="n">min_start</span> <span class="ow">or</span> <span class="n">min_start</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">min_start</span> <span class="o">=</span> <span class="n">start</span>
            <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">max_end</span> <span class="ow">or</span> <span class="n">max_end</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">max_end</span> <span class="o">=</span> <span class="n">end</span>

        <span class="c1"># Take the first line as template and replace things</span>
        <span class="n">new_gene</span> <span class="o">=</span> <span class="n">to_merge</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">new_gene</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">min_start</span><span class="p">)</span>
        <span class="n">new_gene</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">max_end</span><span class="p">)</span>

        <span class="n">attrs</span> <span class="o">=</span> <span class="n">new_gene</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="n">attrs</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;;is_ordered=true&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;;part=\d+/\d+&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">attrs</span><span class="p">)</span>
        <span class="n">new_gene</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrs</span>

        <span class="k">return</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFGeneMerger.merge" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">merge</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFGeneMerger.merge" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Merge genes in a gff that are split in multiple lines.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>in_gff_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Input GFF3 that may have split merge.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>out_gff_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Output GFF3 with those genes merged.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>List of all merged genes, each represented as a string of the GFF3 lines of all their parts.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/gene_merger.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Merge genes in a gff that are split in multiple lines.</span>

<span class="sd">    Args:</span>
<span class="sd">        in_gff_path: Input GFF3 that may have split merge.</span>
<span class="sd">        out_gff_path: Output GFF3 with those genes merged.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of all merged genes, each represented as a string of the GFF3 lines of all their parts.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">to_merge</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">merged</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">in_gff_fh</span><span class="p">:</span>
            <span class="c1"># Skip comments</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;##FASTA&quot;</span><span class="p">):</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;This GFF3 file contains FASTA sequences&quot;</span><span class="p">)</span>
                    <span class="k">break</span>
                <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Parse one line</span>
                <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span>
                <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">attr_fields</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
                <span class="n">attrs</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">attr_fields</span><span class="p">:</span>
                    <span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">)</span>
                    <span class="n">attrs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

                <span class="c1"># Check this is a gene to merge; cache it then</span>
                <span class="k">if</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span> <span class="ow">and</span> <span class="p">(</span>
                    <span class="s2">&quot;part&quot;</span> <span class="ow">in</span> <span class="n">attrs</span> <span class="ow">or</span> <span class="s2">&quot;is_ordered&quot;</span> <span class="ow">in</span> <span class="n">attrs</span>
                <span class="p">):</span>
                    <span class="n">to_merge</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span>

                <span class="c1"># If not, merge previous gene if needed, and print the line</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">to_merge</span><span class="p">:</span>
                        <span class="n">merged_str</span> <span class="o">=</span> <span class="p">[]</span>
                        <span class="k">for</span> <span class="n">line_to_merge</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
                            <span class="n">merged_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_to_merge</span><span class="p">))</span>
                        <span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

                        <span class="n">new_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge_genes</span><span class="p">(</span><span class="n">to_merge</span><span class="p">)</span>
                        <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>
                        <span class="n">to_merge</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Print last merged gene if there is one</span>
        <span class="k">if</span> <span class="n">to_merge</span><span class="p">:</span>
            <span class="n">merged_str</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">line_to_merge</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
                <span class="n">merged_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_to_merge</span><span class="p">))</span>
            <span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">new_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge_genes</span><span class="p">(</span><span class="n">to_merge</span><span class="p">)</span>
            <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged lines: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">merged</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">merged</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.GFFParserError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>GFFParserError</code>


<a href="#ensembl.io.genomio.gff3.GFFParserError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>Error when parsing a GFF3 file.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">56</span>
<span class="normal">57</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GFFParserError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Error when parsing a GFF3 file.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.GFFSimplifier" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>GFFSimplifier</code>


<a href="#ensembl.io.genomio.gff3.GFFSimplifier" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">

  
      <p>Parse a GGF3 file and output a cleaned up GFF3 + annotation json file.</p>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="ensembl.io.genomio.gff3.simplifier.GFFParserError" href="#ensembl.io.genomio.gff3.GFFParserError">GFFParserError</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If an error cannot be automatically fixed.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GFFSimplifier</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse a GGF3 file and output a cleaned up GFF3 + annotation json file.</span>

<span class="sd">    Raises:</span>
<span class="sd">        GFFParserError: If an error cannot be automatically fixed.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># pylint: disable=too-many-public-methods</span>
    <span class="c1"># Some rework needed at some point</span>

    <span class="c1"># Multiple parameters to automate various fixes</span>
    <span class="n">skip_unrecognized</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">gene_cds_skip_others</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">allow_pseudogene_with_CDS</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">exclude_seq_regions</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">fail_types</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">validate_gene_id</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">min_id_length</span> <span class="o">=</span> <span class="mi">8</span>
    <span class="n">stable_id_prefix</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">current_stable_id_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">make_missing_stable_ids</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="n">biotypes_json</span> <span class="o">=</span> <span class="n">files</span><span class="p">(</span><span class="n">ensembl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">genomio</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">gff3</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;biotypes.json&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">biotypes_json</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">records</span> <span class="o">=</span> <span class="n">Records</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="n">FunctionalAnnotations</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="n">genome_path</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">genome_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">genome_fh</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">genome_fh</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">make_missing_stable_ids</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="n">make_missing_stable_ids</span>

    <span class="k">def</span> <span class="nf">simpler_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Loads a GFF3 from INSDC and rewrites it in a simpler version, whilst also writing a</span>
<span class="sd">        functional annotation file.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">in_gff_fh</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">exclude_seq_regions</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skip seq_region </span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">continue</span>

                <span class="c1"># Clean all root features and make clean record</span>
                <span class="n">clean_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                    <span class="n">clean_feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">simpler_gff3_feature</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">clean_feature</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">clean_record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clean_feature</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clean_record</span><span class="p">)</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">:</span>
                <span class="n">fail_errors</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">   &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized types found:</span><span class="se">\n</span><span class="s2">   </span><span class="si">{</span><span class="n">fail_errors</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">simpler_gff3_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Creates a simpler version of a GFF3 feature.</span>

<span class="sd">        If the feature is invalid/skippable, returns None.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ignored_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;ignored&quot;</span><span class="p">]</span>
        <span class="n">allowed_non_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;non_gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>
        <span class="n">allowed_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>
        <span class="n">transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>

        <span class="c1"># Skip explictly ignored features</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">ignored_gene_types</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="c1"># Special processing of non-gene features</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">allowed_non_gene_types</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">):</span>
                <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">format_mobile_element</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">feat</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="c1"># From here we expect only genes</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">feat</span>

        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;protein_coding_gene&quot;</span><span class="p">:</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>

        <span class="c1"># Create actual genes from transcripts/CDS top level features</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">transcript_types</span><span class="p">:</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cds_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># What to do with unsupported gene types</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_gene_types</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">[</span><span class="s2">&quot;gene=&quot;</span> <span class="o">+</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported gene type: </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> (for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">None</span>

        <span class="c1"># Normalize, store annotation, and return the cleaned up gene</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">store_gene_annotations</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">store_gene_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Record the functional_annotations of the gene and its children features.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)</span>

        <span class="n">cds_found</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">transcript</span><span class="p">,</span> <span class="s2">&quot;transcript&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="c1"># Store CDS functional annotation only once</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">cds_found</span><span class="p">:</span>
                    <span class="n">cds_found</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="s2">&quot;translation&quot;</span><span class="p">,</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">clean_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the same gene without qualifiers unrelated to the gene structure.&quot;&quot;&quot;</span>

        <span class="n">old_gene_qualifiers</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_gene_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]}</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing source for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>
        <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="c1"># Replace qualifiers</span>
            <span class="n">old_transcript_qualifiers</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="s2">&quot;source&quot;</span> <span class="ow">in</span> <span class="n">old_transcript_qualifiers</span><span class="p">:</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_transcript_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>

            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                <span class="n">old_qualifiers</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">],</span>
                <span class="p">}</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_qualifiers</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span>
                    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Missing phase for gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">, CDS </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">old_qualifiers</span><span class="si">}</span><span class="s2">)&quot;</span>
                        <span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">format_mobile_element</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Given a mobile_genetic_element feature, transform it into a transposable_element&quot;&quot;&quot;</span>

        <span class="c1"># Change mobile_genetic_element into a transposable_element feature</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">:</span>
            <span class="n">mobile_element_type</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mobile_element_type&quot;</span><span class="p">,</span> <span class="p">[])</span>
            <span class="k">if</span> <span class="n">mobile_element_type</span><span class="p">:</span>
                <span class="c1"># Get the type (and name) from the attrib</span>
                <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                    <span class="n">element_type</span><span class="p">,</span> <span class="n">element_name</span> <span class="o">=</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
                    <span class="n">description</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">element_type</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">element_name</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">element_type</span> <span class="o">=</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">description</span> <span class="o">=</span> <span class="n">element_type</span>

                <span class="c1"># Keep the metadata in the description if the type is known</span>
                <span class="k">if</span> <span class="n">element_type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;transposon&quot;</span><span class="p">,</span> <span class="s2">&quot;retrotransposon&quot;</span><span class="p">):</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;transposable_element&quot;</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;product&quot;</span><span class="p">):</span>
                        <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;product&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">description</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Mobile genetic element &#39;mobile_element_type&#39; is not transposon: </span><span class="si">{</span><span class="n">element_type</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                    <span class="k">return</span> <span class="n">feat</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Mobile genetic element does not have a &#39;mobile_element_type&#39; tag&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">feat</span>
        <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Feature </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> is not a supported TE feature </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">feat</span>

        <span class="c1"># Generate ID if needed and add it to the functional annotation</span>
        <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">)</span>
        <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">}</span>

        <span class="k">return</span> <span class="n">feat</span>

    <span class="k">def</span> <span class="nf">format_gene_segments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the equivalent Ensembl biotype feature for gene segment transcript features.</span>

<span class="sd">        Supported features: &quot;C_gene_segment&quot; and &quot;V_gene_segment&quot;.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript: Gene segment transcript feature.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Change V/C_gene_segment into a its corresponding transcript names</span>
        <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;C_gene_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;V_gene_segment&quot;</span><span class="p">):</span>
            <span class="n">standard_name</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;standard_name&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">biotype</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\b(immunoglobulin|ig)\b&quot;</span><span class="p">,</span> <span class="n">standard_name</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
                <span class="n">biotype</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;IG_</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">elif</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\bt[- _]cell\b&quot;</span><span class="p">,</span> <span class="n">standard_name</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
                <span class="n">biotype</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;TR_</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unexpected &#39;standard_name&#39; content for feature </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">standard_name</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="n">transcript</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="n">biotype</span>
        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">normalize_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a normalized gene structure, separate from the functional elements.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene: Gene object to normalize.</span>
<span class="sd">            functional_annotation: List of feature annotations (appended by this method).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># New gene ID</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># Gene with no subfeatures: need to create a transcript at least</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript for lone gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>

        <span class="c1"># Count features</span>
        <span class="n">fcounter</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">([</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">])</span>

        <span class="c1"># Transform gene - CDS to gene-transcript-exon-CDS</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fcounter</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">):</span>
                <span class="n">num_subs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript-exon feats for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">num_subs</span><span class="si">}</span><span class="s2"> CDSs)&quot;</span><span class="p">)</span>
                <span class="n">transcripts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">transcripts</span>

            <span class="c1"># Transform gene - exon to gene-transcript-exon</span>
            <span class="k">elif</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">):</span>
                <span class="n">num_subs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">num_subs</span><span class="si">}</span><span class="s2"> exons)&quot;</span><span class="p">)</span>
                <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_exon</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Check that we don&#39;t mix</span>
            <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">):</span>
                <span class="c1"># Move CDS(s) from parent gene to parent mRNA if needed</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">move_cds_to_mrna</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">):</span>
                <span class="c1"># Special case with extra exons</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_extra_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># Remove CDS from pseudogenes</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_CDS</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">remove_cds_from_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># TRANSCRIPTS</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_transcripts</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># PSEUDOGENE CDS IDs</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_CDS</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">normalize_pseudogene_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">_normalize_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a normalized transcript.&quot;&quot;&quot;</span>

        <span class="n">allowed_transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>
        <span class="n">ignored_transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;ignored&quot;</span><span class="p">]</span>
        <span class="n">skip_unrecognized</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span>

        <span class="n">transcripts_to_delete</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_transcript_types</span>
                <span class="ow">and</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ignored_transcript_types</span>
            <span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">[</span><span class="s2">&quot;transcript=&quot;</span> <span class="o">+</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unrecognized transcript type: </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span> <span class="sa">f</span><span class="s2">&quot; for </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">skip_unrecognized</span><span class="p">:</span>
                    <span class="n">transcripts_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
                    <span class="k">continue</span>

            <span class="c1"># New transcript ID</span>
            <span class="n">transcript_number</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_transcript_id</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">transcript_number</span><span class="p">)</span>

            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">format_gene_segments</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>

            <span class="c1"># EXONS AND CDS</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_transcript_subfeatures</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">transcript</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">transcripts_to_delete</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">transcripts_to_delete</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">_normalize_transcript_subfeatures</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a transcript with normalized sub-features.&quot;&quot;&quot;</span>
        <span class="n">ignored_transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;ignored&quot;</span><span class="p">]</span>
        <span class="n">exons_to_delete</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">exon_number</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">for</span> <span class="n">tcount</span><span class="p">,</span> <span class="n">feat</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                <span class="c1"># New exon ID</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">-E</span><span class="si">{</span><span class="n">exon_number</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="n">exon_number</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="c1"># Replace qualifiers</span>
                <span class="n">old_exon_qualifiers</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">}</span>
                <span class="k">if</span> <span class="s2">&quot;source&quot;</span> <span class="ow">in</span> <span class="n">old_exon_qualifiers</span><span class="p">:</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_exon_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="c1"># New CDS ID</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_cds_id</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">_cds&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">ignored_transcript_types</span><span class="p">:</span>
                    <span class="n">exons_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tcount</span><span class="p">)</span>
                    <span class="k">continue</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;sub_transcript=</span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unrecognized exon type for </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; (for transcript </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> of type </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span><span class="p">:</span>
                    <span class="n">exons_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tcount</span><span class="p">)</span>
                    <span class="k">continue</span>

        <span class="k">if</span> <span class="n">exons_to_delete</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">exons_to_delete</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">transcript_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ncrna</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create a gene for lone transcripts: &#39;gene&#39; for tRNA/rRNA, and &#39;ncRNA&#39; for all others</span>

<span class="sd">        Args:</span>
<span class="sd">            ncrna: the transcript for which we want to create a gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The gene that contains the transcript.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;ncRNA_gene&quot;</span>
        <span class="k">if</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;tRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">):</span>
            <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the transcript </span><span class="si">{</span><span class="n">ncrna</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> in a </span><span class="si">{</span><span class="n">new_type</span><span class="si">}</span><span class="s2"> parent feature&quot;</span><span class="p">)</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">ncrna</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">new_type</span><span class="p">)</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">ncrna</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">id</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">cds_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cds</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a gene created for a lone CDS.&quot;&quot;&quot;</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the lone CDS in gene-mRNA parent features for </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Create a transcript, add the CDS</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">cds</span><span class="p">]</span>

        <span class="c1"># Add an exon too</span>
        <span class="n">exon</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>
        <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

        <span class="c1"># Create a gene, add the transcript</span>
        <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;pseudo&quot;</span> <span class="ow">in</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;true&quot;</span><span class="p">):</span>
            <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;pseudogene&quot;</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">gene_type</span><span class="p">)</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_stable_id</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">transcript_for_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a transcript, from a gene without one.&quot;&quot;&quot;</span>

        <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;transcript&quot;</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">gene_to_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a list of transcripts (with exons), from a gene with only CDS children.&quot;&quot;&quot;</span>

        <span class="n">gene_cds_skip_others</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_cds_skip_others</span>
        <span class="n">transcripts_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">del_transcript</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">cds</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">gene_cds_skip_others</span><span class="p">:</span>
                    <span class="n">del_transcript</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
                    <span class="k">continue</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
                    <span class="s2">&quot;Can not create a chain &#39;transcript - exon - CDS&#39;&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; when the gene children are not all CDSs&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> of type </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> is child of gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>

            <span class="n">exon</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>

            <span class="c1"># Add to transcript or create a new one</span>
            <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">transcripts_dict</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Create new mRNA for </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_transcript</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript</span>
            <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
            <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>
            <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">del_transcript</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>

        <span class="n">transcripts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">transcripts_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="k">return</span> <span class="n">transcripts</span>

    <span class="k">def</span> <span class="nf">build_transcript</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a transcript with same metadata as the gene provided.&quot;&quot;&quot;</span>

        <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">move_cds_to_mrna</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Move CDS child features of a gene, to the mRNA.</span>

<span class="sd">        This is to fix the case where we have the following structure:</span>
<span class="sd">        gene -&gt; [ mRNA, CDSs ]</span>
<span class="sd">        and change it to</span>
<span class="sd">        gene -&gt; [ mRNA -&gt; [ CDSs ] ]</span>
<span class="sd">        The mRNA might have exons, in which case check that they match the CDS coordinates.</span>

<span class="sd">        Raises an exception if the feature structure is not recognized.</span>

<span class="sd">        Args:</span>
<span class="sd">            A gene with only one transcript, to check and fix.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The gene where the CDSs have been moved, if needed.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># First, count the types</span>
        <span class="n">mrnas</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">cdss</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">gene_subf_clean</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
                <span class="n">mrnas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">cdss</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gene_subf_clean</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Nothing to fix here, no CDSs to move</span>
            <span class="k">return</span> <span class="n">gene</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mrnas</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Can&#39;t fix gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: contains several mRNAs and CDSs, all children of the gene&quot;</span>
            <span class="p">)</span>

        <span class="n">mrna</span> <span class="o">=</span> <span class="n">mrnas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># Check if there are exons (or CDSs) under the mRNA</span>
        <span class="n">sub_exons</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">sub_cdss</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">mrna</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">sub_cdss</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                <span class="n">sub_exons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_check_sub_cdss</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">sub_cdss</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_check_sub_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">cdss</span><span class="p">,</span> <span class="n">sub_exons</span><span class="p">)</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: move </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">)</span><span class="si">}</span><span class="s2"> CDSs to the mRNA&quot;</span><span class="p">)</span>
        <span class="c1"># No more issues? move the CDSs</span>
        <span class="n">mrna</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">+=</span> <span class="n">cdss</span>
        <span class="c1"># And remove them from the gene</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">gene_subf_clean</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mrna</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_check_sub_cdss</span><span class="p">(</span><span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">sub_cdss</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_cdss</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> has CDSs as children of the gene and mRNA&quot;</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_check_sub_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">cdss</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">sub_exons</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check that the exons of the mRNA and the CDSs match&quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_exons</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Check that they match the CDS outside</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_exons</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">):</span>
                <span class="c1"># Now that all coordinates are the same</span>
                <span class="n">coord_exons</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">exon</span><span class="o">.</span><span class="n">location</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">sub_exons</span><span class="p">]</span>
                <span class="n">coord_cdss</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span><span class="p">]</span>

                <span class="k">if</span> <span class="n">coord_exons</span> <span class="o">!=</span> <span class="n">coord_cdss</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> CDSs and exons under the mRNA do not match&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> CDSs and exons under the mRNA do not match (different count)&quot;</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="nf">clean_extra_exons</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove extra exons, already existing in the mRNA.</span>

<span class="sd">        This is a special case where a gene contains proper mRNAs, etc. but also</span>
<span class="sd">        extra exons for the same features. Those exons usually have an ID starting with</span>
<span class="sd">        &quot;id-&quot;, so that&#39;s what we use to detect them.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">exons</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mrnas</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">others</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                <span class="n">exons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
                <span class="n">mrnas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">others</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">exons</span> <span class="ow">and</span> <span class="n">mrnas</span><span class="p">:</span>
            <span class="n">exon_has_id</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="c1"># Check if the exon ids start with &quot;id-&quot;, which is an indication that they do not belong here</span>
            <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;id-&quot;</span><span class="p">):</span>
                    <span class="n">exon_has_id</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">exon_has_id</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">exon_has_id</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">exons</span><span class="p">):</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove </span><span class="si">{</span><span class="n">exon_has_id</span><span class="si">}</span><span class="s2"> extra exons from </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">mrnas</span>
                    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">+=</span> <span class="n">others</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t remove extra exons for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">, not all start with &#39;id-&#39;&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">gene_to_exon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns an intermediary transcript for a gene with direct exon children.&quot;&quot;&quot;</span>

        <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">normalize_gene_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove any unnecessary prefixes around the gene ID.</span>

<span class="sd">        Generate a new stable id if it is not recognized as valid.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene: Gene feature to normalize.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A normalized gene id.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">prefixes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene-&quot;</span><span class="p">,</span> <span class="s2">&quot;gene:&quot;</span><span class="p">]</span>
        <span class="n">new_gene_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span>

        <span class="c1"># In case the gene id is not valid, use the GeneID</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_id</span><span class="p">(</span><span class="n">new_gene_id</span><span class="p">):</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene id is not valid: </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">qual</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
            <span class="k">if</span> <span class="s2">&quot;Dbxref&quot;</span> <span class="ow">in</span> <span class="n">qual</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">xref</span> <span class="ow">in</span> <span class="n">qual</span><span class="p">[</span><span class="s2">&quot;Dbxref&quot;</span><span class="p">]:</span>
                    <span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="o">=</span> <span class="n">xref</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">db</span> <span class="o">==</span> <span class="s2">&quot;GeneID&quot;</span><span class="p">:</span>
                        <span class="n">new_gene_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using GeneID </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2"> for stable_id instead of </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">return</span> <span class="n">new_gene_id</span>

            <span class="c1"># Make a new stable_id</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_missing_stable_ids</span><span class="p">:</span>
                <span class="n">new_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_stable_id</span><span class="p">()</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;New id: </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">new_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">new_id</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t use invalid gene id for </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">new_gene_id</span>

    <span class="k">def</span> <span class="nf">generate_stable_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a new unique gene stable_id with a prefix.</span>

<span class="sd">        The id is made up of a prefix and a number, which is auto incremented.</span>
<span class="sd">        Define the prefix with the param &quot;stable_id_prefix&quot;,</span>
<span class="sd">        or use the genome organism_abbrev and prepend &quot;TMP_&quot; to it.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">:</span>
                <span class="n">org</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;BRC4&quot;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;organism_abbrev&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">org</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;TMP_PREFIX_&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;TMP_&quot;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span> <span class="o">=</span> <span class="n">prefix</span>

        <span class="n">number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">current_stable_id_number</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">new_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">number</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">current_stable_id_number</span> <span class="o">=</span> <span class="n">number</span>

        <span class="k">return</span> <span class="n">new_id</span>

    <span class="k">def</span> <span class="nf">valid_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check that the format of a stable id is valid.&quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">validate_gene_id</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>

        <span class="n">min_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_id_length</span>

        <span class="c1"># Trna (from tRNAscan)</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^Trna&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable ID is a tRNA from tRNA-scan: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="c1"># Coordinates</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^.+:\d+..\d+&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is a coordinate: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="c1"># Special characters</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[ |]&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id contains special characters: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="c1"># Min length</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_length</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is too short (&lt;</span><span class="si">{</span><span class="n">min_length</span><span class="si">}</span><span class="s2">) </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="k">return</span> <span class="kc">True</span>

    <span class="k">def</span> <span class="nf">normalize_transcript_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">number</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Use a gene ID and a number to make a formatted transcript ID.&quot;&quot;&quot;</span>

        <span class="n">transcript_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene_id</span><span class="si">}</span><span class="s2">_t</span><span class="si">{</span><span class="n">number</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">transcript_id</span>

    <span class="k">def</span> <span class="nf">normalize_cds_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cds_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a normalised version of the provided CDS ID.</span>

<span class="sd">        The normalisation implies to remove any unnecessary prefixes around the CDS ID. However, if</span>
<span class="sd">        the CDS ID is still not proper, an empty string will be returned.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">prefixes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;cds-&quot;</span><span class="p">,</span> <span class="s2">&quot;cds:&quot;</span><span class="p">]</span>
        <span class="n">cds_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">cds_id</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span>

        <span class="c1"># Special case: if the ID doesn&#39;t look like one, remove it</span>
        <span class="c1"># It needs to be regenerated</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_id</span><span class="p">(</span><span class="n">cds_id</span><span class="p">):</span>
            <span class="n">cds_id</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="k">return</span> <span class="n">cds_id</span>

    <span class="k">def</span> <span class="nf">normalize_pseudogene_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Normalises the CDS ID of the provided pseudogene.</span>

<span class="sd">        Ensure CDS from a pseudogene have a proper ID:</span>
<span class="sd">        - Different from the gene</span>
<span class="sd">        - Derived from the gene if it is not proper</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_cds_id</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
                        <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">_cds&quot;</span>
                        <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>

    <span class="k">def</span> <span class="nf">remove_cds_from_pseudogene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Removes the CDS from a pseudogene.</span>

<span class="sd">        This assumes the CDSs are sub features of the transcript or the gene.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gene_subfeats</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove pseudo CDS </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="n">new_subfeats</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove pseudo CDS </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">continue</span>
                <span class="n">new_subfeats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">new_subfeats</span>
            <span class="n">gene_subfeats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">gene_subfeats</span>

    <span class="k">def</span> <span class="nf">remove_prefixes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the identifier after removing all the prefixes found in it (if any).&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="n">prefixes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">identifier</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">prefix</span><span class="p">):</span>
                <span class="n">identifier</span> <span class="o">=</span> <span class="n">identifier</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="p">:]</span>
        <span class="k">return</span> <span class="n">identifier</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.build_transcript" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">build_transcript</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.build_transcript" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns a transcript with same metadata as the gene provided.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">build_transcript</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a transcript with same metadata as the gene provided.&quot;&quot;&quot;</span>

    <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">return</span> <span class="n">transcript</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.cds_gene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">cds_gene</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.cds_gene" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns a gene created for a lone CDS.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">cds_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cds</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a gene created for a lone CDS.&quot;&quot;&quot;</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the lone CDS in gene-mRNA parent features for </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Create a transcript, add the CDS</span>
    <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">cds</span><span class="p">]</span>

    <span class="c1"># Add an exon too</span>
    <span class="n">exon</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>
    <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

    <span class="c1"># Create a gene, add the transcript</span>
    <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;pseudo&quot;</span> <span class="ow">in</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;true&quot;</span><span class="p">):</span>
        <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;pseudogene&quot;</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">gene_type</span><span class="p">)</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_stable_id</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.clean_extra_exons" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">clean_extra_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.clean_extra_exons" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Remove extra exons, already existing in the mRNA.</p>
<p>This is a special case where a gene contains proper mRNAs, etc. but also
extra exons for the same features. Those exons usually have an ID starting with
"id-", so that's what we use to detect them.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">clean_extra_exons</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove extra exons, already existing in the mRNA.</span>

<span class="sd">    This is a special case where a gene contains proper mRNAs, etc. but also</span>
<span class="sd">    extra exons for the same features. Those exons usually have an ID starting with</span>
<span class="sd">    &quot;id-&quot;, so that&#39;s what we use to detect them.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">exons</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mrnas</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">others</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
            <span class="n">exons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
            <span class="n">mrnas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">others</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">exons</span> <span class="ow">and</span> <span class="n">mrnas</span><span class="p">:</span>
        <span class="n">exon_has_id</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># Check if the exon ids start with &quot;id-&quot;, which is an indication that they do not belong here</span>
        <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;id-&quot;</span><span class="p">):</span>
                <span class="n">exon_has_id</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">exon_has_id</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">exon_has_id</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">exons</span><span class="p">):</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove </span><span class="si">{</span><span class="n">exon_has_id</span><span class="si">}</span><span class="s2"> extra exons from </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">mrnas</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">+=</span> <span class="n">others</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t remove extra exons for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">, not all start with &#39;id-&#39;&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.clean_gene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">clean_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.clean_gene" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Return the same gene without qualifiers unrelated to the gene structure.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">clean_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the same gene without qualifiers unrelated to the gene structure.&quot;&quot;&quot;</span>

    <span class="n">old_gene_qualifiers</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_gene_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]}</span>
    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing source for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>
    <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="c1"># Replace qualifiers</span>
        <span class="n">old_transcript_qualifiers</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="s2">&quot;source&quot;</span> <span class="ow">in</span> <span class="n">old_transcript_qualifiers</span><span class="p">:</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_transcript_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="n">old_qualifiers</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span>
            <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">],</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_qualifiers</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span>
                <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Missing phase for gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">, CDS </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">old_qualifiers</span><span class="si">}</span><span class="s2">)&quot;</span>
                    <span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.format_gene_segments" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">format_gene_segments</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.format_gene_segments" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns the equivalent Ensembl biotype feature for gene segment transcript features.</p>
<p>Supported features: "C_gene_segment" and "V_gene_segment".</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>transcript</code></td>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Gene segment transcript feature.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">format_gene_segments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the equivalent Ensembl biotype feature for gene segment transcript features.</span>

<span class="sd">    Supported features: &quot;C_gene_segment&quot; and &quot;V_gene_segment&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        transcript: Gene segment transcript feature.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Change V/C_gene_segment into a its corresponding transcript names</span>
    <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;C_gene_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;V_gene_segment&quot;</span><span class="p">):</span>
        <span class="n">standard_name</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;standard_name&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">biotype</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\b(immunoglobulin|ig)\b&quot;</span><span class="p">,</span> <span class="n">standard_name</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
            <span class="n">biotype</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;IG_</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">elif</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\bt[- _]cell\b&quot;</span><span class="p">,</span> <span class="n">standard_name</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
            <span class="n">biotype</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;TR_</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Unexpected &#39;standard_name&#39; content for feature </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">standard_name</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="n">transcript</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="n">biotype</span>
    <span class="k">return</span> <span class="n">transcript</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.format_mobile_element" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">format_mobile_element</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.format_mobile_element" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Given a mobile_genetic_element feature, transform it into a transposable_element</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">format_mobile_element</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a mobile_genetic_element feature, transform it into a transposable_element&quot;&quot;&quot;</span>

    <span class="c1"># Change mobile_genetic_element into a transposable_element feature</span>
    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">:</span>
        <span class="n">mobile_element_type</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mobile_element_type&quot;</span><span class="p">,</span> <span class="p">[])</span>
        <span class="k">if</span> <span class="n">mobile_element_type</span><span class="p">:</span>
            <span class="c1"># Get the type (and name) from the attrib</span>
            <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="n">element_type</span><span class="p">,</span> <span class="n">element_name</span> <span class="o">=</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
                <span class="n">description</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">element_type</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">element_name</span><span class="si">}</span><span class="s2">)&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">element_type</span> <span class="o">=</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">description</span> <span class="o">=</span> <span class="n">element_type</span>

            <span class="c1"># Keep the metadata in the description if the type is known</span>
            <span class="k">if</span> <span class="n">element_type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;transposon&quot;</span><span class="p">,</span> <span class="s2">&quot;retrotransposon&quot;</span><span class="p">):</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;transposable_element&quot;</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;product&quot;</span><span class="p">):</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;product&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">description</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Mobile genetic element &#39;mobile_element_type&#39; is not transposon: </span><span class="si">{</span><span class="n">element_type</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="n">feat</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Mobile genetic element does not have a &#39;mobile_element_type&#39; tag&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">feat</span>
    <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Feature </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> is not a supported TE feature </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">feat</span>

    <span class="c1"># Generate ID if needed and add it to the functional annotation</span>
    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">)</span>
    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">}</span>

    <span class="k">return</span> <span class="n">feat</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.gene_to_cds" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">gene_to_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.gene_to_cds" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns a list of transcripts (with exons), from a gene with only CDS children.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">gene_to_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a list of transcripts (with exons), from a gene with only CDS children.&quot;&quot;&quot;</span>

    <span class="n">gene_cds_skip_others</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_cds_skip_others</span>
    <span class="n">transcripts_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">del_transcript</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">cds</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">gene_cds_skip_others</span><span class="p">:</span>
                <span class="n">del_transcript</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
                <span class="s2">&quot;Can not create a chain &#39;transcript - exon - CDS&#39;&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; when the gene children are not all CDSs&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> of type </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> is child of gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
            <span class="p">)</span>

        <span class="n">exon</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>

        <span class="c1"># Add to transcript or create a new one</span>
        <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">transcripts_dict</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Create new mRNA for </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_transcript</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript</span>
        <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>
        <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">del_transcript</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>

    <span class="n">transcripts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">transcripts_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">transcripts</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.gene_to_exon" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">gene_to_exon</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.gene_to_exon" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns an intermediary transcript for a gene with direct exon children.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">gene_to_exon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns an intermediary transcript for a gene with direct exon children.&quot;&quot;&quot;</span>

    <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">transcript</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.generate_stable_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">generate_stable_id</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.generate_stable_id" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns a new unique gene stable_id with a prefix.</p>
<p>The id is made up of a prefix and a number, which is auto incremented.
Define the prefix with the param "stable_id_prefix",
or use the genome organism_abbrev and prepend "TMP_" to it.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_stable_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a new unique gene stable_id with a prefix.</span>

<span class="sd">    The id is made up of a prefix and a number, which is auto incremented.</span>
<span class="sd">    Define the prefix with the param &quot;stable_id_prefix&quot;,</span>
<span class="sd">    or use the genome organism_abbrev and prepend &quot;TMP_&quot; to it.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span><span class="p">:</span>
        <span class="n">prefix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">:</span>
            <span class="n">org</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;BRC4&quot;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;organism_abbrev&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">org</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;TMP_PREFIX_&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;TMP_&quot;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span> <span class="o">=</span> <span class="n">prefix</span>

    <span class="n">number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">current_stable_id_number</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">new_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">number</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">current_stable_id_number</span> <span class="o">=</span> <span class="n">number</span>

    <span class="k">return</span> <span class="n">new_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.move_cds_to_mrna" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">move_cds_to_mrna</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.move_cds_to_mrna" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Move CDS child features of a gene, to the mRNA.</p>
<p>This is to fix the case where we have the following structure:
gene -&gt; [ mRNA, CDSs ]
and change it to
gene -&gt; [ mRNA -&gt; [ CDSs ] ]
The mRNA might have exons, in which case check that they match the CDS coordinates.</p>
<p>Raises an exception if the feature structure is not recognized.</p>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The gene where the CDSs have been moved, if needed.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">move_cds_to_mrna</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Move CDS child features of a gene, to the mRNA.</span>

<span class="sd">    This is to fix the case where we have the following structure:</span>
<span class="sd">    gene -&gt; [ mRNA, CDSs ]</span>
<span class="sd">    and change it to</span>
<span class="sd">    gene -&gt; [ mRNA -&gt; [ CDSs ] ]</span>
<span class="sd">    The mRNA might have exons, in which case check that they match the CDS coordinates.</span>

<span class="sd">    Raises an exception if the feature structure is not recognized.</span>

<span class="sd">    Args:</span>
<span class="sd">        A gene with only one transcript, to check and fix.</span>

<span class="sd">    Returns:</span>
<span class="sd">        The gene where the CDSs have been moved, if needed.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First, count the types</span>
    <span class="n">mrnas</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">cdss</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">gene_subf_clean</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
            <span class="n">mrnas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="n">cdss</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">gene_subf_clean</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Nothing to fix here, no CDSs to move</span>
        <span class="k">return</span> <span class="n">gene</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mrnas</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Can&#39;t fix gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: contains several mRNAs and CDSs, all children of the gene&quot;</span>
        <span class="p">)</span>

    <span class="n">mrna</span> <span class="o">=</span> <span class="n">mrnas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Check if there are exons (or CDSs) under the mRNA</span>
    <span class="n">sub_exons</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">sub_cdss</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">mrna</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="n">sub_cdss</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
            <span class="n">sub_exons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_check_sub_cdss</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">sub_cdss</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_check_sub_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">cdss</span><span class="p">,</span> <span class="n">sub_exons</span><span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: move </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">)</span><span class="si">}</span><span class="s2"> CDSs to the mRNA&quot;</span><span class="p">)</span>
    <span class="c1"># No more issues? move the CDSs</span>
    <span class="n">mrna</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">+=</span> <span class="n">cdss</span>
    <span class="c1"># And remove them from the gene</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">gene_subf_clean</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mrna</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.normalize_cds_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">normalize_cds_id</span><span class="p">(</span><span class="n">cds_id</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_cds_id" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns a normalised version of the provided CDS ID.</p>
<p>The normalisation implies to remove any unnecessary prefixes around the CDS ID. However, if
the CDS ID is still not proper, an empty string will be returned.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_cds_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cds_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a normalised version of the provided CDS ID.</span>

<span class="sd">    The normalisation implies to remove any unnecessary prefixes around the CDS ID. However, if</span>
<span class="sd">    the CDS ID is still not proper, an empty string will be returned.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">prefixes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;cds-&quot;</span><span class="p">,</span> <span class="s2">&quot;cds:&quot;</span><span class="p">]</span>
    <span class="n">cds_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">cds_id</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span>

    <span class="c1"># Special case: if the ID doesn&#39;t look like one, remove it</span>
    <span class="c1"># It needs to be regenerated</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_id</span><span class="p">(</span><span class="n">cds_id</span><span class="p">):</span>
        <span class="n">cds_id</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="k">return</span> <span class="n">cds_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.normalize_gene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">normalize_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_gene" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns a normalized gene structure, separate from the functional elements.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gene</code></td>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Gene object to normalize.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>functional_annotation</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>List of feature annotations (appended by this method).</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a normalized gene structure, separate from the functional elements.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene: Gene object to normalize.</span>
<span class="sd">        functional_annotation: List of feature annotations (appended by this method).</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># New gene ID</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># Gene with no subfeatures: need to create a transcript at least</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript for lone gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>

    <span class="c1"># Count features</span>
    <span class="n">fcounter</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">([</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">])</span>

    <span class="c1"># Transform gene - CDS to gene-transcript-exon-CDS</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fcounter</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">):</span>
            <span class="n">num_subs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript-exon feats for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">num_subs</span><span class="si">}</span><span class="s2"> CDSs)&quot;</span><span class="p">)</span>
            <span class="n">transcripts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">transcripts</span>

        <span class="c1"># Transform gene - exon to gene-transcript-exon</span>
        <span class="k">elif</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">):</span>
            <span class="n">num_subs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">num_subs</span><span class="si">}</span><span class="s2"> exons)&quot;</span><span class="p">)</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_exon</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Check that we don&#39;t mix</span>
        <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">):</span>
            <span class="c1"># Move CDS(s) from parent gene to parent mRNA if needed</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">move_cds_to_mrna</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">):</span>
            <span class="c1"># Special case with extra exons</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_extra_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># Remove CDS from pseudogenes</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_CDS</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">remove_cds_from_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># TRANSCRIPTS</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_transcripts</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># PSEUDOGENE CDS IDs</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_CDS</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">normalize_pseudogene_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.normalize_gene_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_gene_id" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Remove any unnecessary prefixes around the gene ID.</p>
<p>Generate a new stable id if it is not recognized as valid.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gene</code></td>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Gene feature to normalize.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A normalized gene id.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_gene_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove any unnecessary prefixes around the gene ID.</span>

<span class="sd">    Generate a new stable id if it is not recognized as valid.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene: Gene feature to normalize.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A normalized gene id.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">prefixes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene-&quot;</span><span class="p">,</span> <span class="s2">&quot;gene:&quot;</span><span class="p">]</span>
    <span class="n">new_gene_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span>

    <span class="c1"># In case the gene id is not valid, use the GeneID</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_id</span><span class="p">(</span><span class="n">new_gene_id</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene id is not valid: </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">qual</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="k">if</span> <span class="s2">&quot;Dbxref&quot;</span> <span class="ow">in</span> <span class="n">qual</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">xref</span> <span class="ow">in</span> <span class="n">qual</span><span class="p">[</span><span class="s2">&quot;Dbxref&quot;</span><span class="p">]:</span>
                <span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="o">=</span> <span class="n">xref</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">db</span> <span class="o">==</span> <span class="s2">&quot;GeneID&quot;</span><span class="p">:</span>
                    <span class="n">new_gene_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using GeneID </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2"> for stable_id instead of </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">return</span> <span class="n">new_gene_id</span>

        <span class="c1"># Make a new stable_id</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_missing_stable_ids</span><span class="p">:</span>
            <span class="n">new_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_stable_id</span><span class="p">()</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;New id: </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">new_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">new_id</span>
        <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t use invalid gene id for </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">new_gene_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.normalize_pseudogene_cds" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">normalize_pseudogene_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_pseudogene_cds" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Normalises the CDS ID of the provided pseudogene.</p>
<p>Ensure CDS from a pseudogene have a proper ID:
- Different from the gene
- Derived from the gene if it is not proper</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_pseudogene_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Normalises the CDS ID of the provided pseudogene.</span>

<span class="sd">    Ensure CDS from a pseudogene have a proper ID:</span>
<span class="sd">    - Different from the gene</span>
<span class="sd">    - Derived from the gene if it is not proper</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_cds_id</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">_cds&quot;</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.normalize_transcript_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">normalize_transcript_id</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">number</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.normalize_transcript_id" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Use a gene ID and a number to make a formatted transcript ID.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_transcript_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">number</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Use a gene ID and a number to make a formatted transcript ID.&quot;&quot;&quot;</span>

    <span class="n">transcript_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene_id</span><span class="si">}</span><span class="s2">_t</span><span class="si">{</span><span class="n">number</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">return</span> <span class="n">transcript_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.remove_cds_from_pseudogene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">remove_cds_from_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.remove_cds_from_pseudogene" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Removes the CDS from a pseudogene.</p>
<p>This assumes the CDSs are sub features of the transcript or the gene.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">remove_cds_from_pseudogene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Removes the CDS from a pseudogene.</span>

<span class="sd">    This assumes the CDSs are sub features of the transcript or the gene.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene_subfeats</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove pseudo CDS </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="n">new_subfeats</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove pseudo CDS </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="n">new_subfeats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">new_subfeats</span>
        <span class="n">gene_subfeats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">gene_subfeats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.remove_prefixes" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">identifier</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.remove_prefixes" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns the identifier after removing all the prefixes found in it (if any).</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">remove_prefixes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the identifier after removing all the prefixes found in it (if any).&quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="n">prefixes</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">identifier</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">prefix</span><span class="p">):</span>
            <span class="n">identifier</span> <span class="o">=</span> <span class="n">identifier</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="p">:]</span>
    <span class="k">return</span> <span class="n">identifier</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.simpler_gff3" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">simpler_gff3</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.simpler_gff3" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Loads a GFF3 from INSDC and rewrites it in a simpler version, whilst also writing a
functional annotation file.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">simpler_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Loads a GFF3 from INSDC and rewrites it in a simpler version, whilst also writing a</span>
<span class="sd">    functional annotation file.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">in_gff_fh</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">exclude_seq_regions</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skip seq_region </span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>

            <span class="c1"># Clean all root features and make clean record</span>
            <span class="n">clean_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="n">clean_feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">simpler_gff3_feature</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">clean_feature</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">clean_record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clean_feature</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clean_record</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">:</span>
            <span class="n">fail_errors</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">   &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized types found:</span><span class="se">\n</span><span class="s2">   </span><span class="si">{</span><span class="n">fail_errors</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.simpler_gff3_feature" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">simpler_gff3_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.simpler_gff3_feature" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Creates a simpler version of a GFF3 feature.</p>
<p>If the feature is invalid/skippable, returns None.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">simpler_gff3_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creates a simpler version of a GFF3 feature.</span>

<span class="sd">    If the feature is invalid/skippable, returns None.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">ignored_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;ignored&quot;</span><span class="p">]</span>
    <span class="n">allowed_non_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;non_gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>
    <span class="n">allowed_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>
    <span class="n">transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_biotypes</span><span class="p">[</span><span class="s2">&quot;transcript&quot;</span><span class="p">][</span><span class="s2">&quot;supported&quot;</span><span class="p">]</span>

    <span class="c1"># Skip explictly ignored features</span>
    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">ignored_gene_types</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># Special processing of non-gene features</span>
    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">allowed_non_gene_types</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">):</span>
            <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">format_mobile_element</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">feat</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># From here we expect only genes</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="n">feat</span>

    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;protein_coding_gene&quot;</span><span class="p">:</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>

    <span class="c1"># Create actual genes from transcripts/CDS top level features</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">transcript_types</span><span class="p">:</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cds_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># What to do with unsupported gene types</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_gene_types</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fail_types</span><span class="p">[</span><span class="s2">&quot;gene=&quot;</span> <span class="o">+</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported gene type: </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> (for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># Normalize, store annotation, and return the cleaned up gene</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">store_gene_annotations</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.store_gene_annotations" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">store_gene_annotations</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.store_gene_annotations" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Record the functional_annotations of the gene and its children features.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">store_gene_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Record the functional_annotations of the gene and its children features.&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)</span>

    <span class="n">cds_found</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">transcript</span><span class="p">,</span> <span class="s2">&quot;transcript&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="c1"># Store CDS functional annotation only once</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">cds_found</span><span class="p">:</span>
                <span class="n">cds_found</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="s2">&quot;translation&quot;</span><span class="p">,</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.transcript_for_gene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">transcript_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.transcript_for_gene" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns a transcript, from a gene without one.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">transcript_for_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a transcript, from a gene without one.&quot;&quot;&quot;</span>

    <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;transcript&quot;</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">return</span> <span class="n">transcript</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.transcript_gene" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">transcript_gene</span><span class="p">(</span><span class="n">ncrna</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.transcript_gene" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Create a gene for lone transcripts: 'gene' for tRNA/rRNA, and 'ncRNA' for all others</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>ncrna</code></td>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the transcript for which we want to create a gene.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The gene that contains the transcript.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">transcript_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ncrna</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a gene for lone transcripts: &#39;gene&#39; for tRNA/rRNA, and &#39;ncRNA&#39; for all others</span>

<span class="sd">    Args:</span>
<span class="sd">        ncrna: the transcript for which we want to create a gene.</span>

<span class="sd">    Returns:</span>
<span class="sd">        The gene that contains the transcript.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;ncRNA_gene&quot;</span>
    <span class="k">if</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;tRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">):</span>
        <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the transcript </span><span class="si">{</span><span class="n">ncrna</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> in a </span><span class="si">{</span><span class="n">new_type</span><span class="si">}</span><span class="s2"> parent feature&quot;</span><span class="p">)</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">ncrna</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">new_type</span><span class="p">)</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">ncrna</span><span class="p">]</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">id</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.GFFSimplifier.valid_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">valid_id</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.GFFSimplifier.valid_id" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Check that the format of a stable id is valid.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">valid_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check that the format of a stable id is valid.&quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">validate_gene_id</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="n">min_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_id_length</span>

    <span class="c1"># Trna (from tRNAscan)</span>
    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^Trna&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable ID is a tRNA from tRNA-scan: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># Coordinates</span>
    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^.+:\d+..\d+&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is a coordinate: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># Special characters</span>
    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[ |]&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id contains special characters: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># Min length</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_length</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is too short (&lt;</span><span class="si">{</span><span class="n">min_length</span><span class="si">}</span><span class="s2">) </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="k">return</span> <span class="kc">True</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.MissingParentError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>MissingParentError</code>


<a href="#ensembl.io.genomio.gff3.MissingParentError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>Trying to add a feature without an expected parent.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/extract_annotation.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">MissingParentError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Trying to add a feature without an expected parent.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.gff3.Records" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>Records</code>


<a href="#ensembl.io.genomio.gff3.Records" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>list</code></p>

  
      <p>List of GFF3 SeqRecords.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Records</span><span class="p">(</span><span class="nb">list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List of GFF3 SeqRecords.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">to_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Print out the current list of records in a GFF3 file.</span>

<span class="sd">        Args:</span>
<span class="sd">            out_gff_path: Path to GFF3 file where to write the records.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
            <span class="n">GFF</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_fh</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.gff3.Records.to_gff" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">to_gff</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.Records.to_gff" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Print out the current list of records in a GFF3 file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>out_gff_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to GFF3 file where to write the records.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/simplifier.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">to_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Print out the current list of records in a GFF3 file.</span>

<span class="sd">    Args:</span>
<span class="sd">        out_gff_path: Path to GFF3 file where to write the records.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
        <span class="n">GFF</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_fh</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>



<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.gff3.init_logging_with_args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">init_logging_with_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.gff3.init_logging_with_args" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Processes the Namespace object provided to call <code>init_logging()</code> with the right arguments.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>args</code></td>
          <td>
                <code><span title="argparse.Namespace">Namespace</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Namespace populated by an argument parser.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/logging.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">init_logging_with_args</span><span class="p">(</span><span class="n">args</span><span class="p">:</span> <span class="n">argparse</span><span class="o">.</span><span class="n">Namespace</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Processes the Namespace object provided to call `init_logging()` with the right arguments.</span>

<span class="sd">    Args:</span>
<span class="sd">        args: Namespace populated by an argument parser.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args_dict</span> <span class="o">=</span> <span class="nb">vars</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
    <span class="n">log_args</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="n">args_dict</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span> <span class="s2">&quot;log_file&quot;</span><span class="p">,</span> <span class="s2">&quot;log_file_level&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">args_dict</span><span class="p">}</span>
    <span class="n">init_logging</span><span class="p">(</span><span class="o">**</span><span class="n">log_args</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.gff3.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.gff3.main" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Main script entry-point.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main script entry-point.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Standardize the gene model representation of a GFF3 file, and extract the functional &quot;</span>
            <span class="s2">&quot;annotation in a separate file.&quot;</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_src_path</span><span class="p">(</span><span class="s2">&quot;--in_gff_path&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Input GFF3 file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_src_path</span><span class="p">(</span><span class="s2">&quot;--genome_data&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Genome JSON file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--make_missing_stable_ids&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Generate stable IDs when missing or invalid&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_dst_path</span><span class="p">(</span><span class="s2">&quot;--out_gff_path&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;gene_models.gff3&quot;</span><span class="p">),</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output GFF3 file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_dst_path</span><span class="p">(</span>
        <span class="s2">&quot;--out_func_path&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;functional_annotation.json&quot;</span><span class="p">),</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output functional annotation JSON file&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_log_arguments</span><span class="p">(</span><span class="n">add_log_file</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">init_logging_with_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="c1"># Merge multiline gene features in a separate file</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Checking for genes to merge...&quot;</span><span class="p">)</span>
    <span class="n">interim_gff_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">in_gff_path</span><span class="si">}</span><span class="s2">_INTERIM_MERGE&quot;</span><span class="p">)</span>
    <span class="n">merger</span> <span class="o">=</span> <span class="n">GFFGeneMerger</span><span class="p">()</span>
    <span class="n">merged_genes</span> <span class="o">=</span> <span class="n">merger</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">in_gff_path</span><span class="p">,</span> <span class="n">interim_gff_path</span><span class="p">)</span>
    <span class="n">num_merged_genes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">merged_genes</span><span class="p">)</span>
    <span class="n">in_gff_path</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">in_gff_path</span>
    <span class="c1"># If there are split genes, decide to merge, or just die</span>
    <span class="k">if</span> <span class="n">num_merged_genes</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Report the list of merged genes in case something does not look right</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">num_merged_genes</span><span class="si">}</span><span class="s2"> genes merged&quot;</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_genes</span><span class="p">))</span>
        <span class="c1"># Use the GFF with the merged genes for the next part</span>
        <span class="n">in_gff_path</span> <span class="o">=</span> <span class="n">interim_gff_path</span>

    <span class="c1"># Load GFF3 data and write a simpler version that follows our specifications as well as a</span>
    <span class="c1"># functional annotation JSON file</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Simplify and fix GFF3&quot;</span><span class="p">)</span>
    <span class="n">gff_data</span> <span class="o">=</span> <span class="n">GFFSimplifier</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">genome_data</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">make_missing_stable_ids</span><span class="p">)</span>
    <span class="n">gff_data</span><span class="o">.</span><span class="n">simpler_gff3</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span>
    <span class="n">gff_data</span><span class="o">.</span><span class="n">records</span><span class="o">.</span><span class="n">to_gff</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_gff_path</span><span class="p">)</span>
    <span class="n">gff_data</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">to_json</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_func_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-2023] EMBL-European Bioinformatics Institute
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.8fd75fb4.min.js"></script>
      
    
  </body>
</html>