
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genomio/reference/genomio/manifest/">
      
      
        <link rel="prev" href="../gff3/">
      
      
        <link rel="next" href="../schemas/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.8">
    
    
      
        <title>manifest - Ensembl-GenomIO</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.f2e4d321.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ensembl.io.genomio.manifest" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Ensembl-GenomIO" class="md-header__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../img/ebang.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-GenomIO
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              manifest
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M272 384c9.6-31.9 29.5-59.1 49.2-86.2 5.2-7.1 10.4-14.2 15.4-21.4 19.8-28.5 31.4-63 31.4-100.3C368 78.8 289.2 0 192 0S16 78.8 16 176c0 37.3 11.6 71.9 31.4 100.3 5 7.2 10.2 14.3 15.4 21.4 19.8 27.1 39.7 54.4 49.2 86.2h160zm-80 128c44.2 0 80-35.8 80-80v-16H112v16c0 44.2 35.8 80 80 80zm-80-336c0 8.8-7.2 16-16 16s-16-7.2-16-16c0-61.9 50.1-112 112-112 8.8 0 16 7.2 16 16s-7.2 16-16 16c-44.2 0-80 35.8-80 80z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M297.2 248.9c14.4-20.6 22.8-45.7 22.8-72.9 0-70.7-57.3-128-128-128S64 105.3 64 176c0 27.2 8.4 52.3 22.8 72.9 3.7 5.3 8.1 11.3 12.8 17.7 12.9 17.7 28.3 38.9 39.8 59.8 10.4 19 15.7 38.8 18.3 57.5H109c-2.2-12-5.9-23.7-11.8-34.5-9.9-18-22.2-34.9-34.5-51.8-5.2-7.1-10.4-14.2-15.4-21.4C27.6 247.9 16 213.3 16 176 16 78.8 94.8 0 192 0s176 78.8 176 176c0 37.3-11.6 71.9-31.4 100.3-5 7.2-10.2 14.3-15.4 21.4-12.3 16.8-24.6 33.7-34.5 51.8-5.9 10.8-9.6 22.5-11.8 34.5h-48.5c2.6-18.7 7.9-38.6 18.3-57.5 11.5-20.9 26.9-42.1 39.8-59.8 4.7-6.4 9-12.4 12.7-17.7zM192 128c-26.5 0-48 21.5-48 48 0 8.8-7.2 16-16 16s-16-7.2-16-16c0-44.2 35.8-80 80-80 8.8 0 16 7.2 16 16s-7.2 16-16 16zm0 384c-44.2 0-80-35.8-80-80v-16h160v16c0 44.2-35.8 80-80 80z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl-GenomIO
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Ensembl-GenomIO" class="md-nav__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../img/ebang.png" alt="logo">

    </a>
    Ensembl-GenomIO
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl-GenomIO
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ensembl
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    brc4
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            brc4
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../runnable/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    runnable
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_1" id="__nav_4_1_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            runnable
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/compare_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_fasta.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/compare_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_report.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/core_server/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    core_server.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/download_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download_genbank.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/dump_stable_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_stable_ids.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/extract_from_gb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_gb.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/fill_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fill_metadata.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/gff3_specifier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3_specifier.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/load_sequence_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load_sequence_data.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/read_json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    read_json.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/say_accession/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    say_accession.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/seqregion_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seqregion_parser.py
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    io
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            io
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_1" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genomio
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_1" id="__nav_4_1_2_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            genomio
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    database
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../events/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fasta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genome_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    manifest
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    manifest
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;manifest
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" manifest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;ArgumentParser
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" ArgumentParser">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;__init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument_dst_path" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_dst_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument_src_path" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_src_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument_url" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_url
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_log_arguments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_log_arguments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_server_arguments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_server_arguments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.BiotypeCounter" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;BiotypeCounter
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" BiotypeCounter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.BiotypeCounter.add_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.BiotypeCounter.unique_count" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;unique_count
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;IntegrityTool
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" IntegrityTool">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.add_errors" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_errors
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.check_ids" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_ids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.check_integrity" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_integrity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.check_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.check_seq_region_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_seq_region_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.set_brc_mode" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;set_brc_mode
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.set_ignore_final_stops" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;set_ignore_final_stops
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.InvalidIntegrityError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;InvalidIntegrityError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;Manifest
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" Manifest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_agp_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_agp_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_circular" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_circular
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_fasta_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_fasta_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_functional_annotation" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_functional_annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_gff3" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_gff3
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_manifest
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.has_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;has_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.prepare_integrity_data" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;prepare_integrity_data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ManifestMaker" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;ManifestMaker
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" ManifestMaker">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ManifestMaker.create_manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;create_manifest
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ManifestMaker.get_files_checksums" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_files_checksums
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.StatsError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;StatsError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;manifest_stats
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" manifest_stats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.biotypes_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;biotypes_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.check_ncbi_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_ncbi_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.compare_ncbi_counts" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;compare_ncbi_counts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.coord_systems_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;coord_systems_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.count_biotypes" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;count_biotypes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.get_gff3_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_gff3_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.get_manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_manifest
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.get_seq_region_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_seq_region_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.increment_biotype" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;increment_biotype
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.run" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;run
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.seq_region_special_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;seq_region_special_stats
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.init_logging_with_args" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;init_logging_with_args
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.main" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.open_gz_file" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;open_gz_file
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_1_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../schemas/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    schemas
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_1_10" id="__nav_4_1_2_1_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_2_1_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2_1_10">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seq_region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Nextflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;manifest
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" manifest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;ArgumentParser
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" ArgumentParser">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;__init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument_dst_path" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_dst_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument_src_path" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_src_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument_url" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_argument_url
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_log_arguments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_log_arguments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.add_server_arguments" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_server_arguments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ArgumentParser.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.BiotypeCounter" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;BiotypeCounter
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" BiotypeCounter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.BiotypeCounter.add_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.BiotypeCounter.unique_count" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;unique_count
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;IntegrityTool
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" IntegrityTool">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.add_errors" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;add_errors
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.check_ids" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_ids
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.check_integrity" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_integrity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.check_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.check_seq_region_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_seq_region_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.set_brc_mode" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;set_brc_mode
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.IntegrityTool.set_ignore_final_stops" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;set_ignore_final_stops
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.InvalidIntegrityError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;InvalidIntegrityError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;Manifest
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" Manifest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_agp_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_agp_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_circular" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_circular
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_fasta_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_fasta_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_functional_annotation" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_functional_annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_gff3" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_gff3
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.get_manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_manifest
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.has_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;has_lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.Manifest.prepare_integrity_data" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;prepare_integrity_data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ManifestMaker" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;ManifestMaker
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" ManifestMaker">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ManifestMaker.create_manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;create_manifest
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.ManifestMaker.get_files_checksums" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_files_checksums
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.StatsError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;StatsError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;manifest_stats
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" manifest_stats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.biotypes_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;biotypes_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.check_ncbi_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_ncbi_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.compare_ncbi_counts" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;compare_ncbi_counts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.coord_systems_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;coord_systems_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.count_biotypes" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;count_biotypes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.get_gff3_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_gff3_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.get_manifest" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_manifest
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.get_seq_region_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;get_seq_region_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.increment_biotype" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;increment_biotype
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.run" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;run
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.manifest_stats.seq_region_special_stats" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;seq_region_special_stats
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.init_logging_with_args" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;init_logging_with_args
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.main" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.manifest.open_gz_file" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;open_gz_file
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>manifest</h1>

<div class="doc doc-object doc-module">



<h2 id="ensembl.io.genomio.manifest" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>          <code>ensembl.io.genomio.manifest</code>


<a href="#ensembl.io.genomio.manifest" class="headerlink" title="Permanent link">&para;</a></h2>

  <div class="doc doc-contents first">
  
      <p>Manifest files handling module.</p>

  

  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.manifest.ArgumentParser" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>ArgumentParser</code>


<a href="#ensembl.io.genomio.manifest.ArgumentParser" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><span title="argparse.ArgumentParser">ArgumentParser</span></code></p>

  
      <p>Extends :class:<code>argparse.ArgumentParser</code> with additional methods and functionality.</p>
<p>The default behaviour of the help text will be to display the default values on every non-required
argument, i.e. optional arguments with "required=False".</p>

            <details class="quote">
              <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">ArgumentParser</span><span class="p">(</span><span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extends :class:`argparse.ArgumentParser` with additional methods and functionality.</span>

<span class="sd">    The default behaviour of the help text will be to display the default values on every non-required</span>
<span class="sd">    argument, i.e. optional arguments with &quot;required=False&quot;.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extends the base class to include the information about default argument values by default.&quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">formatter_class</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentDefaultsHelpFormatter</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="nf">_validate_src_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">src_path</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the path if exists and it is readable, raises an error through the parser otherwise.</span>

<span class="sd">        Args:</span>
<span class="sd">            src_path: File or directory path to check.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">src_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">src_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">src_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">src_path</span><span class="si">}</span><span class="s2">&#39; not found&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">src_path</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">R_OK</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">src_path</span><span class="si">}</span><span class="s2">&#39; not readable&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">src_path</span>

    <span class="k">def</span> <span class="nf">_validate_dst_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dst_path</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">PathLike</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the path if it is writable, raises an error through the parser otherwise.</span>

<span class="sd">        Args:</span>
<span class="sd">            dst_path: File or directory path to check.</span>
<span class="sd">            exists_ok: Do not raise an error if the destination path already exists.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">dst_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">dst_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">dst_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">dst_path</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">W_OK</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">exists_ok</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">dst_path</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">dst_path</span><span class="si">}</span><span class="s2">&#39; already exists&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">dst_path</span><span class="si">}</span><span class="s2">&#39; is not writable&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">parent_path</span> <span class="ow">in</span> <span class="n">dst_path</span><span class="o">.</span><span class="n">parents</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">parent_path</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">parent_path</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">W_OK</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">dst_path</span><span class="si">}</span><span class="s2">&#39; is not writable&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">dst_path</span>

    <span class="k">def</span> <span class="nf">add_argument</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extends the parent function by excluding the default value in the help text when not provided.</span>

<span class="sd">        Only applied to required arguments without a default value, i.e. positional arguments or optional</span>
<span class="sd">        arguments with &quot;required=True&quot;.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;required&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
            <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;default&quot;</span><span class="p">,</span> <span class="n">argparse</span><span class="o">.</span><span class="n">SUPPRESS</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_argument_src_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds :class:`pathlib.Path` argument, checking if it exists and it is readable at parsing time.</span>

<span class="sd">        If &quot;metavar&quot; is not defined it is added with &quot;PATH&quot; as value to improve help text readability.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;PATH&quot;</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_src_path</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_argument_dst_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds :class:`pathlib.Path` argument, checking if it is writable at parsing time.</span>

<span class="sd">        If &quot;metavar&quot; is not defined it is added with &quot;PATH&quot; as value to improve help text readability.</span>

<span class="sd">        Args:</span>
<span class="sd">            exists_ok: Do not raise an error if the destination path already exists.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;PATH&quot;</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dst_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_argument_url</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds :class:`sqlalchemy.engine.URL` argument.</span>

<span class="sd">        If &quot;metavar&quot; is not defined it is added with &quot;URI&quot; as value to improve help text readability.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;URI&quot;</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">make_url</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_server_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">include_database</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds the usual set of arguments needed to connect to a server.</span>

<span class="sd">        This means: &quot;--host&quot;, &quot;--port&quot;, &quot;--user&quot; and &quot;--password&quot; (optional).</span>

<span class="sd">        Args:</span>
<span class="sd">            prefix: Prefix to add the each argument, e.g. if prefix is &quot;src_&quot;, the arguments will be</span>
<span class="sd">                &quot;--src_host&quot;, etc.</span>
<span class="sd">            include_database: Include &quot;--database&quot; argument.</span>
<span class="sd">            help: Description message to include for this set of arguments.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">group</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">server connection arguments&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="n">help</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">host&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;HOST&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;host name&quot;</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">port&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PORT&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;port number&quot;</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">user&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;USER&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;user name&quot;</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">password&quot;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PWD&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;host password&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">include_database</span><span class="p">:</span>
            <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">database&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;NAME&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;database name&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_log_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">add_log_file</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds the usual set of arguments required to set and initialise a logging system.</span>

<span class="sd">        The current set includes a mutually exclusive group for the default logging level: &quot;--verbose&quot;,</span>
<span class="sd">        &quot;--debug&quot; or &quot;--log LEVEL&quot;.</span>

<span class="sd">        Args:</span>
<span class="sd">            add_log_file: Add arguments to allow storing messages into a file, i.e. &quot;--log_file&quot; and</span>
<span class="sd">                &quot;--log_file_level&quot;.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Define the list of log levels available</span>
        <span class="n">log_levels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span> <span class="s2">&quot;INFO&quot;</span><span class="p">,</span> <span class="s2">&quot;WARNING&quot;</span><span class="p">,</span> <span class="s2">&quot;ERROR&quot;</span><span class="p">,</span> <span class="s2">&quot;CRITICAL&quot;</span><span class="p">]</span>
        <span class="c1"># NOTE: from 3.11 this list can be changed to: logging.getLevelNamesMapping().keys()</span>
        <span class="c1"># Create logging arguments group</span>
        <span class="n">group</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;logging arguments&quot;</span><span class="p">)</span>
        <span class="c1"># Add 3 mutually exclusive options to set the logging level</span>
        <span class="n">subgroup</span> <span class="o">=</span> <span class="n">group</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">()</span>
        <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;-v&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--verbose&quot;</span><span class="p">,</span>
            <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
            <span class="n">const</span><span class="o">=</span><span class="s2">&quot;INFO&quot;</span><span class="p">,</span>
            <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;verbose mode, i.e. &#39;INFO&#39; log level&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--debug&quot;</span><span class="p">,</span>
            <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
            <span class="n">const</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
            <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;debugging mode, i.e. &#39;DEBUG&#39; log level&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--log&quot;</span><span class="p">,</span>
            <span class="n">choices</span><span class="o">=</span><span class="n">log_levels</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">,</span>
            <span class="n">default</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
            <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;LEVEL&quot;</span><span class="p">,</span>
            <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;level of the events to track: </span><span class="si">%(choices)s</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">subgroup</span><span class="o">.</span><span class="n">set_defaults</span><span class="p">(</span><span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">add_log_file</span><span class="p">:</span>
            <span class="c1"># Add log file-related arguments</span>
            <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
                <span class="s2">&quot;--log_file&quot;</span><span class="p">,</span>
                <span class="nb">type</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dst_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">exists_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PATH&quot;</span><span class="p">,</span>
                <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">help</span><span class="o">=</span><span class="s2">&quot;log file path&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
                <span class="s2">&quot;--log_file_level&quot;</span><span class="p">,</span>
                <span class="n">choices</span><span class="o">=</span><span class="n">log_levels</span><span class="p">,</span>
                <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">,</span>
                <span class="n">default</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
                <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;LEVEL&quot;</span><span class="p">,</span>
                <span class="n">help</span><span class="o">=</span><span class="s2">&quot;level of the events to track in the log file: </span><span class="si">%(choices)s</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">parse_args</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">argparse</span><span class="o">.</span><span class="n">Namespace</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extends the parent function by adding a new URL argument for every server group added.</span>

<span class="sd">        The type of this new argument will be :class:`sqlalchemy.engine.URL`. It also logs all the parsed</span>
<span class="sd">        arguments for debugging purposes when logging arguments have been added.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">args</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="c1"># Build and add an sqlalchemy.engine.URL object for every server group added</span>
        <span class="k">for</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span><span class="p">:</span>
            <span class="c1"># Raise an error rather than overwriting when the URL argument is already present</span>
            <span class="k">if</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&quot;</span> <span class="ow">in</span> <span class="n">args</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;argument &#39;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&#39; is already present&quot;</span><span class="p">)</span>
            <span class="n">server_url</span> <span class="o">=</span> <span class="n">URL</span><span class="o">.</span><span class="n">create</span><span class="p">(</span>
                <span class="s2">&quot;mysql&quot;</span><span class="p">,</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">user&quot;</span><span class="p">),</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">password&quot;</span><span class="p">),</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">host&quot;</span><span class="p">),</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">port&quot;</span><span class="p">),</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">database&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&quot;</span><span class="p">,</span> <span class="n">server_url</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">args</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ArgumentParser.__init__" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.ArgumentParser.__init__" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Extends the base class to include the information about default argument values by default.</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extends the base class to include the information about default argument values by default.&quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">formatter_class</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentDefaultsHelpFormatter</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span> <span class="o">=</span> <span class="p">[]</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ArgumentParser.add_argument" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Extends the parent function by excluding the default value in the help text when not provided.</p>
<p>Only applied to required arguments without a default value, i.e. positional arguments or optional
arguments with "required=True".</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_argument</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extends the parent function by excluding the default value in the help text when not provided.</span>

<span class="sd">    Only applied to required arguments without a default value, i.e. positional arguments or optional</span>
<span class="sd">    arguments with &quot;required=True&quot;.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;required&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;default&quot;</span><span class="p">,</span> <span class="n">argparse</span><span class="o">.</span><span class="n">SUPPRESS</span><span class="p">)</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ArgumentParser.add_argument_dst_path" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_argument_dst_path</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">exists_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument_dst_path" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds :class:<code>pathlib.Path</code> argument, checking if it is writable at parsing time.</p>
<p>If "metavar" is not defined it is added with "PATH" as value to improve help text readability.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>exists_ok</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Do not raise an error if the destination path already exists.</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_argument_dst_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds :class:`pathlib.Path` argument, checking if it is writable at parsing time.</span>

<span class="sd">    If &quot;metavar&quot; is not defined it is added with &quot;PATH&quot; as value to improve help text readability.</span>

<span class="sd">    Args:</span>
<span class="sd">        exists_ok: Do not raise an error if the destination path already exists.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;PATH&quot;</span><span class="p">)</span>
    <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dst_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">exists_ok</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ArgumentParser.add_argument_src_path" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_argument_src_path</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument_src_path" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds :class:<code>pathlib.Path</code> argument, checking if it exists and it is readable at parsing time.</p>
<p>If "metavar" is not defined it is added with "PATH" as value to improve help text readability.</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_argument_src_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds :class:`pathlib.Path` argument, checking if it exists and it is readable at parsing time.</span>

<span class="sd">    If &quot;metavar&quot; is not defined it is added with &quot;PATH&quot; as value to improve help text readability.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;PATH&quot;</span><span class="p">)</span>
    <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_src_path</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ArgumentParser.add_argument_url" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_argument_url</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.ArgumentParser.add_argument_url" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds :class:<code>sqlalchemy.engine.URL</code> argument.</p>
<p>If "metavar" is not defined it is added with "URI" as value to improve help text readability.</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_argument_url</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds :class:`sqlalchemy.engine.URL` argument.</span>

<span class="sd">    If &quot;metavar&quot; is not defined it is added with &quot;URI&quot; as value to improve help text readability.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">kwargs</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;metavar&quot;</span><span class="p">,</span> <span class="s2">&quot;URI&quot;</span><span class="p">)</span>
    <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">make_url</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ArgumentParser.add_log_arguments" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_log_arguments</span><span class="p">(</span><span class="n">add_log_file</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.ArgumentParser.add_log_arguments" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds the usual set of arguments required to set and initialise a logging system.</p>
<p>The current set includes a mutually exclusive group for the default logging level: "--verbose",
"--debug" or "--log LEVEL".</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>add_log_file</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Add arguments to allow storing messages into a file, i.e. "--log_file" and
"--log_file_level".</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_log_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">add_log_file</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the usual set of arguments required to set and initialise a logging system.</span>

<span class="sd">    The current set includes a mutually exclusive group for the default logging level: &quot;--verbose&quot;,</span>
<span class="sd">    &quot;--debug&quot; or &quot;--log LEVEL&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        add_log_file: Add arguments to allow storing messages into a file, i.e. &quot;--log_file&quot; and</span>
<span class="sd">            &quot;--log_file_level&quot;.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define the list of log levels available</span>
    <span class="n">log_levels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span> <span class="s2">&quot;INFO&quot;</span><span class="p">,</span> <span class="s2">&quot;WARNING&quot;</span><span class="p">,</span> <span class="s2">&quot;ERROR&quot;</span><span class="p">,</span> <span class="s2">&quot;CRITICAL&quot;</span><span class="p">]</span>
    <span class="c1"># NOTE: from 3.11 this list can be changed to: logging.getLevelNamesMapping().keys()</span>
    <span class="c1"># Create logging arguments group</span>
    <span class="n">group</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;logging arguments&quot;</span><span class="p">)</span>
    <span class="c1"># Add 3 mutually exclusive options to set the logging level</span>
    <span class="n">subgroup</span> <span class="o">=</span> <span class="n">group</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">()</span>
    <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-v&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--verbose&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
        <span class="n">const</span><span class="o">=</span><span class="s2">&quot;INFO&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;verbose mode, i.e. &#39;INFO&#39; log level&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--debug&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
        <span class="n">const</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;debugging mode, i.e. &#39;DEBUG&#39; log level&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">subgroup</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--log&quot;</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="n">log_levels</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;LEVEL&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;level of the events to track: </span><span class="si">%(choices)s</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">subgroup</span><span class="o">.</span><span class="n">set_defaults</span><span class="p">(</span><span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">add_log_file</span><span class="p">:</span>
        <span class="c1"># Add log file-related arguments</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--log_file&quot;</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dst_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">exists_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
            <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PATH&quot;</span><span class="p">,</span>
            <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;log file path&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--log_file_level&quot;</span><span class="p">,</span>
            <span class="n">choices</span><span class="o">=</span><span class="n">log_levels</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">,</span>
            <span class="n">default</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
            <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;LEVEL&quot;</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;level of the events to track in the log file: </span><span class="si">%(choices)s</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ArgumentParser.add_server_arguments" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_server_arguments</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">include_database</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.ArgumentParser.add_server_arguments" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Adds the usual set of arguments needed to connect to a server.</p>
<p>This means: "--host", "--port", "--user" and "--password" (optional).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>prefix</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Prefix to add the each argument, e.g. if prefix is "src_", the arguments will be
"--src_host", etc.</p>
            </div>
          </td>
          <td>
                <code>&#39;&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>include_database</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Include "--database" argument.</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>help</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Description message to include for this set of arguments.</p>
            </div>
          </td>
          <td>
                <code>&#39;&#39;</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_server_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">include_database</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the usual set of arguments needed to connect to a server.</span>

<span class="sd">    This means: &quot;--host&quot;, &quot;--port&quot;, &quot;--user&quot; and &quot;--password&quot; (optional).</span>

<span class="sd">    Args:</span>
<span class="sd">        prefix: Prefix to add the each argument, e.g. if prefix is &quot;src_&quot;, the arguments will be</span>
<span class="sd">            &quot;--src_host&quot;, etc.</span>
<span class="sd">        include_database: Include &quot;--database&quot; argument.</span>
<span class="sd">        help: Description message to include for this set of arguments.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">group</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">server connection arguments&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="n">help</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">host&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;HOST&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;host name&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">port&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PORT&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;port number&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">user&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;USER&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;user name&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">password&quot;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;PWD&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;host password&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">include_database</span><span class="p">:</span>
        <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">database&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;NAME&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;database name&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ArgumentParser.parse_args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">parse_args</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.ArgumentParser.parse_args" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Extends the parent function by adding a new URL argument for every server group added.</p>
<p>The type of this new argument will be :class:<code>sqlalchemy.engine.URL</code>. It also logs all the parsed
arguments for debugging purposes when logging arguments have been added.</p>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/argparse.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">parse_args</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">argparse</span><span class="o">.</span><span class="n">Namespace</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extends the parent function by adding a new URL argument for every server group added.</span>

<span class="sd">    The type of this new argument will be :class:`sqlalchemy.engine.URL`. It also logs all the parsed</span>
<span class="sd">    arguments for debugging purposes when logging arguments have been added.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="c1"># Build and add an sqlalchemy.engine.URL object for every server group added</span>
    <span class="k">for</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">__server_groups</span><span class="p">:</span>
        <span class="c1"># Raise an error rather than overwriting when the URL argument is already present</span>
        <span class="k">if</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&quot;</span> <span class="ow">in</span> <span class="n">args</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;argument &#39;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&#39; is already present&quot;</span><span class="p">)</span>
        <span class="n">server_url</span> <span class="o">=</span> <span class="n">URL</span><span class="o">.</span><span class="n">create</span><span class="p">(</span>
            <span class="s2">&quot;mysql&quot;</span><span class="p">,</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">user&quot;</span><span class="p">),</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">password&quot;</span><span class="p">),</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">host&quot;</span><span class="p">),</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">port&quot;</span><span class="p">),</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">database&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">url&quot;</span><span class="p">,</span> <span class="n">server_url</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">args</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.manifest.BiotypeCounter" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>BiotypeCounter</code>


<a href="#ensembl.io.genomio.manifest.BiotypeCounter" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">

  
      <p>A counter for a given biotype, given a list of features.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">BiotypeCounter</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A counter for a given biotype, given a list of features.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">ids</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">example</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">count</span>
        <span class="k">if</span> <span class="n">ids</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">ids</span>
        <span class="k">if</span> <span class="n">example</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">example</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">example</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">example</span>

    <span class="k">def</span> <span class="nf">add_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a feature to the counter.</span>

<span class="sd">        Args:</span>
<span class="sd">            feature_id (str): Feature id to add.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">unique_count</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Total number feature ids added to the counter so far.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: number of features in the counter.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.BiotypeCounter.add_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_id</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.BiotypeCounter.add_id" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Add a feature to the counter.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>feature_id</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Feature id to add.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add a feature to the counter.</span>

<span class="sd">    Args:</span>
<span class="sd">        feature_id (str): Feature id to add.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.BiotypeCounter.unique_count" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">unique_count</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.manifest.BiotypeCounter.unique_count" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Total number feature ids added to the counter so far.</p>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>int</code></td>          <td>
                <code>int</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of features in the counter.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">unique_count</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Total number feature ids added to the counter so far.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: number of features in the counter.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.manifest.IntegrityTool" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>IntegrityTool</code>


<a href="#ensembl.io.genomio.manifest.IntegrityTool" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">

  
      <p>Check the integrity of sequence and annotation files in the genome</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">IntegrityTool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check the integrity of sequence and annotation files in the genome&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">brc_mode</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span> <span class="o">=</span> <span class="n">Manifest</span><span class="p">(</span><span class="n">manifest_file</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_brc_mode</span><span class="p">(</span><span class="n">brc_mode</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_ignore_final_stops</span><span class="p">(</span><span class="n">ignore_final_stops</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="nf">add_errors</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Store the given errors in the list.&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">args</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">errors</span> <span class="o">+=</span> <span class="n">error</span>

    <span class="k">def</span> <span class="nf">check_integrity</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Load files listed in the manifest.json and check the integrity.</span>
<span class="sd">        Check if the files are correct by verifying the MD5 hash.</span>
<span class="sd">        Check if translation, functional annotation and sequence region ids</span>
<span class="sd">        and lengths are consistent with the information in gff.</span>
<span class="sd">        Compare sequence length from fasta_dna file to seq_region.json metadata.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Load the manifest integrity counts</span>
        <span class="n">manifest</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span>
        <span class="n">manifest</span><span class="o">.</span><span class="n">prepare_integrity_data</span><span class="p">()</span>

        <span class="n">genome</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">genome</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">seq_regions</span>

        <span class="n">dna</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;dna_sequences&quot;</span><span class="p">)</span>
        <span class="n">pep</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">)</span>
        <span class="n">seq_lengths</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">)</span>
        <span class="n">seq_circular</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_circular</span><span class="p">(</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">)</span>

        <span class="n">agp_seqr</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;agp&quot;</span><span class="p">)</span>

        <span class="c1"># Then, run the checks</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_check_genome</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

        <span class="c1"># Check gff3</span>
        <span class="k">if</span> <span class="n">manifest</span><span class="o">.</span><span class="n">has_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_genes&quot;</span><span class="p">):</span>
            <span class="n">gff_genes</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_genes&quot;</span><span class="p">)</span>
            <span class="n">gff_seq_regions</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_seq_regions&quot;</span><span class="p">)</span>
            <span class="n">gff_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_translations&quot;</span><span class="p">)</span>
            <span class="n">gff_all_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_all_translations&quot;</span><span class="p">)</span>
            <span class="n">gff_transposable_elements</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_transposable_elements&quot;</span><span class="p">)</span>

            <span class="n">ann_genes</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_genes&quot;</span><span class="p">)</span>
            <span class="n">ann_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_translations&quot;</span><span class="p">)</span>
            <span class="n">ann_transposable_elements</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">)</span>

            <span class="c1"># Check fasta_pep.fa integrity</span>
            <span class="c1"># The sequence length and id retrieved from the fasta_pep file</span>
            <span class="c1"># and compared to the translated CDS id and length in the gff</span>
            <span class="c1"># We don&#39;t compare the peptide lengths because of seqedits</span>
            <span class="k">if</span> <span class="n">pep</span><span class="p">:</span>
                <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_lengths</span><span class="p">(</span>
                    <span class="n">pep</span><span class="p">,</span> <span class="n">gff_translations</span><span class="p">,</span> <span class="s2">&quot;Fasta translations vs gff&quot;</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">True</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tr_errors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># The pseudo CDSs are included in this check</span>
                    <span class="c1"># Pseudo CDSs are not translated, if the pseudo translation ids are not ignored</span>
                    <span class="c1"># in the gff it will give an error</span>
                    <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_lengths</span><span class="p">(</span>
                        <span class="n">pep</span><span class="p">,</span>
                        <span class="n">gff_all_translations</span><span class="p">,</span>
                        <span class="s2">&quot;Fasta translations vs gff (include pseudo CDS)&quot;</span><span class="p">,</span>
                        <span class="n">special_diff</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">tr_errors</span><span class="p">)</span>

            <span class="c1"># Check functional_annotation.json integrity</span>
            <span class="c1"># Gene ids, translated CDS ids and translated CDSs</span>
            <span class="c1"># including pseudogenes are compared to the gff</span>
            <span class="k">if</span> <span class="n">ann_genes</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span><span class="n">ann_genes</span><span class="p">,</span> <span class="n">gff_genes</span><span class="p">,</span> <span class="s2">&quot;Gene ids metadata vs gff&quot;</span><span class="p">)</span>
                <span class="n">found_tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                    <span class="n">ann_translations</span><span class="p">,</span> <span class="n">gff_translations</span><span class="p">,</span> <span class="s2">&quot;Translation ids metadata vs gff&quot;</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">found_tr_errors</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                        <span class="n">ann_translations</span><span class="p">,</span>
                        <span class="n">gff_all_translations</span><span class="p">,</span>
                        <span class="s2">&quot;Translation ids metadata vs gff (include pseudo CDS)&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                    <span class="n">ann_transposable_elements</span><span class="p">,</span>
                    <span class="n">gff_transposable_elements</span><span class="p">,</span>
                    <span class="s2">&quot;TE ids metadata vs gff&quot;</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="c1"># Check the seq.json intregrity</span>
            <span class="c1"># Compare the length and id retrieved from seq.json to the gff</span>
            <span class="k">if</span> <span class="n">seq_regions</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span>
                    <span class="n">seq_lengths</span><span class="p">,</span> <span class="n">gff_seq_regions</span><span class="p">,</span> <span class="s2">&quot;Seq_regions metadata vs gff&quot;</span><span class="p">,</span> <span class="n">seq_circular</span>
                <span class="p">)</span>

        <span class="c1"># Check fasta dna and seq_region integrity</span>
        <span class="k">if</span> <span class="n">dna</span> <span class="ow">and</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seq_lengths</span><span class="p">,</span> <span class="n">dna</span><span class="p">,</span> <span class="s2">&quot;seq_regions json vs dna&quot;</span><span class="p">)</span>

        <span class="c1"># Check agp and seq_region integrity</span>
        <span class="k">if</span> <span class="n">agp_seqr</span> <span class="ow">and</span> <span class="n">seq_lengths</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seq_lengths</span><span class="p">,</span> <span class="n">agp_seqr</span><span class="p">,</span> <span class="s2">&quot;seq_regions json vs agps&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">manifest</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span>
            <span class="n">errors_str</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">manifest</span><span class="o">.</span><span class="n">errors</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">InvalidIntegrityError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Integrity test failed:</span><span class="se">\n</span><span class="si">{</span><span class="n">errors_str</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">set_brc_mode</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">brc_mode</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set brc mode for this tool and the manifest.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="n">brc_mode</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="n">brc_mode</span>

    <span class="k">def</span> <span class="nf">set_ignore_final_stops</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set ignore_final_stops (when calculating peptide length) for this tool and the manifest.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="n">ignore_final_stops</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="n">ignore_final_stops</span>

    <span class="k">def</span> <span class="nf">_check_genome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if the accession is correct in genome.json.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">genome</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;assembly&quot;</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">:</span>
                <span class="n">genome_ass</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">]</span>
                <span class="k">if</span> <span class="s2">&quot;accession&quot;</span> <span class="ow">in</span> <span class="n">genome_ass</span><span class="p">:</span>
                    <span class="n">genome_acc</span> <span class="o">=</span> <span class="n">genome_ass</span><span class="p">[</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;GC[AF]_\d</span><span class="si">{9}</span><span class="s2">(\.\d+)?&quot;</span><span class="p">,</span> <span class="n">genome_acc</span><span class="p">):</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Genome assembly accession is wrong: &#39;</span><span class="si">{</span><span class="n">genome_acc</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">check_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compare the ids in list1 and list2.</span>

<span class="sd">        Args:</span>
<span class="sd">            list1: dict containing sequence ids retrieved from functional.json.</span>
<span class="sd">            list2: dict containing length and id in the retrieved from the gff.</span>
<span class="sd">            name:  string</span>

<span class="sd">        Return:</span>
<span class="sd">            Whether the checks found errors.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">only1</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">only2</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">common</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list1</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list2</span><span class="p">:</span>
                <span class="n">common</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">only1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list2</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">item_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">common</span><span class="p">:</span>
                <span class="n">only2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>

        <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">only1</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only1</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in first list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only1</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in first list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">only2</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only2</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only1</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">errors</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">errors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>

    <span class="k">def</span> <span class="nf">check_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">allowed_len_diff</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check the difference in ids and length between list1 and list2.</span>
<span class="sd">            There are a few special cases here where we allow a certain asymmetry</span>
<span class="sd">            by changing the values of the arguments.</span>

<span class="sd">        Args:</span>
<span class="sd">            list1: dict containing length and id of the sequence from fasta files.</span>
<span class="sd">            list2: dict containing length and id in the retrieved from the gff.</span>
<span class="sd">            name:  string</span>

<span class="sd">        allowed_len_diff : None to to not accept differences in length between list1 and list2.</span>
<span class="sd">            The value can be changed based on how much difference in sequence length we are wanting to accept.</span>

<span class="sd">        special_diff: set as False when no special length difference is expected between the lists.</span>
<span class="sd">                    This can be changed if we want to report common sequences with 1 BP difference.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Error if there is a difference in length or ids between the lists.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># check list diffferences, checks if abs(values diff) &lt; allowed_len_diff</span>

        <span class="n">set1</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">list1</span><span class="p">)</span>
        <span class="n">set2</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">list2</span><span class="p">)</span>
        <span class="n">list1_2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">set1</span> <span class="o">-</span> <span class="n">set2</span><span class="p">)</span>
        <span class="n">list2_1</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">set2</span> <span class="o">-</span> <span class="n">set1</span><span class="p">)</span>

        <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list1_2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">list1_2</span><span class="p">)</span><span class="si">}</span><span class="s2"> from the first list only (i.e. </span><span class="si">{</span><span class="n">list1_2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list2_1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">list2_1</span><span class="p">)</span><span class="si">}</span><span class="s2"> from the second list only (i.e. </span><span class="si">{</span><span class="n">list2_1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

        <span class="n">common_len</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">allowed_len_diff</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">common_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">set1</span> <span class="o">&amp;</span> <span class="n">set2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># check for the sequence length difference</span>
            <span class="n">diff_len_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">diff_len_special_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">set1</span> <span class="o">&amp;</span> <span class="n">set2</span><span class="p">:</span>
                <span class="n">dl12</span> <span class="o">=</span> <span class="n">list1</span><span class="p">[</span><span class="n">e</span><span class="p">]</span> <span class="o">-</span> <span class="n">list2</span><span class="p">[</span><span class="n">e</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">dl12</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">allowed_len_diff</span><span class="p">:</span>
                    <span class="n">common_len</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">_dlist</span> <span class="o">=</span> <span class="n">diff_len_list</span>
                    <span class="c1"># Special case: 1 AA /BP shorter,</span>
                    <span class="c1">#   so assuming the stop codon is not included in the CDS (when it should be)</span>
                    <span class="k">if</span> <span class="n">dl12</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">special_diff</span><span class="p">:</span>
                        <span class="n">_dlist</span> <span class="o">=</span> <span class="n">diff_len_special_list</span>
                    <span class="n">_dlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">list1</span><span class="p">[</span><span class="n">e</span><span class="p">]</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">list2</span><span class="p">[</span><span class="n">e</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">diff_len_special_list</span><span class="p">:</span>
                <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_len_special_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with one BP/AA length diff for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;(e.g. </span><span class="si">{</span><span class="n">diff_len_special_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="n">diff_len_list</span><span class="p">:</span>
                <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_len_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with length diff for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;(e.g. </span><span class="si">{</span><span class="n">diff_len_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>
                    <span class="p">)</span>
                <span class="p">)</span>
        <span class="k">if</span> <span class="n">common_len</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">common_len</span><span class="si">}</span><span class="s2"> common elements between lists for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">errors</span>

    <span class="k">def</span> <span class="nf">check_seq_region_lengths</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">seqrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
        <span class="n">feats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
        <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">circular</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check the integrity of seq_region.json file by comparing the length of the sequence</span>
<span class="sd">            to fasta files and the gff.</span>

<span class="sd">            Seq_region file is in json format containing the metadata of the sequence.</span>
<span class="sd">            It contains sequence id, length, location and the synonyms for the sequence name</span>
<span class="sd">            from different sources.</span>

<span class="sd">        Args:</span>
<span class="sd">            seqs: Sequence name and length retrieved from seq_region.json file.</span>
<span class="sd">            feats: Sequence name and length retrieved from the fasta and gff file.</span>
<span class="sd">            name: Name of the check to show in the logs.</span>
<span class="sd">            circular: Whether any sequence is circular.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Error if there are common sequences with difference in ids</span>
<span class="sd">            and if the sequences are not consistent in the files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">comp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compare_seqs</span><span class="p">(</span><span class="n">seqrs</span><span class="p">,</span> <span class="n">feats</span><span class="p">,</span> <span class="n">circular</span><span class="p">)</span>

        <span class="n">common</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;common&quot;</span><span class="p">]</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">]</span>
        <span class="n">diff_circular</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff_circular&quot;</span><span class="p">]</span>
        <span class="n">only_seqr</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_seqr&quot;</span><span class="p">]</span>
        <span class="n">only_feat</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_feat&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_circular</span><span class="p">:</span>
            <span class="n">example</span> <span class="o">=</span> <span class="n">diff_circular</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_circular</span><span class="p">)</span><span class="si">}</span><span class="s2"> differences for circular elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (e.g. </span><span class="si">{</span><span class="n">example</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">diff</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with higher length in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (e.g. </span><span class="si">{</span><span class="n">diff</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">only_seqr</span><span class="p">:</span>
            <span class="c1"># Not an error!</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only_seqr</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in seq_region list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only_seqr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">only_feat</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only_feat</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only_feat</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_compare_seqs</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">seqrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="n">feats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="n">circular</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Give the intersection and other comparison between two groups of sequences.</span>

<span class="sd">        Args:</span>
<span class="sd">            seqs: Sequence name and length retrieved from seq_region.json file.</span>
<span class="sd">            feats: Sequence name and length retrieved from the fasta and gff file.</span>
<span class="sd">            circular: Whether any sequence is circular.</span>

<span class="sd">        Returns: Dict with 5 stats:</span>
<span class="sd">            common: Common elements.</span>
<span class="sd">            only_seqr: Elements only in the first one.</span>
<span class="sd">            only_feat: Elements only in the second one.</span>
<span class="sd">            diff: Elements that differ.</span>
<span class="sd">            diff_circular: Elements that differ in a circular sequence.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">comp</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;common&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;only_seqr&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;only_feat&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;diff&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;diff_circular&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">seq_id</span> <span class="ow">in</span> <span class="n">seqrs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">seq_id</span> <span class="ow">in</span> <span class="n">feats</span><span class="p">:</span>
                <span class="c1"># Check that feature is within the seq_region length</span>
                <span class="k">if</span> <span class="n">feats</span><span class="p">[</span><span class="n">seq_id</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">seqrs</span><span class="p">[</span><span class="n">seq_id</span><span class="p">]:</span>
                    <span class="n">diff_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seq_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">seqrs</span><span class="p">[</span><span class="n">seq_id</span><span class="p">]</span><span class="si">}</span><span class="s2"> vs </span><span class="si">{</span><span class="n">feats</span><span class="p">[</span><span class="n">seq_id</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="k">if</span> <span class="n">circular</span> <span class="ow">and</span> <span class="n">circular</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">seq_id</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
                        <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff_circular&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">diff_str</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">diff_str</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;common&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_id</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_seqr&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_id</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">seq_id</span> <span class="ow">in</span> <span class="n">feats</span><span class="p">:</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">seq_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;common&quot;</span><span class="p">]</span>
                <span class="ow">and</span> <span class="n">seq_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">]</span>
                <span class="ow">and</span> <span class="n">seq_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff_circular&quot;</span><span class="p">]</span>
            <span class="p">):</span>
                <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_feat&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_id</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">comp</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.IntegrityTool.add_errors" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.IntegrityTool.add_errors" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Store the given errors in the list.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_errors</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Store the given errors in the list.&quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">args</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">errors</span> <span class="o">+=</span> <span class="n">error</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.IntegrityTool.check_ids" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">check_ids</span><span class="p">(</span><span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.IntegrityTool.check_ids" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Compare the ids in list1 and list2.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>list1</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing sequence ids retrieved from functional.json.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>list2</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing length and id in the retrieved from the gff.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>name</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>string</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="return" open>
  <summary>Return</summary>
  <p>Whether the checks found errors.</p>
</details>
          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compare the ids in list1 and list2.</span>

<span class="sd">    Args:</span>
<span class="sd">        list1: dict containing sequence ids retrieved from functional.json.</span>
<span class="sd">        list2: dict containing length and id in the retrieved from the gff.</span>
<span class="sd">        name:  string</span>

<span class="sd">    Return:</span>
<span class="sd">        Whether the checks found errors.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">only1</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">only2</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">common</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list2</span><span class="p">:</span>
            <span class="n">common</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">only1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list2</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">item_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">common</span><span class="p">:</span>
            <span class="n">only2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>

    <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">only1</span><span class="p">:</span>
        <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only1</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in first list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only1</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in first list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">only2</span><span class="p">:</span>
        <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only2</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only1</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">errors</span><span class="p">)</span>

    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">errors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.IntegrityTool.check_integrity" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">check_integrity</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.manifest.IntegrityTool.check_integrity" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Load files listed in the manifest.json and check the integrity.
Check if the files are correct by verifying the MD5 hash.
Check if translation, functional annotation and sequence region ids
and lengths are consistent with the information in gff.
Compare sequence length from fasta_dna file to seq_region.json metadata.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_integrity</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load files listed in the manifest.json and check the integrity.</span>
<span class="sd">    Check if the files are correct by verifying the MD5 hash.</span>
<span class="sd">    Check if translation, functional annotation and sequence region ids</span>
<span class="sd">    and lengths are consistent with the information in gff.</span>
<span class="sd">    Compare sequence length from fasta_dna file to seq_region.json metadata.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Load the manifest integrity counts</span>
    <span class="n">manifest</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span>
    <span class="n">manifest</span><span class="o">.</span><span class="n">prepare_integrity_data</span><span class="p">()</span>

    <span class="n">genome</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">genome</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">seq_regions</span>

    <span class="n">dna</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;dna_sequences&quot;</span><span class="p">)</span>
    <span class="n">pep</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">)</span>
    <span class="n">seq_lengths</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">)</span>
    <span class="n">seq_circular</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_circular</span><span class="p">(</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">)</span>

    <span class="n">agp_seqr</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;agp&quot;</span><span class="p">)</span>

    <span class="c1"># Then, run the checks</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_check_genome</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

    <span class="c1"># Check gff3</span>
    <span class="k">if</span> <span class="n">manifest</span><span class="o">.</span><span class="n">has_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_genes&quot;</span><span class="p">):</span>
        <span class="n">gff_genes</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_genes&quot;</span><span class="p">)</span>
        <span class="n">gff_seq_regions</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_seq_regions&quot;</span><span class="p">)</span>
        <span class="n">gff_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_translations&quot;</span><span class="p">)</span>
        <span class="n">gff_all_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_all_translations&quot;</span><span class="p">)</span>
        <span class="n">gff_transposable_elements</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff3_transposable_elements&quot;</span><span class="p">)</span>

        <span class="n">ann_genes</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_genes&quot;</span><span class="p">)</span>
        <span class="n">ann_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_translations&quot;</span><span class="p">)</span>
        <span class="n">ann_transposable_elements</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">)</span>

        <span class="c1"># Check fasta_pep.fa integrity</span>
        <span class="c1"># The sequence length and id retrieved from the fasta_pep file</span>
        <span class="c1"># and compared to the translated CDS id and length in the gff</span>
        <span class="c1"># We don&#39;t compare the peptide lengths because of seqedits</span>
        <span class="k">if</span> <span class="n">pep</span><span class="p">:</span>
            <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_lengths</span><span class="p">(</span>
                <span class="n">pep</span><span class="p">,</span> <span class="n">gff_translations</span><span class="p">,</span> <span class="s2">&quot;Fasta translations vs gff&quot;</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tr_errors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># The pseudo CDSs are included in this check</span>
                <span class="c1"># Pseudo CDSs are not translated, if the pseudo translation ids are not ignored</span>
                <span class="c1"># in the gff it will give an error</span>
                <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_lengths</span><span class="p">(</span>
                    <span class="n">pep</span><span class="p">,</span>
                    <span class="n">gff_all_translations</span><span class="p">,</span>
                    <span class="s2">&quot;Fasta translations vs gff (include pseudo CDS)&quot;</span><span class="p">,</span>
                    <span class="n">special_diff</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">tr_errors</span><span class="p">)</span>

        <span class="c1"># Check functional_annotation.json integrity</span>
        <span class="c1"># Gene ids, translated CDS ids and translated CDSs</span>
        <span class="c1"># including pseudogenes are compared to the gff</span>
        <span class="k">if</span> <span class="n">ann_genes</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span><span class="n">ann_genes</span><span class="p">,</span> <span class="n">gff_genes</span><span class="p">,</span> <span class="s2">&quot;Gene ids metadata vs gff&quot;</span><span class="p">)</span>
            <span class="n">found_tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                <span class="n">ann_translations</span><span class="p">,</span> <span class="n">gff_translations</span><span class="p">,</span> <span class="s2">&quot;Translation ids metadata vs gff&quot;</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">found_tr_errors</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                    <span class="n">ann_translations</span><span class="p">,</span>
                    <span class="n">gff_all_translations</span><span class="p">,</span>
                    <span class="s2">&quot;Translation ids metadata vs gff (include pseudo CDS)&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                <span class="n">ann_transposable_elements</span><span class="p">,</span>
                <span class="n">gff_transposable_elements</span><span class="p">,</span>
                <span class="s2">&quot;TE ids metadata vs gff&quot;</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="c1"># Check the seq.json intregrity</span>
        <span class="c1"># Compare the length and id retrieved from seq.json to the gff</span>
        <span class="k">if</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span>
                <span class="n">seq_lengths</span><span class="p">,</span> <span class="n">gff_seq_regions</span><span class="p">,</span> <span class="s2">&quot;Seq_regions metadata vs gff&quot;</span><span class="p">,</span> <span class="n">seq_circular</span>
            <span class="p">)</span>

    <span class="c1"># Check fasta dna and seq_region integrity</span>
    <span class="k">if</span> <span class="n">dna</span> <span class="ow">and</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seq_lengths</span><span class="p">,</span> <span class="n">dna</span><span class="p">,</span> <span class="s2">&quot;seq_regions json vs dna&quot;</span><span class="p">)</span>

    <span class="c1"># Check agp and seq_region integrity</span>
    <span class="k">if</span> <span class="n">agp_seqr</span> <span class="ow">and</span> <span class="n">seq_lengths</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seq_lengths</span><span class="p">,</span> <span class="n">agp_seqr</span><span class="p">,</span> <span class="s2">&quot;seq_regions json vs agps&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">manifest</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span>
        <span class="n">errors_str</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">manifest</span><span class="o">.</span><span class="n">errors</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">InvalidIntegrityError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Integrity test failed:</span><span class="se">\n</span><span class="si">{</span><span class="n">errors_str</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.IntegrityTool.check_lengths" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">check_lengths</span><span class="p">(</span><span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">allowed_len_diff</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.IntegrityTool.check_lengths" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Check the difference in ids and length between list1 and list2.
    There are a few special cases here where we allow a certain asymmetry
    by changing the values of the arguments.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>list1</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing length and id of the sequence from fasta files.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>list2</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing length and id in the retrieved from the gff.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>name</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>string</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="allowed_len_diff-" open>
  <summary>None to to not accept differences in length between list1 and list2.</summary>
  <p>The value can be changed based on how much difference in sequence length we are wanting to accept.</p>
</details>
<details class="special_diff" open>
  <summary>set as False when no special length difference is expected between the lists.</summary>
  <p>This can be changed if we want to report common sequences with 1 BP difference.</p>
</details>


  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Error if there is a difference in length or ids between the lists.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">allowed_len_diff</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check the difference in ids and length between list1 and list2.</span>
<span class="sd">        There are a few special cases here where we allow a certain asymmetry</span>
<span class="sd">        by changing the values of the arguments.</span>

<span class="sd">    Args:</span>
<span class="sd">        list1: dict containing length and id of the sequence from fasta files.</span>
<span class="sd">        list2: dict containing length and id in the retrieved from the gff.</span>
<span class="sd">        name:  string</span>

<span class="sd">    allowed_len_diff : None to to not accept differences in length between list1 and list2.</span>
<span class="sd">        The value can be changed based on how much difference in sequence length we are wanting to accept.</span>

<span class="sd">    special_diff: set as False when no special length difference is expected between the lists.</span>
<span class="sd">                This can be changed if we want to report common sequences with 1 BP difference.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Error if there is a difference in length or ids between the lists.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># check list diffferences, checks if abs(values diff) &lt; allowed_len_diff</span>

    <span class="n">set1</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">list1</span><span class="p">)</span>
    <span class="n">set2</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">list2</span><span class="p">)</span>
    <span class="n">list1_2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">set1</span> <span class="o">-</span> <span class="n">set2</span><span class="p">)</span>
    <span class="n">list2_1</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">set2</span> <span class="o">-</span> <span class="n">set1</span><span class="p">)</span>

    <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list1_2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">list1_2</span><span class="p">)</span><span class="si">}</span><span class="s2"> from the first list only (i.e. </span><span class="si">{</span><span class="n">list1_2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list2_1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">list2_1</span><span class="p">)</span><span class="si">}</span><span class="s2"> from the second list only (i.e. </span><span class="si">{</span><span class="n">list2_1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

    <span class="n">common_len</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">allowed_len_diff</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">common_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">set1</span> <span class="o">&amp;</span> <span class="n">set2</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># check for the sequence length difference</span>
        <span class="n">diff_len_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">diff_len_special_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">set1</span> <span class="o">&amp;</span> <span class="n">set2</span><span class="p">:</span>
            <span class="n">dl12</span> <span class="o">=</span> <span class="n">list1</span><span class="p">[</span><span class="n">e</span><span class="p">]</span> <span class="o">-</span> <span class="n">list2</span><span class="p">[</span><span class="n">e</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">dl12</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">allowed_len_diff</span><span class="p">:</span>
                <span class="n">common_len</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">_dlist</span> <span class="o">=</span> <span class="n">diff_len_list</span>
                <span class="c1"># Special case: 1 AA /BP shorter,</span>
                <span class="c1">#   so assuming the stop codon is not included in the CDS (when it should be)</span>
                <span class="k">if</span> <span class="n">dl12</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">special_diff</span><span class="p">:</span>
                    <span class="n">_dlist</span> <span class="o">=</span> <span class="n">diff_len_special_list</span>
                <span class="n">_dlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">list1</span><span class="p">[</span><span class="n">e</span><span class="p">]</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">list2</span><span class="p">[</span><span class="n">e</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_len_special_list</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_len_special_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with one BP/AA length diff for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;(e.g. </span><span class="si">{</span><span class="n">diff_len_special_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_len_list</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_len_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with length diff for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;(e.g. </span><span class="si">{</span><span class="n">diff_len_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
            <span class="p">)</span>
    <span class="k">if</span> <span class="n">common_len</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">common_len</span><span class="si">}</span><span class="s2"> common elements between lists for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">errors</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.IntegrityTool.check_seq_region_lengths" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seqrs</span><span class="p">,</span> <span class="n">feats</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.IntegrityTool.check_seq_region_lengths" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Check the integrity of seq_region.json file by comparing the length of the sequence
    to fasta files and the gff.</p>
<pre><code>Seq_region file is in json format containing the metadata of the sequence.
It contains sequence id, length, location and the synonyms for the sequence name
from different sources.
</code></pre>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seqs</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence name and length retrieved from seq_region.json file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>feats</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence name and length retrieved from the fasta and gff file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>name</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Name of the check to show in the logs.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>circular</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Whether any sequence is circular.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>None</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Error if there are common sequences with difference in ids</p>
            </div>
          </td>
        </tr>
        <tr>
          <td>
                <code>None</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>and if the sequences are not consistent in the files.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_seq_region_lengths</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">seqrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">feats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">circular</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check the integrity of seq_region.json file by comparing the length of the sequence</span>
<span class="sd">        to fasta files and the gff.</span>

<span class="sd">        Seq_region file is in json format containing the metadata of the sequence.</span>
<span class="sd">        It contains sequence id, length, location and the synonyms for the sequence name</span>
<span class="sd">        from different sources.</span>

<span class="sd">    Args:</span>
<span class="sd">        seqs: Sequence name and length retrieved from seq_region.json file.</span>
<span class="sd">        feats: Sequence name and length retrieved from the fasta and gff file.</span>
<span class="sd">        name: Name of the check to show in the logs.</span>
<span class="sd">        circular: Whether any sequence is circular.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Error if there are common sequences with difference in ids</span>
<span class="sd">        and if the sequences are not consistent in the files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">comp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compare_seqs</span><span class="p">(</span><span class="n">seqrs</span><span class="p">,</span> <span class="n">feats</span><span class="p">,</span> <span class="n">circular</span><span class="p">)</span>

    <span class="n">common</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;common&quot;</span><span class="p">]</span>
    <span class="n">diff</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">]</span>
    <span class="n">diff_circular</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff_circular&quot;</span><span class="p">]</span>
    <span class="n">only_seqr</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_seqr&quot;</span><span class="p">]</span>
    <span class="n">only_feat</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_feat&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">diff_circular</span><span class="p">:</span>
        <span class="n">example</span> <span class="o">=</span> <span class="n">diff_circular</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_circular</span><span class="p">)</span><span class="si">}</span><span class="s2"> differences for circular elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (e.g. </span><span class="si">{</span><span class="n">example</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">diff</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with higher length in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (e.g. </span><span class="si">{</span><span class="n">diff</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">only_seqr</span><span class="p">:</span>
        <span class="c1"># Not an error!</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only_seqr</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in seq_region list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only_seqr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">only_feat</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only_feat</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only_feat</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.IntegrityTool.set_brc_mode" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">set_brc_mode</span><span class="p">(</span><span class="n">brc_mode</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.IntegrityTool.set_brc_mode" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Set brc mode for this tool and the manifest.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">set_brc_mode</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">brc_mode</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Set brc mode for this tool and the manifest.&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="n">brc_mode</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="n">brc_mode</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.IntegrityTool.set_ignore_final_stops" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">set_ignore_final_stops</span><span class="p">(</span><span class="n">ignore_final_stops</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.IntegrityTool.set_ignore_final_stops" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Set ignore_final_stops (when calculating peptide length) for this tool and the manifest.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">set_ignore_final_stops</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Set ignore_final_stops (when calculating peptide length) for this tool and the manifest.&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="n">ignore_final_stops</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="n">ignore_final_stops</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.manifest.InvalidIntegrityError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>InvalidIntegrityError</code>


<a href="#ensembl.io.genomio.manifest.InvalidIntegrityError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>When a file integrity check fails</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">42</span>
<span class="normal">43</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">InvalidIntegrityError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;When a file integrity check fails&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.manifest.Manifest" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>Manifest</code>


<a href="#ensembl.io.genomio.manifest.Manifest" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">

  
      <p>Representation of the manifest and its files.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Manifest</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Representation of the manifest and its files.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_manifest</span><span class="p">(</span><span class="n">manifest_path</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_regions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;dna_sequences&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;peptide_sequences&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;seq_region_levels&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;annotations&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;agp&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_seq_regions&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_genes&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_translations&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_all_translations&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_transposable_elements&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;ann_genes&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;ann_translations&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;seq_regions&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;seq_regions&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">has_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if a given name has lengths records.</span>

<span class="sd">        Raise KeyError if the name is not supported.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                <span class="k">return</span> <span class="kc">True</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There is no length record for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>

    <span class="k">def</span> <span class="nf">get_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a dict associating IDs with their length from a given file name.&quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There is no length record for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>

    <span class="k">def</span> <span class="nf">get_circular</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a dict associating IDs with their is_circular flag from a given file name.&quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No length available for key </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>

    <span class="k">def</span> <span class="nf">_add_error</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">error</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">error</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Load the content of a manifest file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict: Content of the manifest file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">manifest_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">manifest_path</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">manifest_fh</span><span class="p">:</span>
            <span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">manifest_fh</span><span class="p">)</span>

            <span class="c1"># Use dir name from the manifest</span>
            <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
                <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                    <span class="n">file_path</span> <span class="o">=</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                    <span class="c1"># check if the md5sum is correct</span>
                    <span class="n">md5sum</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;md5sum&quot;</span><span class="p">]</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_check_md5sum</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">)</span>

                    <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_path</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                        <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]:</span>
                            <span class="n">file_path</span> <span class="o">=</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                            <span class="c1"># check if the md5sum is correct</span>
                            <span class="n">md5sum</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;md5sum&quot;</span><span class="p">]</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_check_md5sum</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">)</span>

                            <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_path</span>
            <span class="k">return</span> <span class="n">manifest</span>

    <span class="k">def</span> <span class="nf">_check_md5sum</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Verify the integrity of the files in manifest.json.</span>

<span class="sd">        An MD5 hash is generated using the path provided which is then compared to the hash in manifest.json.</span>
<span class="sd">        Errors are stored in ``self.errors``.</span>

<span class="sd">        Args:</span>
<span class="sd">            file_path: Path to a genome file.</span>
<span class="sd">            md5sum: MD5 hash for the files.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="n">file_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">bytes_obj</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
            <span class="n">readable_hash</span> <span class="o">=</span> <span class="n">hashlib</span><span class="o">.</span><span class="n">md5</span><span class="p">(</span><span class="n">bytes_obj</span><span class="p">)</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">readable_hash</span> <span class="o">!=</span> <span class="n">md5sum</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">InvalidIntegrityError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid md5 checksum for </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">prepare_integrity_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Read all the files and keep a record (IDs and their lengths)</span>
<span class="sd">        for each cases to be compared later.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># First, get the Data</span>
        <span class="k">if</span> <span class="s2">&quot;gff3&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains GFF3&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_gff3</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;fasta_dna&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains DNA fasta&quot;</span><span class="p">)</span>
            <span class="c1"># Verify if the length and id for the sequence is unique</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;dna_sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_fasta_lengths</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;fasta_dna&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;fasta_pep&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains Peptide fasta&quot;</span><span class="p">)</span>
            <span class="c1"># Verify if the length and id for the sequence is unique</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_fasta_lengths</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;fasta_pep&quot;</span><span class="p">],</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;seq_region&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains seq_region JSON&quot;</span><span class="p">)</span>
            <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]))</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No sequences found in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s1">&#39;seq_region&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">seqr_seqlevel</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">seq_lengths</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">seq_circular</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="c1"># Store the length as int</span>
                <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
                    <span class="n">seq_lengths</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>
                    <span class="n">seq_circular</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;circular&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;contig&quot;</span><span class="p">:</span>
                        <span class="n">seqr_seqlevel</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_lengths</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">[</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_circular</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">seq_regions</span> <span class="o">=</span> <span class="n">seq_regions</span>
        <span class="k">if</span> <span class="s2">&quot;functional_annotation&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains functional annotation(s)&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_functional_annotation</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;functional_annotation&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;agp&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains AGP files&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;agp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_agp_seq_regions</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;agp&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;genome&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains genome JSON&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]))</span>

    <span class="k">def</span> <span class="nf">get_fasta_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if the fasta files have the correct ids and no stop codon.</span>

<span class="sd">        Args:</span>
<span class="sd">            fasta_path: Path to fasta_dna and fasta_pep files.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Error if any empty ids, non-unique ids or stop codons are found in the fasta files.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">non_unique</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">non_unique_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">empty_id_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">contains_stop_codon</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">rec_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
            <span class="n">rec_count</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># Flag empty ids</span>
            <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">empty_id_count</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Flag redundant ids</span>
                <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
                    <span class="n">non_unique</span><span class="p">[</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="n">non_unique_count</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="c1"># Store sequence id and length</span>
                <span class="n">data</span><span class="p">[</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
                <span class="n">stops</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">stops</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">contains_stop_codon</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">stops</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">ignore_final_stops</span><span class="p">:</span>
                        <span class="n">contains_stop_codon</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">empty_id_count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">empty_id_count</span><span class="si">}</span><span class="s2"> sequences with empty ids in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">non_unique_count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">non_unique_count</span><span class="si">}</span><span class="s2"> non unique sequence ids in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">contains_stop_codon</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">contains_stop_codon</span><span class="si">}</span><span class="s2"> sequences with stop codons in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">rec_count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No sequences found in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">data</span>

    <span class="k">def</span> <span class="nf">get_functional_annotation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">json_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Load the functional annotation file to retrieve the gene_id and translation id.</span>
<span class="sd">            A functional annotation file contains information about a gene.</span>
<span class="sd">            The functional annotation file is stored in a json format containing</span>
<span class="sd">            the description, id and object type (eg: &quot;gene&quot;, &quot;transcript&quot;, &quot;translation&quot;).</span>

<span class="sd">        Args:</span>
<span class="sd">            json_path: Path to functional_annotation.json.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict with gene and translation ids.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Load the json file</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

        <span class="c1"># Get gene ids and translation ids</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">translations</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">transposons</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">:</span>
                <span class="n">genes</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;translation&quot;</span><span class="p">:</span>
                <span class="n">translations</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
                <span class="n">transposons</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;ann_genes&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span>
            <span class="s2">&quot;ann_translations&quot;</span><span class="p">:</span> <span class="n">translations</span><span class="p">,</span>
            <span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">:</span> <span class="n">transposons</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">,</span> <span class="o">**</span><span class="n">stats</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">get_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;A GFF parser is used to retrieve information in the GFF file such as</span>
<span class="sd">           gene and CDS ids and their corresponding lengths.</span>

<span class="sd">        Args:</span>
<span class="sd">            gff3_path: Path to gff3 file.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">seqs</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">genes</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">peps</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">all_peps</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">tes</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff3_handle</span><span class="p">:</span>
            <span class="n">gff</span> <span class="o">=</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gff3_handle</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">gff</span><span class="p">:</span>
                <span class="n">seqs</span><span class="p">[</span><span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

                <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                    <span class="n">feat_length</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                    <span class="c1"># Store gene id and length</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;ncRNA_gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">]:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_retrieve_gff_gene_lengths</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">peps</span><span class="p">,</span> <span class="n">all_peps</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
                        <span class="n">tes</span><span class="p">[</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_length</span>

        <span class="n">stats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;gff3_seq_regions&quot;</span><span class="p">:</span> <span class="n">seqs</span><span class="p">,</span>
            <span class="s2">&quot;gff3_genes&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span>
            <span class="s2">&quot;gff3_translations&quot;</span><span class="p">:</span> <span class="n">peps</span><span class="p">,</span>
            <span class="s2">&quot;gff3_all_translations&quot;</span><span class="p">:</span> <span class="n">all_peps</span><span class="p">,</span>
            <span class="s2">&quot;gff3_transposable_elements&quot;</span><span class="p">:</span> <span class="n">tes</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">,</span> <span class="o">**</span><span class="n">stats</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">_retrieve_gff_gene_lengths</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="n">Lengths</span><span class="p">,</span> <span class="n">peps</span><span class="p">:</span> <span class="n">Lengths</span><span class="p">,</span> <span class="n">all_peps</span><span class="p">:</span> <span class="n">Lengths</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Record genes and peptides lengths from a feature.</span>

<span class="sd">        Args:</span>
<span class="sd">            feat : Gene feature to check.</span>
<span class="sd">            genes: Record of genes lengths to update.</span>
<span class="sd">            peps: Record of peptides lengths to update.</span>
<span class="sd">            all_peps: Record of all peptides lengths to update (include pseudogenes).</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gene_id</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span><span class="p">:</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">gene_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;gene:&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">genes</span><span class="p">[</span><span class="n">gene_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
        <span class="c1"># Get CDS id and length</span>
        <span class="k">for</span> <span class="n">feat2</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat2</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogenic_transcript&quot;</span><span class="p">):</span>
                <span class="n">length</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">feat3</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">feat3</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                        <span class="n">pep_id</span> <span class="o">=</span> <span class="n">feat3</span><span class="o">.</span><span class="n">id</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span><span class="p">:</span>
                            <span class="n">pep_id</span> <span class="o">=</span> <span class="n">pep_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;CDS:&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">pep_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">length</span><span class="p">:</span>
                            <span class="n">length</span><span class="p">[</span><span class="n">pep_id</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
                        <span class="n">length</span><span class="p">[</span><span class="n">pep_id</span><span class="p">]</span> <span class="o">+=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">feat3</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">feat3</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">pep_id</span><span class="p">,</span> <span class="n">pep_length</span> <span class="ow">in</span> <span class="n">length</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="c1"># Store length for translations, add pseudo translations separately</span>
                    <span class="n">pep_length</span> <span class="o">=</span> <span class="n">floor</span><span class="p">(</span><span class="n">pep_length</span> <span class="o">/</span> <span class="mi">3</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">:</span>
                        <span class="n">peps</span><span class="p">[</span><span class="n">pep_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">pep_length</span>
                    <span class="n">all_peps</span><span class="p">[</span><span class="n">pep_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">pep_length</span>

    <span class="k">def</span> <span class="nf">get_agp_seq_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">agp_dict</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;AGP files describe the assembly of larger sequence objects using smaller objects.</span>
<span class="sd">            Eg: describes the assembly of scaffolds from contigs.</span>

<span class="sd">        Args:</span>
<span class="sd">            agp_dict: dict containing the information about the sequence.</span>

<span class="sd">        Note:</span>
<span class="sd">            AGP file is only used in the older builds, not used for current processing.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">seqr</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">agp</span> <span class="ow">in</span> <span class="n">agp_dict</span><span class="p">:</span>
            <span class="n">agp_path</span> <span class="o">=</span> <span class="n">agp_dict</span><span class="p">[</span><span class="n">agp</span><span class="p">]</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">agp_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">agph</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">agph</span><span class="p">:</span>
                    <span class="p">(</span>
                        <span class="n">asm_id</span><span class="p">,</span>
                        <span class="n">_</span><span class="p">,</span>  <span class="c1"># asm_start</span>
                        <span class="n">asm_end</span><span class="p">,</span>
                        <span class="n">_</span><span class="p">,</span>  <span class="c1"># asm_part</span>
                        <span class="n">typ</span><span class="p">,</span>
                        <span class="n">cmp_id</span><span class="p">,</span>
                        <span class="n">_</span><span class="p">,</span>  <span class="c1"># cmp_start</span>
                        <span class="n">cmp_end</span><span class="p">,</span>
                        <span class="n">_</span><span class="p">,</span>  <span class="c1"># cmp_strand</span>
                    <span class="p">)</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="c1"># Ignore WGS contig</span>
                    <span class="k">if</span> <span class="n">typ</span> <span class="o">!=</span> <span class="s2">&quot;W&quot;</span><span class="p">:</span>
                        <span class="k">continue</span>

                    <span class="c1"># Assembled seq length</span>
                    <span class="k">if</span> <span class="n">asm_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="n">asm_id</span><span class="p">]</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">asm_end</span><span class="p">):</span>
                        <span class="n">seqr</span><span class="p">[</span><span class="n">asm_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">asm_end</span><span class="p">)</span>

                    <span class="c1"># Composite seq length</span>
                    <span class="k">if</span> <span class="n">cmp_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="n">cmp_id</span><span class="p">]</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">cmp_end</span><span class="p">):</span>
                        <span class="n">seqr</span><span class="p">[</span><span class="n">cmp_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cmp_end</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">seqr</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.Manifest.get_agp_seq_regions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_agp_seq_regions</span><span class="p">(</span><span class="n">agp_dict</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.Manifest.get_agp_seq_regions" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>AGP files describe the assembly of larger sequence objects using smaller objects.
    Eg: describes the assembly of scaffolds from contigs.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>agp_dict</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing the information about the sequence.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="note" open>
  <summary>Note</summary>
  <p>AGP file is only used in the older builds, not used for current processing.</p>
</details>
          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_agp_seq_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">agp_dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;AGP files describe the assembly of larger sequence objects using smaller objects.</span>
<span class="sd">        Eg: describes the assembly of scaffolds from contigs.</span>

<span class="sd">    Args:</span>
<span class="sd">        agp_dict: dict containing the information about the sequence.</span>

<span class="sd">    Note:</span>
<span class="sd">        AGP file is only used in the older builds, not used for current processing.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">seqr</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">agp</span> <span class="ow">in</span> <span class="n">agp_dict</span><span class="p">:</span>
        <span class="n">agp_path</span> <span class="o">=</span> <span class="n">agp_dict</span><span class="p">[</span><span class="n">agp</span><span class="p">]</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">agp_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">agph</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">agph</span><span class="p">:</span>
                <span class="p">(</span>
                    <span class="n">asm_id</span><span class="p">,</span>
                    <span class="n">_</span><span class="p">,</span>  <span class="c1"># asm_start</span>
                    <span class="n">asm_end</span><span class="p">,</span>
                    <span class="n">_</span><span class="p">,</span>  <span class="c1"># asm_part</span>
                    <span class="n">typ</span><span class="p">,</span>
                    <span class="n">cmp_id</span><span class="p">,</span>
                    <span class="n">_</span><span class="p">,</span>  <span class="c1"># cmp_start</span>
                    <span class="n">cmp_end</span><span class="p">,</span>
                    <span class="n">_</span><span class="p">,</span>  <span class="c1"># cmp_strand</span>
                <span class="p">)</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="c1"># Ignore WGS contig</span>
                <span class="k">if</span> <span class="n">typ</span> <span class="o">!=</span> <span class="s2">&quot;W&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="c1"># Assembled seq length</span>
                <span class="k">if</span> <span class="n">asm_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="n">asm_id</span><span class="p">]</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">asm_end</span><span class="p">):</span>
                    <span class="n">seqr</span><span class="p">[</span><span class="n">asm_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">asm_end</span><span class="p">)</span>

                <span class="c1"># Composite seq length</span>
                <span class="k">if</span> <span class="n">cmp_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="n">cmp_id</span><span class="p">]</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">cmp_end</span><span class="p">):</span>
                    <span class="n">seqr</span><span class="p">[</span><span class="n">cmp_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cmp_end</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">seqr</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.Manifest.get_circular" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_circular</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.Manifest.get_circular" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns a dict associating IDs with their is_circular flag from a given file name.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_circular</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a dict associating IDs with their is_circular flag from a given file name.&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No length available for key </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.Manifest.get_fasta_lengths" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_fasta_lengths</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.Manifest.get_fasta_lengths" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Check if the fasta files have the correct ids and no stop codon.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>fasta_path</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to fasta_dna and fasta_pep files.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Error if any empty ids, non-unique ids or stop codons are found in the fasta files.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_fasta_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if the fasta files have the correct ids and no stop codon.</span>

<span class="sd">    Args:</span>
<span class="sd">        fasta_path: Path to fasta_dna and fasta_pep files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Error if any empty ids, non-unique ids or stop codons are found in the fasta files.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">data</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">non_unique</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">non_unique_count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">empty_id_count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">contains_stop_codon</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">rec_count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
        <span class="n">rec_count</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="c1"># Flag empty ids</span>
        <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">empty_id_count</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Flag redundant ids</span>
            <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
                <span class="n">non_unique</span><span class="p">[</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">non_unique_count</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="c1"># Store sequence id and length</span>
            <span class="n">data</span><span class="p">[</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">stops</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">stops</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">contains_stop_codon</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">stops</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">ignore_final_stops</span><span class="p">:</span>
                    <span class="n">contains_stop_codon</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">if</span> <span class="n">empty_id_count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">empty_id_count</span><span class="si">}</span><span class="s2"> sequences with empty ids in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">non_unique_count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">non_unique_count</span><span class="si">}</span><span class="s2"> non unique sequence ids in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">contains_stop_codon</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">contains_stop_codon</span><span class="si">}</span><span class="s2"> sequences with stop codons in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">rec_count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No sequences found in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.Manifest.get_functional_annotation" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_functional_annotation</span><span class="p">(</span><span class="n">json_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.Manifest.get_functional_annotation" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Load the functional annotation file to retrieve the gene_id and translation id.
    A functional annotation file contains information about a gene.
    The functional annotation file is stored in a json format containing
    the description, id and object type (eg: "gene", "transcript", "translation").</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>json_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to functional_annotation.json.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>None</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict with gene and translation ids.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_functional_annotation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">json_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load the functional annotation file to retrieve the gene_id and translation id.</span>
<span class="sd">        A functional annotation file contains information about a gene.</span>
<span class="sd">        The functional annotation file is stored in a json format containing</span>
<span class="sd">        the description, id and object type (eg: &quot;gene&quot;, &quot;transcript&quot;, &quot;translation&quot;).</span>

<span class="sd">    Args:</span>
<span class="sd">        json_path: Path to functional_annotation.json.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict with gene and translation ids.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Load the json file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

    <span class="c1"># Get gene ids and translation ids</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">translations</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">transposons</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">:</span>
            <span class="n">genes</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;translation&quot;</span><span class="p">:</span>
            <span class="n">translations</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
            <span class="n">transposons</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="n">stats</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ann_genes&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span>
        <span class="s2">&quot;ann_translations&quot;</span><span class="p">:</span> <span class="n">translations</span><span class="p">,</span>
        <span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">:</span> <span class="n">transposons</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">,</span> <span class="o">**</span><span class="n">stats</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.Manifest.get_gff3" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_gff3</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.Manifest.get_gff3" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>A GFF parser is used to retrieve information in the GFF file such as
   gene and CDS ids and their corresponding lengths.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gff3_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to gff3 file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A GFF parser is used to retrieve information in the GFF file such as</span>
<span class="sd">       gene and CDS ids and their corresponding lengths.</span>

<span class="sd">    Args:</span>
<span class="sd">        gff3_path: Path to gff3 file.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">seqs</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">peps</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">all_peps</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">tes</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff3_handle</span><span class="p">:</span>
        <span class="n">gff</span> <span class="o">=</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gff3_handle</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">gff</span><span class="p">:</span>
            <span class="n">seqs</span><span class="p">[</span><span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="n">feat_length</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="c1"># Store gene id and length</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;ncRNA_gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">]:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_retrieve_gff_gene_lengths</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">peps</span><span class="p">,</span> <span class="n">all_peps</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
                    <span class="n">tes</span><span class="p">[</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_length</span>

    <span class="n">stats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;gff3_seq_regions&quot;</span><span class="p">:</span> <span class="n">seqs</span><span class="p">,</span>
        <span class="s2">&quot;gff3_genes&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span>
        <span class="s2">&quot;gff3_translations&quot;</span><span class="p">:</span> <span class="n">peps</span><span class="p">,</span>
        <span class="s2">&quot;gff3_all_translations&quot;</span><span class="p">:</span> <span class="n">all_peps</span><span class="p">,</span>
        <span class="s2">&quot;gff3_transposable_elements&quot;</span><span class="p">:</span> <span class="n">tes</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">,</span> <span class="o">**</span><span class="n">stats</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.Manifest.get_lengths" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_lengths</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.Manifest.get_lengths" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Returns a dict associating IDs with their length from a given file name.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a dict associating IDs with their length from a given file name.&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There is no length record for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.Manifest.get_manifest" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_manifest</span><span class="p">(</span><span class="n">manifest_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.Manifest.get_manifest" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Load the content of a manifest file.</p>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>Dict</code></td>          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Content of the manifest file.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load the content of a manifest file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict: Content of the manifest file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">manifest_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">manifest_path</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">manifest_fh</span><span class="p">:</span>
        <span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">manifest_fh</span><span class="p">)</span>

        <span class="c1"># Use dir name from the manifest</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                <span class="c1"># check if the md5sum is correct</span>
                <span class="n">md5sum</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;md5sum&quot;</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_check_md5sum</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">)</span>

                <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_path</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]:</span>
                        <span class="n">file_path</span> <span class="o">=</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                        <span class="c1"># check if the md5sum is correct</span>
                        <span class="n">md5sum</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;md5sum&quot;</span><span class="p">]</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_check_md5sum</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">)</span>

                        <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_path</span>
        <span class="k">return</span> <span class="n">manifest</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.Manifest.has_lengths" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">has_lengths</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.Manifest.has_lengths" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Check if a given name has lengths records.</p>
<p>Raise KeyError if the name is not supported.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">has_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if a given name has lengths records.</span>

<span class="sd">    Raise KeyError if the name is not supported.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
            <span class="k">return</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There is no length record for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.Manifest.prepare_integrity_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">prepare_integrity_data</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.manifest.Manifest.prepare_integrity_data" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Read all the files and keep a record (IDs and their lengths)
for each cases to be compared later.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/check_integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">prepare_integrity_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Read all the files and keep a record (IDs and their lengths)</span>
<span class="sd">    for each cases to be compared later.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First, get the Data</span>
    <span class="k">if</span> <span class="s2">&quot;gff3&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains GFF3&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">get_gff3</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;fasta_dna&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains DNA fasta&quot;</span><span class="p">)</span>
        <span class="c1"># Verify if the length and id for the sequence is unique</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;dna_sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_fasta_lengths</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;fasta_dna&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;fasta_pep&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains Peptide fasta&quot;</span><span class="p">)</span>
        <span class="c1"># Verify if the length and id for the sequence is unique</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_fasta_lengths</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;fasta_pep&quot;</span><span class="p">],</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;seq_region&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains seq_region JSON&quot;</span><span class="p">)</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No sequences found in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s1">&#39;seq_region&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">seqr_seqlevel</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">seq_lengths</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">seq_circular</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="c1"># Store the length as int</span>
            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
                <span class="n">seq_lengths</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>
                <span class="n">seq_circular</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;circular&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;contig&quot;</span><span class="p">:</span>
                    <span class="n">seqr_seqlevel</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_lengths</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">[</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_circular</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">seq_regions</span> <span class="o">=</span> <span class="n">seq_regions</span>
    <span class="k">if</span> <span class="s2">&quot;functional_annotation&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains functional annotation(s)&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">get_functional_annotation</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;functional_annotation&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;agp&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains AGP files&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;agp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_agp_seq_regions</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;agp&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;genome&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Manifest contains genome JSON&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]))</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.manifest.ManifestMaker" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>ManifestMaker</code>


<a href="#ensembl.io.genomio.manifest.ManifestMaker" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">

  
      <p>Given a directory with genomic files, create a manifest json file for them.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/generate.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">ManifestMaker</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a directory with genomic files, create a manifest json file for them.&quot;&quot;&quot;</span>

    <span class="n">same_names</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;gff3&quot;</span><span class="p">,</span>
        <span class="s2">&quot;fasta_dna&quot;</span><span class="p">,</span>
        <span class="s2">&quot;fasta_pep&quot;</span><span class="p">,</span>
        <span class="s2">&quot;functional_annotation&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genome&quot;</span><span class="p">,</span>
        <span class="s2">&quot;seq_attrib&quot;</span><span class="p">,</span>
        <span class="s2">&quot;seq_region&quot;</span><span class="p">,</span>
        <span class="s2">&quot;agp&quot;</span><span class="p">,</span>
        <span class="s2">&quot;events&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">alias_names</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;gene_models&quot;</span><span class="p">:</span> <span class="s2">&quot;gff3&quot;</span><span class="p">,</span>
        <span class="s2">&quot;dna&quot;</span><span class="p">:</span> <span class="s2">&quot;fasta_dna&quot;</span><span class="p">,</span>
        <span class="s2">&quot;pep&quot;</span><span class="p">:</span> <span class="s2">&quot;fasta_pep&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">names</span> <span class="o">=</span> <span class="p">{</span><span class="n">name</span><span class="p">:</span> <span class="n">name</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">same_names</span><span class="p">}</span>
    <span class="n">names</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="n">names</span><span class="p">,</span> <span class="o">**</span><span class="n">alias_names</span><span class="p">}</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dir</span> <span class="o">=</span> <span class="n">manifest_dir</span>

    <span class="k">def</span> <span class="nf">create_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create the manifest file.&quot;&quot;&quot;</span>
        <span class="n">manifest_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_files_checksums</span><span class="p">()</span>
        <span class="n">manifest_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dir</span> <span class="o">/</span> <span class="s2">&quot;manifest.json&quot;</span>
        <span class="k">with</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_out</span><span class="p">:</span>
            <span class="n">json_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">manifest_data</span><span class="p">,</span> <span class="n">sort_keys</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">manifest_path</span>

    <span class="k">def</span> <span class="nf">get_files_checksums</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compute the checksum of all the files in the directory.&quot;&quot;&quot;</span>
        <span class="n">manifest_files</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">subfile</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dir</span><span class="o">.</span><span class="n">iterdir</span><span class="p">():</span>
            <span class="n">used_file</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="n">subfile</span><span class="o">.</span><span class="n">is_dir</span><span class="p">():</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Can&#39;t create manifest for subdirectory&quot;</span><span class="p">)</span>
                <span class="k">continue</span>

            <span class="c1"># Delete and skip empty files</span>
            <span class="k">if</span> <span class="n">subfile</span><span class="o">.</span><span class="n">stat</span><span class="p">()</span><span class="o">.</span><span class="n">st_size</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">subfile</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
                <span class="k">continue</span>

            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">standard_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">names</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">subfile</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">name</span><span class="p">):</span>
                    <span class="n">used_file</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">md5</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_md5sum</span><span class="p">(</span><span class="n">subfile</span><span class="p">)</span>
                    <span class="n">file_obj</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;file&quot;</span><span class="p">:</span> <span class="n">subfile</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;md5sum&quot;</span><span class="p">:</span> <span class="n">md5</span><span class="p">}</span>
                    <span class="k">if</span> <span class="n">standard_name</span> <span class="ow">in</span> <span class="n">manifest_files</span><span class="p">:</span>
                        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">],</span> <span class="nb">list</span><span class="p">):</span>
                            <span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">file_obj</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># Convert to a list</span>
                            <span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">],</span> <span class="n">file_obj</span><span class="p">]</span>

                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_obj</span>
                    <span class="k">break</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">used_file</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">subfile</span><span class="si">}</span><span class="s2"> was not included in the manifest&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">manifest_files</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_get_md5sum</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">file_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">data_bytes</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
            <span class="k">return</span> <span class="n">hashlib</span><span class="o">.</span><span class="n">md5</span><span class="p">(</span><span class="n">data_bytes</span><span class="p">)</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ManifestMaker.create_manifest" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">create_manifest</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.manifest.ManifestMaker.create_manifest" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Create the manifest file.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/generate.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create the manifest file.&quot;&quot;&quot;</span>
    <span class="n">manifest_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_files_checksums</span><span class="p">()</span>
    <span class="n">manifest_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dir</span> <span class="o">/</span> <span class="s2">&quot;manifest.json&quot;</span>
    <span class="k">with</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_out</span><span class="p">:</span>
        <span class="n">json_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">manifest_data</span><span class="p">,</span> <span class="n">sort_keys</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">manifest_path</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.ManifestMaker.get_files_checksums" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_files_checksums</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.manifest.ManifestMaker.get_files_checksums" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Compute the checksum of all the files in the directory.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/generate.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_files_checksums</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the checksum of all the files in the directory.&quot;&quot;&quot;</span>
    <span class="n">manifest_files</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">subfile</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dir</span><span class="o">.</span><span class="n">iterdir</span><span class="p">():</span>
        <span class="n">used_file</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">subfile</span><span class="o">.</span><span class="n">is_dir</span><span class="p">():</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Can&#39;t create manifest for subdirectory&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># Delete and skip empty files</span>
        <span class="k">if</span> <span class="n">subfile</span><span class="o">.</span><span class="n">stat</span><span class="p">()</span><span class="o">.</span><span class="n">st_size</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">subfile</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
            <span class="k">continue</span>

        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">standard_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">names</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">subfile</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">name</span><span class="p">):</span>
                <span class="n">used_file</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">md5</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_md5sum</span><span class="p">(</span><span class="n">subfile</span><span class="p">)</span>
                <span class="n">file_obj</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;file&quot;</span><span class="p">:</span> <span class="n">subfile</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;md5sum&quot;</span><span class="p">:</span> <span class="n">md5</span><span class="p">}</span>
                <span class="k">if</span> <span class="n">standard_name</span> <span class="ow">in</span> <span class="n">manifest_files</span><span class="p">:</span>
                    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">],</span> <span class="nb">list</span><span class="p">):</span>
                        <span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">file_obj</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Convert to a list</span>
                        <span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">],</span> <span class="n">file_obj</span><span class="p">]</span>

                <span class="k">else</span><span class="p">:</span>
                    <span class="n">manifest_files</span><span class="p">[</span><span class="n">standard_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_obj</span>
                <span class="k">break</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">used_file</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">subfile</span><span class="si">}</span><span class="s2"> was not included in the manifest&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">manifest_files</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.manifest.StatsError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>StatsError</code>


<a href="#ensembl.io.genomio.manifest.StatsError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>Raised when stats could not be computed.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">62</span>
<span class="normal">63</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">StatsError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Raised when stats could not be computed.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.manifest.manifest_stats" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>manifest_stats</code>


<a href="#ensembl.io.genomio.manifest.manifest_stats" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">

  
      <p>Representation of the statistics of the set of files listed in the manifest file provided.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">manifest_stats</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Representation of the statistics of the set of files listed in the manifest file provided.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">accession</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">datasets_bin</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">manifest_dir</span><span class="si">}</span><span class="s2">/manifest.json&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">accession</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">accession</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">errors_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">manifest_dir</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;stats_diff.log&quot;</span>
        <span class="k">if</span> <span class="n">datasets_bin</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">datasets_bin</span> <span class="o">=</span> <span class="s2">&quot;datasets&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">datasets_bin</span> <span class="o">=</span> <span class="n">datasets_bin</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest_parent</span> <span class="o">=</span> <span class="n">manifest_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stats_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compute stats in the files and output a stats.txt file in the same folder.</span>

<span class="sd">        Raises:</span>
<span class="sd">            StatsError: Could not compute some stats.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">manifest</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_manifest</span><span class="p">()</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accession</span><span class="p">)</span>

        <span class="c1"># Compute the stats from the GFF3 file</span>
        <span class="k">if</span> <span class="s2">&quot;gff3&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gff3_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">]))</span>

        <span class="c1"># Compute the stats from the seq_region file</span>
        <span class="k">if</span> <span class="s2">&quot;seq_region&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq_region_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]))</span>

        <span class="c1"># Print out the stats in a separate file</span>
        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">stats_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stats_out</span><span class="p">:</span>
            <span class="n">stats_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">stats</span><span class="p">))</span>

        <span class="c1"># Die if there were errors in stats comparison</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span>
            <span class="k">with</span> <span class="bp">self</span><span class="o">.</span><span class="n">errors_file</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">errors_fh</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">error_line</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span>
                    <span class="n">errors_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">error_line</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the files metadata from the manifest json file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict: A representation of the manifest json data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_json</span><span class="p">:</span>
            <span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f_json</span><span class="p">)</span>
            <span class="n">manifest_root</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_parent</span>

        <span class="c1"># Use dir name from the manifest</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">manifest_root</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]:</span>
                        <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                        <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest_root</span><span class="p">,</span> <span class="n">file_name</span>
                        <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>

        <span class="k">return</span> <span class="n">manifest</span>

    <span class="k">def</span> <span class="nf">get_seq_region_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_region_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compute stats from the seq_region json file.</span>

<span class="sd">        Args:</span>
<span class="sd">            seq_region_path (Path): the seq_region json file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: Stats from the seq_regions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">seq_region_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
            <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

        <span class="c1"># Get basic data</span>
        <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">circular</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">locations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">codon_tables</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="c1"># Get readable seq_region name:</span>
            <span class="c1"># either use a Genbank synonym, or just the provided seq_region name</span>
            <span class="n">genbank</span> <span class="o">=</span> <span class="s2">&quot;synonyms&quot;</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">and</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GenBank&quot;</span><span class="p">]</span>
            <span class="n">seqr_name</span> <span class="o">=</span> <span class="n">genbank</span> <span class="ow">and</span> <span class="n">genbank</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>

            <span class="c1"># Record the lengths of the elements of each coord_system</span>
            <span class="n">coord_level</span> <span class="o">=</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">coord_level</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="p">:</span>
                <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>

            <span class="c1"># Additional metadata records to count</span>
            <span class="k">if</span> <span class="s2">&quot;circular&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="n">circular</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="s2">&quot;codon_table&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="n">codon_tables</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;codon_table&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;location&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="n">locations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;location&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Stats</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_region_path</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">coord_systems_stats</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_region_special_stats</span><span class="p">(</span><span class="n">circular</span><span class="p">,</span> <span class="n">locations</span><span class="p">,</span> <span class="n">codon_tables</span><span class="p">)</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">coord_systems_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;For each coord_system compute various stats:</span>
<span class="sd">            - number of sequences</span>
<span class="sd">            - sequence length sum, minimum, maximum, mean</span>

<span class="sd">        Args:</span>
<span class="sd">            coord_systems: Coordinate system dictionary of lengths.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list with the computed statistics in a printable format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total coord_systems </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">coord_name</span><span class="p">,</span> <span class="n">lengths</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Coord_system: </span><span class="si">{</span><span class="n">coord_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">stat_counts</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;Number of sequences&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length sum&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length minimum&quot;</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length maximum&quot;</span><span class="p">:</span> <span class="nb">max</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length mean&quot;</span><span class="p">:</span> <span class="n">mean</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="p">}</span>

            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">stat_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">seq_region_special_stats</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">circular</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">locations</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">codon_tables</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Prepare stats in case there are circular regions, specific locations and codon_tables.</span>
<span class="sd">                stats.append(f&quot;{count: 9f}\t{name}&quot;)</span>

<span class="sd">        Args:</span>
<span class="sd">            circular: Number of circular regions. Defaults to 0.</span>
<span class="sd">            locations: The regions and their location. Defaults to None.</span>
<span class="sd">            codon_tables: The regions and their codon_table. Defaults to None.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list with the computed statistics in a printable format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">circular</span> <span class="ow">or</span> <span class="n">locations</span> <span class="ow">or</span> <span class="n">codon_tables</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Special&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">circular</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">circular</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="s2">circular sequences&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">locations</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">locations</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with location&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">locations</span><span class="p">:</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">loc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">codon_tables</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">codon_tables</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with codon_table&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="n">codon_tables</span><span class="p">:</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">table</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">get_gff3_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract the gene models from the GFF3 file and compute stats.</span>

<span class="sd">        Args:</span>
<span class="sd">            gff3_path (Path): the GFF3 file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List: Stats from the gene model.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">biotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_biotypes</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span>
        <span class="c1"># Compile final stats</span>
        <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biotypes_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">count_biotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Count the biotypes in a GFF3 file.</span>

<span class="sd">        Args:</span>
<span class="sd">            gff3_path: Path to the GFF3 file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dictionary of biotype counters.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff3_handle</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gff3_handle</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">feat1</span> <span class="ow">in</span> <span class="n">rec</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                    <span class="c1"># Check if the gene contains proteins (CDSs),</span>
                    <span class="c1"># and keep a count of all hierarchies (e.g. gene-mRNA-CDS)</span>
                    <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">for</span> <span class="n">feat2</span> <span class="ow">in</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">feat2</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
                            <span class="n">types2</span> <span class="o">=</span> <span class="p">{</span><span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">}</span>
                            <span class="k">if</span> <span class="s2">&quot;CDS&quot;</span> <span class="ow">in</span> <span class="n">types2</span><span class="p">:</span>
                                <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat2</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">feat3</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">feat3</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                                <span class="k">continue</span>
                            <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span>
                                <span class="n">biotypes</span><span class="p">,</span> <span class="n">feat3</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat3</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span>
                            <span class="p">)</span>

                    <span class="c1"># Main categories counts</span>
                    <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">:</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">)</span>
                    <span class="k">elif</span> <span class="n">is_protein</span><span class="p">:</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;PROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Special case, undefined gene-transcript</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span>
                            <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span>
                            <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transcript&quot;</span>
                        <span class="p">):</span>
                            <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;NONPROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                    <span class="c1"># Total</span>
                    <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">):</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">biotypes</span>

    <span class="k">def</span> <span class="nf">biotypes_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Prepare biotype stats in order of their name.</span>

<span class="sd">        Args:</span>
<span class="sd">            biotypes: Biotypes counters.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list with the computed statistics in a printable format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sorted_biotypes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">biotypes</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">data</span><span class="p">:</span> <span class="n">BiotypeCounter</span> <span class="o">=</span> <span class="n">biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">sorted_biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">unique_count</span><span class="p">()</span><span class="si">:</span><span class="s2">&gt;9</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">biotype</span><span class="si">:</span><span class="s2">&lt;20</span><span class="si">}</span><span class="se">\t</span><span class="s2">ID = </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">example</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">biotype</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sorted_biotypes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">]</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">check_ncbi_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Use the dataset tool from NCBI to get stats and compare with what we have&quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">stats</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">stats</span>

        <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span>

        <span class="n">datasets_bin</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datasets_bin</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">which</span><span class="p">(</span><span class="n">datasets_bin</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">stats</span>

        <span class="c1"># Get the dataset summary from NCBI</span>
        <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">datasets_bin</span><span class="p">,</span> <span class="s2">&quot;summary&quot;</span><span class="p">,</span> <span class="s2">&quot;genome&quot;</span><span class="p">,</span> <span class="s2">&quot;accession&quot;</span><span class="p">,</span> <span class="n">accession</span><span class="p">]</span>
        <span class="n">result_out</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">result_out</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>

        <span class="c1"># Get stats</span>
        <span class="k">if</span> <span class="s2">&quot;reports&quot;</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
            <span class="n">genome</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;reports&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="s2">&quot;annotation_info&quot;</span> <span class="ow">in</span> <span class="n">genome</span> <span class="ow">and</span> <span class="s2">&quot;stats&quot;</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">]:</span>
                <span class="n">ncbi_stats</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">][</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span>

                <span class="k">if</span> <span class="s2">&quot;gene_counts&quot;</span> <span class="ow">in</span> <span class="n">ncbi_stats</span><span class="p">:</span>
                    <span class="n">counts</span> <span class="o">=</span> <span class="n">ncbi_stats</span><span class="p">[</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">]</span>
                    <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compare_ncbi_counts</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">compare_ncbi_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">ncbi</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compare specific gene stats from NCBI&quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">maps</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span><span class="s2">&quot;total&quot;</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;protein_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;PROT_gene&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;pseudogene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;non_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;NONPROT_gene&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;other&quot;</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">],</span>
        <span class="p">]</span>

        <span class="k">for</span> <span class="n">count_map</span> <span class="ow">in</span> <span class="n">maps</span><span class="p">:</span>
            <span class="n">ncbi_name</span><span class="p">,</span> <span class="n">prep_name</span> <span class="o">=</span> <span class="n">count_map</span>
            <span class="n">ncbi_count</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">ncbi_name</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
            <span class="n">preped</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="n">biotypes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">prep_name</span><span class="p">)</span>
            <span class="n">prep_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">preped</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">prep_count</span> <span class="o">=</span> <span class="n">preped</span><span class="o">.</span><span class="n">count</span>

            <span class="k">if</span> <span class="n">prep_count</span> <span class="o">!=</span> <span class="n">ncbi_count</span><span class="p">:</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">prep_count</span> <span class="o">-</span> <span class="n">ncbi_count</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DIFF gene count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">ncbi_count</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">diff</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Same count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">stats</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">feature_biotype</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add the feature to their respective biotype counter.</span>

<span class="sd">        Args:</span>
<span class="sd">            biotypes (Dict[str, BiotypeCounter]): All current biotypes, with their counter.</span>
<span class="sd">            feature_id (str): Feature id to be counted.</span>
<span class="sd">            feature_biotype (str): The biotype of the feature.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">feature_biotype</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">biotypes</span><span class="p">:</span>
            <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span> <span class="o">=</span> <span class="n">BiotypeCounter</span><span class="p">(</span><span class="n">example</span><span class="o">=</span><span class="n">feature_id</span><span class="p">)</span>
        <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span><span class="o">.</span><span class="n">add_id</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.biotypes_stats" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">biotypes_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.biotypes_stats" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Prepare biotype stats in order of their name.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>biotypes</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <a class="autorefs autorefs-internal" title="ensembl.io.genomio.manifest.compute_stats.BiotypeCounter" href="#ensembl.io.genomio.manifest.BiotypeCounter">BiotypeCounter</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Biotypes counters.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A list with the computed statistics in a printable format.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">biotypes_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prepare biotype stats in order of their name.</span>

<span class="sd">    Args:</span>
<span class="sd">        biotypes: Biotypes counters.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A list with the computed statistics in a printable format.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sorted_biotypes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">biotypes</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
        <span class="n">data</span><span class="p">:</span> <span class="n">BiotypeCounter</span> <span class="o">=</span> <span class="n">biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="n">sorted_biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

    <span class="n">stats</span> <span class="o">=</span> <span class="p">[</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">unique_count</span><span class="p">()</span><span class="si">:</span><span class="s2">&gt;9</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">biotype</span><span class="si">:</span><span class="s2">&lt;20</span><span class="si">}</span><span class="se">\t</span><span class="s2">ID = </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">example</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">biotype</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sorted_biotypes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.check_ncbi_stats" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">check_ncbi_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.check_ncbi_stats" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Use the dataset tool from NCBI to get stats and compare with what we have</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_ncbi_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Use the dataset tool from NCBI to get stats and compare with what we have&quot;&quot;&quot;</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span>

    <span class="n">datasets_bin</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datasets_bin</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">which</span><span class="p">(</span><span class="n">datasets_bin</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="c1"># Get the dataset summary from NCBI</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">datasets_bin</span><span class="p">,</span> <span class="s2">&quot;summary&quot;</span><span class="p">,</span> <span class="s2">&quot;genome&quot;</span><span class="p">,</span> <span class="s2">&quot;accession&quot;</span><span class="p">,</span> <span class="n">accession</span><span class="p">]</span>
    <span class="n">result_out</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">result_out</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>

    <span class="c1"># Get stats</span>
    <span class="k">if</span> <span class="s2">&quot;reports&quot;</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;reports&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="s2">&quot;annotation_info&quot;</span> <span class="ow">in</span> <span class="n">genome</span> <span class="ow">and</span> <span class="s2">&quot;stats&quot;</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">]:</span>
            <span class="n">ncbi_stats</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">][</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span>

            <span class="k">if</span> <span class="s2">&quot;gene_counts&quot;</span> <span class="ow">in</span> <span class="n">ncbi_stats</span><span class="p">:</span>
                <span class="n">counts</span> <span class="o">=</span> <span class="n">ncbi_stats</span><span class="p">[</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">]</span>
                <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compare_ncbi_counts</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.compare_ncbi_counts" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">compare_ncbi_counts</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">ncbi</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.compare_ncbi_counts" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Compare specific gene stats from NCBI</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compare_ncbi_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">ncbi</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compare specific gene stats from NCBI&quot;&quot;&quot;</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">maps</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span><span class="s2">&quot;total&quot;</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;protein_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;PROT_gene&quot;</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;pseudogene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;non_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;NONPROT_gene&quot;</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;other&quot;</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">],</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">count_map</span> <span class="ow">in</span> <span class="n">maps</span><span class="p">:</span>
        <span class="n">ncbi_name</span><span class="p">,</span> <span class="n">prep_name</span> <span class="o">=</span> <span class="n">count_map</span>
        <span class="n">ncbi_count</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">ncbi_name</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">preped</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="n">biotypes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">prep_name</span><span class="p">)</span>
        <span class="n">prep_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">preped</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">prep_count</span> <span class="o">=</span> <span class="n">preped</span><span class="o">.</span><span class="n">count</span>

        <span class="k">if</span> <span class="n">prep_count</span> <span class="o">!=</span> <span class="n">ncbi_count</span><span class="p">:</span>
            <span class="n">diff</span> <span class="o">=</span> <span class="n">prep_count</span> <span class="o">-</span> <span class="n">ncbi_count</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DIFF gene count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">ncbi_count</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">diff</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Same count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.coord_systems_stats" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">coord_systems_stats</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.coord_systems_stats" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  

<details class="for-each-coord_system-compute-various-stats" open>
  <summary>For each coord_system compute various stats</summary>
  <ul>
<li>number of sequences</li>
<li>sequence length sum, minimum, maximum, mean</li>
</ul>
</details>


  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>coord_systems</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.List">List</span>[int]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Coordinate system dictionary of lengths.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A list with the computed statistics in a printable format.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">coord_systems_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;For each coord_system compute various stats:</span>
<span class="sd">        - number of sequences</span>
<span class="sd">        - sequence length sum, minimum, maximum, mean</span>

<span class="sd">    Args:</span>
<span class="sd">        coord_systems: Coordinate system dictionary of lengths.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A list with the computed statistics in a printable format.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total coord_systems </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">coord_name</span><span class="p">,</span> <span class="n">lengths</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Coord_system: </span><span class="si">{</span><span class="n">coord_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">stat_counts</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;Number of sequences&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="s2">&quot;Sequence length sum&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="s2">&quot;Sequence length minimum&quot;</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="s2">&quot;Sequence length maximum&quot;</span><span class="p">:</span> <span class="nb">max</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="s2">&quot;Sequence length mean&quot;</span><span class="p">:</span> <span class="n">mean</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">stat_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.count_biotypes" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">count_biotypes</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.count_biotypes" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Count the biotypes in a GFF3 file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gff3_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to the GFF3 file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <a class="autorefs autorefs-internal" title="ensembl.io.genomio.manifest.compute_stats.BiotypeCounter" href="#ensembl.io.genomio.manifest.BiotypeCounter">BiotypeCounter</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Dictionary of biotype counters.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">count_biotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Count the biotypes in a GFF3 file.</span>

<span class="sd">    Args:</span>
<span class="sd">        gff3_path: Path to the GFF3 file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dictionary of biotype counters.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff3_handle</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gff3_handle</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">feat1</span> <span class="ow">in</span> <span class="n">rec</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="c1"># Check if the gene contains proteins (CDSs),</span>
                <span class="c1"># and keep a count of all hierarchies (e.g. gene-mRNA-CDS)</span>
                <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">for</span> <span class="n">feat2</span> <span class="ow">in</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">feat2</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
                        <span class="n">types2</span> <span class="o">=</span> <span class="p">{</span><span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">}</span>
                        <span class="k">if</span> <span class="s2">&quot;CDS&quot;</span> <span class="ow">in</span> <span class="n">types2</span><span class="p">:</span>
                            <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat2</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">feat3</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">feat3</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                            <span class="k">continue</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span>
                            <span class="n">biotypes</span><span class="p">,</span> <span class="n">feat3</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat3</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">)</span>

                <span class="c1"># Main categories counts</span>
                <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">:</span>
                    <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">is_protein</span><span class="p">:</span>
                    <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;PROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Special case, undefined gene-transcript</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span>
                        <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span>
                        <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transcript&quot;</span>
                    <span class="p">):</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;NONPROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="c1"># Total</span>
                <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">):</span>
                    <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">biotypes</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.get_gff3_stats" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_gff3_stats</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.get_gff3_stats" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Extract the gene models from the GFF3 file and compute stats.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gff3_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the GFF3 file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>List</code></td>          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Stats from the gene model.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_gff3_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract the gene models from the GFF3 file and compute stats.</span>

<span class="sd">    Args:</span>
<span class="sd">        gff3_path (Path): the GFF3 file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List: Stats from the gene model.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">biotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_biotypes</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span>
    <span class="c1"># Compile final stats</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biotypes_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
    <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.get_manifest" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_manifest</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.get_manifest" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Get the files metadata from the manifest json file.</p>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>Dict</code></td>          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A representation of the manifest json data.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the files metadata from the manifest json file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict: A representation of the manifest json data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_json</span><span class="p">:</span>
        <span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f_json</span><span class="p">)</span>
        <span class="n">manifest_root</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_parent</span>

    <span class="c1"># Use dir name from the manifest</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">manifest_root</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]:</span>
                    <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                    <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest_root</span><span class="p">,</span> <span class="n">file_name</span>
                    <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>

    <span class="k">return</span> <span class="n">manifest</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.get_seq_region_stats" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">get_seq_region_stats</span><span class="p">(</span><span class="n">seq_region_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.get_seq_region_stats" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Compute stats from the seq_region json file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_region_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the seq_region json file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>List[str]: Stats from the seq_regions.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_seq_region_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_region_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute stats from the seq_region json file.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_region_path (Path): the seq_region json file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[str]: Stats from the seq_regions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">seq_region_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

    <span class="c1"># Get basic data</span>
    <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">circular</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">locations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">codon_tables</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="c1"># Get readable seq_region name:</span>
        <span class="c1"># either use a Genbank synonym, or just the provided seq_region name</span>
        <span class="n">genbank</span> <span class="o">=</span> <span class="s2">&quot;synonyms&quot;</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">and</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GenBank&quot;</span><span class="p">]</span>
        <span class="n">seqr_name</span> <span class="o">=</span> <span class="n">genbank</span> <span class="ow">and</span> <span class="n">genbank</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>

        <span class="c1"># Record the lengths of the elements of each coord_system</span>
        <span class="n">coord_level</span> <span class="o">=</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">coord_level</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="p">:</span>
            <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>

        <span class="c1"># Additional metadata records to count</span>
        <span class="k">if</span> <span class="s2">&quot;circular&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
            <span class="n">circular</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="s2">&quot;codon_table&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
            <span class="n">codon_tables</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;codon_table&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;location&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
            <span class="n">locations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;location&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Stats</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_region_path</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">coord_systems_stats</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span>
    <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_region_special_stats</span><span class="p">(</span><span class="n">circular</span><span class="p">,</span> <span class="n">locations</span><span class="p">,</span> <span class="n">codon_tables</span><span class="p">)</span>
    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.increment_biotype" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">,</span> <span class="n">feature_biotype</span><span class="p">)</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#ensembl.io.genomio.manifest.manifest_stats.increment_biotype" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Add the feature to their respective biotype counter.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>biotypes</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <a class="autorefs autorefs-internal" title="ensembl.io.genomio.manifest.compute_stats.BiotypeCounter" href="#ensembl.io.genomio.manifest.BiotypeCounter">BiotypeCounter</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>All current biotypes, with their counter.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>feature_id</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Feature id to be counted.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>feature_biotype</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The biotype of the feature.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">feature_biotype</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add the feature to their respective biotype counter.</span>

<span class="sd">    Args:</span>
<span class="sd">        biotypes (Dict[str, BiotypeCounter]): All current biotypes, with their counter.</span>
<span class="sd">        feature_id (str): Feature id to be counted.</span>
<span class="sd">        feature_biotype (str): The biotype of the feature.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">feature_biotype</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">biotypes</span><span class="p">:</span>
        <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span> <span class="o">=</span> <span class="n">BiotypeCounter</span><span class="p">(</span><span class="n">example</span><span class="o">=</span><span class="n">feature_id</span><span class="p">)</span>
    <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span><span class="o">.</span><span class="n">add_id</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.run" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">run</span><span class="p">(</span><span class="n">stats_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.run" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Compute stats in the files and output a stats.txt file in the same folder.</p>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="ensembl.io.genomio.manifest.compute_stats.StatsError" href="#ensembl.io.genomio.manifest.StatsError">StatsError</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Could not compute some stats.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stats_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute stats in the files and output a stats.txt file in the same folder.</span>

<span class="sd">    Raises:</span>
<span class="sd">        StatsError: Could not compute some stats.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">manifest</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_manifest</span><span class="p">()</span>

    <span class="n">stats</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accession</span><span class="p">)</span>

    <span class="c1"># Compute the stats from the GFF3 file</span>
    <span class="k">if</span> <span class="s2">&quot;gff3&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gff3_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">]))</span>

    <span class="c1"># Compute the stats from the seq_region file</span>
    <span class="k">if</span> <span class="s2">&quot;seq_region&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq_region_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]))</span>

    <span class="c1"># Print out the stats in a separate file</span>
    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">stats_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stats_out</span><span class="p">:</span>
        <span class="n">stats_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">stats</span><span class="p">))</span>

    <span class="c1"># Die if there were errors in stats comparison</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span>
        <span class="k">with</span> <span class="bp">self</span><span class="o">.</span><span class="n">errors_file</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">errors_fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">error_line</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span>
                <span class="n">errors_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">error_line</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h4 id="ensembl.io.genomio.manifest.manifest_stats.seq_region_special_stats" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>          <code class="highlight language-python"><span class="n">seq_region_special_stats</span><span class="p">(</span><span class="n">circular</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">locations</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">codon_tables</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.manifest_stats.seq_region_special_stats" class="headerlink" title="Permanent link">&para;</a></h4>


  <div class="doc doc-contents ">
  
      <p>Prepare stats in case there are circular regions, specific locations and codon_tables.
        stats.append(f"{count: 9f}      {name}")</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>circular</code></td>
          <td>
                <code>int</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Number of circular regions. Defaults to 0.</p>
            </div>
          </td>
          <td>
                <code>0</code>
          </td>
        </tr>
        <tr>
          <td><code>locations</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The regions and their location. Defaults to None.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>codon_tables</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The regions and their codon_table. Defaults to None.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A list with the computed statistics in a printable format.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">seq_region_special_stats</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">circular</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">locations</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">codon_tables</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prepare stats in case there are circular regions, specific locations and codon_tables.</span>
<span class="sd">            stats.append(f&quot;{count: 9f}\t{name}&quot;)</span>

<span class="sd">    Args:</span>
<span class="sd">        circular: Number of circular regions. Defaults to 0.</span>
<span class="sd">        locations: The regions and their location. Defaults to None.</span>
<span class="sd">        codon_tables: The regions and their codon_table. Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A list with the computed statistics in a printable format.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">circular</span> <span class="ow">or</span> <span class="n">locations</span> <span class="ow">or</span> <span class="n">codon_tables</span><span class="p">:</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Special&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">circular</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">circular</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="s2">circular sequences&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">locations</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">locations</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with location&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">locations</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">loc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">codon_tables</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">codon_tables</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with codon_table&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="n">codon_tables</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">table</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>


</div>



<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.manifest.init_logging_with_args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">init_logging_with_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.init_logging_with_args" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Processes the Namespace object provided to call <code>init_logging()</code> with the right arguments.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>args</code></td>
          <td>
                <code><span title="argparse.Namespace">Namespace</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Namespace populated by an argument parser.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>ensembl/utils/logging.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">init_logging_with_args</span><span class="p">(</span><span class="n">args</span><span class="p">:</span> <span class="n">argparse</span><span class="o">.</span><span class="n">Namespace</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Processes the Namespace object provided to call `init_logging()` with the right arguments.</span>

<span class="sd">    Args:</span>
<span class="sd">        args: Namespace populated by an argument parser.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args_dict</span> <span class="o">=</span> <span class="nb">vars</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
    <span class="n">log_args</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="n">args_dict</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;log_level&quot;</span><span class="p">,</span> <span class="s2">&quot;log_file&quot;</span><span class="p">,</span> <span class="s2">&quot;log_file_level&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">args_dict</span><span class="p">}</span>
    <span class="n">init_logging</span><span class="p">(</span><span class="o">**</span><span class="n">log_args</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.manifest.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.io.genomio.manifest.main" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Main entrypoint.</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/manifest/compute_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main entrypoint.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Compute stats from the current genome files associated with the manifest.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_src_path</span><span class="p">(</span>
        <span class="s2">&quot;--manifest_dir&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Manifest directory where &#39;manifest.json&#39; file is located&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--accession&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Sequence accession ID to compare stats with NCBI&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--datasets_bin&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Datasets bin status&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_dst_path</span><span class="p">(</span><span class="s2">&quot;--stats_file&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output file with the stats&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_log_arguments</span><span class="p">()</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">init_logging_with_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="n">mstats</span> <span class="o">=</span> <span class="n">manifest_stats</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">manifest_dir</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">accession</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">datasets_bin</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mstats</span><span class="o">.</span><span class="n">check_ncbi</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">stats_file</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">stats_file</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">stats_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">args</span><span class="o">.</span><span class="n">manifest_dir</span> <span class="o">/</span> <span class="s2">&quot;stats.txt&quot;</span>
    <span class="n">mstats</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">stats_file</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.manifest.open_gz_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">open_gz_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.manifest.open_gz_file" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Yields an open file object, even if the file is compressed with gzip.</p>
<p>The file is expected to contain a text, and this can be used with the usual "with".</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>file_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A file path to open.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/utils/archive_utils.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@contextmanager</span>
<span class="k">def</span> <span class="nf">open_gz_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Generator</span><span class="p">[</span><span class="n">TextIO</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Yields an open file object, even if the file is compressed with gzip.</span>

<span class="sd">    The file is expected to contain a text, and this can be used with the usual &quot;with&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        file_path: A file path to open.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">this_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">this_file</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.gz&quot;</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">this_file</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">fh</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">this_file</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;rt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">fh</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-2023] EMBL-European Bioinformatics Institute
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.8fd75fb4.min.js"></script>
      
    
  </body>
</html>