
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genomio/reference/genomio/seq_region/">
      
      
        <link rel="prev" href="../schemas/json/">
      
      
        <link rel="next" href="../utils/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.8">
    
    
      
        <title>seq_region - Ensembl-GenomIO</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.f2e4d321.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ensembl.io.genomio.seq_region" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Ensembl-GenomIO" class="md-header__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../img/ebang.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-GenomIO
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              seq_region
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M272 384c9.6-31.9 29.5-59.1 49.2-86.2 5.2-7.1 10.4-14.2 15.4-21.4 19.8-28.5 31.4-63 31.4-100.3C368 78.8 289.2 0 192 0S16 78.8 16 176c0 37.3 11.6 71.9 31.4 100.3 5 7.2 10.2 14.3 15.4 21.4 19.8 27.1 39.7 54.4 49.2 86.2h160zm-80 128c44.2 0 80-35.8 80-80v-16H112v16c0 44.2 35.8 80 80 80zm-80-336c0 8.8-7.2 16-16 16s-16-7.2-16-16c0-61.9 50.1-112 112-112 8.8 0 16 7.2 16 16s-7.2 16-16 16c-44.2 0-80 35.8-80 80z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M297.2 248.9c14.4-20.6 22.8-45.7 22.8-72.9 0-70.7-57.3-128-128-128S64 105.3 64 176c0 27.2 8.4 52.3 22.8 72.9 3.7 5.3 8.1 11.3 12.8 17.7 12.9 17.7 28.3 38.9 39.8 59.8 10.4 19 15.7 38.8 18.3 57.5H109c-2.2-12-5.9-23.7-11.8-34.5-9.9-18-22.2-34.9-34.5-51.8-5.2-7.1-10.4-14.2-15.4-21.4C27.6 247.9 16 213.3 16 176 16 78.8 94.8 0 192 0s176 78.8 176 176c0 37.3-11.6 71.9-31.4 100.3-5 7.2-10.2 14.3-15.4 21.4-12.3 16.8-24.6 33.7-34.5 51.8-5.9 10.8-9.6 22.5-11.8 34.5h-48.5c2.6-18.7 7.9-38.6 18.3-57.5 11.5-20.9 26.9-42.1 39.8-59.8 4.7-6.4 9-12.4 12.7-17.7zM192 128c-26.5 0-48 21.5-48 48 0 8.8-7.2 16-16 16s-16-7.2-16-16c0-44.2 35.8-80 80-80 8.8 0 16 7.2 16 16s-7.2 16-16 16zm0 384c-44.2 0-80-35.8-80-80v-16h160v16c0 44.2-35.8 80-80 80z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl-GenomIO
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Ensembl-GenomIO" class="md-nav__button md-logo" aria-label="Ensembl-GenomIO" data-md-component="logo">
      
  <img src="../../../img/ebang.png" alt="logo">

    </a>
    Ensembl-GenomIO
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genomio" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    Ensembl-GenomIO
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ensembl
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    brc4
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            brc4
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../runnable/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    runnable
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_1" id="__nav_4_1_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            runnable
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/compare_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_fasta.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/compare_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_report.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/core_server/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    core_server.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/download_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download_genbank.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/dump_stable_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_stable_ids.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/extract_from_gb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_gb.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/fill_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fill_metadata.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/gff3_specifier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3_specifier.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/load_sequence_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load_sequence_data.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/read_json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    read_json.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/say_accession/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    say_accession.py
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../runnable/seqregion_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seqregion_parser.py
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    io
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            io
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_1" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genomio
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_1" id="__nav_4_1_2_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            genomio
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    database
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../events/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fasta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genome_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_1_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../schemas/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    schemas
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_1_10" id="__nav_4_1_2_1_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_2_1_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2_1_10">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    seq_region
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    seq_region
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;seq_region
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" seq_region">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.LOCATION_CODON" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;LOCATION_CODON
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.MOLECULE_LOCATION" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;MOLECULE_LOCATION
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.SYNONYM_MAP" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SYNONYM_MAP
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.SYNONYM_RESOURCES" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SYNONYM_RESOURCES
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.SeqRegion" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SeqRegion
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.SeqRegionDict" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SeqRegionDict
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.MapFormatError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MapFormatError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.UnknownMetadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;UnknownMetadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.add_attribs" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_attribs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.add_insdc_seq_region_name" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_insdc_seq_region_name
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.add_mitochondrial_codon_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_mitochondrial_codon_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.add_translation_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_translation_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.exclude_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;exclude_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_attribs" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_attribs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_codon_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_codon_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_coord_systems" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_coord_systems
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_external_db_map" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_external_db_map
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_gbff_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_gbff_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_genbank_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_genbank_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_karyotype" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_karyotype
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_organelle" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_organelle
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_report_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_report_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_synonyms" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_synonyms
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.make_seq_region" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;make_seq_region
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.merge_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;merge_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.prepare_seq_region_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;prepare_seq_region_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.report_to_csv" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;report_to_csv
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nextflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Nextflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;seq_region
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" seq_region">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.LOCATION_CODON" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;LOCATION_CODON
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.MOLECULE_LOCATION" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;MOLECULE_LOCATION
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.SYNONYM_MAP" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SYNONYM_MAP
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.SYNONYM_RESOURCES" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SYNONYM_RESOURCES
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.SeqRegion" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SeqRegion
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.SeqRegionDict" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;SeqRegionDict
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.MapFormatError" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;MapFormatError
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.UnknownMetadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;UnknownMetadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.add_attribs" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_attribs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.add_insdc_seq_region_name" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_insdc_seq_region_name
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.add_mitochondrial_codon_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_mitochondrial_codon_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.add_translation_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_translation_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.exclude_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;exclude_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_attribs" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_attribs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_codon_table" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_codon_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_coord_systems" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_coord_systems
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_external_db_map" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_external_db_map
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_gbff_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_gbff_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_genbank_id" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_genbank_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_karyotype" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_karyotype
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_organelle" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_organelle
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_report_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_report_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.get_synonyms" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_synonyms
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.make_seq_region" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;make_seq_region
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.merge_seq_regions" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;merge_seq_regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.prepare_seq_region_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;prepare_seq_region_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.io.genomio.seq_region.report_to_csv" class="md-nav__link">
    <span class="md-ellipsis">
      <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;report_to_csv
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>seq_region</h1>

<div class="doc doc-object doc-module">



<h2 id="ensembl.io.genomio.seq_region" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>          <code>ensembl.io.genomio.seq_region</code>


<a href="#ensembl.io.genomio.seq_region" class="headerlink" title="Permanent link">&para;</a></h2>

  <div class="doc doc-contents first">
  
      <p>Sequence regions handling module.</p>

  

  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.seq_region.LOCATION_CODON" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>          <code class="highlight language-python"><span class="n">LOCATION_CODON</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;apicoplast_chromosome&#39;</span><span class="p">:</span> <span class="mi">4</span><span class="p">}</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.LOCATION_CODON" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.seq_region.MOLECULE_LOCATION" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>          <code class="highlight language-python"><span class="n">MOLECULE_LOCATION</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;apicoplast&#39;</span><span class="p">:</span> <span class="s1">&#39;apicoplast_chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="s1">&#39;nuclear_chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;kinetoplast&#39;</span><span class="p">:</span> <span class="s1">&#39;kinetoplast_chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;linkage group&#39;</span><span class="p">:</span> <span class="s1">&#39;linkage_group&#39;</span><span class="p">,</span> <span class="s1">&#39;mitochondrion&#39;</span><span class="p">:</span> <span class="s1">&#39;mitochondrial_chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;plasmid&#39;</span><span class="p">:</span> <span class="s1">&#39;plasmid&#39;</span><span class="p">}</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.MOLECULE_LOCATION" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.seq_region.SYNONYM_MAP" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>          <code class="highlight language-python"><span class="n">SYNONYM_MAP</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Assigned-Molecule&#39;</span><span class="p">:</span> <span class="s1">&#39;INSDC&#39;</span><span class="p">,</span> <span class="s1">&#39;GenBank-Accn&#39;</span><span class="p">:</span> <span class="s1">&#39;GenBank&#39;</span><span class="p">,</span> <span class="s1">&#39;RefSeq-Accn&#39;</span><span class="p">:</span> <span class="s1">&#39;RefSeq&#39;</span><span class="p">,</span> <span class="s1">&#39;Sequence-Name&#39;</span><span class="p">:</span> <span class="s1">&#39;INSDC_submitted_name&#39;</span><span class="p">}</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.SYNONYM_MAP" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.seq_region.SYNONYM_RESOURCES" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>          <code class="highlight language-python"><span class="n">SYNONYM_RESOURCES</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;GenBank&#39;</span><span class="p">,</span> <span class="s1">&#39;RefSeq&#39;</span><span class="p">,</span> <span class="s1">&#39;INSDC&#39;</span><span class="p">]</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.SYNONYM_RESOURCES" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.seq_region.SeqRegion" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>          <code class="highlight language-python"><span class="n">SeqRegion</span> <span class="o">=</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.SeqRegion" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.io.genomio.seq_region.SeqRegionDict" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>          <code class="highlight language-python"><span class="n">SeqRegionDict</span> <span class="o">=</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqRegion</span><span class="p">]</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.io.genomio.seq_region.SeqRegionDict" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  </div>

</div>


<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.seq_region.MapFormatError" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>MapFormatError</code>


<a href="#ensembl.io.genomio.seq_region.MapFormatError" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>Error when parsing the db map file.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/dump.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">MapFormatError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Error when parsing the db map file.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>

<div class="doc doc-object doc-class">



<h3 id="ensembl.io.genomio.seq_region.UnknownMetadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>          <code>UnknownMetadata</code>


<a href="#ensembl.io.genomio.seq_region.UnknownMetadata" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>If a metadata if not supported or recognized.</p>

            <details class="quote">
              <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">79</span>
<span class="normal">80</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">UnknownMetadata</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;If a metadata if not supported or recognized.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>


</div>



<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.add_attribs" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">add_attribs</span><span class="p">(</span><span class="n">seq_region</span><span class="p">,</span> <span class="n">attrib_dict</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.add_attribs" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Map seq_regions attribs to a specific name and type defined below.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_region</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A seq_region dict to modify.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>attrib_dict</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The attribs for this seq_region.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_attribs</span><span class="p">(</span><span class="n">seq_region</span><span class="p">:</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">attrib_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map seq_regions attribs to a specific name and type defined below.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_region (Dict): A seq_region dict to modify.</span>
<span class="sd">        attrib_dict (Dict): The attribs for this seq_region.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bool_attribs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;circular_seq&quot;</span><span class="p">:</span> <span class="s2">&quot;circular&quot;</span><span class="p">,</span>
        <span class="s2">&quot;non_ref&quot;</span><span class="p">:</span> <span class="s2">&quot;non_ref&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">int_attribs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;codon_table&quot;</span><span class="p">:</span> <span class="s2">&quot;codon_table&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">string_attribs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;BRC4_seq_region_name&quot;</span><span class="p">:</span> <span class="s2">&quot;BRC4_seq_region_name&quot;</span><span class="p">,</span>
        <span class="s2">&quot;EBI_seq_region_name&quot;</span><span class="p">:</span> <span class="s2">&quot;EBI_seq_region_name&quot;</span><span class="p">,</span>
        <span class="s2">&quot;coord_system_tag&quot;</span><span class="p">:</span> <span class="s2">&quot;coord_system_level&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sequence_location&quot;</span><span class="p">:</span> <span class="s2">&quot;location&quot;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">bool_attribs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">value</span> <span class="o">=</span> <span class="n">attrib_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">value</span><span class="p">:</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">bool</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">int_attribs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">value</span> <span class="o">=</span> <span class="n">attrib_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">value</span><span class="p">:</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">string_attribs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">value</span> <span class="o">=</span> <span class="n">attrib_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">value</span><span class="p">:</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.add_insdc_seq_region_name" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">add_insdc_seq_region_name</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">synonym_sources</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.add_insdc_seq_region_name" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns the list of sequence regions with their corresponding INSDC sequence region names.</p>
<p>"BRC4_seq_region_name" and "EBI_seq_region_name" fields are added to each sequence region: the
former will contain the corresponding INSDC name whilst the latter will contain the current name.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_regions</code></td>
          <td>
                <code><span title="typing.List">List</span>[<a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.SeqRegion" href="#ensembl.io.genomio.seq_region.SeqRegion">SeqRegion</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence regions.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>synonym_sources</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Synonym sources to use for the BRC4 name, in order of preference.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.UnknownMetadata" href="#ensembl.io.genomio.seq_region.UnknownMetadata">UnknownMetadata</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If no synonym name is found for a sequence region.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_insdc_seq_region_name</span><span class="p">(</span>
    <span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">synonym_sources</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the list of sequence regions with their corresponding INSDC sequence region names.</span>

<span class="sd">    &quot;BRC4_seq_region_name&quot; and &quot;EBI_seq_region_name&quot; fields are added to each sequence region: the</span>
<span class="sd">    former will contain the corresponding INSDC name whilst the latter will contain the current name.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        synonym_sources: Synonym sources to use for the BRC4 name, in order of preference.</span>

<span class="sd">    Raises:</span>
<span class="sd">        UnknownMetadata: If no synonym name is found for a sequence region.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">synonym_sources</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">synonym_sources</span> <span class="o">=</span> <span class="n">SYNONYM_RESOURCES</span>
    <span class="n">new_seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="p">{</span><span class="n">synonym</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]:</span> <span class="n">synonym</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">synonym</span> <span class="ow">in</span> <span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;synonyms&quot;</span><span class="p">,</span> <span class="p">[])}</span>
        <span class="c1"># Choose the synonym to use as the BRC name, using the first valid name from the source list</span>
        <span class="k">for</span> <span class="n">source_name</span> <span class="ow">in</span> <span class="n">synonym_sources</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">source_name</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
                <span class="n">brc_name</span> <span class="o">=</span> <span class="n">names</span><span class="p">[</span><span class="n">source_name</span><span class="p">]</span>
                <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Cannot set BRC4 sequence region name for </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">brc_name</span> <span class="o">=</span> <span class="n">brc_name</span><span class="o">.</span><span class="n">partition</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;BRC4_seq_region_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">brc_name</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;EBI_seq_region_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
        <span class="n">new_seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">new_seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.add_mitochondrial_codon_table" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">add_mitochondrial_codon_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.add_mitochondrial_codon_table" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Adds the mitochondrial codon table to each sequence region (when missing) based on the taxon ID.</p>
<p>If no mitochondrial genetic code can be found for the given taxon ID nothing will be changed.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_regions</code></td>
          <td>
                <code><span title="typing.List">List</span>[<a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.SeqRegion" href="#ensembl.io.genomio.seq_region.SeqRegion">SeqRegion</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence regions.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>taxon_id</code></td>
          <td>
                <code>int</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The species taxon ID.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_mitochondrial_codon_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the mitochondrial codon table to each sequence region (when missing) based on the taxon ID.</span>

<span class="sd">    If no mitochondrial genetic code can be found for the given taxon ID nothing will be changed.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        taxon_id: The species taxon ID.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/data/taxonomy/v1/taxon/tax-id/</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">headers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Content-Type&quot;</span><span class="p">:</span> <span class="s2">&quot;application/json&quot;</span><span class="p">},</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">60</span><span class="p">)</span>
    <span class="n">decoded</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
    <span class="k">if</span> <span class="s2">&quot;mitochondrialGeneticCode&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">decoded</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No mitochondria genetic code found for taxon </span><span class="si">{taxon_id}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">genetic_code</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">decoded</span><span class="p">[</span><span class="s2">&quot;mitochondrialGeneticCode&quot;</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="n">location</span> <span class="o">=</span> <span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;location&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="c1"># Do not overwrite any existing mitochondrial codon table</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">location</span> <span class="o">==</span> <span class="s2">&quot;mitochondrial_chromosome&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;codon_table&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">):</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genetic_code</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.add_translation_table" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">add_translation_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">location_codon</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.add_translation_table" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Adds the translation codon table to each sequence region (when missing) based on its location.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_regions</code></td>
          <td>
                <code><span title="typing.List">List</span>[<a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.SeqRegion" href="#ensembl.io.genomio.seq_region.SeqRegion">SeqRegion</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence regions.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>location_codon</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>[str, int]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Map of known codon tables for known locations.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_translation_table</span><span class="p">(</span>
    <span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">location_codon</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the translation codon table to each sequence region (when missing) based on its location.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        location_codon: Map of known codon tables for known locations.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">location_codon</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">location_codon</span> <span class="o">=</span> <span class="n">LOCATION_CODON</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="c1"># Do not overwrite any existing codon table</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;codon_table&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;location&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">location_codon</span><span class="p">):</span>
            <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location_codon</span><span class="p">[</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]]</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.exclude_seq_regions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">exclude_seq_regions</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">to_exclude</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.exclude_seq_regions" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns the list of sequence regions with the ones from the exclusion list removed.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_regions</code></td>
          <td>
                <code><span title="typing.List">List</span>[<a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.SeqRegion" href="#ensembl.io.genomio.seq_region.SeqRegion">SeqRegion</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence regions.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>to_exclude</code></td>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence region names to exclude.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">exclude_seq_regions</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">to_exclude</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the list of sequence regions with the ones from the exclusion list removed.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        to_exclude: Sequence region names to exclude.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filtered_seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;name&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">to_exclude</span><span class="p">):</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Not considering seq_region </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">filtered_seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">filtered_seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_attribs" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_attribs</span><span class="p">(</span><span class="n">seq_region</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_attribs" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Given a seq_region, extract the attribs as value-source items.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_region</code></td>
          <td>
                <code><span title="ensembl.core.models.SeqRegion">SeqRegion</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The seq_region from which the attribs are extracted.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>All attributes as a list of dictionaries with 'value' and 'source' keys.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_attribs</span><span class="p">(</span><span class="n">seq_region</span><span class="p">:</span> <span class="n">SeqRegion</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a seq_region, extract the attribs as value-source items.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_region (SeqRegion): The seq_region from which the attribs are extracted.</span>

<span class="sd">    Returns:</span>
<span class="sd">        All attributes as a list of dictionaries with &#39;value&#39; and &#39;source&#39; keys.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">attribs</span> <span class="o">=</span> <span class="n">seq_region</span><span class="o">.</span><span class="n">seq_region_attrib</span>
    <span class="n">atts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">attribs</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">attrib</span> <span class="ow">in</span> <span class="n">attribs</span><span class="p">:</span>
            <span class="n">att_obj</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;value&quot;</span><span class="p">:</span> <span class="n">attrib</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">attrib</span><span class="o">.</span><span class="n">attrib_type</span><span class="o">.</span><span class="n">code</span><span class="p">}</span>
            <span class="n">atts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">att_obj</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">atts</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_codon_table" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_codon_table</span><span class="p">(</span><span class="n">record</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_codon_table" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns the codon table number from a given a GenBank sequence record (if present).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>record</code></td>
          <td>
                <code><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>GenBank sequence record.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_codon_table</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the codon table number from a given a GenBank sequence record (if present).</span>

<span class="sd">    Args:</span>
<span class="sd">        record: GenBank sequence record.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">table_number</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;transl_table&quot;</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
            <span class="n">table_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;transl_table&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">table_number</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_coord_systems" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_coord_systems</span><span class="p">(</span><span class="n">session</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_coord_systems" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve the coord_system metadata from the current core.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>session</code></td>
          <td>
                <code><span title="sqlalchemy.orm.Session">Session</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Session for the current core.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="ensembl.core.models.CoordSystem">CoordSystem</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>List[CoordSystem]: All coord_systems in the core.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_coord_systems</span><span class="p">(</span><span class="n">session</span><span class="p">:</span> <span class="n">Session</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">CoordSystem</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieve the coord_system metadata from the current core.</span>

<span class="sd">    Args:</span>
<span class="sd">        session (Session): Session for the current core.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[CoordSystem]: All coord_systems in the core.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">coord_systems</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">CoordSystem</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">coord_stmt</span> <span class="o">=</span> <span class="n">select</span><span class="p">(</span><span class="n">CoordSystem</span><span class="p">)</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">CoordSystem</span><span class="o">.</span><span class="n">attrib</span><span class="o">.</span><span class="n">like</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2">efault_version%&quot;</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">session</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">coord_stmt</span><span class="p">)</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
        <span class="n">coord_systems</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">coord_systems</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_external_db_map" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_external_db_map</span><span class="p">(</span><span class="n">map_file</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_external_db_map" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Class method, set up the map for all SeqRegion objects</p>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_external_db_map</span><span class="p">(</span><span class="n">map_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class method, set up the map for all SeqRegion objects&quot;&quot;&quot;</span>
    <span class="n">db_map</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="n">map_file</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">map_fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">map_fh</span><span class="p">:</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">line</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">parts</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="n">MapFormatError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;External db file is not formatted correctly for: </span><span class="si">{</span><span class="n">line</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">db_map</span><span class="p">[</span><span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">db_map</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_gbff_seq_regions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_gbff_seq_regions</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_gbff_seq_regions" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns the sequence regions found in the GBFF file (if any).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gbff_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to GBFF file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.SeqRegionDict" href="#ensembl.io.genomio.seq_region.SeqRegionDict">SeqRegionDict</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A dict of SeqRegions, with their name as the key.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_gbff_seq_regions</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the sequence regions found in the GBFF file (if any).</span>

<span class="sd">    Args:</span>
<span class="sd">        gbff_path: Path to GBFF file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dict of SeqRegions, with their name as the key.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gbff_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gbff_file</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
            <span class="n">seqr</span><span class="p">:</span> <span class="n">SeqRegion</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;length&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)}</span>
            <span class="c1"># Is the seq_region circular?</span>
            <span class="n">annotations</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span>
            <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;topology&quot;</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;topology&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;circular&quot;</span><span class="p">):</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;circular&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1"># Is there a genetic code defined?</span>
            <span class="n">codon_table</span> <span class="o">=</span> <span class="n">get_codon_table</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">codon_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">codon_table</span>
            <span class="c1"># Is it an organelle?</span>
            <span class="n">location</span> <span class="o">=</span> <span class="n">get_organelle</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">location</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location</span>
            <span class="c1"># Is there a comment stating the Genbank record this is based on?</span>
            <span class="n">genbank_id</span> <span class="o">=</span> <span class="n">get_genbank_id</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">genbank_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="s2">&quot;INSDC&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">genbank_id</span><span class="p">}]</span>
            <span class="c1"># Store the seq_region</span>
            <span class="n">seq_regions</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqr</span>
    <span class="k">return</span> <span class="n">seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_genbank_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_genbank_id</span><span class="p">(</span><span class="n">record</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_genbank_id" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns the GenBank accession from a given sequence record (if present).</p>
<p>Only useful for RefSeq sequence records, where the GenBank accession is stored in a comment.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>record</code></td>
          <td>
                <code><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence record.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_genbank_id</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the GenBank accession from a given sequence record (if present).</span>

<span class="sd">    Only useful for RefSeq sequence records, where the GenBank accession is stored in a comment.</span>

<span class="sd">    Args:</span>
<span class="sd">        record: Sequence record.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genbank_id</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="s2">&quot;comment&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
        <span class="n">comment</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;comment&quot;</span><span class="p">]</span>
        <span class="n">comment</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[ \n\r]+&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">comment</span><span class="p">)</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;The reference sequence was derived from ([^\.]+)\.&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
            <span class="n">genbank_id</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genbank_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_karyotype" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_karyotype</span><span class="p">(</span><span class="n">seq_region</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_karyotype" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Given a seq_region, extract the karyotype bands.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_region</code></td>
          <td>
                <code><span title="ensembl.core.models.SeqRegion">SeqRegion</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The seq_region from which the karyotype bands are extracted.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>List</code></td>          <td>
                <code><span title="typing.List">List</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>All karyotype bands as a dict with values 'start', 'end', 'name' 'stain', 'structure'.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_karyotype</span><span class="p">(</span><span class="n">seq_region</span><span class="p">:</span> <span class="n">SeqRegion</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a seq_region, extract the karyotype bands.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_region (SeqRegion): The seq_region from which the karyotype bands are extracted.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List: All karyotype bands as a dict with values &#39;start&#39;, &#39;end&#39;, &#39;name&#39; &#39;stain&#39;, &#39;structure&#39;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bands</span> <span class="o">=</span> <span class="n">seq_region</span><span class="o">.</span><span class="n">karyotype</span>
    <span class="n">kars</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">bands</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">band</span> <span class="ow">in</span> <span class="n">bands</span><span class="p">:</span>
            <span class="n">kar</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">band</span><span class="o">.</span><span class="n">seq_region_start</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">band</span><span class="o">.</span><span class="n">seq_region_end</span><span class="p">}</span>
            <span class="k">if</span> <span class="n">band</span><span class="o">.</span><span class="n">band</span><span class="p">:</span>
                <span class="n">kar</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">band</span><span class="o">.</span><span class="n">band</span>
            <span class="k">if</span> <span class="n">band</span><span class="o">.</span><span class="n">stain</span><span class="p">:</span>
                <span class="n">kar</span><span class="p">[</span><span class="s2">&quot;stain&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">band</span><span class="o">.</span><span class="n">stain</span>
                <span class="n">structure</span> <span class="o">=</span> <span class="n">_KARYOTYPE_STRUCTURE</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">band</span><span class="o">.</span><span class="n">stain</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">structure</span><span class="p">:</span>
                    <span class="n">kar</span><span class="p">[</span><span class="s2">&quot;structure&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">structure</span>
            <span class="n">kars</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">kar</span><span class="p">)</span>

    <span class="n">kars</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">kars</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">kar</span><span class="p">:</span> <span class="n">kar</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">kars</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_organelle" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_organelle</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">molecule_location</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_organelle" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns the organelle location from the given GenBank record (if present).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>record</code></td>
          <td>
                <code><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>GenBank sequence record.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>molecule_location</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Map of sequence type to SO location.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>KeyError</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If the location is not part of the controlled vocabulary.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_organelle</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">,</span> <span class="n">molecule_location</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the organelle location from the given GenBank record (if present).</span>

<span class="sd">    Args:</span>
<span class="sd">        record: GenBank sequence record.</span>
<span class="sd">        molecule_location: Map of sequence type to SO location.</span>

<span class="sd">    Raises:</span>
<span class="sd">        KeyError: If the location is not part of the controlled vocabulary.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">molecule_location</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">molecule_location</span> <span class="o">=</span> <span class="n">MOLECULE_LOCATION</span>
    <span class="n">location</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;organelle&quot;</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
            <span class="n">organelle</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;organelle&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">organelle</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="c1"># Remove plastid prefix</span>
            <span class="n">with_prefix</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^(plastid|mitochondrion):(.+)$&quot;</span><span class="p">,</span> <span class="n">organelle</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">with_prefix</span><span class="p">:</span>
                <span class="n">organelle</span> <span class="o">=</span> <span class="n">with_prefix</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="c1"># Get controlled name</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">location</span> <span class="o">=</span> <span class="n">molecule_location</span><span class="p">[</span><span class="n">organelle</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence location: </span><span class="si">{</span><span class="n">organelle</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">exc</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">location</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_report_regions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_report_regions</span><span class="p">(</span><span class="n">report_path</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_report_regions" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns the sequence regions found in the report file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>report_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to the sequence regions report file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>is_refseq</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>True if the source of the report is RefSeq, false if INSDC.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.SeqRegionDict" href="#ensembl.io.genomio.seq_region.SeqRegionDict">SeqRegionDict</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A dict of SeqRegions, with their name as the key.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_report_regions</span><span class="p">(</span><span class="n">report_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the sequence regions found in the report file.</span>

<span class="sd">    Args:</span>
<span class="sd">        report_path: Path to the sequence regions report file.</span>
<span class="sd">        is_refseq: True if the source of the report is RefSeq, false if INSDC.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dict of SeqRegions, with their name as the key.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get the report in a CSV format</span>
    <span class="n">report_csv</span> <span class="o">=</span> <span class="n">report_to_csv</span><span class="p">(</span><span class="n">report_path</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># Feed the csv string to the CSV reader</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">report_csv</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_NONE</span><span class="p">)</span>
    <span class="c1"># Create the seq_regions</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">reader</span><span class="p">:</span>
        <span class="n">seq_region</span> <span class="o">=</span> <span class="n">make_seq_region</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">seq_region</span><span class="p">:</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
            <span class="n">seq_regions</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_region</span>
    <span class="k">return</span> <span class="n">seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_seq_regions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_seq_regions</span><span class="p">(</span><span class="n">session</span><span class="p">,</span> <span class="n">external_db_map</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_seq_regions" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns all the sequence regions from the current core database.</p>
<p>Include synonyms, attribs and karyotypes. Only the top level sequences are exported.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>session</code></td>
          <td>
                <code><span title="sqlalchemy.orm.Session">Session</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Session from the current core.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>external_db_map</code></td>
          <td>
                <code>dict</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Mapping of external_db names for the synonyms.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_seq_regions</span><span class="p">(</span><span class="n">session</span><span class="p">:</span> <span class="n">Session</span><span class="p">,</span> <span class="n">external_db_map</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns all the sequence regions from the current core database.</span>

<span class="sd">    Include synonyms, attribs and karyotypes. Only the top level sequences are exported.</span>

<span class="sd">    Args:</span>
<span class="sd">        session (Session): Session from the current core.</span>
<span class="sd">        external_db_map (dict): Mapping of external_db names for the synonyms.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">coord_systems</span> <span class="o">=</span> <span class="n">get_coord_systems</span><span class="p">(</span><span class="n">session</span><span class="p">)</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">coord_system</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Dump coord </span><span class="si">{</span><span class="n">coord_system</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">seqr_stmt</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">select</span><span class="p">(</span><span class="n">SeqRegion</span><span class="p">)</span>
            <span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">SeqRegion</span><span class="o">.</span><span class="n">coord_system_id</span> <span class="o">==</span> <span class="n">coord_system</span><span class="o">.</span><span class="n">coord_system_id</span><span class="p">)</span>
            <span class="o">.</span><span class="n">options</span><span class="p">(</span>
                <span class="n">joinedload</span><span class="p">(</span><span class="n">SeqRegion</span><span class="o">.</span><span class="n">seq_region_synonym</span><span class="p">)</span><span class="o">.</span><span class="n">joinedload</span><span class="p">(</span><span class="n">SeqRegionSynonym</span><span class="o">.</span><span class="n">external_db</span><span class="p">),</span>
                <span class="n">joinedload</span><span class="p">(</span><span class="n">SeqRegion</span><span class="o">.</span><span class="n">seq_region_attrib</span><span class="p">)</span><span class="o">.</span><span class="n">joinedload</span><span class="p">(</span><span class="n">SeqRegionAttrib</span><span class="o">.</span><span class="n">attrib_type</span><span class="p">),</span>
                <span class="n">joinedload</span><span class="p">(</span><span class="n">SeqRegion</span><span class="o">.</span><span class="n">karyotype</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">session</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">seqr_stmt</span><span class="p">)</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
            <span class="n">seqr</span><span class="p">:</span> <span class="n">SeqRegion</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">seq_region</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">seq_region</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">seqr</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;length&quot;</span><span class="p">:</span> <span class="n">seqr</span><span class="o">.</span><span class="n">length</span><span class="p">}</span>
            <span class="n">synonyms</span> <span class="o">=</span> <span class="n">get_synonyms</span><span class="p">(</span><span class="n">seqr</span><span class="p">,</span> <span class="n">external_db_map</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">synonyms</span><span class="p">:</span>
                <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">synonyms</span>

            <span class="n">attribs</span> <span class="o">=</span> <span class="n">get_attribs_dict</span><span class="p">(</span><span class="n">seqr</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">attribs</span><span class="p">:</span>
                <span class="k">if</span> <span class="s2">&quot;toplevel&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">attribs</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">add_attribs</span><span class="p">(</span><span class="n">seq_region</span><span class="p">,</span> <span class="n">attribs</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Skip seq_region without attribs, not toplevel</span>
                <span class="k">continue</span>

            <span class="n">karyotype</span> <span class="o">=</span> <span class="n">get_karyotype</span><span class="p">(</span><span class="n">seqr</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">karyotype</span><span class="p">:</span>
                <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;karyotype_bands&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">karyotype</span>

            <span class="n">added_seq</span> <span class="o">=</span> <span class="n">get_added_sequence</span><span class="p">(</span><span class="n">seqr</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">added_seq</span><span class="p">:</span>
                <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;added_sequence&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">added_seq</span>

            <span class="k">if</span> <span class="s2">&quot;coord_system_level&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seq_region</span><span class="p">:</span>
                <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">coord_system</span><span class="o">.</span><span class="n">name</span>

            <span class="n">seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_region</span><span class="p">)</span>

    <span class="n">seq_regions</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">seqr</span><span class="p">:</span> <span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">],</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]))</span>
    <span class="k">return</span> <span class="n">seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.get_synonyms" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">get_synonyms</span><span class="p">(</span><span class="n">seq_region</span><span class="p">,</span> <span class="n">external_db_map</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.get_synonyms" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Get all synonyms for a given seq_region. Use the mapping for synonym source names.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_region</code></td>
          <td>
                <code><span title="ensembl.core.models.SeqRegion">SeqRegion</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Seq_region from which the synonyms are extracted.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>external_db_map</code></td>
          <td>
                <code>dict</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>To map the synonym source names.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>List</code></td>          <td>
                <code><span title="typing.List">List</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>All synonyms as a dict with 'name' and 'source' keys.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/dump.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_synonyms</span><span class="p">(</span><span class="n">seq_region</span><span class="p">:</span> <span class="n">SeqRegion</span><span class="p">,</span> <span class="n">external_db_map</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get all synonyms for a given seq_region. Use the mapping for synonym source names.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_region (SeqRegion): Seq_region from which the synonyms are extracted.</span>
<span class="sd">        external_db_map (dict): To map the synonym source names.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List: All synonyms as a dict with &#39;name&#39; and &#39;source&#39; keys.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">synonyms</span> <span class="o">=</span> <span class="n">seq_region</span><span class="o">.</span><span class="n">seq_region_synonym</span>
    <span class="n">syns</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">synonyms</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">syn</span> <span class="ow">in</span> <span class="n">synonyms</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">syn</span><span class="o">.</span><span class="n">external_db</span><span class="p">:</span>
                <span class="n">source</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">external_db</span><span class="o">.</span><span class="n">db_name</span>
                <span class="k">if</span> <span class="n">source</span> <span class="ow">in</span> <span class="n">external_db_map</span><span class="p">:</span>
                    <span class="n">source</span> <span class="o">=</span> <span class="n">external_db_map</span><span class="p">[</span><span class="n">source</span><span class="p">]</span>
                <span class="n">syn_obj</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">syn</span><span class="o">.</span><span class="n">synonym</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">source</span><span class="p">}</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">syn_obj</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">syn</span><span class="o">.</span><span class="n">synonym</span><span class="p">}</span>
            <span class="n">syns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">syn_obj</span><span class="p">)</span>

    <span class="n">syns</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">syns</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">syn</span><span class="p">:</span> <span class="p">(</span><span class="n">syn</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="n">syn</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;source&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">syns</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.make_seq_region" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">make_seq_region</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">,</span> <span class="n">synonym_map</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">molecule_location</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.make_seq_region" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns a sequence region from the information provided.</p>
<p>An empty sequence region will be returned if not accession information is found.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>data</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>a dict from the report representing one line, where the key is the column name.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>is_refseq</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>True if the source is RefSeq, false if INSDC.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>synonym_map</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>[str, str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Map of INSDC report column names to sequence region field names.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>molecule_location</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>[str, str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Map of sequence type to SO location.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>KeyError</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If the sequence location is not recognised.</p>
            </div>
          </td>
        </tr>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.UnknownMetadata" href="#ensembl.io.genomio.seq_region.UnknownMetadata">UnknownMetadata</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>if the sequence role is not recognised.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">make_seq_region</span><span class="p">(</span>
    <span class="n">data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">,</span>
    <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">synonym_map</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">molecule_location</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegion</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a sequence region from the information provided.</span>

<span class="sd">    An empty sequence region will be returned if not accession information is found.</span>

<span class="sd">    Args:</span>
<span class="sd">        data: a dict from the report representing one line, where the key is the column name.</span>
<span class="sd">        is_refseq: True if the source is RefSeq, false if INSDC.</span>
<span class="sd">        synonym_map: Map of INSDC report column names to sequence region field names.</span>
<span class="sd">        molecule_location: Map of sequence type to SO location.</span>

<span class="sd">    Raises:</span>
<span class="sd">        KeyError: If the sequence location is not recognised.</span>
<span class="sd">        UnknownMetadata: if the sequence role is not recognised.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">synonym_map</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">synonym_map</span> <span class="o">=</span> <span class="n">SYNONYM_MAP</span>
    <span class="k">if</span> <span class="n">molecule_location</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">molecule_location</span> <span class="o">=</span> <span class="n">MOLECULE_LOCATION</span>
    <span class="n">seq_region</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Set accession as the sequence region name</span>
    <span class="n">src</span> <span class="o">=</span> <span class="s2">&quot;RefSeq&quot;</span> <span class="k">if</span> <span class="n">is_refseq</span> <span class="k">else</span> <span class="s2">&quot;GenBank&quot;</span>
    <span class="n">accession_id</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s2">-Accn&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">accession_id</span> <span class="ow">and</span> <span class="p">(</span><span class="n">accession_id</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">accession_id</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;No </span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s1"> accession ID found for </span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;Sequence-Name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{}</span>
    <span class="c1"># Add synonyms</span>
    <span class="n">synonyms</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">field</span><span class="p">,</span> <span class="n">source</span> <span class="ow">in</span> <span class="n">synonym_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
            <span class="n">synonym</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">]}</span>
            <span class="n">synonyms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">synonym</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">synonyms</span><span class="p">:</span>
        <span class="n">synonyms</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">])</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">synonyms</span>
    <span class="c1"># Add sequence length</span>
    <span class="n">field</span> <span class="o">=</span> <span class="s2">&quot;Sequence-Length&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">])</span>
    <span class="c1"># Add coordinate system and location</span>
    <span class="n">seq_role</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Sequence-Role&quot;</span><span class="p">]</span>
    <span class="c1"># Scaffold?</span>
    <span class="k">if</span> <span class="n">seq_role</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;unplaced-scaffold&quot;</span><span class="p">,</span> <span class="s2">&quot;unlocalized-scaffold&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;scaffold&quot;</span>
    <span class="c1"># Chromosome? Check location</span>
    <span class="k">elif</span> <span class="n">seq_role</span> <span class="o">==</span> <span class="s2">&quot;assembled-molecule&quot;</span><span class="p">:</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;chromosome&quot;</span>
        <span class="n">location</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Assigned-Molecule-Location/Type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="c1"># Get location metadata</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">molecule_location</span><span class="p">[</span><span class="n">location</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence location: </span><span class="si">{</span><span class="n">location</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">exc</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence role: </span><span class="si">{</span><span class="n">seq_role</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seq_region</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.merge_seq_regions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">merge_seq_regions</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.merge_seq_regions" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Merges sequence regions from different sources.</p>
<p>When combining two regions matching the same key, the "right" seq region data will take precedence
over the "left".</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>left</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.SeqRegionDict" href="#ensembl.io.genomio.seq_region.SeqRegionDict">SeqRegionDict</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Dictionary of sequence regions with names as keys.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>right</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.SeqRegionDict" href="#ensembl.io.genomio.seq_region.SeqRegionDict">SeqRegionDict</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Dictionary of sequence regions with names as keys.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<a class="autorefs autorefs-internal" title="ensembl.io.genomio.seq_region.prepare.SeqRegion" href="#ensembl.io.genomio.seq_region.SeqRegion">SeqRegion</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A list of merged sequence regions, sorted by their name.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">merge_seq_regions</span><span class="p">(</span>
    <span class="n">left</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqRegionDict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">right</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqRegionDict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Merges sequence regions from different sources.</span>

<span class="sd">    When combining two regions matching the same key, the &quot;right&quot; seq region data will take precedence</span>
<span class="sd">    over the &quot;left&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        left: Dictionary of sequence regions with names as keys.</span>
<span class="sd">        right: Dictionary of sequence regions with names as keys.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A list of merged sequence regions, sorted by their name.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">left</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">left</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">right</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">right</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Get all the names</span>
    <span class="n">all_names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">left</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">right</span><span class="p">)</span>
    <span class="c1"># Create the seq_regions, merge if needed</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">all_names</span><span class="p">:</span>
        <span class="n">left_seqr</span> <span class="o">=</span> <span class="n">left</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{})</span>
        <span class="n">right_seqr</span> <span class="o">=</span> <span class="n">right</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{})</span>
        <span class="k">if</span> <span class="n">left_seqr</span> <span class="ow">and</span> <span class="n">right_seqr</span><span class="p">:</span>
            <span class="n">final_seqr</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="n">left_seqr</span><span class="p">,</span> <span class="o">**</span><span class="n">right_seqr</span><span class="p">}</span>
        <span class="k">elif</span> <span class="n">left_seqr</span><span class="p">:</span>
            <span class="n">final_seqr</span> <span class="o">=</span> <span class="n">left_seqr</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">final_seqr</span> <span class="o">=</span> <span class="n">right_seqr</span>
        <span class="n">seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">final_seqr</span><span class="p">)</span>
    <span class="n">seq_regions</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.prepare_seq_region_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">prepare_seq_region_metadata</span><span class="p">(</span><span class="n">genome_file</span><span class="p">,</span> <span class="n">report_file</span><span class="p">,</span> <span class="n">dst_file</span><span class="p">,</span> <span class="n">gbff_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">brc_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">to_exclude</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.prepare_seq_region_metadata" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Prepares the sequence region metadata found in the INSDC/RefSeq report and GBFF files.</p>
<p>The sequence region information is loaded from both sources and combined. Elements are added/excluded
as requested, and the final sequence region metadata is dumped in a JSON file that follows the schema
defined in "src/python/ensembl/io/genomio/data/schemas/seq_region.json".</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>genome_file</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Genome metadata JSON file path.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>report_file</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>INSDC/RefSeq sequences report file path to parse.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>gbff_file</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="os.PathLike">PathLike</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>INSDC/RefSeq GBFF file path to parse.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>dst_file</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>JSON file output for the processed sequence regions JSON.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>brc_mode</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Include INSDC sequence region names.</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>to_exclude</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence region names to exclude.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">prepare_seq_region_metadata</span><span class="p">(</span>
    <span class="n">genome_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">report_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">dst_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">gbff_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">brc_mode</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">to_exclude</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prepares the sequence region metadata found in the INSDC/RefSeq report and GBFF files.</span>

<span class="sd">    The sequence region information is loaded from both sources and combined. Elements are added/excluded</span>
<span class="sd">    as requested, and the final sequence region metadata is dumped in a JSON file that follows the schema</span>
<span class="sd">    defined in &quot;src/python/ensembl/io/genomio/data/schemas/seq_region.json&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        genome_file: Genome metadata JSON file path.</span>
<span class="sd">        report_file: INSDC/RefSeq sequences report file path to parse.</span>
<span class="sd">        gbff_file: INSDC/RefSeq GBFF file path to parse.</span>
<span class="sd">        dst_file: JSON file output for the processed sequence regions JSON.</span>
<span class="sd">        brc_mode: Include INSDC sequence region names.</span>
<span class="sd">        to_exclude: Sequence region names to exclude.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome_data</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">genome_file</span><span class="p">)</span>
    <span class="n">dst_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">dst_file</span><span class="p">)</span>
    <span class="n">is_refseq</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GCF_&quot;</span><span class="p">)</span>

    <span class="c1"># Get the sequence regions from the report and gbff files, and merge them</span>
    <span class="n">report_regions</span> <span class="o">=</span> <span class="n">get_report_regions</span><span class="p">(</span><span class="n">report_file</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">gbff_file</span><span class="p">:</span>
        <span class="n">gbff_regions</span> <span class="o">=</span> <span class="n">get_gbff_seq_regions</span><span class="p">(</span><span class="n">gbff_file</span><span class="p">)</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">merge_seq_regions</span><span class="p">(</span><span class="n">report_regions</span><span class="p">,</span> <span class="n">gbff_regions</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">report_regions</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="c1"># Exclude seq_regions from a list</span>
    <span class="k">if</span> <span class="n">to_exclude</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">exclude_seq_regions</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">to_exclude</span><span class="p">)</span>

    <span class="c1"># Setup the BRC4_seq_region_name</span>
    <span class="k">if</span> <span class="n">brc_mode</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">add_insdc_seq_region_name</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">)</span>

    <span class="c1"># Add translation and mitochondrial codon tables</span>
    <span class="n">add_translation_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">)</span>
    <span class="n">add_mitochondrial_codon_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;species&quot;</span><span class="p">][</span><span class="s2">&quot;taxonomy_id&quot;</span><span class="p">])</span>

    <span class="c1"># Print out the file</span>
    <span class="n">print_json</span><span class="p">(</span><span class="n">dst_file</span><span class="p">,</span> <span class="n">seq_regions</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="ensembl.io.genomio.seq_region.report_to_csv" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>          <code class="highlight language-python"><span class="n">report_to_csv</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span></code>

<a href="#ensembl.io.genomio.seq_region.report_to_csv" class="headerlink" title="Permanent link">&para;</a></h3>


  <div class="doc doc-contents ">
  
      <p>Returns an assembly report as a CSV string.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>report_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>path to a seq_region file from INSDC/RefSeq</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Tuple">Tuple</span>[str, <span title="typing.Dict">Dict</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The data as a string in CSV format, and the head metadata as a dictionary.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/python/ensembl/io/genomio/seq_region/prepare.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">report_to_csv</span><span class="p">(</span><span class="n">report_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns an assembly report as a CSV string.</span>

<span class="sd">    Args:</span>
<span class="sd">        report_path: path to a seq_region file from INSDC/RefSeq</span>

<span class="sd">    Returns:</span>
<span class="sd">        The data as a string in CSV format, and the head metadata as a dictionary.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">report</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">metadata</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">last_head</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">report</span><span class="p">:</span>
            <span class="c1"># Ignore header</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="c1"># Get metadata values if possible</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;# (.+?): (.+?)$&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">metadata</span><span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)]</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
                <span class="n">last_head</span> <span class="o">=</span> <span class="n">line</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">last_head</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">+=</span> <span class="n">last_head</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="n">last_head</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="n">data</span> <span class="o">+=</span> <span class="n">line</span>
        <span class="k">return</span> <span class="n">data</span><span class="p">,</span> <span class="n">metadata</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-2023] EMBL-European Bioinformatics Institute
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.8fd75fb4.min.js"></script>
      
    
  </body>
</html>