<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ensembl.io.genomio.seq_region.prepare &#8212; ensembl-genomio 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../../../_static/agogo.css?v=cc331ede" />
    <script data-url_root="../../../../../" id="documentation_options" src="../../../../../_static/documentation_options.js?v=e031e9a9"></script>
    <script src="../../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <link rel="index" title="Index" href="../../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../../search.html" /> 
  </head><body>
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="../../../../../index.html">ensembl-genomio 0.1 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="../../../../../py-modindex.html" title="Python Module Index"
             >modules</a> |
          <a href="../../../../../genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for ensembl.io.genomio.seq_region.prepare</h1><div class="highlight"><pre>
<span></span><span class="c1"># See the NOTICE file distributed with this work for additional information</span>
<span class="c1"># regarding copyright ownership.</span>
<span class="c1">#</span>
<span class="c1"># Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);</span>
<span class="c1"># you may not use this file except in compliance with the License.</span>
<span class="c1"># You may obtain a copy of the License at</span>
<span class="c1">#</span>
<span class="c1">#      http://www.apache.org/licenses/LICENSE-2.0</span>
<span class="c1">#</span>
<span class="c1"># Unless required by applicable law or agreed to in writing, software</span>
<span class="c1"># distributed under the License is distributed on an &quot;AS IS&quot; BASIS,</span>
<span class="c1"># WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.</span>
<span class="c1"># See the License for the specific language governing permissions and</span>
<span class="c1"># limitations under the License.</span>
<span class="sd">&quot;&quot;&quot;Construct a seq_region metadata file from INSDC files.&quot;&quot;&quot;</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;SeqRegion&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SeqRegionDict&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LOCATION_CODON&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MOLECULE_LOCATION&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SYNONYM_MAP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SYNONYM_RESOURCES&quot;</span><span class="p">,</span>
    <span class="s2">&quot;UnknownMetadata&quot;</span><span class="p">,</span>
    <span class="s2">&quot;add_insdc_seq_region_name&quot;</span><span class="p">,</span>
    <span class="s2">&quot;add_mitochondrial_codon_table&quot;</span><span class="p">,</span>
    <span class="s2">&quot;add_translation_table&quot;</span><span class="p">,</span>
    <span class="s2">&quot;exclude_seq_regions&quot;</span><span class="p">,</span>
    <span class="s2">&quot;get_codon_table&quot;</span><span class="p">,</span>
    <span class="s2">&quot;get_gbff_seq_regions&quot;</span><span class="p">,</span>
    <span class="s2">&quot;get_genbank_id&quot;</span><span class="p">,</span>
    <span class="s2">&quot;get_organelle&quot;</span><span class="p">,</span>
    <span class="s2">&quot;get_report_regions&quot;</span><span class="p">,</span>
    <span class="s2">&quot;make_seq_region&quot;</span><span class="p">,</span>
    <span class="s2">&quot;merge_seq_regions&quot;</span><span class="p">,</span>
    <span class="s2">&quot;prepare_seq_region_metadata&quot;</span><span class="p">,</span>
    <span class="s2">&quot;report_to_csv&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">from</span> <span class="nn">os</span> <span class="kn">import</span> <span class="n">PathLike</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span>

<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">import</span> <span class="nn">requests</span>

<span class="kn">from</span> <span class="nn">ensembl.io.genomio.utils</span> <span class="kn">import</span> <span class="n">get_json</span><span class="p">,</span> <span class="n">open_gz_file</span><span class="p">,</span> <span class="n">print_json</span>
<span class="kn">from</span> <span class="nn">ensembl.utils.argparse</span> <span class="kn">import</span> <span class="n">ArgumentParser</span>


<span class="c1"># Definition of SeqRegion types</span>
<span class="n">SeqRegion</span> <span class="o">=</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span>
<span class="n">SeqRegionDict</span> <span class="o">=</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqRegion</span><span class="p">]</span>


<span class="n">SYNONYM_RESOURCES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GenBank&quot;</span><span class="p">,</span> <span class="s2">&quot;RefSeq&quot;</span><span class="p">,</span> <span class="s2">&quot;INSDC&quot;</span><span class="p">]</span>
<span class="n">SYNONYM_MAP</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;Assigned-Molecule&quot;</span><span class="p">:</span> <span class="s2">&quot;INSDC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GenBank-Accn&quot;</span><span class="p">:</span> <span class="s2">&quot;GenBank&quot;</span><span class="p">,</span>
    <span class="s2">&quot;RefSeq-Accn&quot;</span><span class="p">:</span> <span class="s2">&quot;RefSeq&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Sequence-Name&quot;</span><span class="p">:</span> <span class="s2">&quot;INSDC_submitted_name&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">MOLECULE_LOCATION</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;apicoplast&quot;</span><span class="p">:</span> <span class="s2">&quot;apicoplast_chromosome&quot;</span><span class="p">,</span>
    <span class="s2">&quot;chromosome&quot;</span><span class="p">:</span> <span class="s2">&quot;nuclear_chromosome&quot;</span><span class="p">,</span>
    <span class="s2">&quot;kinetoplast&quot;</span><span class="p">:</span> <span class="s2">&quot;kinetoplast_chromosome&quot;</span><span class="p">,</span>
    <span class="s2">&quot;linkage group&quot;</span><span class="p">:</span> <span class="s2">&quot;linkage_group&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mitochondrion&quot;</span><span class="p">:</span> <span class="s2">&quot;mitochondrial_chromosome&quot;</span><span class="p">,</span>
    <span class="s2">&quot;plasmid&quot;</span><span class="p">:</span> <span class="s2">&quot;plasmid&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">LOCATION_CODON</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;apicoplast_chromosome&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">}</span>


<div class="viewcode-block" id="UnknownMetadata"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.UnknownMetadata">[docs]</a><span class="k">class</span> <span class="nc">UnknownMetadata</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;If a metadata if not supported or recognized.&quot;&quot;&quot;</span></div>


<div class="viewcode-block" id="exclude_seq_regions"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.exclude_seq_regions">[docs]</a><span class="k">def</span> <span class="nf">exclude_seq_regions</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">to_exclude</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the list of sequence regions with the ones from the exclusion list removed.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        to_exclude: Sequence region names to exclude.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filtered_seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;name&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">to_exclude</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Remove seq_region </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">filtered_seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">filtered_seq_regions</span></div>


<div class="viewcode-block" id="add_translation_table"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.add_translation_table">[docs]</a><span class="k">def</span> <span class="nf">add_translation_table</span><span class="p">(</span>
    <span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">location_codon</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the translation codon table to each sequence region (when missing) based on its location.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        location_codon: Map of known codon tables for known locations.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">location_codon</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">location_codon</span> <span class="o">=</span> <span class="n">LOCATION_CODON</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="c1"># Do not overwrite any existing codon table</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;codon_table&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;location&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">location_codon</span><span class="p">):</span>
            <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location_codon</span><span class="p">[</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]]</span></div>


<div class="viewcode-block" id="add_insdc_seq_region_name"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.add_insdc_seq_region_name">[docs]</a><span class="k">def</span> <span class="nf">add_insdc_seq_region_name</span><span class="p">(</span>
    <span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">synonym_sources</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the list of sequence regions with their corresponding INSDC sequence region names.</span>

<span class="sd">    &quot;BRC4_seq_region_name&quot; and &quot;EBI_seq_region_name&quot; fields are added to each sequence region: the</span>
<span class="sd">    former will contain the corresponding INSDC name whilst the latter will contain the current name.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        synonym_sources: Synonym sources to use for the BRC4 name, in order of preference.</span>

<span class="sd">    Raises:</span>
<span class="sd">        UnknownMetadata: If no synonym name is found for a sequence region.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">synonym_sources</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">synonym_sources</span> <span class="o">=</span> <span class="n">SYNONYM_RESOURCES</span>
    <span class="n">new_seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="p">{</span><span class="n">synonym</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]:</span> <span class="n">synonym</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">synonym</span> <span class="ow">in</span> <span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;synonyms&quot;</span><span class="p">,</span> <span class="p">[])}</span>
        <span class="c1"># Choose the synonym to use as the BRC name, using the first valid name from the source list</span>
        <span class="k">for</span> <span class="n">source_name</span> <span class="ow">in</span> <span class="n">synonym_sources</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">source_name</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
                <span class="n">brc_name</span> <span class="o">=</span> <span class="n">names</span><span class="p">[</span><span class="n">source_name</span><span class="p">]</span>
                <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Cannot set BRC4 sequence region name for </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">brc_name</span> <span class="o">=</span> <span class="n">brc_name</span><span class="o">.</span><span class="n">partition</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;BRC4_seq_region_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">brc_name</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;EBI_seq_region_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
        <span class="n">new_seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">new_seq_regions</span></div>


<div class="viewcode-block" id="add_mitochondrial_codon_table"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.add_mitochondrial_codon_table">[docs]</a><span class="k">def</span> <span class="nf">add_mitochondrial_codon_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the mitochondrial codon table to each sequence region (when missing) based on the taxon ID.</span>

<span class="sd">    If no mitochondrial genetic code can be found for the given taxon ID nothing will be changed.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        taxon_id: The species taxon ID.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/data/taxonomy/v1/taxon/tax-id/</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">headers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Content-Type&quot;</span><span class="p">:</span> <span class="s2">&quot;application/json&quot;</span><span class="p">},</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">60</span><span class="p">)</span>
    <span class="n">decoded</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
    <span class="k">if</span> <span class="s2">&quot;mitochondrialGeneticCode&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">decoded</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No mitochondria genetic code found for taxon </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">genetic_code</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">decoded</span><span class="p">[</span><span class="s2">&quot;mitochondrialGeneticCode&quot;</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="n">location</span> <span class="o">=</span> <span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;location&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="c1"># Do not overwrite any existing mitochondrial codon table</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">location</span> <span class="o">==</span> <span class="s2">&quot;mitochondrial_chromosome&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;codon_table&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">):</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genetic_code</span></div>


<div class="viewcode-block" id="merge_seq_regions"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.merge_seq_regions">[docs]</a><span class="k">def</span> <span class="nf">merge_seq_regions</span><span class="p">(</span>
    <span class="n">left</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqRegionDict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">right</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqRegionDict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Merges sequence regions from different sources.</span>

<span class="sd">    When combining two regions matching the same key, the &quot;right&quot; seq region data will take precedence</span>
<span class="sd">    over the &quot;left&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        left: Dictionary of sequence regions with names as keys.</span>
<span class="sd">        right: Dictionary of sequence regions with names as keys.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A list of merged sequence regions, sorted by their name.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">left</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">left</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">right</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">right</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Get all the names</span>
    <span class="n">all_names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">left</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">right</span><span class="p">)</span>
    <span class="c1"># Create the seq_regions, merge if needed</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">all_names</span><span class="p">:</span>
        <span class="n">left_seqr</span> <span class="o">=</span> <span class="n">left</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{})</span>
        <span class="n">right_seqr</span> <span class="o">=</span> <span class="n">right</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{})</span>
        <span class="k">if</span> <span class="n">left_seqr</span> <span class="ow">and</span> <span class="n">right_seqr</span><span class="p">:</span>
            <span class="n">final_seqr</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="n">left_seqr</span><span class="p">,</span> <span class="o">**</span><span class="n">right_seqr</span><span class="p">}</span>
        <span class="k">elif</span> <span class="n">left_seqr</span><span class="p">:</span>
            <span class="n">final_seqr</span> <span class="o">=</span> <span class="n">left_seqr</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">final_seqr</span> <span class="o">=</span> <span class="n">right_seqr</span>
        <span class="n">seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">final_seqr</span><span class="p">)</span>
    <span class="n">seq_regions</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">seq_regions</span></div>


<div class="viewcode-block" id="get_gbff_seq_regions"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.get_gbff_seq_regions">[docs]</a><span class="k">def</span> <span class="nf">get_gbff_seq_regions</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the sequence regions found in the GBFF file (if any).</span>

<span class="sd">    Args:</span>
<span class="sd">        gbff_path: Path to GBFF file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dict of SeqRegions, with their name as the key.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gbff_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gbff_file</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
            <span class="n">seqr</span><span class="p">:</span> <span class="n">SeqRegion</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;length&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)}</span>
            <span class="c1"># Is the seq_region circular?</span>
            <span class="n">annotations</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span>
            <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;topology&quot;</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;topology&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;circular&quot;</span><span class="p">):</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;circular&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1"># Is there a genetic code defined?</span>
            <span class="n">codon_table</span> <span class="o">=</span> <span class="n">get_codon_table</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">codon_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">codon_table</span>
            <span class="c1"># Is it an organelle?</span>
            <span class="n">location</span> <span class="o">=</span> <span class="n">get_organelle</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">location</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location</span>
            <span class="c1"># Is there a comment stating the Genbank record this is based on?</span>
            <span class="n">genbank_id</span> <span class="o">=</span> <span class="n">get_genbank_id</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">genbank_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="s2">&quot;INSDC&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">genbank_id</span><span class="p">}]</span>
            <span class="c1"># Store the seq_region</span>
            <span class="n">seq_regions</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqr</span>
    <span class="k">return</span> <span class="n">seq_regions</span></div>


<div class="viewcode-block" id="get_genbank_id"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.get_genbank_id">[docs]</a><span class="k">def</span> <span class="nf">get_genbank_id</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the GenBank accession from a given sequence record (if present).</span>

<span class="sd">    Only useful for RefSeq sequence records, where the GenBank accession is stored in a comment.</span>

<span class="sd">    Args:</span>
<span class="sd">        record: Sequence record.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genbank_id</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="s2">&quot;comment&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
        <span class="n">comment</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;comment&quot;</span><span class="p">]</span>
        <span class="n">comment</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[ \n\r]+&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">comment</span><span class="p">)</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;The reference sequence was derived from ([^\.]+)\.&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
            <span class="n">genbank_id</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genbank_id</span></div>


<div class="viewcode-block" id="get_codon_table"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.get_codon_table">[docs]</a><span class="k">def</span> <span class="nf">get_codon_table</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the codon table number from a given a GenBank sequence record (if present).</span>

<span class="sd">    Args:</span>
<span class="sd">        record: GenBank sequence record.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">table_number</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;transl_table&quot;</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
            <span class="n">table_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;transl_table&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">table_number</span></div>


<div class="viewcode-block" id="get_organelle"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.get_organelle">[docs]</a><span class="k">def</span> <span class="nf">get_organelle</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">,</span> <span class="n">molecule_location</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the organelle location from the given GenBank record (if present).</span>

<span class="sd">    Args:</span>
<span class="sd">        record: GenBank sequence record.</span>
<span class="sd">        molecule_location: Map of sequence type to SO location.</span>

<span class="sd">    Raises:</span>
<span class="sd">        KeyError: If the location is not part of the controlled vocabulary.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">molecule_location</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">molecule_location</span> <span class="o">=</span> <span class="n">MOLECULE_LOCATION</span>
    <span class="n">location</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;organelle&quot;</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
            <span class="n">organelle</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;organelle&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">organelle</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="c1"># Remove plastid prefix</span>
            <span class="n">with_prefix</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^(plastid|mitochondrion):(.+)$&quot;</span><span class="p">,</span> <span class="n">organelle</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">with_prefix</span><span class="p">:</span>
                <span class="n">organelle</span> <span class="o">=</span> <span class="n">with_prefix</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="c1"># Get controlled name</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">location</span> <span class="o">=</span> <span class="n">molecule_location</span><span class="p">[</span><span class="n">organelle</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence location: </span><span class="si">{</span><span class="n">organelle</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">exc</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">location</span></div>


<div class="viewcode-block" id="get_report_regions"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.get_report_regions">[docs]</a><span class="k">def</span> <span class="nf">get_report_regions</span><span class="p">(</span><span class="n">report_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the sequence regions found in the report file.</span>

<span class="sd">    Args:</span>
<span class="sd">        report_path: Path to the sequence regions report file.</span>
<span class="sd">        is_refseq: True if the source of the report is RefSeq, false if INSDC.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dict of SeqRegions, with their name as the key.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get the report in a CSV format</span>
    <span class="n">report_csv</span> <span class="o">=</span> <span class="n">report_to_csv</span><span class="p">(</span><span class="n">report_path</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># Feed the csv string to the CSV reader</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">report_csv</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_NONE</span><span class="p">)</span>
    <span class="c1"># Create the seq_regions</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">reader</span><span class="p">:</span>
        <span class="n">seq_region</span> <span class="o">=</span> <span class="n">make_seq_region</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">seq_region</span><span class="p">:</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
            <span class="n">seq_regions</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_region</span>
    <span class="k">return</span> <span class="n">seq_regions</span></div>


<div class="viewcode-block" id="make_seq_region"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.make_seq_region">[docs]</a><span class="k">def</span> <span class="nf">make_seq_region</span><span class="p">(</span>
    <span class="n">data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">,</span>
    <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">synonym_map</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">molecule_location</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegion</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a sequence region from the information provided.</span>

<span class="sd">    An empty sequence region will be returned if not accession information is found.</span>

<span class="sd">    Args:</span>
<span class="sd">        data: a dict from the report representing one line, where the key is the column name.</span>
<span class="sd">        is_refseq: True if the source is RefSeq, false if INSDC.</span>
<span class="sd">        synonym_map: Map of INSDC report column names to sequence region field names.</span>
<span class="sd">        molecule_location: Map of sequence type to SO location.</span>

<span class="sd">    Raises:</span>
<span class="sd">        KeyError: If the sequence location is not recognised.</span>
<span class="sd">        UnknownMetadata: if the sequence role is not recognised.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">synonym_map</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">synonym_map</span> <span class="o">=</span> <span class="n">SYNONYM_MAP</span>
    <span class="k">if</span> <span class="n">molecule_location</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">molecule_location</span> <span class="o">=</span> <span class="n">MOLECULE_LOCATION</span>
    <span class="n">seq_region</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Set accession as the sequence region name</span>
    <span class="n">src</span> <span class="o">=</span> <span class="s2">&quot;RefSeq&quot;</span> <span class="k">if</span> <span class="n">is_refseq</span> <span class="k">else</span> <span class="s2">&quot;GenBank&quot;</span>
    <span class="n">accession_id</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s2">-Accn&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">accession_id</span> <span class="ow">and</span> <span class="p">(</span><span class="n">accession_id</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">accession_id</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;No </span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s1"> accession ID found for </span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;Sequence-Name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{}</span>
    <span class="c1"># Add synonyms</span>
    <span class="n">synonyms</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">field</span><span class="p">,</span> <span class="n">source</span> <span class="ow">in</span> <span class="n">synonym_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
            <span class="n">synonym</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">]}</span>
            <span class="n">synonyms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">synonym</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">synonyms</span><span class="p">:</span>
        <span class="n">synonyms</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">])</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">synonyms</span>
    <span class="c1"># Add sequence length</span>
    <span class="n">field</span> <span class="o">=</span> <span class="s2">&quot;Sequence-Length&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">])</span>
    <span class="c1"># Add coordinate system and location</span>
    <span class="n">seq_role</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Sequence-Role&quot;</span><span class="p">]</span>
    <span class="c1"># Scaffold?</span>
    <span class="k">if</span> <span class="n">seq_role</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;unplaced-scaffold&quot;</span><span class="p">,</span> <span class="s2">&quot;unlocalized-scaffold&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;scaffold&quot;</span>
    <span class="c1"># Chromosome? Check location</span>
    <span class="k">elif</span> <span class="n">seq_role</span> <span class="o">==</span> <span class="s2">&quot;assembled-molecule&quot;</span><span class="p">:</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;chromosome&quot;</span>
        <span class="n">location</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Assigned-Molecule-Location/Type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="c1"># Get location metadata</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">molecule_location</span><span class="p">[</span><span class="n">location</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence location: </span><span class="si">{</span><span class="n">location</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">exc</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence role: </span><span class="si">{</span><span class="n">seq_role</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seq_region</span></div>


<div class="viewcode-block" id="report_to_csv"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.report_to_csv">[docs]</a><span class="k">def</span> <span class="nf">report_to_csv</span><span class="p">(</span><span class="n">report_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns an assembly report as a CSV string.</span>

<span class="sd">    Args:</span>
<span class="sd">        report_path: path to a seq_region file from INSDC/RefSeq</span>

<span class="sd">    Returns:</span>
<span class="sd">        The data as a string in CSV format, and the head metadata as a dictionary.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">report</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">metadata</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">last_head</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">report</span><span class="p">:</span>
            <span class="c1"># Ignore header</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="c1"># Get metadata values if possible</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;# (.+?): (.+?)$&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">metadata</span><span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)]</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
                <span class="n">last_head</span> <span class="o">=</span> <span class="n">line</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">last_head</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">+=</span> <span class="n">last_head</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="n">last_head</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="n">data</span> <span class="o">+=</span> <span class="n">line</span>
        <span class="k">return</span> <span class="n">data</span><span class="p">,</span> <span class="n">metadata</span></div>


<div class="viewcode-block" id="prepare_seq_region_metadata"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.seq_region.html#ensembl.io.genomio.seq_region.prepare.prepare_seq_region_metadata">[docs]</a><span class="k">def</span> <span class="nf">prepare_seq_region_metadata</span><span class="p">(</span>
    <span class="n">genome_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">report_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">dst_dir</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">gbff_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">brc_mode</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">to_exclude</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prepares the sequence region metadata found in the INSDC/RefSeq report and GBFF files.</span>

<span class="sd">    The sequence region information is loaded from both sources and combined. Elements are added/excluded</span>
<span class="sd">    as requested, and the final sequence region metadata is dumped in a JSON file that follows the schema</span>
<span class="sd">    defined in &quot;schemas/seq_region_schema.json&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        genome_file: Genome metadata JSON file path.</span>
<span class="sd">        report_file: INSDC/RefSeq sequences report file path to parse.</span>
<span class="sd">        dst_dir: Output folder for the processed sequence regions JSON file.</span>
<span class="sd">        gbff_file: INSDC/RefSeq GBFF file path to parse.</span>
<span class="sd">        brc_mode: Include INSDC sequence region names.</span>
<span class="sd">        to_exclude: Sequence region names to exclude.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome_data</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">genome_file</span><span class="p">)</span>
    <span class="n">dst_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">dst_dir</span><span class="p">)</span>
    <span class="n">dst_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Final file name</span>
    <span class="n">metadata_type</span> <span class="o">=</span> <span class="s2">&quot;seq_region&quot;</span>
    <span class="n">metadata_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metadata_type</span><span class="si">}</span><span class="s2">.json&quot;</span>
    <span class="n">final_path</span> <span class="o">=</span> <span class="n">dst_dir</span> <span class="o">/</span> <span class="n">metadata_file</span>
    <span class="n">is_refseq</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GCF_&quot;</span><span class="p">)</span>
    <span class="c1"># Get the sequence regions from the report and gbff files, and merge them</span>
    <span class="n">report_regions</span> <span class="o">=</span> <span class="n">get_report_regions</span><span class="p">(</span><span class="n">report_file</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">gbff_file</span><span class="p">:</span>
        <span class="n">gbff_regions</span> <span class="o">=</span> <span class="n">get_gbff_seq_regions</span><span class="p">(</span><span class="n">gbff_file</span><span class="p">)</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">merge_seq_regions</span><span class="p">(</span><span class="n">report_regions</span><span class="p">,</span> <span class="n">gbff_regions</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">report_regions</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="c1"># Exclude seq_regions from a list</span>
    <span class="k">if</span> <span class="n">to_exclude</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">exclude_seq_regions</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">to_exclude</span><span class="p">)</span>
    <span class="c1"># Setup the BRC4_seq_region_name</span>
    <span class="k">if</span> <span class="n">brc_mode</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">add_insdc_seq_region_name</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">)</span>
    <span class="c1"># Add translation and mitochondrial codon tables</span>
    <span class="n">add_translation_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">)</span>
    <span class="n">add_mitochondrial_codon_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;species&quot;</span><span class="p">][</span><span class="s2">&quot;taxonomy_id&quot;</span><span class="p">])</span>
    <span class="c1"># Print out the file</span>
    <span class="n">print_json</span><span class="p">(</span><span class="n">final_path</span><span class="p">,</span> <span class="n">seq_regions</span><span class="p">)</span></div>


<span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Module&#39;s entry-point.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s2">&quot;Construct a sequence region metadata file from INSDC files.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_src_path</span><span class="p">(</span><span class="s2">&quot;--genome_file&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Genome metadata JSON file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_src_path</span><span class="p">(</span>
        <span class="s2">&quot;--report_file&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;INSDC/RefSeq sequences report file to parse&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_src_path</span><span class="p">(</span><span class="s2">&quot;--gbff_file&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;INSDC/RefSeq GBFF file to parse&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_dst_path</span><span class="p">(</span>
        <span class="s2">&quot;--dst_dir&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">(),</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output folder for the processed sequence regions JSON file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--brc_mode&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Enable BRC mode&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--to_exclude&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;SEQ_REGION_NAME&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Sequence region names to exclude&quot;</span>
    <span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="n">prepare_seq_region_metadata</span><span class="p">(</span><span class="o">**</span><span class="nb">vars</span><span class="p">(</span><span class="n">args</span><span class="p">))</span>
</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          
          <h3>Table of Contents</h3>
          <p class="caption" role="heading"><span class="caption-text">Index</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../install.html">API Setup and installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules.html">ensembl</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../license.html">License</a></li>
</ul>

          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="../../../../../search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
            </form>
          </div>

        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="../../../../../py-modindex.html" title="Python Module Index"
              >modules</a> |
            <a href="../../../../../genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
    &#169; Copyright [2016-2023] EMBL-European Bioinformatics Institute.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.1.2.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>