<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ensembl.io.genomio.manifest.compute_stats &#8212; ensembl-genomio 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../../../_static/agogo.css?v=cc331ede" />
    <script data-url_root="../../../../../" id="documentation_options" src="../../../../../_static/documentation_options.js?v=e031e9a9"></script>
    <script src="../../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <link rel="index" title="Index" href="../../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../../search.html" /> 
  </head><body>
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="../../../../../index.html">ensembl-genomio 0.1 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="../../../../../py-modindex.html" title="Python Module Index"
             >modules</a> |
          <a href="../../../../../genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for ensembl.io.genomio.manifest.compute_stats</h1><div class="highlight"><pre>
<span></span><span class="c1"># See the NOTICE file distributed with this work for additional information</span>
<span class="c1"># regarding copyright ownership.</span>
<span class="c1">#</span>
<span class="c1"># Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);</span>
<span class="c1"># you may not use this file except in compliance with the License.</span>
<span class="c1"># You may obtain a copy of the License at</span>
<span class="c1">#</span>
<span class="c1">#      http://www.apache.org/licenses/LICENSE-2.0</span>
<span class="c1">#</span>
<span class="c1"># Unless required by applicable law or agreed to in writing, software</span>
<span class="c1"># distributed under the License is distributed on an &quot;AS IS&quot; BASIS,</span>
<span class="c1"># WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.</span>
<span class="c1"># See the License for the specific language governing permissions and</span>
<span class="c1"># limitations under the License.</span>
<span class="sd">&quot;&quot;&quot;Compute stats from the current genome files associated with the manifest.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">from</span> <span class="nn">os</span> <span class="kn">import</span> <span class="n">PathLike</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">shutil</span> <span class="kn">import</span> <span class="n">which</span>
<span class="kn">from</span> <span class="nn">statistics</span> <span class="kn">import</span> <span class="n">mean</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Set</span><span class="p">,</span> <span class="n">Union</span>

<span class="kn">from</span> <span class="nn">BCBio</span> <span class="kn">import</span> <span class="n">GFF</span>

<span class="kn">from</span> <span class="nn">ensembl.io.genomio.utils</span> <span class="kn">import</span> <span class="n">open_gz_file</span>
<span class="kn">from</span> <span class="nn">ensembl.utils.argparse</span> <span class="kn">import</span> <span class="n">ArgumentParser</span>


<div class="viewcode-block" id="BiotypeCounter"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.BiotypeCounter">[docs]</a><span class="k">class</span> <span class="nc">BiotypeCounter</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A counter for a given biotype, given a list of features.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">ids</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">example</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">count</span>
        <span class="k">if</span> <span class="n">ids</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">ids</span>
        <span class="k">if</span> <span class="n">example</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">example</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">example</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">example</span>

<div class="viewcode-block" id="BiotypeCounter.add_id"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.BiotypeCounter.add_id">[docs]</a>    <span class="k">def</span> <span class="nf">add_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a feature to the counter.</span>

<span class="sd">        Args:</span>
<span class="sd">            feature_id (str): Feature id to add.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span></div>

<div class="viewcode-block" id="BiotypeCounter.unique_count"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.BiotypeCounter.unique_count">[docs]</a>    <span class="k">def</span> <span class="nf">unique_count</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Total number feature ids added to the counter so far.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: number of features in the counter.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="StatsError"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.StatsError">[docs]</a><span class="k">class</span> <span class="nc">StatsError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Raised when stats could not be computed.&quot;&quot;&quot;</span></div>


<div class="viewcode-block" id="manifest_stats"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats">[docs]</a><span class="k">class</span> <span class="nc">manifest_stats</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Representation of the statistics of the set of files listed in the manifest file provided.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">accession</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">datasets_bin</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">manifest_dir</span><span class="si">}</span><span class="s2">/manifest.json&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">accession</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">accession</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">error</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">datasets_bin</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">datasets_bin</span> <span class="o">=</span> <span class="s2">&quot;datasets&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">datasets_bin</span> <span class="o">=</span> <span class="n">datasets_bin</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest_parent</span> <span class="o">=</span> <span class="n">manifest_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi</span> <span class="o">=</span> <span class="kc">False</span>

<div class="viewcode-block" id="manifest_stats.run"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.run">[docs]</a>    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stats_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compute stats in the files and output a stats.txt file in the same folder.</span>

<span class="sd">        Raises:</span>
<span class="sd">            StatsError: Could not compute some stats.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">manifest</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_manifest</span><span class="p">()</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accession</span><span class="p">)</span>

        <span class="c1"># Compute the stats from the GFF3 file</span>
        <span class="k">if</span> <span class="s2">&quot;gff3&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gff3_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">]))</span>

        <span class="c1"># Compute the stats from the seq_region file</span>
        <span class="k">if</span> <span class="s2">&quot;seq_region&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq_region_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]))</span>

        <span class="c1"># Print out the stats in a separate file</span>
        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">stats_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stats_out</span><span class="p">:</span>
            <span class="n">stats_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">stats</span><span class="p">))</span>

        <span class="c1"># Die if there were errors in stats comparison</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">StatsError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stats count errors, check the file </span><span class="si">{</span><span class="n">stats_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="manifest_stats.get_manifest"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.get_manifest">[docs]</a>    <span class="k">def</span> <span class="nf">get_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the files metadata from the manifest json file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict: A representation of the manifest json data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_json</span><span class="p">:</span>
            <span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f_json</span><span class="p">)</span>
            <span class="n">manifest_root</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_parent</span>

        <span class="c1"># Use dir name from the manifest</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">manifest_root</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]:</span>
                        <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                        <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest_root</span><span class="p">,</span> <span class="n">file_name</span>
                        <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>

        <span class="k">return</span> <span class="n">manifest</span></div>

<div class="viewcode-block" id="manifest_stats.get_seq_region_stats"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.get_seq_region_stats">[docs]</a>    <span class="k">def</span> <span class="nf">get_seq_region_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_region_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compute stats from the seq_region json file.</span>

<span class="sd">        Args:</span>
<span class="sd">            seq_region_path (Path): the seq_region json file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: Stats from the seq_regions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">seq_region_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
            <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

        <span class="c1"># Get basic data</span>
        <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">circular</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">locations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">codon_tables</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="c1"># Get readable seq_region name:</span>
            <span class="c1"># either use a Genbank synonym, or just the provided seq_region name</span>
            <span class="n">genbank</span> <span class="o">=</span> <span class="s2">&quot;synonyms&quot;</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">and</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GenBank&quot;</span><span class="p">]</span>
            <span class="n">seqr_name</span> <span class="o">=</span> <span class="n">genbank</span> <span class="ow">and</span> <span class="n">genbank</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>

            <span class="c1"># Record the lengths of the elements of each coord_system</span>
            <span class="n">coord_level</span> <span class="o">=</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">coord_level</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="p">:</span>
                <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>

            <span class="c1"># Additional metadata records to count</span>
            <span class="k">if</span> <span class="s2">&quot;circular&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="n">circular</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="s2">&quot;codon_table&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="n">codon_tables</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;codon_table&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;location&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="n">locations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;location&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Stats</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_region_path</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">coord_systems_stats</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_region_special_stats</span><span class="p">(</span><span class="n">circular</span><span class="p">,</span> <span class="n">locations</span><span class="p">,</span> <span class="n">codon_tables</span><span class="p">)</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span></div>

<div class="viewcode-block" id="manifest_stats.coord_systems_stats"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.coord_systems_stats">[docs]</a>    <span class="k">def</span> <span class="nf">coord_systems_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;For each coord_system compute various stats:</span>
<span class="sd">            - number of sequences</span>
<span class="sd">            - sequence length sum, minimum, maximum, mean</span>

<span class="sd">        Args:</span>
<span class="sd">            coord_systems: Coordinate system dictionary of lengths.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list with the computed statistics in a printable format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total coord_systems </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">coord_name</span><span class="p">,</span> <span class="n">lengths</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Coord_system: </span><span class="si">{</span><span class="n">coord_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">stat_counts</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;Number of sequences&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length sum&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length minimum&quot;</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length maximum&quot;</span><span class="p">:</span> <span class="nb">max</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length mean&quot;</span><span class="p">:</span> <span class="n">mean</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="p">}</span>

            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">stat_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span></div>

<div class="viewcode-block" id="manifest_stats.seq_region_special_stats"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.seq_region_special_stats">[docs]</a>    <span class="k">def</span> <span class="nf">seq_region_special_stats</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">circular</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">locations</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">codon_tables</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Prepare stats in case there are circular regions, specific locations and codon_tables.</span>
<span class="sd">                stats.append(f&quot;{count: 9f}\t{name}&quot;)</span>

<span class="sd">        Args:</span>
<span class="sd">            circular: Number of circular regions. Defaults to 0.</span>
<span class="sd">            locations: The regions and their location. Defaults to None.</span>
<span class="sd">            codon_tables: The regions and their codon_table. Defaults to None.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list with the computed statistics in a printable format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">circular</span> <span class="ow">or</span> <span class="n">locations</span> <span class="ow">or</span> <span class="n">codon_tables</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Special&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">circular</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">circular</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="s2">circular sequences&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">locations</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">locations</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with location&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">locations</span><span class="p">:</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">loc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">codon_tables</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">codon_tables</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with codon_table&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="n">codon_tables</span><span class="p">:</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">table</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span></div>

<div class="viewcode-block" id="manifest_stats.get_gff3_stats"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.get_gff3_stats">[docs]</a>    <span class="k">def</span> <span class="nf">get_gff3_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract the gene models from the GFF3 file and compute stats.</span>

<span class="sd">        Args:</span>
<span class="sd">            gff3_path (Path): the GFF3 file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List: Stats from the gene model.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">biotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_biotypes</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span>
        <span class="c1"># Compile final stats</span>
        <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biotypes_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span></div>

<div class="viewcode-block" id="manifest_stats.count_biotypes"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.count_biotypes">[docs]</a>    <span class="k">def</span> <span class="nf">count_biotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Count the biotypes in a GFF3 file.</span>

<span class="sd">        Args:</span>
<span class="sd">            gff3_path: Path to the GFF3 file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dictionary of biotype counters.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff3_handle</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gff3_handle</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">feat1</span> <span class="ow">in</span> <span class="n">rec</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                    <span class="c1"># Check if the gene contains proteins (CDSs),</span>
                    <span class="c1"># and keep a count of all hierarchies (e.g. gene-mRNA-CDS)</span>
                    <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">for</span> <span class="n">feat2</span> <span class="ow">in</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">feat2</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
                            <span class="n">types2</span> <span class="o">=</span> <span class="p">{</span><span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">}</span>
                            <span class="k">if</span> <span class="s2">&quot;CDS&quot;</span> <span class="ow">in</span> <span class="n">types2</span><span class="p">:</span>
                                <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat2</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">feat3</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">feat3</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                                <span class="k">continue</span>
                            <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span>
                                <span class="n">biotypes</span><span class="p">,</span> <span class="n">feat3</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat3</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span>
                            <span class="p">)</span>

                    <span class="c1"># Main categories counts</span>
                    <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">:</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">)</span>
                    <span class="k">elif</span> <span class="n">is_protein</span><span class="p">:</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;PROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Special case, undefined gene-transcript</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span>
                            <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span>
                            <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transcript&quot;</span>
                        <span class="p">):</span>
                            <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;NONPROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                    <span class="c1"># Total</span>
                    <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">):</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">biotypes</span></div>

<div class="viewcode-block" id="manifest_stats.biotypes_stats"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.biotypes_stats">[docs]</a>    <span class="k">def</span> <span class="nf">biotypes_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Prepare biotype stats in order of their name.</span>

<span class="sd">        Args:</span>
<span class="sd">            biotypes: Biotypes counters.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list with the computed statistics in a printable format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sorted_biotypes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">biotypes</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">data</span><span class="p">:</span> <span class="n">BiotypeCounter</span> <span class="o">=</span> <span class="n">biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">sorted_biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">unique_count</span><span class="p">()</span><span class="si">:</span><span class="s2">&gt;9</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">biotype</span><span class="si">:</span><span class="s2">&lt;20</span><span class="si">}</span><span class="se">\t</span><span class="s2">ID = </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">example</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">biotype</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sorted_biotypes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">]</span>
        <span class="k">return</span> <span class="n">stats</span></div>

<div class="viewcode-block" id="manifest_stats.check_ncbi_stats"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.check_ncbi_stats">[docs]</a>    <span class="k">def</span> <span class="nf">check_ncbi_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Use the dataset tool from NCBI to get stats and compare with what we have&quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">stats</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">stats</span>

        <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span>

        <span class="n">datasets_bin</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datasets_bin</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">which</span><span class="p">(</span><span class="n">datasets_bin</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">stats</span>

        <span class="c1"># Get the dataset summary from NCBI</span>
        <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">datasets_bin</span><span class="p">,</span> <span class="s2">&quot;summary&quot;</span><span class="p">,</span> <span class="s2">&quot;genome&quot;</span><span class="p">,</span> <span class="s2">&quot;accession&quot;</span><span class="p">,</span> <span class="n">accession</span><span class="p">]</span>
        <span class="n">result_out</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">result_out</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>

        <span class="c1"># Get stats</span>
        <span class="k">if</span> <span class="s2">&quot;reports&quot;</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
            <span class="n">genome</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;reports&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="s2">&quot;annotation_info&quot;</span> <span class="ow">in</span> <span class="n">genome</span> <span class="ow">and</span> <span class="s2">&quot;stats&quot;</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">]:</span>
                <span class="n">ncbi_stats</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">][</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span>

                <span class="k">if</span> <span class="s2">&quot;gene_counts&quot;</span> <span class="ow">in</span> <span class="n">ncbi_stats</span><span class="p">:</span>
                    <span class="n">counts</span> <span class="o">=</span> <span class="n">ncbi_stats</span><span class="p">[</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">]</span>
                    <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compare_ncbi_counts</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span></div>

<div class="viewcode-block" id="manifest_stats.compare_ncbi_counts"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.compare_ncbi_counts">[docs]</a>    <span class="k">def</span> <span class="nf">compare_ncbi_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">ncbi</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compare specific gene stats from NCBI&quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">maps</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span><span class="s2">&quot;total&quot;</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;protein_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;PROT_gene&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;pseudogene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;non_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;NONPROT_gene&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;other&quot;</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">],</span>
        <span class="p">]</span>

        <span class="k">for</span> <span class="n">count_map</span> <span class="ow">in</span> <span class="n">maps</span><span class="p">:</span>
            <span class="n">ncbi_name</span><span class="p">,</span> <span class="n">prep_name</span> <span class="o">=</span> <span class="n">count_map</span>
            <span class="n">ncbi_count</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">ncbi_name</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
            <span class="n">preped</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="n">biotypes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">prep_name</span><span class="p">)</span>
            <span class="n">prep_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">preped</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">prep_count</span> <span class="o">=</span> <span class="n">preped</span><span class="o">.</span><span class="n">count</span>

            <span class="k">if</span> <span class="n">prep_count</span> <span class="o">!=</span> <span class="n">ncbi_count</span><span class="p">:</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">prep_count</span> <span class="o">-</span> <span class="n">ncbi_count</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DIFF gene count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">ncbi_count</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">diff</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">error</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Same count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">stats</span></div>

<div class="viewcode-block" id="manifest_stats.increment_biotype"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.manifest_stats.increment_biotype">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">feature_biotype</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add the feature to their respective biotype counter.</span>

<span class="sd">        Args:</span>
<span class="sd">            biotypes (Dict[str, BiotypeCounter]): All current biotypes, with their counter.</span>
<span class="sd">            feature_id (str): Feature id to be counted.</span>
<span class="sd">            feature_biotype (str): The biotype of the feature.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">feature_biotype</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">biotypes</span><span class="p">:</span>
            <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span> <span class="o">=</span> <span class="n">BiotypeCounter</span><span class="p">(</span><span class="n">example</span><span class="o">=</span><span class="n">feature_id</span><span class="p">)</span>
        <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span><span class="o">.</span><span class="n">add_id</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../../../ensembl.io.genomio.manifest.html#ensembl.io.genomio.manifest.compute_stats.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main entrypoint.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Compute stats from the current genome files associated with the manifest.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_src_path</span><span class="p">(</span>
        <span class="s2">&quot;--manifest_dir&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Manifest directory where &#39;manifest.json&#39; file is located&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--accession&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Sequence accession ID to compare stats with NCBI&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--datasets_bin&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Datasets bin status&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_dst_path</span><span class="p">(</span><span class="s2">&quot;--stats_file&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output file with the stats&quot;</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="n">mstats</span> <span class="o">=</span> <span class="n">manifest_stats</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">manifest_dir</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">accession</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">datasets_bin</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mstats</span><span class="o">.</span><span class="n">check_ncbi</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">stats_file</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">stats_file</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">stats_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">args</span><span class="o">.</span><span class="n">manifest_dir</span> <span class="o">/</span> <span class="s2">&quot;stats.txt&quot;</span>
    <span class="n">mstats</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">stats_file</span><span class="p">)</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          
          <h3>Table of Contents</h3>
          <p class="caption" role="heading"><span class="caption-text">Index</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../install.html">API Setup and installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules.html">ensembl</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../license.html">License</a></li>
</ul>

          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="../../../../../search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
            </form>
          </div>

        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="../../../../../py-modindex.html" title="Python Module Index"
              >modules</a> |
            <a href="../../../../../genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
    &#169; Copyright [2016-2023] EMBL-European Bioinformatics Institute.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.1.2.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>